{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/keviniu/agent-shutton/blob/main/consurf_copy_kevin.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "JVWrnF_a4KIq"
      },
      "source": [
        "Welcome to ConSurf Colab.\n",
        "\n",
        "To run ConSurf press the play button at the left side of screen.\n",
        "\n",
        "The first run might take a bit long to complete because dependencies have to be installed.\n",
        "\n",
        "\n",
        "A temporary output directory will be created. The name of this directory is the same as the job_name given, a number may be added if such a directory already exists.\n",
        "\n",
        "In the results directory there will be a zip with the results. Download and unzip it.\n",
        "\n",
        "If there is a 3D structure (or model structure) you can create an image with the consevation colors projected on it using PyMOL. There are two structure files with conservation rates in the column of the b factor: In the file ending with _ATOMS_section_With_ConSurf_isd.pdb (or _ATOMS_section_With_ConSurf_isd.cif) areas where there is insufficient data will be colored yellow, in the file ending with _ATOMS_section_With_ConSurf.pdb (or _ATOMS_section_With_ConSurf.cif) these areas will be in the regular color. To view the image open PyMOL, drag one of these files to the PyMOL window, now drag to the window either the file pymol.py or pymol_CBS.py (for color blind results).\n",
        "\n",
        "Contacts: barakyariv@gmail.com"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "id": "zBTwgYcIK6Ul",
        "outputId": "1b95618a-865d-4c4f-9a58-c5f4d8430f85",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000,
          "referenced_widgets": [
            "cf766ab06e0f43208f28fbf02f371aea",
            "b3aa93f6e4a94a369ec1dfa8f3dbb688",
            "0d1cdfdf33ae4d2292a6b191554e6e94",
            "4d1c4ce75b924083bb9ee7ebec04b6d1",
            "a6e7cb42440b4dcb8b464ba99c556cdc",
            "799435b1fda343db866c5fdde712931b",
            "87508c5ca57e4b36a40b54f57f43a18c",
            "8bd9d7a10e69492193c32945b7d1c972",
            "0f8180bdfdfd4e9fb96b36b025f4ca5c",
            "16af9184a54848c685bfb3d380d74fd5",
            "7f5a34d7ce184d5cb5e3ce619924c4a5"
          ]
        },
        "collapsed": true
      },
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "<style>\n",
              "    .container { width: 100% !important; }\n",
              "    .CodeMirror, .output_subarea, .output_area { max-width: 100% !important; }\n",
              "    .input { width: 100% !important; }\n",
              "</style>\n"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\n",
            "Do you have a PDB/uniprot ID? (Y/N):\n",
            "Y\n",
            "Please enter your ID:\n",
            "P54132\n",
            "\n",
            "The PDB has only one chain A\n",
            "\n",
            "Please upload your MSA.\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "     <input type=\"file\" id=\"files-3dbfab4c-6d3f-4833-823a-d93999a45d92\" name=\"files[]\" multiple disabled\n",
              "        style=\"border:none\" />\n",
              "     <output id=\"result-3dbfab4c-6d3f-4833-823a-d93999a45d92\">\n",
              "      Upload widget is only available when the cell has been executed in the\n",
              "      current browser session. Please rerun this cell to enable.\n",
              "      </output>\n",
              "      <script>// Copyright 2017 Google LLC\n",
              "//\n",
              "// Licensed under the Apache License, Version 2.0 (the \"License\");\n",
              "// you may not use this file except in compliance with the License.\n",
              "// You may obtain a copy of the License at\n",
              "//\n",
              "//      http://www.apache.org/licenses/LICENSE-2.0\n",
              "//\n",
              "// Unless required by applicable law or agreed to in writing, software\n",
              "// distributed under the License is distributed on an \"AS IS\" BASIS,\n",
              "// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n",
              "// See the License for the specific language governing permissions and\n",
              "// limitations under the License.\n",
              "\n",
              "/**\n",
              " * @fileoverview Helpers for google.colab Python module.\n",
              " */\n",
              "(function(scope) {\n",
              "function span(text, styleAttributes = {}) {\n",
              "  const element = document.createElement('span');\n",
              "  element.textContent = text;\n",
              "  for (const key of Object.keys(styleAttributes)) {\n",
              "    element.style[key] = styleAttributes[key];\n",
              "  }\n",
              "  return element;\n",
              "}\n",
              "\n",
              "// Max number of bytes which will be uploaded at a time.\n",
              "const MAX_PAYLOAD_SIZE = 100 * 1024;\n",
              "\n",
              "function _uploadFiles(inputId, outputId) {\n",
              "  const steps = uploadFilesStep(inputId, outputId);\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  // Cache steps on the outputElement to make it available for the next call\n",
              "  // to uploadFilesContinue from Python.\n",
              "  outputElement.steps = steps;\n",
              "\n",
              "  return _uploadFilesContinue(outputId);\n",
              "}\n",
              "\n",
              "// This is roughly an async generator (not supported in the browser yet),\n",
              "// where there are multiple asynchronous steps and the Python side is going\n",
              "// to poll for completion of each step.\n",
              "// This uses a Promise to block the python side on completion of each step,\n",
              "// then passes the result of the previous step as the input to the next step.\n",
              "function _uploadFilesContinue(outputId) {\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  const steps = outputElement.steps;\n",
              "\n",
              "  const next = steps.next(outputElement.lastPromiseValue);\n",
              "  return Promise.resolve(next.value.promise).then((value) => {\n",
              "    // Cache the last promise value to make it available to the next\n",
              "    // step of the generator.\n",
              "    outputElement.lastPromiseValue = value;\n",
              "    return next.value.response;\n",
              "  });\n",
              "}\n",
              "\n",
              "/**\n",
              " * Generator function which is called between each async step of the upload\n",
              " * process.\n",
              " * @param {string} inputId Element ID of the input file picker element.\n",
              " * @param {string} outputId Element ID of the output display.\n",
              " * @return {!Iterable<!Object>} Iterable of next steps.\n",
              " */\n",
              "function* uploadFilesStep(inputId, outputId) {\n",
              "  const inputElement = document.getElementById(inputId);\n",
              "  inputElement.disabled = false;\n",
              "\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  outputElement.innerHTML = '';\n",
              "\n",
              "  const pickedPromise = new Promise((resolve) => {\n",
              "    inputElement.addEventListener('change', (e) => {\n",
              "      resolve(e.target.files);\n",
              "    });\n",
              "  });\n",
              "\n",
              "  const cancel = document.createElement('button');\n",
              "  inputElement.parentElement.appendChild(cancel);\n",
              "  cancel.textContent = 'Cancel upload';\n",
              "  const cancelPromise = new Promise((resolve) => {\n",
              "    cancel.onclick = () => {\n",
              "      resolve(null);\n",
              "    };\n",
              "  });\n",
              "\n",
              "  // Wait for the user to pick the files.\n",
              "  const files = yield {\n",
              "    promise: Promise.race([pickedPromise, cancelPromise]),\n",
              "    response: {\n",
              "      action: 'starting',\n",
              "    }\n",
              "  };\n",
              "\n",
              "  cancel.remove();\n",
              "\n",
              "  // Disable the input element since further picks are not allowed.\n",
              "  inputElement.disabled = true;\n",
              "\n",
              "  if (!files) {\n",
              "    return {\n",
              "      response: {\n",
              "        action: 'complete',\n",
              "      }\n",
              "    };\n",
              "  }\n",
              "\n",
              "  for (const file of files) {\n",
              "    const li = document.createElement('li');\n",
              "    li.append(span(file.name, {fontWeight: 'bold'}));\n",
              "    li.append(span(\n",
              "        `(${file.type || 'n/a'}) - ${file.size} bytes, ` +\n",
              "        `last modified: ${\n",
              "            file.lastModifiedDate ? file.lastModifiedDate.toLocaleDateString() :\n",
              "                                    'n/a'} - `));\n",
              "    const percent = span('0% done');\n",
              "    li.appendChild(percent);\n",
              "\n",
              "    outputElement.appendChild(li);\n",
              "\n",
              "    const fileDataPromise = new Promise((resolve) => {\n",
              "      const reader = new FileReader();\n",
              "      reader.onload = (e) => {\n",
              "        resolve(e.target.result);\n",
              "      };\n",
              "      reader.readAsArrayBuffer(file);\n",
              "    });\n",
              "    // Wait for the data to be ready.\n",
              "    let fileData = yield {\n",
              "      promise: fileDataPromise,\n",
              "      response: {\n",
              "        action: 'continue',\n",
              "      }\n",
              "    };\n",
              "\n",
              "    // Use a chunked sending to avoid message size limits. See b/62115660.\n",
              "    let position = 0;\n",
              "    do {\n",
              "      const length = Math.min(fileData.byteLength - position, MAX_PAYLOAD_SIZE);\n",
              "      const chunk = new Uint8Array(fileData, position, length);\n",
              "      position += length;\n",
              "\n",
              "      const base64 = btoa(String.fromCharCode.apply(null, chunk));\n",
              "      yield {\n",
              "        response: {\n",
              "          action: 'append',\n",
              "          file: file.name,\n",
              "          data: base64,\n",
              "        },\n",
              "      };\n",
              "\n",
              "      let percentDone = fileData.byteLength === 0 ?\n",
              "          100 :\n",
              "          Math.round((position / fileData.byteLength) * 100);\n",
              "      percent.textContent = `${percentDone}% done`;\n",
              "\n",
              "    } while (position < fileData.byteLength);\n",
              "  }\n",
              "\n",
              "  // All done.\n",
              "  yield {\n",
              "    response: {\n",
              "      action: 'complete',\n",
              "    }\n",
              "  };\n",
              "}\n",
              "\n",
              "scope.google = scope.google || {};\n",
              "scope.google.colab = scope.google.colab || {};\n",
              "scope.google.colab._files = {\n",
              "  _uploadFiles,\n",
              "  _uploadFilesContinue,\n",
              "};\n",
              "})(self);\n",
              "</script> "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[1;30;43mStreaming output truncated to the last 5000 lines.\u001b[0m\n",
            "14456. A0A0D2PS75_9AGAR/71-432\n",
            "14457. SRR5262249_18766031/13-67\n",
            "14458. A0A250XP69_9CHLO/429-551\n",
            "14459. ERR1719323_1242782/71-229\n",
            "14460. SRR5256885_13416807/41-125\n",
            "14461. ERR1719489_182738/345-468\n",
            "14462. SRR5688572_31117669/366-479\n",
            "14463. A0A0K9NSY9_ZOSMR/278-611\n",
            "14464. SRR2546423_381812/78-154\n",
            "14465. ERR1740116_116240/534-743\n",
            "14466. A0A0G1DZV3_9BACT/148-359\n",
            "14467. SRR4051794_4692298/346-481\n",
            "14468. SRR5882724_302418/339-456\n",
            "14469. GraSoiStandDraft_45_1057281.scaffolds.fasta_scaffold4835035_1/181-382\n",
            "14470. GraSoiStandDraft_45_1057281.scaffolds.fasta_scaffold4835035_1/478-544\n",
            "14471. A0A0G4I1M6_9ALVE/316-645\n",
            "14472. ERR1719322_136041/81-128\n",
            "14473. ERR1719357_1402821/488-604\n",
            "14474. ERR1719247_22510/243-569\n",
            "14475. A0A0N4T593_BRUPA/186-536\n",
            "14476. ERR1719341_995786/17-137\n",
            "14477. SRR5579883_88560/31-358\n",
            "14478. C4A095_BRAFL/1-185\n",
            "14479. E6QTP4_9ZZZZ/27-360\n",
            "14480. A0A2E1AYU0_9EURY/27-346\n",
            "14481. A0A0E0CBZ2_9ORYZ/282-616\n",
            "14482. SRR3989338_3832190/245-459\n",
            "14483. SRR6056300_17100/259-385\n",
            "14484. ERR1719247_3074076/13-124\n",
            "14485. SRR4051794_30084031/47-124\n",
            "14486. SRR5262245_51610182/59-180\n",
            "14487. A0A1W9L7W2_9BACT/230-345\n",
            "14488. ERR1700679_2375242/73-158\n",
            "14489. A0A1J1I9X8_9DIPT/492-843\n",
            "14490. A0A1S9DD20_ASPOZ/639-983\n",
            "14491. A0A157ZVC0_9BURK/139-471\n",
            "14492. R9PM89_PSEHS/37-392\n",
            "14493. ERR1711911_116325/43-96\n",
            "14494. A0A1X6NR90_PORUM/446-655\n",
            "14495. A0A151VZK5_HYPMA/39-379\n",
            "14496. ERR1700712_5081950/24-144\n",
            "14497. ERR1719362_2483620/143-500\n",
            "14498. SRR6266516_1537470/107-436\n",
            "14499. ERR1719483_1572988/79-411\n",
            "14500. ERR1719189_810568/45-191\n",
            "14501. SRR5262249_58597493/154-207\n",
            "14502. SRR3712207_3219664/88-153\n",
            "14503. SRR3990167_1963313/62-182\n",
            "14504. SRR5450759_2723970/517-854\n",
            "14505. SRR5436305_1761212/11-123\n",
            "14506. SRR4051812_12498693/95-415\n",
            "14507. A0A060DI11_AZOBR/75-398\n",
            "14508. M1NA97_DESSD/28-356\n",
            "14509. SRR5687768_16733488/110-257\n",
            "14510. B0SYC6_CAUSK/29-351\n",
            "14511. M9SC40_METAX/147-355\n",
            "14512. A0A199UYX3_ANACO/493-830\n",
            "14513. A0A1U8Q7B9_NELNU/473-659\n",
            "14514. A0A1U8Q7B9_NELNU/750-843\n",
            "14515. SRR5690554_5932356/43-168\n",
            "14516. SRR3954452_6224982/97-181\n",
            "14517. ERR1700693_2379450/312-439\n",
            "14518. A0A1C7M4T9_GRIFR/236-444\n",
            "14519. A0A1Z9NBT7_9FLAO/124-469\n",
            "14520. SRR6476646_815663/14-126\n",
            "14521. SRR5436190_3889155/176-251\n",
            "14522. ERR1719193_604806/124-451\n",
            "14523. ERR1719234_375794/160-500\n",
            "14524. A0A094YLD2_9PROT/63-397\n",
            "14525. T1HRP4_RHOPR/63-174\n",
            "14526. SRR5919108_5663970/40-134\n",
            "14527. ERR1719319_1634406/133-274\n",
            "14528. SRR5690606_30488168/62-135\n",
            "14529. A0A058ZGQ9_9EUKA/749-1098\n",
            "14530. SRR6266849_3417406/352-454\n",
            "14531. ERR1719471_989468/267-409\n",
            "14532. ERR1719186_242965/90-417\n",
            "14533. A0A2A4LPF6_9RHIZ/235-364\n",
            "14534. SRR4029450_12659021/2-97\n",
            "14535. A0A1G6KFV5_9ACTN/267-383\n",
            "14536. ERR1712228_195261/21-96\n",
            "14537. SRR5919198_497773/248-361\n",
            "14538. A0A1V9Z8I7_9STRA/302-636\n",
            "14539. J6HWK3_9FIRM/17-354\n",
            "14540. SRR3989344_5842245/25-376\n",
            "14541. ERR1700733_5626884/9-115\n",
            "14542. A0A2E9KY48_9DELT/149-356\n",
            "14543. ERR1719341_549780/231-572\n",
            "14544. ERR1719329_975667/338-465\n",
            "14545. D3NXJ7_AZOS1/36-366\n",
            "14546. I6YI10_LINUS/451-778\n",
            "14547. OM-RGC.v1.025677438/338-547\n",
            "14548. SRR5690606_23353969/1-51\n",
            "14549. ERR1740128_325879/386-508\n",
            "14550. K8EL30_9CHLO/126-204\n",
            "14551. K8EL30_9CHLO/454-527\n",
            "14552. ERR1719330_1598764/96-455\n",
            "14553. ERR1719240_1642501/165-270\n",
            "14554. G7E9L6_MIXOS/461-815\n",
            "14555. SRR5690606_3848226/1-51\n",
            "14556. A0A177QN03_9GAMM/25-355\n",
            "14557. SRR5450830_17147/93-418\n",
            "14558. SRR3989344_4633014/86-444\n",
            "14559. Laugresbdmm110dd_1035094.scaffolds.fasta_scaffold207747_1/497-563\n",
            "14560. ERR1719510_598081/285-484\n",
            "14561. N8YCR3_ACIGI/247-368\n",
            "14562. SRR5438270_10115607/113-473\n",
            "14563. SRR5690606_8433320/93-157\n",
            "14564. SRR5215213_2193173/170-231\n",
            "14565. K1PI60_CRAGI/258-442\n",
            "14566. SRR6266545_1725528/360-479\n",
            "14567. SRR5690348_2027626/190-263\n",
            "14568. SRR5574344_275838/319-424\n",
            "14569. SRR6266404_866394/141-501\n",
            "14570. SRR4051812_11627419/160-307\n",
            "14571. SRR3546814_362911/133-455\n",
            "14572. A0A0V0VM20_9BILA/131-480\n",
            "14573. ERR1740123_2463149/448-553\n",
            "14574. W9SBA1_9ROSA/1585-1726\n",
            "14575. ERR550534_804444/406-527\n",
            "14576. A0A074Y3M4_AURPU/91-420\n",
            "14577. ERR1719460_1782571/67-392\n",
            "14578. SRR5665648_122140/73-424\n",
            "14579. SRR5665213_319946/313-626\n",
            "14580. SRR6187200_3180065/2-116\n",
            "14581. U5EVA7_9DIPT/101-442\n",
            "14582. SRR6266545_1033972/165-253\n",
            "14583. SRR5687768_6826294/15-125\n",
            "14584. SRR5216684_3020224/2-55\n",
            "14585. SRR5271165_2665839/328-531\n",
            "14586. A0A2H0I594_9BACT/220-347\n",
            "14587. SRR6185295_9966143/38-348\n",
            "14588. SRR3989339_611420/102-249\n",
            "14589. SRR3954468_10246000/108-167\n",
            "14590. ERR1740129_458971/240-347\n",
            "14591. SRR3954453_2827586/3-91\n",
            "14592. A0A267DZQ7_9PLAT/49-372\n",
            "14593. ERR1740129_2619048/162-276\n",
            "14594. A0A2E0AC24_9RHOB/20-359\n",
            "14595. SRR6266508_565698/384-495\n",
            "14596. A0A0V0QSQ9_PSEPJ/266-508\n",
            "14597. H1BFQ3_9FIRM/34-364\n",
            "14598. SRR4051812_38882604/195-271\n",
            "14599. B7VS27_VIBTL/205-404\n",
            "14600. ERR1039457_5126401/27-82\n",
            "14601. S7U3I3_9DELT/20-375\n",
            "14602. SRR5262249_25783986/93-155\n",
            "14603. ERR1712071_408676/161-497\n",
            "14604. ERR1719483_1373718/6-345\n",
            "14605. SRR3954470_7780612/299-428\n",
            "14606. ERR1719336_236857/98-264\n",
            "14607. SRR5690349_15716533/98-160\n",
            "14608. SRR5690606_20550180/15-131\n",
            "14609. ERR1719499_1660503/344-465\n",
            "14610. ERR1041385_5669996/10-105\n",
            "14611. ERR1719203_2624963/2-93\n",
            "14612. SRR3546814_1361313/136-205\n",
            "14613. SRR5262245_65981859/21-85\n",
            "14614. SRR3954463_15479335/84-208\n",
            "14615. SRR6266567_7996852/48-115\n",
            "14616. SRR5436305_703020/6-157\n",
            "14617. A0A151ZHW7_9MYCE/153-490\n",
            "14618. A0A151VZP9_HYPMA/198-563\n",
            "14619. V5WHU2_9SPIO/25-354\n",
            "14620. A0A1V5UD28_9BACT/131-362\n",
            "14621. ERR1719354_15280/24-379\n",
            "14622. ERR1719487_2488969/81-204\n",
            "14623. SRR5262245_941964/126-184\n",
            "14624. SRR5436309_15492183/94-143\n",
            "14625. A0A066VH79_9HOMO/43-219\n",
            "14626. SRR5690606_13634515/2-56\n",
            "14627. F4PQ59_DICFS/119-448\n",
            "14628. SRR6266487_5700272/8-115\n",
            "14629. SRR3989449_9929817/35-373\n",
            "14630. A0A1Q3W6E2_9BACT/21-356\n",
            "14631. SRR3989338_2375861/25-329\n",
            "14632. ERR1740123_127907/129-265\n",
            "14633. SRR5262245_38546510/115-272\n",
            "14634. ERR1719491_1349769/242-583\n",
            "14635. SRR5574337_595630/173-377\n",
            "14636. A0A1Q5Q7D5_9EURO/641-781\n",
            "14637. SRR4030042_1844015/80-421\n",
            "14638. ERR1719383_1617442/9-198\n",
            "14639. SRR5919112_5013582/97-161\n",
            "14640. SRR5580765_5016430/323-445\n",
            "14641. A0A094ITQ3_9PEZI/564-903\n",
            "14642. A0A2E6M0M7_9GAMM/17-342\n",
            "14643. SRR3954447_9796456/350-488\n",
            "14644. SRR5260364_245602/283-398\n",
            "14645. A0A2E2EJ44_9PROT/115-306\n",
            "14646. SRR5512146_486220/88-152\n",
            "14647. A0A1J4UXQ4_9ARCH/29-354\n",
            "14648. SRR5262249_49278509/19-143\n",
            "14649. SRR5712664_959740/74-247\n",
            "14650. SRR6266571_4289276/76-405\n",
            "14651. SRR5712691_1552575/192-503\n",
            "14652. X6NB67_RETFI/413-527\n",
            "14653. SRR5215831_16819145/261-588\n",
            "14654. ERR1719387_146310/12-127\n",
            "14655. SRR6478736_4220735/166-500\n",
            "14656. R7J0F9_9BACT/300-433\n",
            "14657. SRR3989339_27479/194-383\n",
            "14658. SRR4249919_418005/487-603\n",
            "14659. SRR5471032_2989980/211-328\n",
            "14660. A0A2D5XWZ5_9ARCH/20-347\n",
            "14661. A0A250WTQ0_9CHLO/153-568\n",
            "14662. ERR1719378_1617397/63-426\n",
            "14663. A0A1T4WTE5_9BACT/13-345\n",
            "14664. SRR5579864_6145510/117-438\n",
            "14665. SRR6266550_2756444/120-430\n",
            "14666. ERR1711977_539780/389-518\n",
            "14667. SRR4051812_27281023/63-392\n",
            "14668. SRR3990167_2326653/10-123\n",
            "14669. GraSoiStandDraft_46_1057282.scaffolds.fasta_scaffold14802_1/4-292\n",
            "14670. SRR4051812_29677266/8-143\n",
            "14671. ERR1719362_635685/699-854\n",
            "14672. ERR1719399_116357/651-766\n",
            "14673. A0A0C9XQJ7_9AGAR/48-177\n",
            "14674. A0A194UYZ7_9PEZI/344-678\n",
            "14675. SRR3989454_1013229/28-132\n",
            "14676. SRR6478735_3564989/330-553\n",
            "14677. SRR3954451_556184/1-45\n",
            "14678. A0A177T0V7_9BASI/536-843\n",
            "14679. SRR5438067_3618874/117-198\n",
            "14680. Q4FSS4_PSYA2/276-388\n",
            "14681. A0A2E1VZP8_9BACT/17-339\n",
            "14682. I5AU62_EUBCE/30-352\n",
            "14683. A0A1A9TIN6_ANOST/330-674\n",
            "14684. SRR5580693_1403689/38-168\n",
            "14685. A0A1Y5IMV7_OSTTA/471-576\n",
            "14686. SRR3989339_171870/329-448\n",
            "14687. A0A1J4KS76_9EUKA/381-499\n",
            "14688. A0A1G0GVQ5_9GAMM/56-378\n",
            "14689. SRR6266568_1283974/80-408\n",
            "14690. OrbCnscriptome_FD_contig_21_4520766_length_238_multi_51_in_0_out_0_1/9-213\n",
            "14691. A0A152A4G4_9MYCE/362-501\n",
            "14692. ERR1719282_174063/68-176\n",
            "14693. ERR1719361_154960/421-557\n",
            "14694. A0A2C6LII0_9APIC/504-604\n",
            "14695. SRR5579871_5571989/151-480\n",
            "14696. SRR5579871_3433152/113-290\n",
            "14697. SRR5665811_572209/177-394\n",
            "14698. ERR1712086_436375/10-181\n",
            "14699. SRR5688500_7418362/328-447\n",
            "14700. SRR3954469_22310943/47-115\n",
            "14701. ERR1719428_1898743/54-154\n",
            "14702. ERR1035437_8037637/336-665\n",
            "14703. SRR4051812_17522793/58-120\n",
            "14704. ERR1719264_633792/38-186\n",
            "14705. ERR1719387_65787/386-533\n",
            "14706. SRR5579872_3738632/143-476\n",
            "14707. A0A0V0QPA3_PSEPJ/313-468\n",
            "14708. SRR5712664_1732700/29-338\n",
            "14709. SRR5690554_6935435/1-139\n",
            "14710. A0A132HKN2_9BURK/100-428\n",
            "14711. SRR3954471_24540767/289-452\n",
            "14712. SRR5690554_6381758/52-372\n",
            "14713. SRR3989338_7650141/284-421\n",
            "14714. SRR3989344_3301158/53-379\n",
            "14715. SRR5690606_14116030/29-138\n",
            "14716. SRR3954454_5255258/124-210\n",
            "14717. SRR6202046_3857662/93-269\n",
            "14718. SRR5579872_4899254/7-125\n",
            "14719. ERR1712000_335380/41-376\n",
            "14720. SRR5690606_13785131/9-307\n",
            "14721. SRR6516162_4074733/92-411\n",
            "14722. SRR5688572_19339376/46-366\n",
            "14723. L1JV98_GUITH/382-508\n",
            "14724. SRR3546814_14837609/15-130\n",
            "14725. ERR1719154_169291/327-447\n",
            "14726. SRR5215472_1756527/116-441\n",
            "14727. A0A2G2GLE8_9RHOB/39-372\n",
            "14728. SRR4051812_5321094/93-271\n",
            "14729. SRR3954454_15241745/114-224\n",
            "14730. A0A098VVW8_9MICR/485-612\n",
            "14731. ERR1719494_1271012/348-639\n",
            "14732. A0A2D9HND9_9FLAO/309-409\n",
            "14733. ERR1719277_1293919/72-181\n",
            "14734. SRR5690554_227405/91-409\n",
            "14735. E7RVF1_9BURK/420-534\n",
            "14736. SRR6476661_740670/108-188\n",
            "14737. A0A1G7DB54_9ACTO/275-387\n",
            "14738. V6TJQ8_GIAIN/47-390\n",
            "14739. ERR1712071_331626/125-446\n",
            "14740. A0A1M3NR92_9RHIZ/30-345\n",
            "14741. ERR1719508_55462/506-845\n",
            "14742. SRR5579884_115701/237-584\n",
            "14743. SRR5580700_2783891/127-452\n",
            "14744. SRR4051794_9206400/35-143\n",
            "14745. SRR3954464_11922432/2-141\n",
            "14746. A0A2D6HNY0_9PROT/241-352\n",
            "14747. A0A100I549_ASPNG/2160-2498\n",
            "14748. ERR1719362_1267009/58-186\n",
            "14749. R1GJ43_BOTPV/578-702\n",
            "14750. ERR1740139_469496/32-173\n",
            "14751. SaaInlStandDraft_3_1057020.scaffolds.fasta_scaffold245356_1/116-201\n",
            "14752. SRR3954451_15812723/73-182\n",
            "14753. ERR1017187_7407415/174-368\n",
            "14754. SRR5947207_8776076/217-538\n",
            "14755. SRR5690606_2396627/146-332\n",
            "14756. MGYP001069204483/16-253\n",
            "14757. MGYP000498184084/435-970\n",
            "14758. MGYP000743211759/251-615\n",
            "14759. MGYP001115952336/11-605\n",
            "14760. MGYP000972599038/38-533\n",
            "14761. MGYP000860347671/750-1296\n",
            "14762. MGYP001047534190/234-787\n",
            "14763. MGYP000371519947/440-971\n",
            "14764. MGYP000182671255/306-851\n",
            "14765. MGYP000630089940/506-1090\n",
            "14766. MGYP001040821303/93-763\n",
            "14767. MGYP000964122108/539-1157\n",
            "14768. MGYP000861073388/463-986\n",
            "14769. MGYP001042544660/468-990\n",
            "14770. MGYP000010366885/458-980\n",
            "14771. MGYP000654293178/1-192\n",
            "14772. MGYP000867206424/579-1171\n",
            "14773. MGYP000726426085/438-984\n",
            "14774. MGYP000192944718/444-988\n",
            "14775. MGYP000679085165/461-1104\n",
            "14776. MGYP001011253141/553-1187\n",
            "14777. MGYP000625261036/83-676\n",
            "14778. MGYP000533639082/20-629\n",
            "14779. MGYP000318162246/786-1237\n",
            "14780. MGYP000682691026/418-922\n",
            "14781. MGYP001139179951/564-1085\n",
            "14782. MGYP000700138449/88-586\n",
            "14783. MGYP000470138932/20-612\n",
            "14784. MGYP000207749160/7-598\n",
            "14785. MGYP000888382522/300-847\n",
            "14786. MGYP001079681137/1-598\n",
            "14787. MGYP000249055056/6-601\n",
            "14788. MGYP000899200778/11-557\n",
            "14789. MGYP000417114301/2-598\n",
            "14790. MGYP001028025687/396-909\n",
            "14791. MGYP000223635268/3-599\n",
            "14792. MGYP000742773876/154-611\n",
            "14793. MGYP000969827279/3-415\n",
            "14794. MGYP000350113285/3-596\n",
            "14795. MGYP000964636743/8-596\n",
            "14796. MGYP000094444657/3-597\n",
            "14797. MGYP000325725071/8-614\n",
            "14798. MGYP000188581518/134-689\n",
            "14799. MGYP000258779985/32-648\n",
            "14800. MGYP001025391946/140-558\n",
            "14801. MGYP001144393048/283-898\n",
            "14802. MGYP000568749291/3-597\n",
            "14803. MGYP000935877864/130-781\n",
            "14804. MGYP001111046254/24-612\n",
            "14805. MGYP000016119721/49-650\n",
            "14806. MGYP000196730768/47-655\n",
            "14807. MGYP000971280713/5-592\n",
            "14808. MGYP000868790940/172-688\n",
            "14809. MGYP001146765718/4-596\n",
            "14810. MGYP000197088239/43-636\n",
            "14811. MGYP000116603901/8-403\n",
            "14812. MGYP000116603901/517-678\n",
            "14813. MGYP000311993347/14-601\n",
            "14814. MGYP000753647968/3-596\n",
            "14815. MGYP000075245938/510-905\n",
            "14816. MGYP001064402101/9-614\n",
            "14817. MGYP000869551680/27-487\n",
            "14818. MGYP000326845602/20-428\n",
            "14819. MGYP000905310364/2-580\n",
            "14820. MGYP000032317760/10-600\n",
            "14821. MGYP000468489795/14-625\n",
            "14822. MGYP000081201288/38-466\n",
            "14823. MGYP000114924935/6-603\n",
            "14824. MGYP000163706289/7-609\n",
            "14825. MGYP000019365035/6-599\n",
            "14826. MGYP000269866767/5-596\n",
            "14827. MGYP000520544397/4-420\n",
            "14828. MGYP000941348668/4-594\n",
            "14829. MGYP000378860562/10-597\n",
            "14830. MGYP000719321874/22-611\n",
            "14831. MGYP000241046400/55-478\n",
            "14832. MGYP000974587901/5-591\n",
            "14833. MGYP000049810212/6-602\n",
            "14834. MGYP000046873087/25-619\n",
            "14835. MGYP000657098888/7-592\n",
            "14836. MGYP001097220225/4-599\n",
            "14837. MGYP000226308047/3-600\n",
            "14838. MGYP000172491421/9-435\n",
            "14839. MGYP000117588604/3-595\n",
            "14840. MGYP000445209150/1-623\n",
            "14841. MGYP000674482237/8-596\n",
            "14842. MGYP001084202102/9-598\n",
            "14843. MGYP000405655188/18-621\n",
            "14844. MGYP000465107667/3-590\n",
            "14845. MGYP000285250978/9-607\n",
            "14846. MGYP000548149653/4-602\n",
            "14847. MGYP000739872203/6-591\n",
            "14848. MGYP001118959114/447-842\n",
            "14849. MGYP001023521727/54-497\n",
            "14850. MGYP000412059808/1-591\n",
            "14851. MGYP000106116212/11-600\n",
            "14852. MGYP000064836475/7-602\n",
            "14853. MGYP000618405225/8-602\n",
            "14854. MGYP000368447309/13-600\n",
            "14855. MGYP000201889859/3-591\n",
            "14856. MGYP000402616025/12-608\n",
            "14857. MGYP000275928458/7-604\n",
            "14858. MGYP000334592856/20-609\n",
            "14859. MGYP000588968227/9-603\n",
            "14860. MGYP000652967266/7-617\n",
            "14861. MGYP000140287925/6-596\n",
            "14862. MGYP000896372299/8-602\n",
            "14863. MGYP000358146703/13-632\n",
            "14864. MGYP001068512799/5-605\n",
            "14865. MGYP001066390229/5-627\n",
            "14866. MGYP000993572013/9-609\n",
            "14867. MGYP000471490357/8-631\n",
            "14868. MGYP000441560349/9-641\n",
            "14869. MGYP000668611734/8-626\n",
            "14870. MGYP000927456580/7-602\n",
            "14871. MGYP000913827829/6-426\n",
            "14872. MGYP000957634137/4-426\n",
            "14873. MGYP000918732400/4-589\n",
            "14874. MGYP000001548730/6-600\n",
            "14875. MGYP000503461356/6-631\n",
            "14876. MGYP000188247539/10-600\n",
            "14877. MGYP000179757317/9-600\n",
            "14878. MGYP001078987651/20-609\n",
            "14879. MGYP000488105273/7-596\n",
            "14880. MGYP000931438741/134-529\n",
            "14881. MGYP000243158564/6-596\n",
            "14882. MGYP000172409058/13-600\n",
            "14883. MGYP000008736148/9-600\n",
            "14884. MGYP000258739870/7-596\n",
            "14885. MGYP000903167928/6-428\n",
            "14886. MGYP000040026194/9-604\n",
            "14887. MGYP001045653981/38-452\n",
            "14888. MGYP000845497178/331-757\n",
            "14889. MGYP000181669741/320-921\n",
            "14890. MGYP000753319624/9-600\n",
            "14891. MGYP000388455571/6-605\n",
            "14892. MGYP000035076653/4-602\n",
            "14893. MGYP000983357245/18-600\n",
            "14894. MGYP000031648801/5-616\n",
            "14895. MGYP000140767254/22-605\n",
            "14896. MGYP000051488293/6-611\n",
            "14897. MGYP000146656824/5-591\n",
            "14898. MGYP000676962417/14-601\n",
            "14899. MGYP000583503195/6-611\n",
            "14900. MGYP000863912994/6-598\n",
            "14901. MGYP000485275415/6-596\n",
            "14902. MGYP000683797698/55-643\n",
            "14903. MGYP001075798426/6-596\n",
            "14904. MGYP000929335215/7-630\n",
            "14905. MGYP000977342239/10-614\n",
            "14906. MGYP000600125407/4-602\n",
            "14907. MGYP000709148474/5-605\n",
            "14908. MGYP000709409977/7-596\n",
            "14909. MGYP000246129731/21-408\n",
            "14910. MGYP000101957425/5-595\n",
            "14911. MGYP000143254840/5-597\n",
            "14912. MGYP001028607816/7-596\n",
            "14913. MGYP000238083196/4-594\n",
            "14914. MGYP000514509981/7-603\n",
            "14915. MGYP000932179809/31-618\n",
            "14916. MGYP001052155644/76-551\n",
            "14917. MGYP001048685078/7-589\n",
            "14918. MGYP000861604380/9-415\n",
            "14919. MGYP000209656116/4-614\n",
            "14920. MGYP000170237660/2-592\n",
            "14921. MGYP000854013140/4-595\n",
            "14922. MGYP000280019102/11-609\n",
            "14923. MGYP000228404532/6-597\n",
            "14924. MGYP001068389985/5-623\n",
            "14925. MGYP000694401357/12-592\n",
            "14926. MGYP000159175600/4-593\n",
            "14927. MGYP000308832785/10-608\n",
            "14928. MGYP000448138530/24-621\n",
            "14929. MGYP000094285882/10-607\n",
            "14930. MGYP000284686321/31-407\n",
            "14931. MGYP000707666007/11-606\n",
            "14932. MGYP000657268776/17-623\n",
            "14933. MGYP001144393264/1-381\n",
            "14934. MGYP000692907659/7-596\n",
            "14935. MGYP001083701249/15-613\n",
            "14936. MGYP000043758592/11-595\n",
            "14937. MGYP000603752853/17-640\n",
            "14938. MGYP000255276657/7-600\n",
            "14939. MGYP001051473237/250-834\n",
            "14940. MGYP000515865641/20-620\n",
            "14941. MGYP000698102525/25-626\n",
            "14942. MGYP000864358693/8-611\n",
            "14943. MGYP000906750885/8-594\n",
            "14944. MGYP000249851559/4-599\n",
            "14945. MGYP000922514686/531-1026\n",
            "14946. MGYP000956029391/6-602\n",
            "14947. MGYP000271157633/9-600\n",
            "14948. MGYP001130189228/18-606\n",
            "14949. MGYP000867201308/18-615\n",
            "14950. MGYP000597445880/2-613\n",
            "14951. MGYP001151621569/9-594\n",
            "14952. MGYP000028112884/1-380\n",
            "14953. MGYP000659733541/3-594\n",
            "14954. MGYP000140283476/6-596\n",
            "14955. MGYP001010197963/23-621\n",
            "14956. MGYP000917657247/35-630\n",
            "14957. MGYP000507026059/8-597\n",
            "14958. MGYP000153256697/12-611\n",
            "14959. MGYP000639724941/8-596\n",
            "14960. MGYP001052503521/44-464\n",
            "14961. MGYP000586646996/9-620\n",
            "14962. MGYP001113677967/8-611\n",
            "14963. MGYP000008524800/3-594\n",
            "14964. MGYP000630775969/7-601\n",
            "14965. MGYP001117449757/4-592\n",
            "14966. MGYP000448171146/8-594\n",
            "14967. MGYP000069543163/9-594\n",
            "14968. MGYP000546461218/17-631\n",
            "14969. MGYP001127357104/9-594\n",
            "14970. MGYP000298518551/35-631\n",
            "14971. MGYP000187931496/3-597\n",
            "14972. MGYP000468065828/7-598\n",
            "14973. MGYP000393643670/6-403\n",
            "14974. MGYP000179723440/31-630\n",
            "14975. MGYP001128248411/7-599\n",
            "14976. MGYP001036742478/10-411\n",
            "14977. MGYP000931461158/15-609\n",
            "14978. MGYP000005854997/24-619\n",
            "14979. MGYP000901167292/11-395\n",
            "14980. MGYP000901167292/501-590\n",
            "14981. MGYP000039673835/19-607\n",
            "14982. MGYP001075950878/11-600\n",
            "14983. MGYP000046905376/28-618\n",
            "14984. MGYP000641245888/7-607\n",
            "14985. MGYP000035104728/7-594\n",
            "14986. MGYP001071293943/9-605\n",
            "14987. MGYP000347479756/8-595\n",
            "14988. MGYP000241327501/27-624\n",
            "14989. MGYP000042718769/9-649\n",
            "14990. MGYP000104565681/8-600\n",
            "14991. MGYP000878721928/7-596\n",
            "14992. MGYP000383299442/11-614\n",
            "14993. MGYP000377363677/8-600\n",
            "14994. MGYP000100245926/5-596\n",
            "14995. MGYP000172392566/7-595\n",
            "14996. MGYP000093349814/45-637\n",
            "14997. MGYP000721664903/8-612\n",
            "14998. MGYP000656834896/27-622\n",
            "14999. MGYP000273968110/4-594\n",
            "15000. MGYP000412430581/8-596\n",
            "15001. MGYP000026483640/53-649\n",
            "15002. MGYP000010533455/256-868\n",
            "15003. MGYP000210768984/3-595\n",
            "15004. MGYP000502842571/8-626\n",
            "15005. MGYP000213759359/12-609\n",
            "15006. MGYP000135559541/9-634\n",
            "15007. MGYP000206351159/8-607\n",
            "15008. MGYP001027067233/13-610\n",
            "15009. MGYP001105025225/32-638\n",
            "15010. MGYP000270575037/4-595\n",
            "15011. MGYP000486491244/3-599\n",
            "15012. MGYP000433393798/3-597\n",
            "15013. MGYP000492333560/21-424\n",
            "15014. MGYP001033455385/5-598\n",
            "15015. MGYP000229971166/3-424\n",
            "15016. MGYP001090398776/7-598\n",
            "15017. MGYP000424527581/15-610\n",
            "15018. MGYP000541776636/8-606\n",
            "15019. MGYP001151579297/19-615\n",
            "15020. MGYP000122570912/9-600\n",
            "15021. MGYP000054539500/31-619\n",
            "15022. MGYP001016892459/7-598\n",
            "15023. MGYP000143161535/7-608\n",
            "15024. MGYP000984352224/6-597\n",
            "15025. MGYP000324057144/14-617\n",
            "15026. MGYP000688405160/11-599\n",
            "15027. MGYP001153418739/8-427\n",
            "15028. MGYP000949072395/4-589\n",
            "15029. MGYP000028633119/4-594\n",
            "15030. MGYP001109133699/7-604\n",
            "15031. MGYP001143048221/14-405\n",
            "15032. MGYP000671601227/8-614\n",
            "15033. MGYP001038416269/11-609\n",
            "15034. MGYP000482833224/4-617\n",
            "15035. MGYP000338085277/8-602\n",
            "15036. MGYP000255702871/2-591\n",
            "15037. MGYP000799824148/21-616\n",
            "15038. MGYP001001557833/28-614\n",
            "15039. MGYP000973489179/7-604\n",
            "15040. MGYP001039593362/6-598\n",
            "15041. MGYP000598445090/5-601\n",
            "15042. MGYP000962049571/13-606\n",
            "15043. MGYP000008366351/7-592\n",
            "15044. MGYP000005832204/5-600\n",
            "15045. MGYP000862941864/7-606\n",
            "15046. MGYP000892873049/6-599\n",
            "15047. MGYP000695839164/7-595\n",
            "15048. MGYP000697295565/10-435\n",
            "15049. MGYP000950623384/3-385\n",
            "15050. MGYP000660448132/19-440\n",
            "15051. MGYP001088526834/7-604\n",
            "15052. MGYP001082137075/2-612\n",
            "15053. MGYP000601204121/5-433\n",
            "15054. MGYP000740102672/16-612\n",
            "15055. MGYP000642052384/5-588\n",
            "15056. MGYP000064517568/10-593\n",
            "15057. MGYP000423218694/10-630\n",
            "15058. MGYP001072872599/3-593\n",
            "15059. MGYP000702113783/7-400\n",
            "15060. MGYP001109304197/64-657\n",
            "15061. MGYP000514368865/12-609\n",
            "15062. MGYP000866838999/11-594\n",
            "15063. MGYP000869301731/16-622\n",
            "15064. MGYP000055139535/52-657\n",
            "15065. MGYP000663315754/5-604\n",
            "15066. MGYP001024305466/8-610\n",
            "15067. MGYP000630194798/23-610\n",
            "15068. MGYP000220369599/6-594\n",
            "15069. MGYP000916981091/12-603\n",
            "15070. MGYP000686232753/5-619\n",
            "15071. MGYP000657412836/5-594\n",
            "15072. MGYP000195988836/10-594\n",
            "15073. MGYP000659736189/8-601\n",
            "15074. MGYP000863937045/8-607\n",
            "15075. MGYP000471446551/15-601\n",
            "15076. MGYP000075398601/7-611\n",
            "15077. MGYP000367812211/6-592\n",
            "15078. MGYP000008366836/11-622\n",
            "15079. MGYP001154399356/8-626\n",
            "15080. MGYP000409785337/2-593\n",
            "15081. MGYP001144234934/5-602\n",
            "15082. MGYP000028723081/3-600\n",
            "15083. MGYP000032770040/5-595\n",
            "15084. MGYP000574643382/15-624\n",
            "15085. MGYP000173662860/5-594\n",
            "15086. MGYP001030769543/11-602\n",
            "15087. MGYP000902860091/19-622\n",
            "15088. MGYP000962765477/18-615\n",
            "15089. MGYP000305426083/89-681\n",
            "15090. MGYP001069028596/6-591\n",
            "15091. MGYP001026485248/14-616\n",
            "15092. MGYP000329481888/9-441\n",
            "15093. MGYP000728389298/120-724\n",
            "15094. MGYP000414991497/32-635\n",
            "15095. MGYP000597363893/7-605\n",
            "15096. MGYP000268057300/7-596\n",
            "15097. MGYP000169717801/25-619\n",
            "15098. MGYP000568271049/10-410\n",
            "15099. MGYP000128168232/2-595\n",
            "15100. MGYP000269708348/14-606\n",
            "15101. MGYP000013474951/26-625\n",
            "15102. MGYP001015718023/5-594\n",
            "15103. MGYP000208572667/10-597\n",
            "15104. MGYP000129670491/12-615\n",
            "15105. MGYP001137124208/5-606\n",
            "15106. MGYP000299906043/4-595\n",
            "15107. MGYP000154680430/11-605\n",
            "15108. MGYP000247101398/7-591\n",
            "15109. MGYP000352245381/3-597\n",
            "15110. MGYP000476905980/3-586\n",
            "15111. MGYP000067666379/13-605\n",
            "15112. MGYP001012106114/21-417\n",
            "15113. MGYP001079219243/13-608\n",
            "15114. MGYP000517405622/20-523\n",
            "15115. MGYP000977950063/15-411\n",
            "15116. MGYP000852413802/6-599\n",
            "15117. MGYP000845351147/13-598\n",
            "15118. MGYP000657264930/14-601\n",
            "15119. MGYP000964149728/64-661\n",
            "15120. MGYP000552875124/20-611\n",
            "15121. MGYP000615825789/11-598\n",
            "15122. MGYP000606606530/5-600\n",
            "15123. MGYP000222571642/4-601\n",
            "15124. MGYP001099501177/7-436\n",
            "15125. MGYP000256510941/21-617\n",
            "15126. MGYP000131202656/46-653\n",
            "15127. MGYP000717106287/24-614\n",
            "15128. MGYP000931658812/8-605\n",
            "15129. MGYP000402995125/15-609\n",
            "15130. MGYP001105685644/2-602\n",
            "15131. MGYP000924628335/46-650\n",
            "15132. MGYP000942085384/8-598\n",
            "15133. MGYP000527032735/43-638\n",
            "15134. MGYP000237489847/12-629\n",
            "15135. MGYP001031971958/21-619\n",
            "15136. MGYP000577121275/13-608\n",
            "15137. MGYP000019870594/15-601\n",
            "15138. MGYP000049749037/42-641\n",
            "15139. MGYP000668105825/5-603\n",
            "15140. MGYP000247614739/13-626\n",
            "15141. MGYP000536585934/20-613\n",
            "15142. MGYP000031663433/16-606\n",
            "15143. MGYP001060074181/2-594\n",
            "15144. MGYP000061661034/10-610\n",
            "15145. MGYP000067076410/6-606\n",
            "15146. MGYP000558780003/28-639\n",
            "15147. MGYP000871420837/15-610\n",
            "15148. MGYP000031180149/11-597\n",
            "15149. MGYP000551371020/70-504\n",
            "15150. MGYP001077655226/5-392\n",
            "15151. MGYP000695854222/10-620\n",
            "15152. MGYP000214469965/4-602\n",
            "15153. MGYP000084116244/9-385\n",
            "15154. MGYP000682784418/4-388\n",
            "15155. MGYP000134292128/4-601\n",
            "15156. MGYP000987175777/18-613\n",
            "15157. MGYP000921509557/15-604\n",
            "15158. MGYP000415725637/2-411\n",
            "15159. MGYP000166815006/271-875\n",
            "15160. MGYP000263184849/4-611\n",
            "15161. MGYP000188556157/10-603\n",
            "15162. MGYP000090059240/34-650\n",
            "15163. MGYP000740085257/13-612\n",
            "15164. MGYP000468544433/6-607\n",
            "15165. MGYP001088377798/28-613\n",
            "15166. MGYP000574214686/6-596\n",
            "15167. MGYP001077192927/27-666\n",
            "15168. MGYP000846088255/5-595\n",
            "15169. MGYP000223318573/12-605\n",
            "15170. MGYP001024811767/7-605\n",
            "15171. MGYP001084762053/13-607\n",
            "15172. MGYP001073475353/7-595\n",
            "15173. MGYP001083858201/473-998\n",
            "15174. MGYP000219900521/256-864\n",
            "15175. MGYP000405660272/4-593\n",
            "15176. MGYP000564917364/10-486\n",
            "15177. MGYP000735599415/5-613\n",
            "15178. MGYP000002833891/19-622\n",
            "15179. MGYP000314258882/10-603\n",
            "15180. MGYP000513702874/2-597\n",
            "15181. MGYP000929381475/3-439\n",
            "15182. MGYP000616135543/10-606\n",
            "15183. MGYP000220796015/11-595\n",
            "15184. MGYP000270521789/7-593\n",
            "15185. MGYP001115902064/17-629\n",
            "15186. MGYP000521568428/4-594\n",
            "15187. MGYP000026404281/59-652\n",
            "15188. MGYP000060323322/4-396\n",
            "15189. MGYP001145894679/83-269\n",
            "15190. MGYP001145894679/290-552\n",
            "15191. MGYP000896994936/5-598\n",
            "15192. MGYP000155037220/18-616\n",
            "15193. MGYP000902418858/9-602\n",
            "15194. MGYP001029084989/7-406\n",
            "15195. MGYP000346563727/11-598\n",
            "15196. MGYP000983248776/45-651\n",
            "15197. MGYP000595606077/7-406\n",
            "15198. MGYP001059044660/12-411\n",
            "15199. MGYP000336065719/6-604\n",
            "15200. MGYP000004288770/10-598\n",
            "15201. MGYP000950667779/28-627\n",
            "15202. MGYP000175406136/18-605\n",
            "15203. MGYP000087677137/10-610\n",
            "15204. MGYP000697297222/46-483\n",
            "15205. MGYP001108965230/6-424\n",
            "15206. MGYP000249776056/12-396\n",
            "15207. MGYP000446766380/5-608\n",
            "15208. MGYP000664818174/14-614\n",
            "15209. MGYP000719231381/20-616\n",
            "15210. MGYP000430378721/29-634\n",
            "15211. MGYP000872080324/12-391\n",
            "15212. MGYP000186510564/21-425\n",
            "15213. MGYP000853231676/15-603\n",
            "15214. MGYP000002813842/4-594\n",
            "15215. MGYP000340469382/8-594\n",
            "15216. MGYP000688527087/41-641\n",
            "15217. MGYP000635681772/12-610\n",
            "15218. MGYP000659728338/79-689\n",
            "15219. MGYP000524033653/8-611\n",
            "15220. MGYP000644531302/6-604\n",
            "15221. MGYP000695649375/6-597\n",
            "15222. MGYP001059742503/7-615\n",
            "15223. MGYP000761716918/1-367\n",
            "15224. MGYP000761716918/653-804\n",
            "15225. MGYP000117165991/16-434\n",
            "15226. MGYP000002665638/11-610\n",
            "15227. MGYP001073047294/7-599\n",
            "15228. MGYP000013512573/5-597\n",
            "15229. MGYP000986002887/6-607\n",
            "15230. MGYP000453355276/10-619\n",
            "15231. MGYP000384928620/30-653\n",
            "15232. MGYP000897043864/16-619\n",
            "15233. MGYP000944860908/10-620\n",
            "15234. MGYP000894048697/57-643\n",
            "15235. MGYP000875059786/8-592\n",
            "15236. MGYP000421625643/4-609\n",
            "15237. MGYP000232911305/15-603\n",
            "15238. MGYP000868133717/18-613\n",
            "15239. MGYP001129098679/3-607\n",
            "15240. MGYP000642256114/7-481\n",
            "15241. MGYP001038924106/5-392\n",
            "15242. MGYP000432211608/76-667\n",
            "15243. MGYP000156215833/12-597\n",
            "15244. MGYP000132217478/6-603\n",
            "15245. MGYP000502878436/42-673\n",
            "15246. MGYP000188552171/6-601\n",
            "15247. MGYP000862239266/11-604\n",
            "15248. MGYP000902439245/10-595\n",
            "15249. MGYP000636464457/19-620\n",
            "15250. MGYP000235617839/9-606\n",
            "15251. MGYP000846771770/19-606\n",
            "15252. MGYP000896127686/7-386\n",
            "15253. MGYP001100004821/9-618\n",
            "15254. MGYP000058668877/11-396\n",
            "15255. MGYP000538117301/8-614\n",
            "15256. MGYP000970496553/14-594\n",
            "15257. MGYP000088747960/29-624\n",
            "15258. MGYP000190845702/13-603\n",
            "15259. MGYP001003415932/20-414\n",
            "15260. MGYP000154875639/7-600\n",
            "15261. MGYP000317175690/10-605\n",
            "15262. MGYP000632679940/10-629\n",
            "15263. MGYP001046974781/8-597\n",
            "15264. MGYP000129480251/9-616\n",
            "15265. MGYP000220319178/4-611\n",
            "15266. MGYP000435942085/5-594\n",
            "15267. MGYP000604153024/9-622\n",
            "15268. MGYP000279670863/8-610\n",
            "15269. MGYP000444488159/8-590\n",
            "15270. MGYP000361440116/5-418\n",
            "15271. MGYP001045281455/8-596\n",
            "15272. MGYP000874968692/22-618\n",
            "15273. MGYP000925232521/11-395\n",
            "15274. MGYP001028808084/9-602\n",
            "15275. MGYP000730524989/9-611\n",
            "15276. MGYP001074385983/7-613\n",
            "15277. MGYP000910481699/7-598\n",
            "15278. MGYP000114954706/9-597\n",
            "15279. MGYP000557178751/14-602\n",
            "15280. MGYP000518182551/29-621\n",
            "15281. MGYP000588148105/5-627\n",
            "15282. MGYP000993789165/5-598\n",
            "15283. MGYP001091890959/8-596\n",
            "15284. MGYP000043456344/56-644\n",
            "15285. MGYP000967567969/6-609\n",
            "15286. MGYP000347345491/17-608\n",
            "15287. MGYP000979592570/6-617\n",
            "15288. MGYP001063356977/1-570\n",
            "15289. MGYP000962058376/5-382\n",
            "15290. MGYP000997818344/12-604\n",
            "15291. MGYP000514732315/104-708\n",
            "15292. MGYP000889588439/6-606\n",
            "15293. MGYP000353580068/463-987\n",
            "15294. MGYP000014721428/10-602\n",
            "15295. MGYP000979410787/37-636\n",
            "15296. MGYP000040489542/12-604\n",
            "15297. MGYP000580441103/6-596\n",
            "15298. MGYP000154715908/20-623\n",
            "15299. MGYP000492079800/4-591\n",
            "15300. MGYP000858921230/9-608\n",
            "15301. MGYP001010789429/32-640\n",
            "15302. MGYP000915923491/8-608\n",
            "15303. MGYP001039412081/7-408\n",
            "15304. MGYP000328841660/8-610\n",
            "15305. MGYP000244275862/13-603\n",
            "15306. MGYP000965737104/14-615\n",
            "15307. MGYP000908837554/8-411\n",
            "15308. MGYP000662722322/9-601\n",
            "15309. MGYP000960779278/38-639\n",
            "15310. MGYP000926591283/7-401\n",
            "15311. MGYP000562431875/11-606\n",
            "15312. MGYP000648419078/391-808\n",
            "15313. MGYP000855187731/7-623\n",
            "15314. MGYP000101642010/11-404\n",
            "15315. MGYP000150124530/6-613\n",
            "15316. MGYP000204796661/9-583\n",
            "15317. MGYP000113733965/5-602\n",
            "15318. MGYP001015643884/11-618\n",
            "15319. MGYP000859667616/7-527\n",
            "15320. MGYP000006186840/6-603\n",
            "15321. MGYP000381810559/11-604\n",
            "15322. MGYP000377402004/25-618\n",
            "15323. MGYP000952165836/14-609\n",
            "15324. MGYP000197484160/36-634\n",
            "15325. MGYP000595957961/142-727\n",
            "15326. MGYP000443453572/5-419\n",
            "15327. MGYP000256223082/6-599\n",
            "15328. MGYP001089476577/4-379\n",
            "15329. MGYP000254407527/4-419\n",
            "15330. MGYP001014719694/29-461\n",
            "15331. MGYP000556074666/2-600\n",
            "15332. MGYP000965353758/19-609\n",
            "15333. MGYP000659734566/6-616\n",
            "15334. MGYP000098321592/15-612\n",
            "15335. MGYP001119808981/7-593\n",
            "15336. MGYP000966993411/6-591\n",
            "15337. MGYP000962354772/6-603\n",
            "15338. MGYP001108621293/7-614\n",
            "15339. MGYP001076980726/5-595\n",
            "15340. MGYP000101638148/17-609\n",
            "15341. MGYP000873817287/6-595\n",
            "15342. MGYP000889773996/46-650\n",
            "15343. MGYP000979410677/5-606\n",
            "15344. MGYP000903258413/6-597\n",
            "15345. MGYP000379637471/11-395\n",
            "15346. MGYP000656068725/7-403\n",
            "15347. MGYP000173182603/10-600\n",
            "15348. MGYP000070633703/18-603\n",
            "15349. MGYP000719411486/6-599\n",
            "15350. MGYP000280085356/5-414\n",
            "15351. MGYP001143600874/16-612\n",
            "15352. MGYP001065211725/19-615\n",
            "15353. MGYP000655914973/13-437\n",
            "15354. MGYP000318395806/7-592\n",
            "15355. MGYP000812446526/22-620\n",
            "15356. MGYP000265404752/12-607\n",
            "15357. MGYP000350431040/3-580\n",
            "15358. MGYP000250565499/35-629\n",
            "15359. MGYP000016312658/10-395\n",
            "15360. MGYP000931690108/10-607\n",
            "15361. MGYP000181418812/7-620\n",
            "15362. MGYP000462813206/18-615\n",
            "15363. MGYP000684039674/7-577\n",
            "15364. MGYP000037091797/14-602\n",
            "15365. MGYP000891448476/26-527\n",
            "15366. MGYP000331640402/8-404\n",
            "15367. MGYP001042843757/14-610\n",
            "15368. MGYP000027917937/4-400\n",
            "15369. MGYP000928747678/4-605\n",
            "15370. MGYP001025380681/7-623\n",
            "15371. MGYP001018889207/9-602\n",
            "15372. MGYP001111174759/10-410\n",
            "15373. MGYP000874197699/10-595\n",
            "15374. MGYP000908130975/14-616\n",
            "15375. MGYP000903849337/24-626\n",
            "15376. MGYP000377742890/8-591\n",
            "15377. MGYP000919806720/8-595\n",
            "15378. MGYP001067676401/4-606\n",
            "15379. MGYP000058194425/7-609\n",
            "15380. MGYP000140753538/11-601\n",
            "15381. MGYP000187966040/4-597\n",
            "15382. MGYP000178629778/20-622\n",
            "15383. MGYP001028945798/11-612\n",
            "15384. MGYP001055461872/8-593\n",
            "15385. MGYP000993251607/21-612\n",
            "15386. MGYP001032770791/22-654\n",
            "15387. MGYP000981702768/7-400\n",
            "15388. MGYP000462632045/5-593\n",
            "15389. MGYP000466447404/19-620\n",
            "15390. MGYP000441424582/6-614\n",
            "15391. MGYP000106663039/20-610\n",
            "15392. MGYP000668412594/15-606\n",
            "15393. MGYP001129216290/6-601\n",
            "15394. MGYP000623576039/3-400\n",
            "15395. MGYP000406156376/13-608\n",
            "15396. MGYP000383653960/13-605\n",
            "15397. MGYP000916186238/8-608\n",
            "15398. MGYP000692900433/2-591\n",
            "15399. MGYP000188287338/5-590\n",
            "15400. MGYP000019242370/6-596\n",
            "15401. MGYP000235149788/5-611\n",
            "15402. MGYP000685496458/6-601\n",
            "15403. MGYP000927932863/8-612\n",
            "15404. MGYP000869287342/12-607\n",
            "15405. MGYP000523469396/27-638\n",
            "15406. MGYP000961877075/6-633\n",
            "15407. MGYP001123581379/24-626\n",
            "15408. MGYP000185407212/24-609\n",
            "15409. MGYP000930729356/8-596\n",
            "15410. MGYP000418409835/14-601\n",
            "15411. MGYP000154800444/4-396\n",
            "15412. MGYP000664031096/11-609\n",
            "15413. MGYP000385519935/7-598\n",
            "15414. MGYP000664052264/12-402\n",
            "15415. MGYP000743029783/7-612\n",
            "15416. MGYP000022800708/5-593\n",
            "15417. MGYP001048609459/19-623\n",
            "15418. MGYP000099895329/11-606\n",
            "15419. MGYP001129225408/8-606\n",
            "15420. MGYP000242163378/13-613\n",
            "15421. MGYP001038818861/7-601\n",
            "15422. MGYP000300745052/8-597\n",
            "15423. MGYP000005421278/11-608\n",
            "15424. MGYP000940158072/9-481\n",
            "15425. MGYP000225167253/7-398\n",
            "15426. MGYP000179972795/3-599\n",
            "15427. MGYP000913764980/12-610\n",
            "15428. MGYP000072936772/6-598\n",
            "15429. MGYP000338274671/3-602\n",
            "15430. MGYP000356483648/10-595\n",
            "15431. MGYP000314290177/12-608\n",
            "15432. MGYP000611845852/9-588\n",
            "15433. MGYP001073309301/10-604\n",
            "15434. MGYP000207968461/5-609\n",
            "15435. MGYP000374417489/33-627\n",
            "15436. MGYP000086894158/4-380\n",
            "15437. MGYP001119826724/53-645\n",
            "15438. MGYP000844287601/6-594\n",
            "15439. MGYP001115213764/5-604\n",
            "15440. MGYP000238787154/8-610\n",
            "15441. MGYP000364289714/10-388\n",
            "15442. MGYP000995866136/6-596\n",
            "15443. MGYP000467514300/8-608\n",
            "15444. MGYP001056433063/47-646\n",
            "15445. MGYP001006510873/22-645\n",
            "15446. MGYP001059935369/26-620\n",
            "15447. MGYP000400741894/10-608\n",
            "15448. MGYP000908794795/13-615\n",
            "15449. MGYP000985141049/17-614\n",
            "15450. MGYP000164972035/9-595\n",
            "15451. MGYP000030761820/1-72\n",
            "15452. MGYP000030761820/117-435\n",
            "15453. MGYP000716080542/28-634\n",
            "15454. MGYP000547127098/14-638\n",
            "15455. MGYP000146653244/2-571\n",
            "15456. MGYP000288712374/14-610\n",
            "15457. MGYP000374141182/112-697\n",
            "15458. MGYP000099368624/14-605\n",
            "15459. MGYP000067073210/5-611\n",
            "15460. MGYP000247338008/7-602\n",
            "15461. MGYP000934978897/4-601\n",
            "15462. MGYP001020439044/34-630\n",
            "15463. MGYP000433266524/8-628\n",
            "15464. MGYP000993102732/8-597\n",
            "15465. MGYP000333129449/8-596\n",
            "15466. MGYP000137838324/5-600\n",
            "15467. MGYP000035196636/10-622\n",
            "15468. MGYP001130698165/6-587\n",
            "15469. MGYP000958406354/7-418\n",
            "15470. MGYP000341898402/19-634\n",
            "15471. MGYP000949907473/6-612\n",
            "15472. MGYP000872350037/5-590\n",
            "15473. MGYP000663122325/11-435\n",
            "15474. MGYP000448018274/9-606\n",
            "15475. MGYP000863711134/8-600\n",
            "15476. MGYP000058317422/12-600\n",
            "15477. MGYP000362379661/10-599\n",
            "15478. MGYP000056810378/5-397\n",
            "15479. MGYP000682525793/6-605\n",
            "15480. MGYP000620590206/42-622\n",
            "15481. MGYP000517571970/13-624\n",
            "15482. MGYP000864570435/9-407\n",
            "15483. MGYP001115949248/29-631\n",
            "15484. MGYP000994058773/6-603\n",
            "15485. MGYP000095721520/27-627\n",
            "15486. MGYP000896486507/18-406\n",
            "15487. MGYP001149189680/10-598\n",
            "15488. MGYP000063899511/15-619\n",
            "15489. MGYP000650884645/7-603\n",
            "15490. MGYP000434891345/4-382\n",
            "15491. MGYP000527050372/12-440\n",
            "15492. MGYP000206349894/9-634\n",
            "15493. MGYP001021282963/8-611\n",
            "15494. MGYP000854516037/38-437\n",
            "15495. MGYP001046688235/31-638\n",
            "15496. MGYP000373546842/16-394\n",
            "15497. MGYP000140278041/18-604\n",
            "15498. MGYP000863308789/8-602\n",
            "15499. MGYP001052141538/17-609\n",
            "15500. MGYP000875721961/7-601\n",
            "15501. MGYP000712331145/37-633\n",
            "15502. MGYP000704404698/7-599\n",
            "15503. MGYP001041098029/29-622\n",
            "15504. MGYP000697985664/5-406\n",
            "15505. MGYP000303652695/12-597\n",
            "15506. MGYP001037873925/6-656\n",
            "15507. MGYP000150224362/12-621\n",
            "15508. MGYP000659682061/15-599\n",
            "15509. MGYP001055038401/54-595\n",
            "15510. MGYP000114238909/37-639\n",
            "15511. MGYP000562833637/10-622\n",
            "15512. MGYP000996762505/6-605\n",
            "15513. MGYP000123778224/6-594\n",
            "15514. MGYP000436213688/10-622\n",
            "15515. MGYP001027166952/6-600\n",
            "15516. MGYP000033954927/4-598\n",
            "15517. MGYP001053538466/7-593\n",
            "15518. MGYP000530576614/25-631\n",
            "15519. MGYP000314754224/4-620\n",
            "15520. MGYP001031150423/20-405\n",
            "15521. MGYP000567616801/8-398\n",
            "15522. MGYP000150354194/45-426\n",
            "15523. MGYP000621433111/15-607\n",
            "15524. MGYP000005525556/9-609\n",
            "15525. MGYP000672000716/9-600\n",
            "15526. MGYP001134536762/7-614\n",
            "15527. MGYP000499665250/6-595\n",
            "15528. MGYP000700963943/10-427\n",
            "15529. MGYP000008768234/5-593\n",
            "15530. MGYP001034112599/6-604\n",
            "15531. MGYP001046970541/14-606\n",
            "15532. MGYP000647967219/4-593\n",
            "15533. MGYP000273972252/101-518\n",
            "15534. MGYP001084837507/8-607\n",
            "15535. MGYP000856957189/12-603\n",
            "15536. MGYP000608796653/17-616\n",
            "15537. MGYP000860403762/6-605\n",
            "15538. MGYP000875159392/8-594\n",
            "15539. MGYP000648645263/15-615\n",
            "15540. MGYP000729886868/24-624\n",
            "15541. MGYP000211528299/46-466\n",
            "15542. MGYP001016442330/24-638\n",
            "15543. MGYP000143708219/47-429\n",
            "15544. MGYP000318394763/11-608\n",
            "15545. MGYP000071802448/9-598\n",
            "15546. MGYP000485768777/28-640\n",
            "15547. MGYP000880954907/3-603\n",
            "15548. MGYP000851468441/6-608\n",
            "15549. MGYP000856607865/8-591\n",
            "15550. MGYP000200334770/2-605\n",
            "15551. MGYP000025021673/22-440\n",
            "15552. MGYP000000302072/45-635\n",
            "15553. MGYP000597089543/68-672\n",
            "15554. MGYP000636799371/11-603\n",
            "15555. MGYP000400472326/67-659\n",
            "15556. MGYP000555049530/4-598\n",
            "15557. MGYP000536508357/18-453\n",
            "15558. MGYP000739397389/28-451\n",
            "15559. MGYP000149031281/7-597\n",
            "15560. MGYP000925355347/2-599\n",
            "15561. MGYP000962512548/9-612\n",
            "15562. MGYP000751623919/4-593\n",
            "15563. MGYP000884786300/6-427\n",
            "15564. MGYP001137426455/4-560\n",
            "15565. MGYP000196209591/8-603\n",
            "15566. MGYP000667857445/64-679\n",
            "15567. MGYP000142882572/65-662\n",
            "15568. MGYP000568725372/10-622\n",
            "15569. MGYP000306594008/14-603\n",
            "15570. MGYP000948395392/9-607\n",
            "15571. MGYP001090573746/12-611\n",
            "15572. MGYP000229842885/10-595\n",
            "15573. MGYP000005839469/10-636\n",
            "15574. MGYP000063601980/18-603\n",
            "15575. MGYP000297058494/16-455\n",
            "15576. MGYP001130922803/13-613\n",
            "15577. MGYP000361910978/7-394\n",
            "15578. MGYP000627642625/7-649\n",
            "15579. MGYP000677441151/67-664\n",
            "15580. MGYP001044719269/8-616\n",
            "15581. MGYP001029580653/13-615\n",
            "15582. MGYP000418357574/41-636\n",
            "15583. MGYP000152560353/14-617\n",
            "15584. MGYP000928835148/5-601\n",
            "15585. MGYP000871420992/8-604\n",
            "15586. MGYP001060795940/55-658\n",
            "15587. MGYP000361947595/9-599\n",
            "15588. MGYP000314255506/5-603\n",
            "15589. MGYP000403208672/7-600\n",
            "15590. MGYP000656453446/3-605\n",
            "15591. MGYP000562742227/43-371\n",
            "15592. MGYP000109006859/8-602\n",
            "15593. MGYP000911723470/11-597\n",
            "15594. MGYP001006346860/9-597\n",
            "15595. MGYP000406033125/11-610\n",
            "15596. MGYP000940781423/8-399\n",
            "15597. MGYP000169625725/10-609\n",
            "15598. MGYP001107278796/8-600\n",
            "15599. MGYP001135962385/10-601\n",
            "15600. MGYP000921140805/19-615\n",
            "15601. MGYP000049652076/11-594\n",
            "15602. MGYP000488217395/32-624\n",
            "15603. MGYP000293221285/17-625\n",
            "15604. MGYP000892046606/10-598\n",
            "15605. MGYP000497365501/6-597\n",
            "15606. MGYP000927456193/3-602\n",
            "15607. MGYP000136572583/11-409\n",
            "15608. MGYP000394721508/15-604\n",
            "15609. MGYP001024174342/7-599\n",
            "15610. MGYP000864720823/9-603\n",
            "15611. MGYP000096535159/5-620\n",
            "15612. MGYP000415532449/20-609\n",
            "15613. MGYP000235854372/9-606\n",
            "15614. MGYP000104657280/19-447\n",
            "15615. MGYP000403888775/13-408\n",
            "15616. MGYP000403888775/565-634\n",
            "15617. MGYP000489547988/8-608\n",
            "15618. MGYP000340685973/11-605\n",
            "15619. MGYP000412409033/15-459\n",
            "15620. MGYP001125971330/7-610\n",
            "15621. MGYP000019115320/12-405\n",
            "15622. MGYP000918859279/14-616\n",
            "15623. MGYP000517660764/13-625\n",
            "15624. MGYP000349514434/6-618\n",
            "15625. MGYP001126062149/2-422\n",
            "15626. MGYP000733884656/10-624\n",
            "15627. MGYP000524819631/3-405\n",
            "15628. MGYP000932416378/9-404\n",
            "15629. MGYP001026031338/8-604\n",
            "15630. MGYP000511515384/16-607\n",
            "15631. MGYP000423367903/55-644\n",
            "15632. MGYP000299824196/9-615\n",
            "15633. MGYP000155253477/15-589\n",
            "15634. MGYP000017991632/5-611\n",
            "15635. MGYP000888400210/7-603\n",
            "15636. MGYP000907464360/11-613\n",
            "15637. MGYP000564551401/91-447\n",
            "15638. MGYP000014376098/18-623\n",
            "15639. MGYP000726978050/12-614\n",
            "15640. MGYP001046678476/8-596\n",
            "15641. MGYP000320584628/10-618\n",
            "15642. MGYP000633238217/23-642\n",
            "15643. MGYP000921233670/3-573\n",
            "15644. MGYP000384912469/32-634\n",
            "15645. MGYP000173227459/4-601\n",
            "15646. MGYP000441520719/8-604\n",
            "15647. MGYP000969610289/38-634\n",
            "15648. MGYP000584695197/20-616\n",
            "15649. MGYP000462061835/11-408\n",
            "15650. MGYP000601485737/23-611\n",
            "15651. MGYP000509360541/13-607\n",
            "15652. MGYP000329905503/13-609\n",
            "15653. MGYP000736868328/14-601\n",
            "15654. MGYP001002756990/7-619\n",
            "15655. MGYP000535010396/13-605\n",
            "15656. MGYP000409402231/13-658\n",
            "15657. MGYP000053388535/3-395\n",
            "15658. MGYP000055800226/21-414\n",
            "15659. MGYP000014611931/7-382\n",
            "15660. MGYP000866692601/2-606\n",
            "15661. MGYP000093516600/10-399\n",
            "15662. MGYP000937869778/36-642\n",
            "15663. MGYP000315486680/22-608\n",
            "15664. MGYP000510069202/9-602\n",
            "15665. MGYP000876374062/7-613\n",
            "15666. MGYP001082191875/13-608\n",
            "15667. MGYP000895088373/5-608\n",
            "15668. MGYP000294175496/11-416\n",
            "15669. MGYP000137209437/12-606\n",
            "15670. MGYP001032674623/19-410\n",
            "15671. MGYP001032674623/496-624\n",
            "15672. MGYP000949187260/10-600\n",
            "15673. MGYP000611744522/6-595\n",
            "15674. MGYP000685732304/8-608\n",
            "15675. MGYP000188557459/16-626\n",
            "15676. MGYP000765001642/6-394\n",
            "15677. MGYP000565450045/4-599\n",
            "15678. MGYP000658938166/10-607\n",
            "15679. MGYP001018958700/5-602\n",
            "15680. MGYP000255117058/5-590\n",
            "15681. MGYP001039400166/8-590\n",
            "15682. MGYP000223968917/29-627\n",
            "15683. MGYP000601479716/11-607\n",
            "15684. MGYP000327273685/5-597\n",
            "15685. MGYP000151697504/4-600\n",
            "15686. MGYP000458397439/18-631\n",
            "15687. MGYP000386209519/23-611\n",
            "15688. MGYP000683189543/11-624\n",
            "15689. MGYP000178607634/7-594\n",
            "15690. MGYP000380024329/9-595\n",
            "15691. MGYP001061344822/11-603\n",
            "15692. MGYP000160665719/9-594\n",
            "15693. MGYP000282609416/4-589\n",
            "15694. MGYP000457003986/6-413\n",
            "15695. MGYP000942053492/8-597\n",
            "15696. MGYP000560189932/11-596\n",
            "15697. MGYP000869340781/12-603\n",
            "15698. MGYP000872849780/8-608\n",
            "15699. MGYP001045152289/11-603\n",
            "15700. MGYP001130174508/12-599\n",
            "15701. MGYP001034379117/10-601\n",
            "15702. MGYP001143321096/7-602\n",
            "15703. MGYP000052302763/2-565\n",
            "15704. MGYP001135991734/52-660\n",
            "15705. MGYP000939233092/55-639\n",
            "15706. MGYP000689031134/27-636\n",
            "15707. MGYP000538907421/12-605\n",
            "15708. MGYP000486462206/8-393\n",
            "15709. MGYP000972456801/12-605\n",
            "15710. MGYP001047540683/3-596\n",
            "15711. MGYP000107619622/6-602\n",
            "15712. MGYP000969540631/12-608\n",
            "15713. MGYP000961969368/15-404\n",
            "15714. MGYP001146280752/12-601\n",
            "15715. MGYP000523272677/10-604\n",
            "15716. MGYP000049353836/12-624\n",
            "15717. MGYP000851612765/7-592\n",
            "15718. MGYP000007581118/6-602\n",
            "15719. MGYP000338352253/8-607\n",
            "15720. MGYP000271128421/7-615\n",
            "15721. MGYP000645069026/6-395\n",
            "15722. MGYP000211062907/34-628\n",
            "15723. MGYP000737297433/8-593\n",
            "15724. MGYP000012183562/26-626\n",
            "15725. MGYP000484974159/7-402\n",
            "15726. MGYP000404728087/4-395\n",
            "15727. MGYP000890722266/7-446\n",
            "15728. MGYP000309579718/121-702\n",
            "15729. MGYP001074050766/10-607\n",
            "15730. MGYP001027977716/6-616\n",
            "15731. MGYP000956825737/2-578\n",
            "15732. MGYP000915200272/12-603\n",
            "15733. MGYP000040964689/6-609\n",
            "15734. MGYP000141076786/15-603\n",
            "15735. MGYP000880039706/7-611\n",
            "15736. MGYP000692533675/9-612\n",
            "15737. MGYP001053773883/11-603\n",
            "15738. MGYP000890745125/7-613\n",
            "15739. MGYP000459640126/74-667\n",
            "15740. MGYP000744431545/12-411\n",
            "15741. MGYP000146820080/4-405\n",
            "15742. MGYP001019365520/13-615\n",
            "15743. MGYP000930795518/22-620\n",
            "15744. MGYP000524894142/4-395\n",
            "15745. MGYP000159077795/7-605\n",
            "15746. MGYP000444019099/4-398\n",
            "15747. MGYP000241617670/6-600\n",
            "15748. MGYP000447271048/38-631\n",
            "15749. MGYP000673785033/12-624\n",
            "15750. MGYP000196811239/12-609\n",
            "15751. MGYP000394664154/9-593\n",
            "15752. MGYP000567759792/5-574\n",
            "15753. MGYP000982492974/7-604\n",
            "15754. MGYP000615580754/8-603\n",
            "15755. MGYP000246117330/4-600\n",
            "15756. MGYP000888316061/35-644\n",
            "15757. MGYP001062499980/9-604\n",
            "15758. MGYP000926004848/14-618\n",
            "15759. MGYP000605845488/7-591\n",
            "15760. MGYP000642790276/21-619\n",
            "15761. MGYP000754805333/9-617\n",
            "15762. MGYP001012323574/11-603\n",
            "15763. MGYP001101252399/20-625\n",
            "15764. MGYP000701778181/29-624\n",
            "15765. MGYP000054423948/16-427\n",
            "15766. MGYP000125220005/35-639\n",
            "15767. MGYP000026060299/7-593\n",
            "15768. MGYP000024912083/26-636\n",
            "15769. MGYP001033296451/19-433\n",
            "15770. MGYP000107781310/23-642\n",
            "15771. MGYP000607090121/18-621\n",
            "15772. MGYP000882906992/21-606\n",
            "15773. MGYP000193211494/10-628\n",
            "15774. MGYP000863132144/39-638\n",
            "15775. MGYP001004209498/3-591\n",
            "15776. MGYP001110004191/19-626\n",
            "15777. MGYP000511782254/16-651\n",
            "15778. MGYP000070372783/7-601\n",
            "15779. MGYP000863747730/14-614\n",
            "15780. MGYP000585272070/4-602\n",
            "15781. MGYP000704710727/7-594\n",
            "15782. MGYP001092978191/10-607\n",
            "15783. MGYP001083384601/9-443\n",
            "15784. MGYP000728257006/3-407\n",
            "15785. MGYP001004528742/25-617\n",
            "15786. MGYP000969599743/12-611\n",
            "15787. MGYP000500354296/2-602\n",
            "15788. MGYP000867288181/12-609\n",
            "15789. MGYP001051481246/14-477\n",
            "15790. MGYP000200435721/11-604\n",
            "15791. MGYP000246913042/11-598\n",
            "15792. MGYP000628011158/21-625\n",
            "15793. MGYP000687099007/4-398\n",
            "15794. MGYP001027002795/13-423\n",
            "15795. MGYP000892636400/7-611\n",
            "15796. MGYP000393740089/21-649\n",
            "15797. MGYP000897121764/72-683\n",
            "15798. MGYP000438474950/3-602\n",
            "15799. MGYP000231437643/3-611\n",
            "15800. MGYP000315445587/13-608\n",
            "15801. MGYP000550072683/6-637\n",
            "15802. MGYP000397281162/12-434\n",
            "15803. MGYP000852363565/28-632\n",
            "15804. MGYP000344539549/5-592\n",
            "15805. MGYP000894525839/7-598\n",
            "15806. MGYP000778246602/52-643\n",
            "15807. MGYP000273815497/7-603\n",
            "15808. MGYP000476936527/4-595\n",
            "15809. MGYP000893310710/18-431\n",
            "15810. MGYP000040502364/9-630\n",
            "15811. MGYP001021295787/7-416\n",
            "15812. MGYP000076320687/17-622\n",
            "15813. MGYP000332667847/61-672\n",
            "15814. MGYP000882892507/15-627\n",
            "15815. MGYP000697533880/11-604\n",
            "15816. MGYP001134468400/9-616\n",
            "15817. MGYP000334758423/11-395\n",
            "15818. MGYP000312412927/10-609\n",
            "15819. MGYP000459657365/22-616\n",
            "15820. MGYP001069396426/10-399\n",
            "15821. MGYP000673678338/10-480\n",
            "15822. MGYP001086386035/14-612\n",
            "15823. MGYP000741732197/12-611\n",
            "15824. MGYP000111893150/9-616\n",
            "15825. MGYP000026417143/13-399\n",
            "15826. MGYP000278552602/11-604\n",
            "15827. MGYP001061959801/15-608\n",
            "15828. MGYP001133249319/49-456\n",
            "15829. MGYP000098995600/5-559\n",
            "15830. MGYP000088764317/11-608\n",
            "15831. MGYP000315128043/12-612\n",
            "15832. MGYP000482026957/11-608\n",
            "15833. MGYP000550696544/8-615\n",
            "15834. MGYP000580761974/8-453\n",
            "15835. MGYP000997146732/6-602\n",
            "15836. MGYP000939490672/2-595\n",
            "15837. MGYP000966976872/7-602\n",
            "15838. MGYP001110479537/9-598\n",
            "15839. MGYP000644985486/6-596\n",
            "15840. MGYP000928277397/21-610\n",
            "15841. MGYP000476438638/6-597\n",
            "15842. MGYP000574977920/3-600\n",
            "15843. MGYP000978277232/3-604\n",
            "15844. MGYP000070248797/15-616\n",
            "15845. MGYP000938894732/9-604\n",
            "15846. MGYP000568757591/7-598\n",
            "15847. MGYP000950667490/15-602\n",
            "15848. MGYP000181215078/6-614\n",
            "15849. MGYP000858060734/10-606\n",
            "15850. MGYP000559509721/13-642\n",
            "15851. MGYP000046371914/12-606\n",
            "15852. MGYP000438355204/17-628\n",
            "15853. MGYP000948681072/6-600\n",
            "15854. MGYP000156264265/8-598\n",
            "15855. MGYP001069794337/5-604\n",
            "15856. MGYP000668118428/7-601\n",
            "15857. MGYP001068408438/12-606\n",
            "15858. MGYP001051343503/4-594\n",
            "15859. MGYP000175528711/8-601\n",
            "15860. MGYP001140991215/10-609\n",
            "15861. MGYP001155044021/21-612\n",
            "15862. MGYP000282665872/13-613\n",
            "15863. MGYP000106518651/10-609\n",
            "15864. MGYP000617983947/20-411\n",
            "15865. MGYP000617983947/557-629\n",
            "15866. MGYP000408681058/29-628\n",
            "15867. MGYP000377390887/5-597\n",
            "15868. MGYP000946443820/20-625\n",
            "15869. MGYP000076201322/9-600\n",
            "15870. MGYP001115692814/11-600\n",
            "15871. MGYP000770164421/3-395\n",
            "15872. MGYP000488256381/5-599\n",
            "15873. MGYP000102730166/6-599\n",
            "15874. MGYP000865103628/11-611\n",
            "15875. MGYP001087300019/15-607\n",
            "15876. MGYP000665023776/12-618\n",
            "15877. MGYP000031682802/18-625\n",
            "15878. MGYP000545502898/8-598\n",
            "15879. MGYP001132201469/17-482\n",
            "15880. MGYP000226305216/5-398\n",
            "15881. MGYP000197420408/6-396\n",
            "15882. MGYP001075986769/49-660\n",
            "15883. MGYP001112079345/16-617\n",
            "15884. MGYP000734773217/17-610\n",
            "15885. MGYP000192091059/15-604\n",
            "15886. MGYP000234497558/9-608\n",
            "15887. MGYP000166597155/14-395\n",
            "15888. MGYP000687030163/6-597\n",
            "15889. MGYP000745099864/4-395\n",
            "15890. MGYP000628863687/4-406\n",
            "15891. MGYP000321048557/10-610\n",
            "15892. MGYP000613526095/4-395\n",
            "15893. MGYP000001433125/7-455\n",
            "15894. MGYP000104766735/4-385\n",
            "15895. MGYP000104766735/429-512\n",
            "15896. MGYP000227099444/4-395\n",
            "15897. MGYP000210949269/14-420\n",
            "15898. MGYP001061683667/14-609\n",
            "15899. MGYP000026708014/11-609\n",
            "15900. MGYP000056286811/13-602\n",
            "15901. MGYP000433357402/10-397\n",
            "15902. MGYP000545269847/8-595\n",
            "15903. MGYP000482165473/51-646\n",
            "15904. MGYP000143083762/10-609\n",
            "15905. MGYP000346489220/4-603\n",
            "15906. MGYP000090670275/11-612\n",
            "15907. MGYP000486429412/14-611\n",
            "15908. MGYP000559607626/28-617\n",
            "15909. MGYP000121317546/61-654\n",
            "15910. MGYP000332619506/10-613\n",
            "15911. MGYP000775837176/23-417\n",
            "15912. MGYP001083636163/21-617\n",
            "15913. MGYP000107719117/5-597\n",
            "15914. MGYP000660056453/9-608\n",
            "15915. MGYP000668956374/8-640\n",
            "15916. MGYP000101606194/11-603\n",
            "15917. MGYP000221107174/5-589\n",
            "15918. MGYP001106402216/6-620\n",
            "15919. MGYP001083951966/9-609\n",
            "15920. MGYP000280721412/2-602\n",
            "15921. MGYP000361547636/12-620\n",
            "15922. MGYP000494914497/6-399\n",
            "15923. MGYP000952681270/9-599\n",
            "15924. MGYP000224384554/35-626\n",
            "15925. MGYP000967673408/2-662\n",
            "15926. MGYP000482323333/5-595\n",
            "15927. MGYP001032098925/8-603\n",
            "15928. MGYP000356716026/31-628\n",
            "15929. MGYP000438648513/9-618\n",
            "15930. MGYP001037997897/33-631\n",
            "15931. MGYP000527065025/13-605\n",
            "15932. MGYP000789111357/4-395\n",
            "15933. MGYP001005533020/20-630\n",
            "15934. MGYP000365053708/26-449\n",
            "15935. MGYP000157961774/188-764\n",
            "15936. MGYP000002195064/5-598\n",
            "15937. MGYP001003673856/10-402\n",
            "15938. MGYP000150534980/28-617\n",
            "15939. MGYP001085399504/17-614\n",
            "15940. MGYP000875655621/8-615\n",
            "15941. MGYP000060601981/10-609\n",
            "15942. MGYP000574933302/16-607\n",
            "15943. MGYP000237275171/1-370\n",
            "15944. MGYP000333051884/8-595\n",
            "15945. MGYP000284691867/5-410\n",
            "15946. MGYP000855396733/5-601\n",
            "15947. MGYP000855554765/35-636\n",
            "15948. MGYP001118545968/16-607\n",
            "15949. MGYP000630877118/12-609\n",
            "15950. MGYP001081831758/3-595\n",
            "15951. MGYP000370813391/22-402\n",
            "15952. MGYP000892258770/5-399\n",
            "15953. MGYP000028648956/7-594\n",
            "15954. MGYP000571209918/9-603\n",
            "15955. MGYP001055271327/5-622\n",
            "15956. MGYP000470196362/12-600\n",
            "15957. MGYP000458430844/3-400\n",
            "15958. MGYP000167121770/18-614\n",
            "15959. MGYP001151306260/33-628\n",
            "15960. MGYP001100203453/17-433\n",
            "15961. MGYP000382622837/15-609\n",
            "15962. MGYP000250603650/17-606\n",
            "15963. MGYP000196702927/15-615\n",
            "15964. MGYP000878421296/13-603\n",
            "15965. MGYP000359056034/14-604\n",
            "15966. MGYP000176450813/9-614\n",
            "15967. MGYP000299935626/8-614\n",
            "15968. MGYP001126665007/6-389\n",
            "15969. MGYP000851886051/9-602\n",
            "15970. MGYP000094123609/4-393\n",
            "15971. MGYP000701218207/41-641\n",
            "15972. MGYP000972689636/9-596\n",
            "15973. MGYP000158972453/5-411\n",
            "15974. MGYP000087218370/12-609\n",
            "15975. MGYP001118997555/11-611\n",
            "15976. MGYP000539585919/2-576\n",
            "15977. MGYP001131206378/3-603\n",
            "15978. MGYP000173200850/10-609\n",
            "15979. MGYP000516548210/30-619\n",
            "15980. MGYP000979817064/12-603\n",
            "15981. MGYP000933576274/6-610\n",
            "15982. MGYP001100985741/13-619\n",
            "15983. MGYP000943571071/7-612\n",
            "15984. MGYP000026117931/15-628\n",
            "15985. MGYP001038004364/6-600\n",
            "15986. MGYP000464330014/10-602\n",
            "15987. MGYP000486378414/7-467\n",
            "15988. MGYP000669225604/6-393\n",
            "15989. MGYP000474274745/7-592\n",
            "15990. MGYP000533250297/10-597\n",
            "15991. MGYP001053778946/9-606\n",
            "15992. MGYP000889085503/7-609\n",
            "15993. MGYP000098667309/8-597\n",
            "15994. MGYP000414426850/32-636\n",
            "15995. MGYP000905525559/27-631\n",
            "15996. MGYP000344015183/6-411\n",
            "15997. MGYP000310419112/12-613\n",
            "15998. MGYP000788275673/9-610\n",
            "15999. MGYP000014264253/9-598\n",
            "16000. MGYP000970280221/20-612\n",
            "16001. MGYP000131333627/6-609\n",
            "16002. MGYP000265339390/7-403\n",
            "16003. MGYP000265339390/494-604\n",
            "16004. MGYP000378186158/13-603\n",
            "16005. MGYP000654216009/12-610\n",
            "16006. MGYP000939215877/13-615\n",
            "16007. MGYP000105841012/15-603\n",
            "16008. MGYP000846018047/30-615\n",
            "16009. MGYP000846219577/27-631\n",
            "16010. MGYP000395900871/4-395\n",
            "16011. MGYP001008605129/35-638\n",
            "16012. MGYP000762567394/32-426\n",
            "16013. MGYP000554511166/4-395\n",
            "16014. MGYP000731113726/8-598\n",
            "16015. MGYP000282480292/4-395\n",
            "16016. MGYP000883500180/5-599\n",
            "16017. MGYP000091217614/21-610\n",
            "16018. MGYP000077948185/13-608\n",
            "16019. MGYP000365473121/6-603\n",
            "16020. MGYP000453384815/4-603\n",
            "16021. MGYP000442238484/16-611\n",
            "16022. MGYP000832128164/4-603\n",
            "16023. MGYP000898487845/7-601\n",
            "16024. MGYP000922363990/6-407\n",
            "16025. MGYP000672244843/14-605\n",
            "16026. MGYP000515630389/20-627\n",
            "16027. MGYP000479169037/13-418\n",
            "16028. MGYP000210655982/6-590\n",
            "16029. MGYP000859412544/17-615\n",
            "16030. MGYP000252835875/14-610\n",
            "16031. MGYP000320621898/12-629\n",
            "16032. MGYP001095302137/14-405\n",
            "16033. MGYP000016370270/7-617\n",
            "16034. MGYP001023279602/7-623\n",
            "16035. MGYP000458375775/16-605\n",
            "16036. MGYP001109886762/4-421\n",
            "16037. MGYP000102367408/9-610\n",
            "16038. MGYP000924480426/26-636\n",
            "16039. MGYP001097326125/4-397\n",
            "16040. MGYP001079734275/6-601\n",
            "16041. MGYP000285681876/11-619\n",
            "16042. MGYP001075871799/17-609\n",
            "16043. MGYP000639559034/7-613\n",
            "16044. MGYP000007952936/10-603\n",
            "16045. MGYP000627893954/5-602\n",
            "16046. MGYP000175511905/2-605\n",
            "16047. MGYP000504616354/4-395\n",
            "16048. MGYP000317717745/4-609\n",
            "16049. MGYP000084763254/6-608\n",
            "16050. MGYP000076797670/11-609\n",
            "16051. MGYP000131916760/2-424\n",
            "16052. MGYP001111928797/4-395\n",
            "16053. MGYP000275842593/30-629\n",
            "16054. MGYP000038236929/4-395\n",
            "16055. MGYP001018917412/13-619\n",
            "16056. MGYP001085555495/7-608\n",
            "16057. MGYP000145010861/16-419\n",
            "16058. MGYP000369381398/4-603\n",
            "16059. MGYP000978645939/4-603\n",
            "16060. MGYP000598913630/4-603\n",
            "16061. MGYP000244765535/4-623\n",
            "16062. MGYP000377810652/4-603\n",
            "16063. MGYP001137651907/12-604\n",
            "16064. MGYP000119475250/7-591\n",
            "16065. MGYP000991138374/10-598\n",
            "16066. MGYP001046313515/10-605\n",
            "16067. MGYP000149053554/10-600\n",
            "16068. MGYP000267444400/15-605\n",
            "16069. MGYP000975971333/21-386\n",
            "16070. MGYP000942076906/18-608\n",
            "16071. MGYP000935809356/11-599\n",
            "16072. MGYP000870404514/7-611\n",
            "16073. MGYP000687369725/11-592\n",
            "16074. MGYP000504040872/22-426\n",
            "16075. MGYP000059980134/3-376\n",
            "16076. MGYP000116681717/14-605\n",
            "16077. MGYP000158436330/22-418\n",
            "16078. MGYP000394513167/14-604\n",
            "16079. MGYP000942036035/6-598\n",
            "16080. MGYP000991661592/11-609\n",
            "16081. MGYP000576623238/6-622\n",
            "16082. MGYP000004086532/16-601\n",
            "16083. MGYP001060556896/9-603\n",
            "16084. MGYP000173864882/7-592\n",
            "16085. MGYP000986781708/14-615\n",
            "16086. MGYP000486453853/9-602\n",
            "16087. MGYP000325911486/9-608\n",
            "16088. MGYP000194560383/13-433\n",
            "16089. MGYP000900674485/4-395\n",
            "16090. MGYP000182125558/10-399\n",
            "16091. MGYP000162740795/4-603\n",
            "16092. MGYP000374638551/4-603\n",
            "16093. MGYP000178511900/3-609\n",
            "16094. MGYP000862312259/33-646\n",
            "16095. MGYP000301450924/3-395\n",
            "16096. MGYP000335369243/8-602\n",
            "16097. MGYP001051294029/7-409\n",
            "16098. MGYP001147954168/7-610\n",
            "16099. MGYP000131285188/10-610\n",
            "16100. MGYP000187715348/3-342\n",
            "16101. MGYP000995824270/6-604\n",
            "16102. MGYP001045392321/2-401\n",
            "16103. MGYP000614347460/8-401\n",
            "16104. MGYP000241646346/9-604\n",
            "16105. MGYP000579315446/4-603\n",
            "16106. MGYP001145617492/4-395\n",
            "16107. MGYP000025747397/7-598\n",
            "16108. MGYP000843365254/4-603\n",
            "16109. MGYP000392601111/12-609\n",
            "16110. MGYP000285262904/9-617\n",
            "16111. MGYP000904838882/14-623\n",
            "16112. MGYP000476947417/7-607\n",
            "16113. MGYP000397603474/16-615\n",
            "16114. MGYP000421976028/12-605\n",
            "16115. MGYP000731767326/18-610\n",
            "16116. MGYP001017627612/17-619\n",
            "16117. MGYP000750407681/10-599\n",
            "16118. MGYP000273204603/4-603\n",
            "16119. MGYP000887264567/25-613\n",
            "16120. MGYP000514633953/3-613\n",
            "16121. MGYP000339455904/3-395\n",
            "16122. MGYP000638492929/10-610\n",
            "16123. MGYP000657614393/3-395\n",
            "16124. MGYP000308233354/7-429\n",
            "16125. MGYP000576286020/8-598\n",
            "16126. MGYP000067563440/6-400\n",
            "16127. MGYP001152811486/10-609\n",
            "16128. MGYP000173658606/5-597\n",
            "16129. MGYP000035076773/6-596\n",
            "16130. MGYP000197081359/9-606\n",
            "16131. MGYP001026122289/7-605\n",
            "16132. MGYP000748636364/14-609\n",
            "16133. MGYP000546377254/4-395\n",
            "16134. MGYP001119767561/13-613\n",
            "16135. MGYP001049771113/6-605\n",
            "16136. MGYP000902488104/26-647\n",
            "16137. MGYP000051456196/17-604\n",
            "16138. MGYP000653675159/5-603\n",
            "16139. MGYP000856132431/2-438\n",
            "16140. MGYP000577491856/13-607\n",
            "16141. MGYP000303189784/24-620\n",
            "16142. MGYP000243421915/11-632\n",
            "16143. MGYP000613436349/4-395\n",
            "16144. MGYP000220374944/1-415\n",
            "16145. MGYP000274577151/4-593\n",
            "16146. MGYP000721126689/10-600\n",
            "16147. MGYP000399666523/4-609\n",
            "16148. MGYP000916993287/14-607\n",
            "16149. MGYP000872400964/7-610\n",
            "16150. MGYP000563930503/4-395\n",
            "16151. MGYP000810671317/4-601\n",
            "16152. MGYP000819180316/3-394\n",
            "16153. MGYP000119292522/14-612\n",
            "16154. MGYP000213767483/9-399\n",
            "16155. MGYP000032186260/4-395\n",
            "16156. MGYP000644944611/10-607\n",
            "16157. MGYP000223564388/3-396\n",
            "16158. MGYP000320162513/10-602\n",
            "16159. MGYP000855374647/73-481\n",
            "16160. MGYP000885559691/9-607\n",
            "16161. MGYP001135504889/19-411\n",
            "16162. MGYP000966797047/5-403\n",
            "16163. MGYP001056516008/27-631\n",
            "16164. MGYP001030164873/7-603\n",
            "16165. MGYP001143105433/35-622\n",
            "16166. MGYP000072386338/10-609\n",
            "16167. MGYP001124873404/4-395\n",
            "16168. MGYP000235428031/27-620\n",
            "16169. MGYP000530025027/23-638\n",
            "16170. MGYP000290331060/8-414\n",
            "16171. MGYP000237277831/18-647\n",
            "16172. MGYP000187994638/4-395\n",
            "16173. MGYP000128230504/12-603\n",
            "16174. MGYP000870226012/27-617\n",
            "16175. MGYP000506347162/7-595\n",
            "16176. MGYP000505917725/4-516\n",
            "16177. MGYP000994505330/92-500\n",
            "16178. MGYP000465617659/10-396\n",
            "16179. MGYP000216874724/22-618\n",
            "16180. MGYP000252817004/7-617\n",
            "16181. MGYP000194358423/3-610\n",
            "16182. MGYP000444785919/5-593\n",
            "16183. MGYP000914704383/9-406\n",
            "16184. MGYP000875005254/10-609\n",
            "16185. MGYP001027714158/37-627\n",
            "16186. MGYP000387626784/6-605\n",
            "16187. MGYP000054764508/37-623\n",
            "16188. MGYP001057462938/11-616\n",
            "16189. MGYP000988656215/6-599\n",
            "16190. MGYP000246210664/16-407\n",
            "16191. MGYP000363432751/2-387\n",
            "16192. MGYP000388525385/4-395\n",
            "16193. MGYP000544727940/2-390\n",
            "16194. MGYP000055970496/10-627\n",
            "16195. MGYP000411577220/8-603\n",
            "16196. MGYP000390011944/4-603\n",
            "16197. MGYP000125406066/21-634\n",
            "16198. MGYP000913939072/79-600\n",
            "16199. MGYP000345285350/16-612\n",
            "16200. MGYP000146608532/4-604\n",
            "16201. MGYP000982549778/23-415\n",
            "16202. MGYP001111758164/7-603\n",
            "16203. MGYP001036097233/4-603\n",
            "16204. MGYP000093059811/14-614\n",
            "16205. MGYP000023548131/5-604\n",
            "16206. MGYP000905561738/1-387\n",
            "16207. MGYP000075801547/12-602\n",
            "16208. MGYP001077761676/15-621\n",
            "16209. MGYP000046363010/12-606\n",
            "16210. MGYP000910213531/22-608\n",
            "16211. MGYP000583384154/12-617\n",
            "16212. MGYP000107838765/9-597\n",
            "16213. MGYP000868712361/11-603\n",
            "16214. MGYP000208097973/3-590\n",
            "16215. MGYP000883777232/11-602\n",
            "16216. MGYP000898636023/11-413\n",
            "16217. MGYP000509207799/4-395\n",
            "16218. MGYP001086956728/13-611\n",
            "16219. MGYP000430149816/10-606\n",
            "16220. MGYP000398039780/5-595\n",
            "16221. MGYP000559357799/33-628\n",
            "16222. MGYP000963978177/13-612\n",
            "16223. MGYP000613041265/12-599\n",
            "16224. MGYP000386208953/10-593\n",
            "16225. MGYP001027414774/67-669\n",
            "16226. MGYP000321054180/39-626\n",
            "16227. MGYP001151716802/3-394\n",
            "16228. MGYP000162084061/10-606\n",
            "16229. MGYP000929612734/3-598\n",
            "16230. MGYP000042604432/10-609\n",
            "16231. MGYP000848376088/4-531\n",
            "16232. MGYP000635741593/4-435\n",
            "16233. MGYP000411931306/15-613\n",
            "16234. MGYP000749380043/4-395\n",
            "16235. MGYP000195184308/4-395\n",
            "16236. MGYP000012699622/16-409\n",
            "16237. MGYP000241323602/9-593\n",
            "16238. MGYP000412830468/4-603\n",
            "16239. MGYP000558877892/2-380\n",
            "16240. MGYP000601043535/4-603\n",
            "16241. MGYP000351240230/8-600\n",
            "16242. MGYP000231687405/13-605\n",
            "16243. MGYP001132206172/9-411\n",
            "16244. MGYP000294521518/30-624\n",
            "16245. MGYP000875646949/24-527\n",
            "16246. MGYP000843967179/4-395\n",
            "16247. MGYP000639415406/10-598\n",
            "16248. MGYP000191536513/15-413\n",
            "16249. MGYP001014445038/10-614\n",
            "16250. MGYP000609925569/14-618\n",
            "16251. MGYP000642092488/81-683\n",
            "16252. MGYP000861039385/14-608\n",
            "16253. MGYP000055259753/7-592\n",
            "16254. MGYP000114696148/11-606\n",
            "16255. MGYP000350584395/18-617\n",
            "16256. MGYP000269967983/7-605\n",
            "16257. MGYP000111999109/13-606\n",
            "16258. MGYP000870132418/7-591\n",
            "16259. MGYP000449798664/4-614\n",
            "16260. MGYP000894289530/12-611\n",
            "16261. MGYP000555986361/4-622\n",
            "16262. MGYP000733645574/6-616\n",
            "16263. MGYP000996613932/22-616\n",
            "16264. MGYP001026897948/11-608\n",
            "16265. MGYP000350484798/16-429\n",
            "16266. MGYP000011317653/10-611\n",
            "16267. MGYP000382924494/11-402\n",
            "16268. MGYP001017106868/38-660\n",
            "16269. MGYP000851370516/27-622\n",
            "16270. MGYP001023594696/4-395\n",
            "16271. MGYP000852730624/7-589\n",
            "16272. MGYP000109651068/3-395\n",
            "16273. MGYP000010793013/3-395\n",
            "16274. MGYP001075749546/8-607\n",
            "16275. MGYP001065306644/25-627\n",
            "16276. MGYP000614815921/3-577\n",
            "16277. MGYP000294506852/24-631\n",
            "16278. MGYP000897095682/17-625\n",
            "16279. MGYP000215371483/4-603\n",
            "16280. MGYP001034835440/5-612\n",
            "16281. MGYP000682527142/11-605\n",
            "16282. MGYP000102716957/7-401\n",
            "16283. MGYP000561161020/45-639\n",
            "16284. MGYP000290356875/12-615\n",
            "16285. MGYP000724799032/94-690\n",
            "16286. MGYP000449963300/13-606\n",
            "16287. MGYP000497863989/10-399\n",
            "16288. MGYP000016968094/14-617\n",
            "16289. MGYP000011105814/4-417\n",
            "16290. MGYP000884148000/33-447\n",
            "16291. MGYP000992184082/3-599\n",
            "16292. MGYP000524827452/65-648\n",
            "16293. MGYP001093443454/34-650\n",
            "16294. MGYP000099772482/4-395\n",
            "16295. MGYP001039598905/29-619\n",
            "16296. MGYP000517731024/10-603\n",
            "16297. MGYP001068715370/6-401\n",
            "16298. MGYP000947771420/57-653\n",
            "16299. MGYP000601519302/3-617\n",
            "16300. MGYP000069262426/4-524\n",
            "16301. MGYP000028892335/15-603\n",
            "16302. MGYP000282304586/7-600\n",
            "16303. MGYP000654231024/6-609\n",
            "16304. MGYP001075638907/17-632\n",
            "16305. MGYP000909558385/13-615\n",
            "16306. MGYP000020537391/9-437\n",
            "16307. MGYP000015119286/20-655\n",
            "16308. MGYP000978790271/7-603\n",
            "16309. MGYP000001589786/16-616\n",
            "16310. MGYP000598135614/35-625\n",
            "16311. MGYP000036697131/16-603\n",
            "16312. MGYP000640208631/3-396\n",
            "16313. MGYP000009498697/4-395\n",
            "16314. MGYP001060206973/4-386\n",
            "16315. MGYP000876355583/3-614\n",
            "16316. MGYP000529754270/15-604\n",
            "16317. MGYP000881523835/9-601\n",
            "16318. MGYP000067357429/4-621\n",
            "16319. MGYP000906072695/14-603\n",
            "16320. MGYP000010222824/3-597\n",
            "16321. MGYP001146126173/9-409\n",
            "16322. MGYP000330886381/4-603\n",
            "16323. MGYP000470860145/51-681\n",
            "16324. MGYP000601003954/41-645\n",
            "16325. MGYP001117793982/12-595\n",
            "16326. MGYP000754295229/22-628\n",
            "16327. MGYP000641435227/17-610\n",
            "16328. MGYP000377370164/11-606\n",
            "16329. MGYP001156377514/4-395\n",
            "16330. MGYP000302153768/17-414\n",
            "16331. MGYP000788330386/2-367\n",
            "16332. MGYP000747349229/3-395\n",
            "16333. MGYP000097350455/3-395\n",
            "16334. MGYP000938670897/7-612\n",
            "16335. MGYP000520576215/10-400\n",
            "16336. MGYP000706595260/3-395\n",
            "16337. MGYP001137000316/9-620\n",
            "16338. MGYP000314274145/7-396\n",
            "16339. MGYP000028109578/10-609\n",
            "16340. MGYP000850529141/12-600\n",
            "16341. MGYP000445624930/3-395\n",
            "16342. MGYP000952898079/30-632\n",
            "16343. MGYP000846269485/6-616\n",
            "16344. MGYP000855361489/8-619\n",
            "16345. MGYP000188555584/11-602\n",
            "16346. MGYP000639104629/6-619\n",
            "16347. MGYP000132913959/3-395\n",
            "16348. MGYP001117948295/7-618\n",
            "16349. MGYP000402366555/7-399\n",
            "16350. MGYP000124028471/6-604\n",
            "16351. MGYP000644397965/12-612\n",
            "16352. MGYP000504189948/17-604\n",
            "16353. MGYP000054446135/14-609\n",
            "16354. MGYP000900752040/11-603\n",
            "16355. MGYP000878608017/2-595\n",
            "16356. MGYP000580874733/12-606\n",
            "16357. MGYP001092792648/3-603\n",
            "16358. MGYP000314286293/8-383\n",
            "16359. MGYP000071439305/4-395\n",
            "16360. MGYP000703752836/3-395\n",
            "16361. MGYP000238012278/14-426\n",
            "16362. MGYP000825365964/6-395\n",
            "16363. MGYP000869319730/5-587\n",
            "16364. MGYP000019864836/6-609\n",
            "16365. MGYP000264746099/4-395\n",
            "16366. MGYP000911866412/32-633\n",
            "16367. MGYP000201891856/4-582\n",
            "16368. MGYP000244670753/5-605\n",
            "16369. MGYP000233530733/26-614\n",
            "16370. MGYP001125504107/41-648\n",
            "16371. MGYP000333053611/6-606\n",
            "16372. MGYP000413389065/3-396\n",
            "16373. MGYP000659809475/22-417\n",
            "16374. MGYP000353586551/7-600\n",
            "16375. MGYP000618945642/3-396\n",
            "16376. MGYP000337249193/4-395\n",
            "16377. MGYP000924520831/4-395\n",
            "16378. MGYP001017649574/7-594\n",
            "16379. MGYP000897529376/11-386\n",
            "16380. MGYP000311810491/10-612\n",
            "16381. MGYP001079956470/3-395\n",
            "16382. MGYP000045076468/5-396\n",
            "16383. MGYP000150529876/3-395\n",
            "16384. MGYP000845419736/8-601\n",
            "16385. MGYP000132235468/2-394\n",
            "16386. MGYP000052319088/48-636\n",
            "16387. MGYP000840026203/3-395\n",
            "16388. MGYP000122504575/22-657\n",
            "16389. MGYP000188174385/3-395\n",
            "16390. MGYP000495790381/4-394\n",
            "16391. MGYP000514102674/6-361\n",
            "16392. MGYP000408946994/3-395\n",
            "16393. MGYP000993133792/14-423\n",
            "16394. MGYP000299420713/15-412\n",
            "16395. MGYP000482642599/23-614\n",
            "16396. MGYP000836602748/4-395\n",
            "16397. MGYP000591136304/10-607\n",
            "16398. MGYP000346543166/12-602\n",
            "16399. MGYP001114132379/9-591\n",
            "16400. MGYP000890756274/9-604\n",
            "16401. MGYP000652968547/10-426\n",
            "16402. MGYP001067873813/16-619\n",
            "16403. MGYP000563169885/14-608\n",
            "16404. MGYP000797466133/2-604\n",
            "16405. MGYP001030943096/10-610\n",
            "16406. MGYP000554046000/9-595\n",
            "16407. MGYP001124516539/16-601\n",
            "16408. MGYP001040086559/12-401\n",
            "16409. MGYP001070828034/11-608\n",
            "16410. MGYP000116552521/45-653\n",
            "16411. MGYP000083593845/15-604\n",
            "16412. MGYP001054524357/15-623\n",
            "16413. MGYP000604468622/15-611\n",
            "16414. MGYP000530416301/15-422\n",
            "16415. MGYP000645792758/3-395\n",
            "16416. MGYP000934169855/3-395\n",
            "16417. MGYP000020267751/20-616\n",
            "16418. MGYP000314588023/3-395\n",
            "16419. MGYP000733768892/6-593\n",
            "16420. MGYP001111878626/6-593\n",
            "16421. MGYP001051663696/9-646\n",
            "16422. MGYP000045822694/111-733\n",
            "16423. MGYP000855183427/11-610\n",
            "16424. MGYP000917907420/4-395\n",
            "16425. MGYP000618938986/15-605\n",
            "16426. MGYP000997124757/19-608\n",
            "16427. MGYP000534301693/3-396\n",
            "16428. MGYP001121721208/4-395\n",
            "16429. MGYP001003910917/3-393\n",
            "16430. MGYP000086855986/6-600\n",
            "16431. MGYP000053918421/4-395\n",
            "16432. MGYP000123233014/3-395\n",
            "16433. MGYP000456887822/22-644\n",
            "16434. MGYP000594814180/17-616\n",
            "16435. MGYP001144860297/19-621\n",
            "16436. MGYP001059279328/2-585\n",
            "16437. MGYP001039282095/7-617\n",
            "16438. MGYP001091429187/10-608\n",
            "16439. MGYP001049360100/11-608\n",
            "16440. MGYP000326063568/6-399\n",
            "16441. MGYP000618776374/4-600\n",
            "16442. MGYP001084210378/33-632\n",
            "16443. MGYP001101517251/16-614\n",
            "16444. MGYP000550939156/22-615\n",
            "16445. MGYP000564805906/3-631\n",
            "16446. MGYP000157993223/3-595\n",
            "16447. MGYP000943020382/4-592\n",
            "16448. MGYP000397571473/16-401\n",
            "16449. MGYP000706822218/52-450\n",
            "16450. MGYP000863863188/40-642\n",
            "16451. MGYP000505809975/5-596\n",
            "16452. MGYP000559792866/13-481\n",
            "16453. MGYP000897557055/10-396\n",
            "16454. MGYP000696732099/20-655\n",
            "16455. MGYP000527764831/2-382\n",
            "16456. MGYP000731248604/4-395\n",
            "16457. MGYP001030306312/16-617\n",
            "16458. MGYP000288192998/7-607\n",
            "16459. MGYP000189439073/3-396\n",
            "16460. MGYP000006226039/4-592\n",
            "16461. MGYP000748480569/35-468\n",
            "16462. MGYP000156208097/13-609\n",
            "16463. MGYP000494634377/8-622\n",
            "16464. MGYP000859434522/12-609\n",
            "16465. MGYP000892261896/12-408\n",
            "16466. MGYP000969869128/3-393\n",
            "16467. MGYP000041533144/11-612\n",
            "16468. MGYP000719026108/10-602\n",
            "16469. MGYP000129782850/4-395\n",
            "16470. MGYP000593667851/3-394\n",
            "16471. MGYP000987977326/25-624\n",
            "16472. MGYP000967817785/10-608\n",
            "16473. MGYP000905345664/9-609\n",
            "16474. MGYP000092780318/6-623\n",
            "16475. MGYP001067229578/4-603\n",
            "16476. MGYP001151063257/73-683\n",
            "16477. MGYP000087619010/15-615\n",
            "16478. MGYP001130278005/11-595\n",
            "16479. MGYP000592233023/18-608\n",
            "16480. MGYP000930998014/14-614\n",
            "16481. MGYP000733736258/3-390\n",
            "16482. MGYP000203333181/12-612\n",
            "16483. MGYP000575313954/11-615\n",
            "16484. MGYP000251987358/11-591\n",
            "16485. MGYP000708614383/9-403\n",
            "16486. MGYP000891892880/25-612\n",
            "16487. MGYP001140993884/30-649\n",
            "16488. MGYP000518600151/10-628\n",
            "16489. MGYP000881513791/7-435\n",
            "16490. MGYP000176085405/12-613\n",
            "16491. MGYP000407595269/3-607\n",
            "16492. MGYP000379547721/30-620\n",
            "16493. MGYP000434990045/12-425\n",
            "16494. MGYP000092949111/13-613\n",
            "16495. MGYP000828042771/56-452\n",
            "16496. MGYP001088789095/21-619\n",
            "16497. MGYP000566011776/23-418\n",
            "16498. MGYP000606431096/4-395\n",
            "16499. MGYP001099801311/6-594\n",
            "16500. MGYP000282601073/4-395\n",
            "16501. MGYP000879908428/9-617\n",
            "16502. MGYP001106970622/11-609\n",
            "16503. MGYP000489391848/8-398\n",
            "16504. MGYP001099180960/29-641\n",
            "16505. MGYP000601127367/8-606\n",
            "16506. MGYP000337604056/13-422\n",
            "16507. MGYP000875654027/6-603\n",
            "16508. MGYP000675185602/8-603\n",
            "16509. MGYP000196604546/8-603\n",
            "16510. MGYP000156205069/6-597\n",
            "16511. MGYP000229643898/18-606\n",
            "16512. MGYP001032569048/7-615\n",
            "16513. MGYP000662083100/5-392\n",
            "16514. MGYP000591762459/3-396\n",
            "16515. MGYP000114647479/7-596\n",
            "16516. MGYP001140945955/2-390\n",
            "16517. MGYP000496718196/7-610\n",
            "16518. MGYP000873196665/14-607\n",
            "16519. MGYP000116652877/18-644\n",
            "16520. MGYP001067128194/25-627\n",
            "16521. MGYP000536311562/11-598\n",
            "16522. MGYP000501100812/15-607\n",
            "16523. MGYP000260860108/10-399\n",
            "16524. MGYP000837716524/16-409\n",
            "16525. MGYP000219588881/16-415\n",
            "16526. MGYP000984400756/5-610\n",
            "16527. MGYP001011698662/1-380\n",
            "16528. MGYP000525349078/6-402\n",
            "16529. MGYP001008498986/9-607\n",
            "16530. MGYP000967030589/7-601\n",
            "16531. MGYP000184930457/9-604\n",
            "16532. MGYP000863142324/20-631\n",
            "16533. MGYP000342030834/7-403\n",
            "16534. MGYP000211780006/5-618\n",
            "16535. MGYP001076981273/31-432\n",
            "16536. MGYP000566107160/16-613\n",
            "16537. MGYP000005755637/7-597\n",
            "16538. MGYP000769726753/19-405\n",
            "16539. MGYP000044039207/10-605\n",
            "16540. MGYP001054880449/14-614\n",
            "16541. MGYP001141411765/3-395\n",
            "16542. MGYP001040741967/9-607\n",
            "16543. MGYP000151949941/10-609\n",
            "16544. MGYP000158392090/9-614\n",
            "16545. MGYP000584184447/3-395\n",
            "16546. MGYP000669209153/39-647\n",
            "16547. MGYP000511725506/26-642\n",
            "16548. MGYP000220457584/8-401\n",
            "16549. MGYP000615454529/7-609\n",
            "16550. MGYP001038045534/7-440\n",
            "16551. MGYP000029213235/10-393\n",
            "16552. MGYP000264446086/12-605\n",
            "16553. MGYP000499230132/30-426\n",
            "16554. MGYP000246106529/12-605\n",
            "16555. MGYP000197527050/38-632\n",
            "16556. MGYP000804728424/17-611\n",
            "16557. MGYP000797155879/26-420\n",
            "16558. MGYP000299208994/4-597\n",
            "16559. MGYP000978264035/27-444\n",
            "16560. MGYP000989710199/12-613\n",
            "16561. MGYP000882809474/7-399\n",
            "16562. MGYP001061083398/5-598\n",
            "16563. MGYP000268253301/12-609\n",
            "16564. MGYP000205877529/25-419\n",
            "16565. MGYP001073112497/13-614\n",
            "16566. MGYP000898387567/25-446\n",
            "16567. MGYP000087220481/7-595\n",
            "16568. MGYP000745532259/12-608\n",
            "16569. MGYP000576708148/12-607\n",
            "16570. MGYP000657542336/7-600\n",
            "16571. MGYP000436270540/25-623\n",
            "16572. MGYP000845494869/12-626\n",
            "16573. MGYP001100014723/203-717\n",
            "16574. MGYP001086445079/16-611\n",
            "16575. MGYP000663107604/17-609\n",
            "16576. MGYP000585422539/9-609\n",
            "16577. MGYP000901088935/9-609\n",
            "16578. MGYP001071099144/7-614\n",
            "16579. MGYP001137883858/9-612\n",
            "16580. MGYP000829734590/3-395\n",
            "16581. MGYP000507110418/425-1006\n",
            "16582. MGYP001133759078/17-619\n",
            "16583. MGYP000042577148/7-610\n",
            "16584. MGYP000096990994/8-597\n",
            "16585. MGYP000929651884/69-462\n",
            "16586. MGYP000351806588/6-395\n",
            "16587. MGYP001038526666/22-617\n",
            "16588. MGYP001025343332/5-597\n",
            "16589. MGYP000976644782/11-609\n",
            "16590. MGYP000958549651/45-641\n",
            "16591. MGYP000751185570/8-602\n",
            "16592. MGYP001126286494/210-846\n",
            "16593. MGYP000456865698/10-406\n",
            "16594. MGYP001144166033/4-603\n",
            "16595. MGYP000042573807/7-399\n",
            "16596. MGYP000858006559/6-599\n",
            "16597. MGYP000017607751/24-612\n",
            "16598. MGYP000487971297/27-510\n",
            "16599. MGYP001125078307/27-614\n",
            "16600. MGYP000110652251/4-395\n",
            "16601. MGYP000033805057/14-422\n",
            "16602. MGYP000179277630/5-620\n",
            "16603. MGYP000198857523/7-605\n",
            "16604. MGYP000969586788/13-409\n",
            "16605. MGYP000022666824/3-395\n",
            "16606. MGYP001120650908/3-635\n",
            "16607. MGYP000184238637/9-599\n",
            "16608. MGYP000252234994/6-409\n",
            "16609. MGYP000495980456/8-399\n",
            "16610. MGYP000068836795/22-618\n",
            "16611. MGYP000751620955/9-595\n",
            "16612. MGYP000920104525/6-600\n",
            "16613. MGYP000066329050/8-606\n",
            "16614. MGYP000739366511/59-442\n",
            "16615. MGYP000922634360/5-611\n",
            "16616. MGYP000916088093/10-609\n",
            "16617. MGYP001130927146/13-602\n",
            "16618. MGYP000576623199/31-631\n",
            "16619. MGYP001060713555/11-603\n",
            "16620. MGYP001068914207/6-591\n",
            "16621. MGYP000042960802/42-626\n",
            "16622. MGYP000538305430/12-605\n",
            "16623. MGYP000187964477/5-399\n",
            "16624. MGYP000226984620/12-602\n",
            "16625. MGYP000149545109/5-593\n",
            "16626. MGYP000160653457/15-608\n",
            "16627. MGYP000636185018/18-622\n",
            "16628. MGYP000010996820/7-601\n",
            "16629. MGYP000346557351/6-402\n",
            "16630. MGYP000237510699/92-718\n",
            "16631. MGYP000184578640/29-633\n",
            "16632. MGYP000274200151/14-611\n",
            "16633. MGYP001100124212/5-392\n",
            "16634. MGYP000402464485/9-605\n",
            "16635. MGYP000492175919/6-576\n",
            "16636. MGYP000400549717/11-617\n",
            "16637. MGYP001032835024/7-589\n",
            "16638. MGYP000437691246/9-606\n",
            "16639. MGYP000218137194/7-401\n",
            "16640. MGYP001155477884/208-829\n",
            "16641. MGYP000003127597/14-609\n",
            "16642. MGYP000176617581/14-400\n",
            "16643. MGYP001137939667/24-644\n",
            "16644. MGYP000049761809/28-629\n",
            "16645. MGYP000583574830/2-396\n",
            "16646. MGYP001079022563/1-161\n",
            "16647. MGYP001000034666/21-613\n",
            "16648. MGYP000864608513/10-606\n",
            "16649. MGYP000605094425/5-432\n",
            "16650. MGYP000630983425/18-622\n",
            "16651. MGYP000679004010/4-395\n",
            "16652. MGYP000721130773/7-600\n",
            "16653. MGYP000334775913/18-622\n",
            "16654. MGYP000107775847/33-639\n",
            "16655. MGYP000907077499/9-598\n",
            "16656. MGYP000809294691/37-431\n",
            "16657. MGYP000001286606/13-610\n",
            "16658. MGYP000505058181/3-399\n",
            "16659. MGYP000913450020/2-595\n",
            "16660. MGYP000957936501/18-643\n",
            "16661. MGYP000376689550/32-644\n",
            "16662. MGYP000426315099/27-610\n",
            "16663. MGYP001092261698/8-611\n",
            "16664. MGYP000167512495/6-612\n",
            "16665. MGYP000852599908/12-612\n",
            "16666. MGYP000548110801/33-639\n",
            "16667. MGYP000191604795/7-399\n",
            "16668. MGYP000928506670/6-605\n",
            "16669. MGYP000296483325/9-609\n",
            "16670. MGYP000166751519/4-625\n",
            "16671. MGYP000179713081/17-609\n",
            "16672. MGYP000647159662/8-594\n",
            "16673. MGYP000971181436/12-607\n",
            "16674. MGYP000441467726/6-414\n",
            "16675. MGYP000844596955/28-623\n",
            "16676. MGYP000098275887/13-407\n",
            "16677. MGYP000551362739/5-610\n",
            "16678. MGYP001026844854/16-603\n",
            "16679. MGYP000026141770/10-610\n",
            "16680. MGYP000953467981/13-619\n",
            "16681. MGYP000003647670/6-601\n",
            "16682. MGYP000046294777/10-591\n",
            "16683. MGYP000994955134/11-595\n",
            "16684. MGYP000252335937/17-405\n",
            "16685. MGYP000217834786/9-384\n",
            "16686. MGYP000143218176/16-602\n",
            "16687. MGYP000038711026/21-413\n",
            "16688. MGYP001011944972/18-609\n",
            "16689. MGYP000242786530/1-394\n",
            "16690. MGYP000123092050/18-399\n",
            "16691. MGYP000151970652/20-607\n",
            "16692. MGYP000631341006/13-608\n",
            "16693. MGYP000061567835/10-599\n",
            "16694. MGYP000958231258/34-465\n",
            "16695. MGYP000064400722/6-609\n",
            "16696. MGYP000315592007/3-395\n",
            "16697. MGYP000184432541/9-609\n",
            "16698. MGYP000459114975/7-416\n",
            "16699. MGYP000971205672/19-454\n",
            "16700. MGYP001067789713/20-621\n",
            "16701. MGYP000854646308/5-605\n",
            "16702. MGYP000729477160/4-395\n",
            "16703. MGYP000344588885/6-414\n",
            "16704. MGYP000984260176/17-623\n",
            "16705. MGYP000441011228/8-610\n",
            "16706. MGYP000461306696/5-395\n",
            "16707. MGYP001081658619/9-592\n",
            "16708. MGYP001107907890/12-610\n",
            "16709. MGYP000011546224/17-610\n",
            "16710. MGYP001063658951/9-609\n",
            "16711. MGYP001015147442/15-608\n",
            "16712. MGYP000924395991/6-605\n",
            "16713. MGYP000138957742/69-462\n",
            "16714. MGYP000098267384/14-604\n",
            "16715. MGYP000707661345/6-605\n",
            "16716. MGYP001094481544/6-602\n",
            "16717. MGYP000336452585/12-602\n",
            "16718. MGYP000555711181/24-529\n",
            "16719. MGYP000977408685/46-473\n",
            "16720. MGYP001027061593/9-608\n",
            "16721. MGYP001046506870/12-607\n",
            "16722. MGYP000153288057/51-457\n",
            "16723. MGYP000682868041/15-409\n",
            "16724. MGYP001112643894/10-611\n",
            "16725. MGYP000787038807/14-406\n",
            "16726. MGYP000886174480/13-608\n",
            "16727. MGYP000110477859/27-615\n",
            "16728. MGYP000963545167/33-408\n",
            "16729. MGYP001104308668/22-609\n",
            "16730. MGYP000108550637/2-394\n",
            "16731. MGYP000427361410/6-400\n",
            "16732. MGYP000743006099/8-608\n",
            "16733. MGYP000244419494/8-427\n",
            "16734. MGYP000625088976/43-629\n",
            "16735. MGYP000006647208/6-401\n",
            "16736. MGYP000984545689/61-448\n",
            "16737. MGYP000364297767/7-608\n",
            "16738. MGYP000029829774/3-395\n",
            "16739. MGYP000308267296/10-609\n",
            "16740. MGYP000095319959/21-414\n",
            "16741. MGYP001132381023/16-615\n",
            "16742. MGYP001067611978/21-411\n",
            "16743. MGYP000194473445/18-609\n",
            "16744. MGYP000479264983/4-609\n",
            "16745. MGYP000656614119/5-587\n",
            "16746. MGYP001076450293/34-429\n",
            "16747. MGYP000858238701/12-607\n",
            "16748. MGYP001012315806/16-392\n",
            "16749. MGYP000289539499/12-603\n",
            "16750. MGYP000789021601/12-403\n",
            "16751. MGYP000054590864/13-597\n",
            "16752. MGYP000957816909/25-616\n",
            "16753. MGYP000243991780/10-401\n",
            "16754. MGYP000011319236/10-607\n",
            "16755. MGYP000846848487/41-647\n",
            "16756. MGYP001027396461/20-622\n",
            "16757. MGYP000843659167/6-396\n",
            "16758. MGYP000975694620/34-454\n",
            "16759. MGYP000667384385/4-395\n",
            "16760. MGYP000948805085/34-454\n",
            "16761. MGYP000594757535/18-614\n",
            "16762. MGYP000485765555/10-610\n",
            "16763. MGYP000430185626/8-395\n",
            "16764. MGYP000973148943/20-411\n",
            "16765. MGYP001038819930/11-594\n",
            "16766. MGYP000995713584/6-407\n",
            "16767. MGYP000191386904/12-423\n",
            "16768. MGYP000653251487/15-601\n",
            "16769. MGYP000061619537/11-605\n",
            "16770. MGYP000088112350/8-392\n",
            "16771. MGYP000342457539/2-398\n",
            "16772. MGYP000855202279/11-603\n",
            "16773. MGYP000978260368/73-665\n",
            "16774. MGYP000454853507/4-601\n",
            "16775. MGYP000682801968/18-620\n",
            "16776. MGYP000013162842/12-607\n",
            "16777. MGYP000726822684/11-611\n",
            "16778. MGYP001096038925/2-398\n",
            "16779. MGYP000688697739/9-409\n",
            "16780. MGYP000250559060/5-609\n",
            "16781. MGYP000327264196/11-596\n",
            "16782. MGYP000087347044/5-396\n",
            "16783. MGYP000943012305/28-634\n",
            "16784. MGYP000890745042/33-636\n",
            "16785. MGYP000737324545/20-411\n",
            "16786. MGYP000540585016/19-417\n",
            "16787. MGYP000004221175/23-619\n",
            "16788. MGYP000480181553/6-595\n",
            "16789. MGYP000161948599/69-462\n",
            "16790. MGYP000906901261/10-432\n",
            "16791. MGYP000291180397/16-407\n",
            "16792. MGYP000014550956/10-407\n",
            "16793. MGYP000884488113/13-625\n",
            "16794. MGYP000921645757/11-396\n",
            "16795. MGYP000229850629/54-455\n",
            "16796. MGYP000417566609/19-407\n",
            "16797. MGYP000729687028/129-735\n",
            "16798. MGYP000340726975/7-607\n",
            "16799. MGYP000252035532/25-618\n",
            "16800. MGYP000883230610/26-617\n",
            "16801. MGYP000022187505/20-368\n",
            "16802. MGYP000272272107/21-413\n",
            "16803. MGYP001043599971/17-611\n",
            "16804. MGYP000238018782/14-609\n",
            "16805. MGYP000855926628/8-612\n",
            "16806. MGYP000002536319/10-407\n",
            "16807. MGYP000607363619/10-594\n",
            "16808. MGYP000870559000/6-396\n",
            "16809. MGYP000426003909/8-388\n",
            "16810. MGYP000414145013/12-610\n",
            "16811. MGYP000952139692/36-629\n",
            "16812. MGYP000148714493/6-624\n",
            "16813. MGYP000114946266/9-401\n",
            "16814. MGYP000588396064/4-611\n",
            "16815. MGYP000194144682/11-613\n",
            "16816. MGYP000058736742/6-601\n",
            "16817. MGYP000168289674/12-607\n",
            "16818. MGYP000585803246/2-600\n",
            "16819. MGYP000366139835/4-394\n",
            "16820. MGYP000106878863/4-393\n",
            "16821. MGYP000373187553/18-651\n",
            "16822. MGYP000784145670/4-601\n",
            "16823. MGYP001134845381/14-600\n",
            "16824. MGYP000108634221/20-609\n",
            "16825. MGYP000947035103/6-599\n",
            "16826. MGYP000515673485/12-408\n",
            "16827. MGYP000553476582/37-632\n",
            "16828. MGYP000464341887/8-614\n",
            "16829. MGYP000252750366/4-602\n",
            "16830. MGYP001003701617/9-607\n",
            "16831. MGYP000972482287/7-424\n",
            "16832. MGYP000393525596/17-625\n",
            "16833. MGYP000365574232/8-606\n",
            "16834. MGYP000850521247/9-603\n",
            "16835. MGYP000852838881/99-692\n",
            "16836. MGYP001089002240/12-610\n",
            "16837. MGYP001138456046/24-651\n",
            "16838. MGYP000660758263/9-616\n",
            "16839. MGYP001038239928/18-615\n",
            "16840. MGYP000468309867/4-395\n",
            "16841. MGYP000971854344/11-599\n",
            "16842. MGYP000285985297/14-611\n",
            "16843. MGYP001138395862/13-408\n",
            "16844. MGYP000740850266/8-388\n",
            "16845. MGYP000899116881/15-604\n",
            "16846. MGYP000036760029/3-410\n",
            "16847. MGYP001043925388/11-618\n",
            "16848. MGYP000312504178/149-528\n",
            "16849. MGYP000402463040/6-588\n",
            "16850. MGYP000033758774/32-624\n",
            "16851. MGYP000465849054/12-614\n",
            "16852. MGYP000139916733/7-619\n",
            "16853. MGYP000850516193/14-616\n",
            "16854. MGYP000716017416/19-606\n",
            "16855. MGYP000008369061/11-606\n",
            "16856. MGYP000609803723/13-607\n",
            "16857. MGYP001122358093/53-650\n",
            "16858. MGYP000871054031/69-462\n",
            "16859. MGYP001107027554/16-408\n",
            "16860. MGYP000383278794/11-591\n",
            "16861. MGYP000871432730/12-604\n",
            "16862. MGYP000918304994/9-605\n",
            "16863. MGYP000856454325/7-397\n",
            "16864. MGYP000436120307/11-612\n",
            "16865. MGYP000869425167/21-604\n",
            "16866. MGYP000016801497/10-603\n",
            "16867. MGYP000004619516/14-594\n",
            "16868. MGYP000743005275/1-604\n",
            "16869. MGYP000167955273/13-610\n",
            "16870. MGYP000152391104/17-608\n",
            "16871. MGYP000421194190/17-618\n",
            "16872. MGYP000193978447/11-603\n",
            "16873. MGYP000622076307/9-573\n",
            "16874. MGYP000990985556/21-414\n",
            "16875. MGYP000911589423/13-609\n",
            "16876. MGYP000879554085/4-589\n",
            "16877. MGYP001054434995/3-390\n",
            "16878. MGYP001133104159/5-395\n",
            "16879. MGYP000718116423/8-602\n",
            "16880. MGYP000884300286/7-602\n",
            "16881. MGYP000503991797/20-637\n",
            "16882. MGYP001071861355/9-598\n",
            "16883. MGYP000329627484/3-395\n",
            "16884. MGYP000833342642/9-374\n",
            "16885. MGYP000639522596/20-607\n",
            "16886. MGYP000679598506/9-429\n",
            "16887. MGYP001128911558/50-648\n",
            "16888. MGYP000876427722/14-620\n",
            "16889. MGYP000856974379/63-652\n",
            "16890. MGYP000032389684/7-599\n",
            "16891. MGYP000185593693/10-612\n",
            "16892. MGYP000679815995/62-475\n",
            "16893. MGYP000175354437/7-624\n",
            "16894. MGYP000935943151/9-433\n",
            "16895. MGYP000477226293/12-612\n",
            "16896. MGYP000294825072/11-598\n",
            "16897. MGYP000863689610/16-617\n",
            "16898. MGYP000911941847/41-639\n",
            "16899. MGYP000441119570/4-395\n",
            "16900. MGYP000918791221/18-616\n",
            "16901. MGYP000295405459/17-608\n",
            "16902. MGYP000864736917/12-609\n",
            "16903. MGYP000732783806/31-447\n",
            "16904. MGYP000184470918/14-417\n",
            "16905. MGYP000650429496/9-408\n",
            "16906. MGYP000883890148/4-575\n",
            "16907. MGYP001078327182/14-604\n",
            "16908. MGYP000044413383/9-406\n",
            "16909. MGYP000088945112/3-394\n",
            "16910. MGYP000532961192/10-597\n",
            "16911. MGYP000053022618/19-602\n",
            "16912. MGYP000847325485/36-625\n",
            "16913. MGYP000998514915/12-398\n",
            "16914. MGYP001043674581/35-415\n",
            "16915. MGYP000223109303/3-395\n",
            "16916. MGYP000992506638/177-826\n",
            "16917. MGYP000857349029/14-604\n",
            "16918. MGYP000916241306/3-395\n",
            "16919. MGYP000940983693/10-391\n",
            "16920. MGYP000080996752/8-605\n",
            "16921. MGYP000468010281/12-607\n",
            "16922. MGYP001008695581/87-483\n",
            "16923. MGYP000541161633/15-407\n",
            "16924. MGYP000862335136/7-602\n",
            "16925. MGYP000500381383/43-419\n",
            "16926. MGYP000181397890/12-599\n",
            "16927. MGYP001079745425/9-603\n",
            "16928. MGYP000645695186/9-401\n",
            "16929. MGYP000897753899/21-614\n",
            "16930. MGYP000874195730/49-460\n",
            "16931. MGYP001142333120/8-596\n",
            "16932. MGYP000739742729/12-607\n",
            "16933. MGYP001016452936/4-608\n",
            "16934. MGYP000882916894/12-610\n",
            "16935. MGYP001142811395/15-617\n",
            "16936. MGYP001147894317/5-606\n",
            "16937. MGYP000576719349/10-389\n",
            "16938. MGYP000471328980/11-613\n",
            "16939. MGYP000239287763/4-394\n",
            "16940. MGYP000036994932/15-406\n",
            "16941. MGYP000700659625/14-612\n",
            "16942. MGYP000033818383/4-597\n",
            "16943. MGYP001138413282/7-605\n",
            "16944. MGYP000231384035/9-611\n",
            "16945. MGYP000864000666/25-406\n",
            "16946. MGYP000137232236/3-606\n",
            "16947. MGYP000533492366/6-599\n",
            "16948. MGYP000692598648/12-605\n",
            "16949. MGYP000004736799/3-609\n",
            "16950. MGYP000193871857/2-392\n",
            "16951. MGYP000123488082/11-393\n",
            "16952. MGYP000689509439/87-678\n",
            "16953. MGYP000046402846/23-635\n",
            "16954. MGYP000656236072/13-411\n",
            "16955. MGYP000614698295/48-438\n",
            "16956. MGYP000560191586/13-606\n",
            "16957. MGYP000645924771/37-640\n",
            "16958. MGYP000854147280/37-630\n",
            "16959. MGYP000174731103/21-413\n",
            "16960. MGYP001128903857/9-610\n",
            "16961. MGYP001041435685/15-602\n",
            "16962. MGYP000300302523/3-553\n",
            "16963. MGYP000660448187/10-605\n",
            "16964. MGYP000043432404/13-605\n",
            "16965. MGYP000375819012/6-392\n",
            "16966. MGYP001055281908/12-606\n",
            "16967. MGYP001122659106/28-436\n",
            "16968. MGYP000949873596/15-614\n",
            "16969. MGYP000940053195/5-387\n",
            "16970. MGYP000819338727/13-588\n",
            "16971. MGYP000438070529/6-599\n",
            "16972. MGYP000000211653/6-608\n",
            "16973. MGYP000712040534/41-428\n",
            "16974. MGYP000119495297/11-390\n",
            "16975. MGYP001054516217/21-399\n",
            "16976. MGYP000704283697/12-613\n",
            "16977. MGYP000267342895/6-601\n",
            "16978. MGYP001042757671/3-608\n",
            "16979. MGYP001042227424/17-608\n",
            "16980. MGYP000317111551/5-618\n",
            "16981. MGYP001134916522/11-427\n",
            "16982. MGYP000254021623/5-592\n",
            "16983. MGYP000910718138/4-394\n",
            "16984. MGYP000764319833/8-400\n",
            "16985. MGYP000055849402/5-600\n",
            "16986. MGYP001100469899/16-412\n",
            "16987. MGYP001115844480/8-605\n",
            "16988. MGYP000464327313/13-599\n",
            "16989. MGYP000364703569/3-595\n",
            "16990. MGYP000396697005/1-623\n",
            "16991. MGYP000110708383/10-620\n",
            "16992. MGYP000053595476/24-413\n",
            "16993. MGYP000854501799/8-387\n",
            "16994. MGYP000850531226/13-609\n",
            "16995. MGYP000391500557/3-395\n",
            "16996. MGYP001149040913/9-402\n",
            "16997. MGYP000958478885/11-605\n",
            "16998. MGYP000036988671/10-609\n",
            "16999. MGYP000193348765/12-608\n",
            "17000. MGYP000114395791/10-597\n",
            "17001. MGYP000809597523/11-392\n",
            "17002. MGYP000874491890/11-602\n",
            "17003. MGYP000371081930/36-430\n",
            "17004. MGYP000321959870/7-616\n",
            "17005. MGYP000692202343/11-601\n",
            "17006. MGYP001054670189/3-393\n",
            "17007. MGYP000446851157/24-630\n",
            "17008. MGYP000050710023/14-612\n",
            "17009. MGYP000435966157/21-612\n",
            "17010. MGYP001085151496/3-616\n",
            "17011. MGYP000592589596/6-431\n",
            "17012. MGYP000017285432/12-598\n",
            "17013. MGYP001099348444/1-557\n",
            "17014. MGYP001098654538/41-420\n",
            "17015. MGYP001098654538/522-631\n",
            "17016. MGYP000597203768/15-418\n",
            "17017. MGYP001059426939/11-605\n",
            "17018. MGYP000939430194/7-610\n",
            "17019. MGYP000604190275/28-623\n",
            "17020. MGYP001145629341/15-649\n",
            "17021. MGYP000630291305/11-601\n",
            "17022. MGYP000792409563/10-391\n",
            "17023. MGYP000514634212/10-632\n",
            "17024. MGYP000007368585/28-628\n",
            "17025. MGYP000564556811/5-600\n",
            "17026. MGYP000629792511/8-607\n",
            "17027. MGYP000973339860/110-727\n",
            "17028. MGYP000146969709/12-609\n",
            "17029. MGYP001092501575/14-602\n",
            "17030. MGYP000938013927/54-450\n",
            "17031. MGYP001045283977/10-393\n",
            "17032. MGYP000253249005/4-389\n",
            "17033. MGYP001014881196/9-429\n",
            "17034. MGYP000871570436/6-606\n",
            "17035. MGYP000335678334/14-617\n",
            "17036. MGYP000044983260/4-617\n",
            "17037. MGYP000926529103/17-458\n",
            "17038. MGYP000601525403/20-424\n",
            "17039. MGYP000485054764/6-617\n",
            "17040. MGYP000939580739/13-608\n",
            "17041. MGYP001034290425/22-617\n",
            "17042. MGYP000074248158/22-415\n",
            "17043. MGYP000270825725/15-612\n",
            "17044. MGYP000294794983/8-399\n",
            "17045. MGYP001017614889/18-613\n",
            "17046. MGYP000269947959/14-449\n",
            "17047. MGYP000061767686/24-611\n",
            "17048. MGYP000922250893/12-609\n",
            "17049. MGYP000079987396/18-414\n",
            "17050. MGYP000133331497/3-395\n",
            "17051. MGYP001102939355/7-592\n",
            "17052. MGYP000893801359/41-432\n",
            "17053. MGYP000766710588/12-379\n",
            "17054. MGYP000589918668/4-395\n",
            "17055. MGYP000955389108/6-400\n",
            "17056. MGYP000423725074/4-604\n",
            "17057. MGYP001123055443/6-424\n",
            "17058. MGYP000745629364/18-622\n",
            "17059. MGYP000847974938/8-638\n",
            "17060. MGYP001035471897/29-644\n",
            "17061. MGYP000663200773/12-643\n",
            "17062. MGYP000290400562/13-606\n",
            "17063. MGYP001007371230/13-607\n",
            "17064. MGYP000013079582/12-596\n",
            "17065. MGYP000014515245/7-609\n",
            "17066. MGYP001155821291/17-606\n",
            "17067. MGYP000888413970/11-603\n",
            "17068. MGYP000689572094/147-538\n",
            "17069. MGYP000686980363/9-609\n",
            "17070. MGYP001012870009/5-399\n",
            "17071. MGYP000100179185/8-600\n",
            "17072. MGYP001091198742/12-602\n",
            "17073. MGYP000661963388/27-420\n",
            "17074. MGYP000141415743/21-609\n",
            "17075. MGYP001117558194/11-612\n",
            "17076. MGYP000935822040/6-618\n",
            "17077. MGYP000099430743/7-593\n",
            "17078. MGYP001041607918/24-666\n",
            "17079. MGYP001128897113/12-406\n",
            "17080. MGYP000381766351/10-602\n",
            "17081. MGYP000125145677/5-396\n",
            "17082. MGYP000146123602/8-604\n",
            "17083. MGYP001080214302/6-426\n",
            "17084. MGYP000675215835/14-635\n",
            "17085. MGYP000748193905/6-600\n",
            "17086. MGYP001034062108/11-603\n",
            "17087. MGYP000981817728/6-609\n",
            "17088. MGYP000008277559/8-409\n",
            "17089. MGYP001138005019/11-400\n",
            "17090. MGYP000297420170/9-606\n",
            "17091. MGYP000020899540/14-607\n",
            "17092. MGYP000507100286/2-397\n",
            "17093. MGYP001099057275/3-395\n",
            "17094. MGYP000163090956/17-601\n",
            "17095. MGYP000859876778/4-394\n",
            "17096. MGYP001084378716/10-391\n",
            "17097. MGYP000716486569/4-396\n",
            "17098. MGYP000666385375/8-590\n",
            "17099. MGYP000474683048/9-402\n",
            "17100. MGYP000140119543/4-403\n",
            "17101. MGYP000171720182/7-396\n",
            "17102. MGYP000017641009/9-437\n",
            "17103. MGYP000403655164/13-610\n",
            "17104. MGYP000299882594/3-511\n",
            "17105. MGYP000030061491/2-395\n",
            "17106. MGYP000956124297/7-408\n",
            "17107. MGYP000008533709/5-600\n",
            "17108. MGYP000754813249/5-607\n",
            "17109. MGYP000374024404/16-605\n",
            "17110. MGYP000998478495/33-622\n",
            "17111. MGYP000034580243/13-403\n",
            "17112. MGYP000942093554/12-614\n",
            "17113. MGYP000986804896/5-611\n",
            "17114. MGYP000856631468/6-395\n",
            "17115. MGYP000124869330/21-608\n",
            "17116. MGYP001058491755/8-393\n",
            "17117. MGYP000757791352/14-607\n",
            "17118. MGYP001063517537/18-611\n",
            "17119. MGYP000802369181/5-395\n",
            "17120. MGYP001017803343/8-616\n",
            "17121. MGYP000668545910/21-613\n",
            "17122. MGYP000858835028/31-635\n",
            "17123. MGYP000938036655/6-424\n",
            "17124. MGYP000303621382/9-610\n",
            "17125. MGYP000453716054/27-646\n",
            "17126. MGYP000408488852/10-441\n",
            "17127. MGYP001078475425/6-602\n",
            "17128. MGYP000818661012/22-432\n",
            "17129. MGYP000040806357/12-614\n",
            "17130. MGYP001118778881/31-625\n",
            "17131. MGYP001037729656/30-432\n",
            "17132. MGYP000084762657/11-603\n",
            "17133. MGYP000214788259/7-593\n",
            "17134. MGYP000285288768/60-667\n",
            "17135. MGYP000690811147/5-602\n",
            "17136. MGYP000406168726/17-459\n",
            "17137. MGYP000138431308/11-606\n",
            "17138. MGYP000762955478/4-395\n",
            "17139. MGYP000353659941/43-456\n",
            "17140. MGYP000917691560/7-385\n",
            "17141. MGYP000618291386/6-420\n",
            "17142. MGYP000529990166/12-400\n",
            "17143. MGYP000152342066/36-432\n",
            "17144. MGYP000861737372/15-602\n",
            "17145. MGYP000083103940/2-396\n",
            "17146. MGYP000098501052/5-393\n",
            "17147. MGYP000387230803/36-463\n",
            "17148. MGYP000408765219/8-400\n",
            "17149. MGYP001096898441/25-426\n",
            "17150. MGYP000265235863/94-650\n",
            "17151. MGYP000388707891/22-614\n",
            "17152. MGYP000730039163/4-567\n",
            "17153. MGYP000055850180/7-598\n",
            "17154. MGYP000612568927/10-610\n",
            "17155. MGYP000618775809/13-620\n",
            "17156. MGYP000872339015/41-432\n",
            "17157. MGYP000039026229/14-604\n",
            "17158. MGYP000028699337/2-387\n",
            "17159. MGYP000733540574/1-562\n",
            "17160. MGYP000371074243/9-600\n",
            "17161. MGYP001087044285/6-396\n",
            "17162. MGYP000733397676/26-435\n",
            "17163. MGYP000644178416/5-604\n",
            "17164. MGYP001137421670/8-605\n",
            "17165. MGYP000671530150/5-602\n",
            "17166. MGYP000978170479/6-394\n",
            "17167. MGYP000070306558/14-607\n",
            "17168. MGYP000558858935/21-409\n",
            "17169. MGYP000085112065/16-617\n",
            "17170. MGYP001098196992/17-420\n",
            "17171. MGYP000403198487/12-422\n",
            "17172. MGYP000517194570/22-414\n",
            "17173. MGYP000978429339/8-609\n",
            "17174. MGYP000627934952/20-416\n",
            "17175. MGYP000917847607/22-624\n",
            "17176. MGYP000852597219/7-607\n",
            "17177. MGYP000663060933/17-612\n",
            "17178. MGYP000146444061/6-585\n",
            "17179. MGYP000904115545/3-393\n",
            "17180. MGYP000410628059/3-395\n",
            "17181. MGYP000498018614/5-613\n",
            "17182. MGYP000095869770/2-397\n",
            "17183. MGYP000439287426/9-612\n",
            "17184. MGYP000562600801/4-394\n",
            "17185. MGYP000977341761/24-445\n",
            "17186. MGYP000836769365/28-428\n",
            "17187. MGYP000417367990/3-599\n",
            "17188. MGYP000961521081/17-410\n",
            "17189. MGYP000152582903/14-609\n",
            "17190. MGYP000026350464/10-619\n",
            "17191. MGYP000907005781/10-588\n",
            "17192. MGYP000090968321/7-614\n",
            "17193. MGYP000911734120/9-597\n",
            "17194. MGYP000042912034/16-437\n",
            "17195. MGYP000123250704/45-435\n",
            "17196. MGYP000794932628/37-439\n",
            "17197. MGYP000992332971/16-424\n",
            "17198. MGYP000126209071/8-402\n",
            "17199. MGYP000651280666/10-426\n",
            "17200. MGYP000685636193/16-605\n",
            "17201. MGYP000002911068/14-605\n",
            "17202. MGYP000766188970/13-394\n",
            "17203. MGYP000983245018/38-491\n",
            "17204. MGYP000396416337/10-603\n",
            "17205. MGYP000114944305/7-594\n",
            "17206. MGYP001045793711/14-617\n",
            "17207. MGYP001079975969/12-606\n",
            "17208. MGYP000331405677/2-403\n",
            "17209. MGYP000852737368/14-428\n",
            "17210. MGYP000734630131/2-395\n",
            "17211. MGYP001077117149/3-574\n",
            "17212. MGYP000546322655/5-394\n",
            "17213. MGYP000344540369/11-403\n",
            "17214. MGYP000501207555/25-610\n",
            "17215. MGYP000004282778/16-627\n",
            "17216. MGYP000021065283/5-395\n",
            "17217. MGYP000178213632/15-415\n",
            "17218. MGYP000857670055/13-421\n",
            "17219. MGYP000309148026/12-612\n",
            "17220. MGYP000586732282/13-430\n",
            "17221. MGYP001048318627/11-607\n",
            "17222. MGYP000667885463/44-424\n",
            "17223. MGYP000059050812/7-605\n",
            "17224. MGYP000848013364/2-398\n",
            "17225. MGYP000057624382/17-446\n",
            "17226. MGYP000980868643/37-641\n",
            "17227. MGYP000919444087/16-400\n",
            "17228. MGYP000025770081/11-619\n",
            "17229. MGYP000782941072/15-393\n",
            "17230. MGYP000696248720/4-601\n",
            "17231. MGYP000906404742/7-594\n",
            "17232. MGYP000855987339/33-628\n",
            "17233. MGYP000661896525/13-604\n",
            "17234. MGYP000097263476/13-607\n",
            "17235. MGYP000412824481/36-630\n",
            "17236. MGYP000661322358/5-395\n",
            "17237. MGYP001090552277/11-408\n",
            "17238. MGYP000885578306/11-625\n",
            "17239. MGYP000277456900/13-615\n",
            "17240. MGYP000185977344/33-623\n",
            "17241. MGYP000159194417/14-615\n",
            "17242. MGYP000415298603/97-696\n",
            "17243. MGYP000923744370/11-609\n",
            "17244. MGYP000011743277/4-386\n",
            "17245. MGYP000496823353/26-510\n",
            "17246. MGYP000904299828/8-396\n",
            "17247. MGYP000370784864/6-598\n",
            "17248. MGYP001123467332/10-427\n",
            "17249. MGYP001068711351/13-608\n",
            "17250. MGYP000079788178/7-394\n",
            "17251. MGYP000169673953/8-503\n",
            "17252. MGYP000196791000/31-439\n",
            "17253. MGYP001041863085/7-426\n",
            "17254. MGYP000596360669/3-399\n",
            "17255. MGYP000539507269/13-625\n",
            "17256. MGYP000535720107/5-631\n",
            "17257. MGYP000079746399/6-608\n",
            "17258. MGYP000812784473/5-375\n",
            "17259. MGYP000212022773/8-602\n",
            "17260. MGYP000147043045/13-611\n",
            "17261. MGYP000078091528/17-611\n",
            "17262. MGYP000885544005/14-625\n",
            "17263. MGYP000911115409/21-407\n",
            "17264. MGYP000318248141/14-409\n",
            "17265. MGYP000015023135/13-609\n",
            "17266. MGYP000505616156/2-340\n",
            "17267. MGYP000229651166/20-626\n",
            "17268. MGYP000880255705/11-602\n",
            "17269. MGYP000284678581/29-628\n",
            "17270. MGYP000029108691/47-663\n",
            "17271. MGYP000583386122/51-689\n",
            "17272. MGYP000917701665/19-617\n",
            "17273. MGYP000083833066/8-397\n",
            "17274. MGYP000515124839/7-649\n",
            "17275. MGYP000713460330/9-609\n",
            "17276. MGYP000421587957/17-598\n",
            "17277. MGYP000291398881/21-413\n",
            "17278. MGYP000182351028/16-617\n",
            "17279. MGYP000240253111/12-618\n",
            "17280. MGYP001020073068/7-613\n",
            "17281. MGYP000871807438/13-608\n",
            "17282. MGYP000216820849/11-603\n",
            "17283. MGYP000011317233/4-592\n",
            "17284. MGYP000060570367/6-613\n",
            "17285. MGYP000865996439/4-603\n",
            "17286. MGYP000183348508/61-660\n",
            "17287. MGYP000604192754/27-419\n",
            "17288. MGYP000825991644/10-405\n",
            "17289. MGYP001033291521/8-602\n",
            "17290. MGYP000701613003/11-617\n",
            "17291. MGYP001113701535/36-622\n",
            "17292. MGYP000672134916/11-437\n",
            "17293. MGYP001101231702/22-628\n",
            "17294. MGYP000984498230/11-621\n",
            "17295. MGYP000251042739/14-413\n",
            "17296. MGYP000406450718/6-596\n",
            "17297. MGYP000305954269/34-621\n",
            "17298. MGYP000059650056/7-615\n",
            "17299. MGYP000005510042/10-619\n",
            "17300. MGYP000855043585/38-458\n",
            "17301. MGYP000585199916/7-605\n",
            "17302. MGYP000176731041/12-437\n",
            "17303. MGYP001103108712/14-409\n",
            "17304. MGYP000246983754/24-477\n",
            "17305. MGYP000413262207/11-408\n",
            "17306. MGYP000515567823/4-604\n",
            "17307. MGYP000333341478/33-626\n",
            "17308. MGYP000162128070/11-596\n",
            "17309. MGYP001011247486/19-631\n",
            "17310. MGYP000224071087/12-604\n",
            "17311. MGYP000791560024/12-406\n",
            "17312. MGYP000850964843/14-610\n",
            "17313. MGYP001089111181/27-624\n",
            "17314. MGYP001043906881/36-625\n",
            "17315. MGYP000437058147/21-414\n",
            "17316. MGYP000809823783/4-395\n",
            "17317. MGYP000663267750/12-604\n",
            "17318. MGYP000423838991/17-622\n",
            "17319. MGYP001153351191/7-595\n",
            "17320. MGYP000064531052/7-607\n",
            "17321. MGYP001104812542/6-615\n",
            "17322. MGYP000019889193/8-600\n",
            "17323. MGYP000865911716/26-619\n",
            "17324. MGYP000930296867/26-457\n",
            "17325. MGYP000865825526/13-615\n",
            "17326. MGYP001003965979/11-608\n",
            "17327. MGYP000666139984/18-597\n",
            "17328. MGYP000791467713/12-602\n",
            "17329. MGYP000108698224/15-597\n",
            "17330. MGYP000594262482/15-603\n",
            "17331. MGYP000093366082/27-624\n",
            "17332. MGYP000936584988/20-434\n",
            "17333. MGYP000070658436/11-610\n",
            "17334. MGYP000184528449/23-623\n",
            "17335. MGYP000934543513/40-637\n",
            "17336. MGYP001076721898/38-637\n",
            "17337. MGYP000199751897/16-406\n",
            "17338. MGYP000984789865/11-607\n",
            "17339. MGYP000701750431/7-600\n",
            "17340. MGYP001137943621/22-603\n",
            "17341. MGYP000327197682/8-402\n",
            "17342. MGYP000473126593/9-615\n",
            "17343. MGYP000235698645/5-492\n",
            "17344. MGYP000730258681/2-398\n",
            "17345. MGYP000148259195/4-389\n",
            "17346. MGYP000668422922/9-651\n",
            "17347. MGYP000963492271/8-608\n",
            "17348. MGYP000892131283/2-574\n",
            "17349. MGYP000563947257/2-397\n",
            "17350. MGYP000543910374/11-399\n",
            "17351. MGYP000887912840/7-595\n",
            "17352. MGYP000834362856/12-418\n",
            "17353. MGYP001042060979/5-592\n",
            "17354. MGYP000858013193/21-526\n",
            "17355. MGYP000186526967/6-395\n",
            "17356. MGYP000876363506/9-610\n",
            "17357. MGYP000099637634/15-607\n",
            "17358. MGYP000862518325/11-604\n",
            "17359. MGYP001002053348/6-599\n",
            "17360. MGYP000223899671/7-394\n",
            "17361. MGYP000331790487/9-607\n",
            "17362. MGYP001009288171/18-404\n",
            "17363. MGYP001115782647/4-587\n",
            "17364. MGYP000425431112/12-419\n",
            "17365. MGYP000183426585/14-607\n",
            "17366. MGYP001129953862/15-607\n",
            "17367. MGYP000916179089/12-617\n",
            "17368. MGYP000975218395/11-632\n",
            "17369. MGYP001112641175/11-607\n",
            "17370. MGYP000051831132/44-435\n",
            "17371. MGYP001109442860/9-596\n",
            "17372. MGYP001028541263/7-398\n",
            "17373. MGYP000811030021/16-411\n",
            "17374. MGYP001020527181/36-633\n",
            "17375. MGYP000893264866/16-599\n",
            "17376. MGYP000662844185/23-424\n",
            "17377. MGYP000184995441/7-609\n",
            "17378. MGYP000468346611/9-610\n",
            "17379. MGYP000995382018/13-610\n",
            "17380. MGYP000435264579/28-421\n",
            "17381. MGYP000364805786/4-611\n",
            "17382. MGYP000873520932/9-611\n",
            "17383. MGYP000971972604/12-605\n",
            "17384. MGYP000434844364/6-402\n",
            "17385. MGYP001129215818/18-614\n",
            "17386. MGYP000297752966/16-607\n",
            "17387. MGYP001135302409/11-614\n",
            "17388. MGYP000887696504/4-395\n",
            "17389. MGYP000412348821/10-608\n",
            "17390. MGYP000903809920/15-608\n",
            "17391. MGYP001050270662/7-382\n",
            "17392. MGYP000704255645/10-405\n",
            "17393. MGYP000989004085/42-644\n",
            "17394. MGYP000156163419/22-607\n",
            "17395. MGYP000188278152/14-423\n",
            "17396. MGYP000946373235/1-476\n",
            "17397. MGYP000236593617/9-397\n",
            "17398. MGYP000968170353/6-615\n",
            "17399. MGYP001024663731/7-394\n",
            "17400. MGYP000149843970/33-623\n",
            "17401. MGYP001094329856/6-598\n",
            "17402. MGYP000965601976/10-486\n",
            "17403. MGYP000080928424/15-607\n",
            "17404. MGYP001138971753/14-399\n",
            "17405. MGYP000488623687/10-589\n",
            "17406. MGYP000843542645/2-406\n",
            "17407. MGYP001142067473/32-638\n",
            "17408. MGYP000113999077/33-439\n",
            "17409. MGYP000276723711/10-614\n",
            "17410. MGYP000847540808/36-632\n",
            "17411. MGYP001062011941/9-437\n",
            "17412. MGYP000439220783/12-433\n",
            "17413. MGYP000019351340/14-607\n",
            "17414. MGYP000180608007/11-623\n",
            "17415. MGYP001130354033/16-613\n",
            "17416. MGYP000534758573/54-651\n",
            "17417. MGYP000921067811/10-392\n",
            "17418. MGYP000194554285/9-604\n",
            "17419. MGYP000939460016/10-599\n",
            "17420. MGYP001154318340/160-548\n",
            "17421. MGYP000651205623/12-401\n",
            "17422. MGYP000653133460/8-400\n",
            "17423. MGYP000701048628/14-613\n",
            "17424. MGYP000055551338/9-613\n",
            "17425. MGYP000131101770/4-600\n",
            "17426. MGYP000846461822/7-598\n",
            "17427. MGYP000909948325/11-611\n",
            "17428. MGYP000526510145/7-596\n",
            "17429. MGYP001031696482/15-411\n",
            "17430. MGYP000104533484/13-616\n",
            "17431. MGYP000024897222/3-391\n",
            "17432. MGYP000936664348/9-597\n",
            "17433. MGYP000785330407/12-420\n",
            "17434. MGYP001047064753/5-406\n",
            "17435. MGYP000982433856/27-620\n",
            "17436. MGYP000179278339/6-412\n",
            "17437. MGYP000857425972/5-402\n",
            "17438. MGYP000480581414/10-606\n",
            "17439. MGYP000844971859/7-611\n",
            "17440. MGYP000922228827/10-610\n",
            "17441. MGYP000755141762/17-426\n",
            "17442. MGYP001055098534/12-607\n",
            "17443. MGYP000284884363/8-602\n",
            "17444. MGYP000686862556/7-394\n",
            "17445. MGYP000525566906/24-416\n",
            "17446. MGYP000082525876/14-606\n",
            "17447. MGYP000232670315/8-606\n",
            "17448. MGYP001037931464/16-603\n",
            "17449. MGYP000706351153/10-613\n",
            "17450. MGYP000942295106/32-634\n",
            "17451. MGYP001085542162/14-600\n",
            "17452. MGYP000170248009/7-602\n",
            "17453. MGYP000182586156/2-600\n",
            "17454. MGYP001139278261/16-405\n",
            "17455. MGYP000059756220/2-602\n",
            "17456. MGYP000269719062/4-610\n",
            "17457. MGYP000585292455/11-410\n",
            "17458. MGYP000268369125/4-606\n",
            "17459. MGYP000630255830/24-621\n",
            "17460. MGYP000153256274/6-383\n",
            "17461. MGYP001087978539/11-397\n",
            "17462. MGYP000096546814/7-595\n",
            "17463. MGYP000238540182/4-610\n",
            "17464. MGYP000655598846/7-406\n",
            "17465. MGYP000918753109/3-441\n",
            "17466. MGYP000283280836/13-391\n",
            "17467. MGYP000491679414/16-613\n",
            "17468. MGYP000456382100/2-397\n",
            "17469. MGYP001063547159/9-393\n",
            "17470. MGYP000138635742/7-394\n",
            "17471. MGYP000575740464/14-606\n",
            "17472. MGYP001101037779/11-612\n",
            "17473. MGYP000889818287/11-604\n",
            "17474. MGYP001061589786/5-630\n",
            "17475. MGYP000958490631/13-609\n",
            "17476. MGYP000154628781/46-639\n",
            "17477. MGYP000686595484/6-592\n",
            "17478. MGYP001039459582/12-613\n",
            "17479. MGYP001153881931/11-605\n",
            "17480. MGYP000096824641/2-404\n",
            "17481. MGYP000077994659/11-601\n",
            "17482. MGYP000675658570/17-423\n",
            "17483. MGYP001050334992/7-607\n",
            "17484. MGYP000060955931/9-588\n",
            "17485. MGYP000364873410/6-594\n",
            "17486. MGYP000121439394/2-395\n",
            "17487. MGYP000929459781/26-424\n",
            "17488. MGYP000000510035/3-377\n",
            "17489. MGYP000792974127/7-380\n",
            "17490. MGYP000875142861/6-597\n",
            "17491. MGYP001108025596/29-615\n",
            "17492. MGYP000354569014/12-418\n",
            "17493. MGYP000081232805/5-590\n",
            "17494. MGYP000691411476/7-394\n",
            "17495. MGYP000435661312/7-399\n",
            "17496. MGYP000020566631/3-599\n",
            "17497. MGYP000597531310/8-605\n",
            "17498. MGYP000065703817/11-405\n",
            "17499. MGYP000244275213/13-598\n",
            "17500. MGYP001047782750/52-653\n",
            "17501. MGYP000207988520/10-390\n",
            "17502. MGYP000095879846/13-609\n",
            "17503. MGYP000490952550/26-510\n",
            "17504. MGYP000168019188/8-613\n",
            "17505. MGYP001149936251/16-613\n",
            "17506. MGYP000947045257/2-545\n",
            "17507. MGYP000379832044/8-408\n",
            "17508. MGYP000934930877/11-428\n",
            "17509. MGYP000292355310/19-617\n",
            "17510. MGYP000463018029/22-448\n",
            "17511. MGYP000358577933/8-421\n",
            "17512. MGYP001077958712/7-601\n",
            "17513. MGYP000583561362/7-394\n",
            "17514. MGYP001042016478/50-657\n",
            "17515. MGYP000743401529/149-537\n",
            "17516. MGYP000430662617/2-397\n",
            "17517. MGYP000171251213/5-399\n",
            "17518. MGYP000065734674/20-403\n",
            "17519. MGYP001000061319/8-608\n",
            "17520. MGYP001065219827/11-358\n",
            "17521. MGYP000949449948/5-398\n",
            "17522. MGYP000684070645/10-601\n",
            "17523. MGYP001030700013/11-418\n",
            "17524. MGYP000850868244/5-399\n",
            "17525. MGYP000872330688/9-597\n",
            "17526. MGYP000011417431/10-400\n",
            "17527. MGYP000128324322/19-405\n",
            "17528. MGYP000300692466/9-401\n",
            "17529. MGYP000371849913/15-428\n",
            "17530. MGYP000740063659/4-595\n",
            "17531. MGYP000296620901/14-610\n",
            "17532. MGYP000219586303/26-657\n",
            "17533. MGYP000859435133/12-606\n",
            "17534. MGYP000866352828/7-394\n",
            "17535. MGYP000896185746/6-605\n",
            "17536. MGYP000545016376/15-608\n",
            "17537. MGYP000187137887/16-599\n",
            "17538. MGYP000085227766/6-606\n",
            "17539. MGYP000404104444/14-609\n",
            "17540. MGYP000081794282/1-350\n",
            "17541. MGYP000061767499/12-608\n",
            "17542. MGYP000806013594/2-397\n",
            "17543. MGYP000477618426/5-602\n",
            "17544. MGYP000379611718/8-611\n",
            "17545. MGYP000851938216/1-595\n",
            "17546. MGYP001119497306/9-402\n",
            "17547. MGYP000131123110/15-613\n",
            "17548. MGYP001131136277/84-446\n",
            "17549. MGYP000388958496/11-429\n",
            "17550. MGYP000134051032/28-406\n",
            "17551. MGYP000562468832/23-635\n",
            "17552. MGYP000852878159/13-609\n",
            "17553. MGYP000922219464/10-400\n",
            "17554. MGYP000009876996/16-602\n",
            "17555. MGYP001039390271/3-609\n",
            "17556. MGYP000140062034/6-399\n",
            "17557. MGYP000594330659/16-463\n",
            "17558. MGYP000053953463/4-401\n",
            "17559. MGYP000099259036/21-613\n",
            "17560. MGYP000148256951/11-607\n",
            "17561. MGYP000858380075/96-693\n",
            "17562. MGYP000267925255/8-600\n",
            "17563. MGYP001142015373/5-404\n",
            "17564. MGYP000548559263/8-400\n",
            "17565. MGYP000482099358/73-488\n",
            "17566. MGYP000024803072/5-384\n",
            "17567. MGYP000252313812/11-604\n",
            "17568. MGYP001031216799/10-604\n",
            "17569. MGYP000027551890/2-599\n",
            "17570. MGYP000889744662/26-626\n",
            "17571. MGYP001082248459/6-598\n",
            "17572. MGYP000363333343/7-394\n",
            "17573. MGYP000028546217/5-600\n",
            "17574. MGYP000944982873/37-632\n",
            "17575. MGYP000993577416/6-597\n",
            "17576. MGYP001037590237/6-394\n",
            "17577. MGYP000757995374/2-392\n",
            "17578. MGYP000288455337/19-609\n",
            "17579. MGYP000504125333/3-476\n",
            "17580. MGYP000849006294/7-595\n",
            "17581. MGYP000250442178/11-605\n",
            "17582. MGYP000657101294/17-616\n",
            "17583. MGYP000876369474/7-591\n",
            "17584. MGYP000209843683/8-425\n",
            "17585. MGYP000357468701/7-394\n",
            "17586. MGYP001071193055/7-394\n",
            "17587. MGYP000929601884/23-630\n",
            "17588. MGYP000625898085/7-394\n",
            "17589. MGYP001107963401/238-735\n",
            "17590. MGYP000921968792/11-608\n",
            "17591. MGYP000888392281/12-602\n",
            "17592. MGYP000339692833/6-401\n",
            "17593. MGYP000397466583/7-394\n",
            "17594. MGYP000850857892/2-397\n",
            "17595. MGYP000138525000/8-480\n",
            "17596. MGYP000681096155/14-610\n",
            "17597. MGYP000232478914/7-439\n",
            "17598. MGYP000439678328/5-432\n",
            "17599. MGYP000300019079/6-613\n",
            "17600. MGYP000047092307/7-403\n",
            "17601. MGYP000076357541/4-608\n",
            "17602. MGYP000247641712/27-595\n",
            "17603. MGYP000747482665/15-610\n",
            "17604. MGYP000029492186/8-588\n",
            "17605. MGYP000146668545/13-608\n",
            "17606. MGYP000573350248/9-599\n",
            "17607. MGYP000350840077/5-404\n",
            "17608. MGYP000254974143/24-634\n",
            "17609. MGYP001032741920/8-606\n",
            "17610. MGYP000978102452/9-591\n",
            "17611. MGYP000944915611/11-605\n",
            "17612. MGYP000044871804/15-437\n",
            "17613. MGYP000293234767/5-399\n",
            "17614. MGYP000879220709/10-602\n",
            "17615. MGYP000756746700/6-398\n",
            "17616. MGYP000173819871/14-603\n",
            "17617. MGYP000927539950/3-394\n",
            "17618. MGYP000210645100/5-397\n",
            "17619. MGYP000941907020/7-394\n",
            "17620. MGYP000315037027/8-611\n",
            "17621. MGYP000880179780/8-466\n",
            "17622. MGYP000321354308/10-608\n",
            "17623. MGYP000521313012/390-779\n",
            "17624. MGYP000061576201/18-612\n",
            "17625. MGYP000695554065/14-621\n",
            "17626. MGYP001048170867/6-596\n",
            "17627. MGYP000080920822/14-613\n",
            "17628. MGYP001095150789/182-512\n",
            "17629. MGYP000703260332/9-412\n",
            "17630. MGYP000058381538/4-418\n",
            "17631. MGYP000356529408/4-392\n",
            "17632. MGYP000055747131/4-593\n",
            "17633. MGYP000529858507/5-397\n",
            "17634. MGYP000542268341/17-413\n",
            "17635. MGYP000955650696/62-572\n",
            "17636. MGYP000933848130/8-589\n",
            "17637. MGYP001154341061/22-612\n",
            "17638. MGYP000853293171/5-605\n",
            "17639. MGYP000295698347/7-394\n",
            "17640. MGYP000100690226/15-603\n",
            "17641. MGYP000512837162/2-412\n",
            "17642. MGYP000145551640/7-381\n",
            "17643. MGYP000076199043/13-403\n",
            "17644. MGYP000936456175/35-636\n",
            "17645. MGYP001091948626/15-395\n",
            "17646. MGYP000957525444/16-623\n",
            "17647. MGYP000894280532/8-424\n",
            "17648. MGYP000980894218/8-591\n",
            "17649. MGYP000632461022/13-592\n",
            "17650. MGYP000132169328/2-397\n",
            "17651. MGYP000220889022/12-615\n",
            "17652. MGYP000978107533/8-406\n",
            "17653. MGYP000477651139/12-416\n",
            "17654. MGYP000003282328/17-411\n",
            "17655. MGYP000689681987/4-397\n",
            "17656. MGYP000860241860/20-618\n",
            "17657. MGYP000071110196/5-395\n",
            "17658. MGYP000085427063/10-410\n",
            "17659. MGYP000921927974/10-590\n",
            "17660. MGYP001154393452/82-468\n",
            "17661. MGYP000058980642/4-584\n",
            "17662. MGYP000156213897/8-385\n",
            "17663. MGYP000879425920/4-598\n",
            "17664. MGYP000186551455/5-598\n",
            "17665. MGYP000326543447/24-429\n",
            "17666. MGYP000873921276/20-468\n",
            "17667. MGYP000071731518/13-406\n",
            "17668. MGYP000386573089/17-620\n",
            "17669. MGYP000881770849/34-639\n",
            "17670. MGYP000492335973/29-430\n",
            "17671. MGYP001087742172/13-607\n",
            "17672. MGYP000045933290/10-595\n",
            "17673. MGYP000562499955/7-595\n",
            "17674. MGYP000456843996/7-396\n",
            "17675. MGYP000004286621/13-405\n",
            "17676. MGYP000924748982/7-398\n",
            "17677. MGYP000038931895/2-398\n",
            "17678. MGYP000966261999/12-412\n",
            "17679. MGYP000473360009/10-453\n",
            "17680. MGYP000066880968/6-605\n",
            "17681. MGYP000209214778/15-602\n",
            "17682. MGYP000387970356/17-607\n",
            "17683. MGYP001027487455/15-603\n",
            "17684. MGYP000058583717/15-463\n",
            "17685. MGYP000135566233/16-615\n",
            "17686. MGYP000394820129/7-612\n",
            "17687. MGYP000600962037/4-402\n",
            "17688. MGYP001029193242/14-604\n",
            "17689. MGYP000867191565/7-597\n",
            "17690. MGYP000272729454/12-605\n",
            "17691. MGYP001047218106/12-414\n",
            "17692. MGYP000446900749/3-623\n",
            "17693. MGYP000623559646/13-609\n",
            "17694. MGYP000276720691/10-614\n",
            "17695. MGYP000150310736/10-600\n",
            "17696. MGYP000234626434/7-610\n",
            "17697. MGYP000862249195/8-608\n",
            "17698. MGYP001025121904/14-606\n",
            "17699. MGYP000841347191/24-624\n",
            "17700. MGYP000647776866/15-617\n",
            "17701. MGYP001143542633/1-146\n",
            "17702. MGYP000935717375/12-604\n",
            "17703. MGYP000057387711/11-400\n",
            "17704. MGYP000168654929/2-397\n",
            "17705. MGYP000253478429/7-596\n",
            "17706. MGYP000147380263/4-401\n",
            "17707. MGYP000421582059/5-596\n",
            "17708. MGYP001010263494/9-396\n",
            "17709. MGYP000417389267/29-429\n",
            "17710. MGYP001116969543/6-406\n",
            "17711. MGYP000273966400/12-611\n",
            "17712. MGYP000202098068/8-416\n",
            "17713. MGYP000340994230/33-429\n",
            "17714. MGYP000002474610/6-442\n",
            "17715. MGYP000329050187/5-397\n",
            "17716. MGYP000724556849/6-603\n",
            "17717. MGYP000862455885/13-600\n",
            "17718. MGYP001147613243/3-395\n",
            "17719. MGYP000011475872/37-632\n",
            "17720. MGYP000259186251/13-398\n",
            "17721. MGYP000790267074/44-440\n",
            "17722. MGYP000689955795/12-609\n",
            "17723. MGYP000400402933/7-394\n",
            "17724. MGYP000691101961/1-395\n",
            "17725. MGYP000731203390/4-600\n",
            "17726. MGYP001090387995/28-631\n",
            "17727. MGYP000855222660/10-593\n",
            "17728. MGYP000050074174/5-601\n",
            "17729. MGYP000716483685/4-590\n",
            "17730. MGYP000939164555/14-618\n",
            "17731. MGYP000872339522/20-421\n",
            "17732. MGYP000046829633/8-400\n",
            "17733. MGYP000567618464/1-381\n",
            "17734. MGYP000164954669/5-606\n",
            "17735. MGYP001087622538/15-434\n",
            "17736. MGYP000118379198/6-600\n",
            "17737. MGYP000241480386/12-612\n",
            "17738. MGYP001097617526/7-402\n",
            "17739. MGYP000444943080/7-606\n",
            "17740. MGYP001073871336/2-418\n",
            "17741. MGYP000029689088/15-608\n",
            "17742. MGYP001117245136/7-611\n",
            "17743. MGYP000097495572/21-622\n",
            "17744. MGYP000845098825/4-611\n",
            "17745. MGYP000414200816/10-616\n",
            "17746. MGYP001098424146/18-612\n",
            "17747. MGYP000086855369/9-335\n",
            "17748. MGYP000485174945/41-638\n",
            "17749. MGYP000555911782/12-409\n",
            "17750. MGYP000849115573/6-605\n",
            "17751. MGYP000902700945/91-684\n",
            "17752. MGYP000465187463/10-414\n",
            "17753. MGYP000228405278/57-661\n",
            "17754. MGYP000904838849/21-636\n",
            "17755. MGYP000626502211/3-600\n",
            "17756. MGYP000859530093/35-423\n",
            "17757. MGYP001027046683/30-622\n",
            "17758. MGYP000415061146/8-511\n",
            "17759. MGYP000260908688/5-397\n",
            "17760. MGYP000609556053/25-416\n",
            "17761. MGYP000885556109/8-606\n",
            "17762. MGYP000099940704/16-401\n",
            "17763. MGYP000019839617/7-595\n",
            "17764. MGYP000645968787/7-608\n",
            "17765. MGYP000428423148/87-696\n",
            "17766. MGYP001105371616/15-432\n",
            "17767. MGYP001016180646/12-608\n",
            "17768. MGYP000220867851/1-561\n",
            "17769. MGYP000945161114/7-394\n",
            "17770. MGYP000552060009/7-394\n",
            "17771. MGYP000272653989/5-592\n",
            "17772. MGYP001010407167/22-417\n",
            "17773. MGYP000232540768/18-412\n",
            "17774. MGYP000587074300/7-394\n",
            "17775. MGYP000332424441/1-378\n",
            "17776. MGYP000747631494/22-414\n",
            "17777. MGYP000601125320/5-397\n",
            "17778. MGYP001079358360/18-615\n",
            "17779. MGYP000553591609/14-445\n",
            "17780. MGYP000350435134/12-611\n",
            "17781. MGYP000991453760/5-396\n",
            "17782. MGYP000050491818/35-626\n",
            "17783. MGYP000714468095/2-365\n",
            "17784. MGYP000117671367/10-607\n",
            "17785. MGYP000140719506/5-607\n",
            "17786. MGYP000898496597/7-605\n",
            "17787. MGYP000284759633/7-591\n",
            "17788. MGYP000850278650/13-610\n",
            "17789. MGYP000850956865/7-387\n",
            "17790. MGYP000984592661/7-601\n",
            "17791. MGYP000903455648/69-483\n",
            "17792. MGYP001020648769/2-390\n",
            "17793. MGYP000408401924/25-624\n",
            "17794. MGYP000741008572/7-381\n",
            "17795. MGYP000550119639/12-420\n",
            "17796. MGYP000530530128/2-395\n",
            "17797. MGYP001068822240/8-593\n",
            "17798. MGYP000627691780/10-603\n",
            "17799. MGYP000892239913/7-394\n",
            "17800. MGYP000055636159/10-645\n",
            "17801. MGYP000520515165/3-601\n",
            "17802. MGYP000015027792/25-623\n",
            "17803. MGYP000910506165/84-687\n",
            "17804. MGYP000881550712/12-612\n",
            "17805. MGYP000785951042/9-348\n",
            "17806. MGYP000906240119/17-617\n",
            "17807. MGYP000620809559/8-413\n",
            "17808. MGYP000216781883/4-596\n",
            "17809. MGYP000673607579/8-442\n",
            "17810. MGYP000375839732/8-604\n",
            "17811. MGYP000176564985/10-613\n",
            "17812. MGYP000228008243/12-607\n",
            "17813. MGYP000856030443/9-600\n",
            "17814. MGYP000461025677/11-603\n",
            "17815. MGYP000148358866/7-383\n",
            "17816. MGYP000476355361/14-411\n",
            "17817. MGYP000850251377/13-612\n",
            "17818. MGYP000659273490/6-397\n",
            "17819. MGYP000953526499/102-699\n",
            "17820. MGYP000922649392/16-603\n",
            "17821. MGYP000928626874/11-423\n",
            "17822. MGYP000008371919/12-619\n",
            "17823. MGYP000139939628/13-603\n",
            "17824. MGYP000550260573/36-632\n",
            "17825. MGYP000021930919/15-496\n",
            "17826. MGYP000716925705/58-667\n",
            "17827. MGYP000364044175/2-601\n",
            "17828. MGYP000957826274/7-438\n",
            "17829. MGYP000017327104/30-622\n",
            "17830. MGYP000506717526/12-604\n",
            "17831. MGYP001007214977/4-650\n",
            "17832. MGYP001044649678/192-590\n",
            "17833. MGYP001037494854/8-591\n",
            "17834. MGYP000213224319/7-398\n",
            "17835. MGYP001036823451/3-560\n",
            "17836. MGYP000253842293/74-478\n",
            "17837. MGYP000456486437/6-609\n",
            "17838. MGYP001119465830/108-711\n",
            "17839. MGYP000716482242/8-394\n",
            "17840. MGYP000840252990/1-569\n",
            "17841. MGYP000666137797/6-609\n",
            "17842. MGYP000349586559/6-411\n",
            "17843. MGYP001083173092/117-712\n",
            "17844. MGYP000489017847/11-619\n",
            "17845. MGYP000071034218/17-608\n",
            "17846. MGYP000045139437/13-405\n",
            "17847. MGYP000413000239/7-607\n",
            "17848. MGYP000636878241/14-612\n",
            "17849. MGYP001099024989/7-394\n",
            "17850. MGYP001042003287/13-611\n",
            "17851. MGYP000180893985/10-604\n",
            "17852. MGYP000468881085/7-394\n",
            "17853. MGYP001103749289/7-394\n",
            "17854. MGYP000551735072/2-592\n",
            "17855. MGYP000202956133/7-394\n",
            "17856. MGYP000218154293/14-606\n",
            "17857. MGYP000302664583/14-607\n",
            "17858. MGYP000297455681/26-445\n",
            "17859. MGYP001055915257/4-609\n",
            "17860. MGYP000023638369/13-604\n",
            "17861. MGYP000480342064/7-397\n",
            "17862. MGYP000262404574/2-598\n",
            "17863. MGYP000916805160/20-621\n",
            "17864. MGYP000058148065/14-610\n",
            "17865. MGYP000487883573/26-632\n",
            "17866. MGYP000087970375/5-395\n",
            "17867. MGYP001032159754/9-607\n",
            "17868. MGYP001016773979/13-606\n",
            "17869. MGYP000319192195/7-342\n",
            "17870. MGYP000046341628/8-606\n",
            "17871. MGYP000855767225/2-594\n",
            "17872. MGYP000055711156/12-589\n",
            "17873. MGYP000780995953/2-388\n",
            "17874. MGYP000017762161/7-394\n",
            "17875. MGYP000360056517/6-351\n",
            "17876. MGYP001045331019/14-437\n",
            "17877. MGYP000002531778/152-723\n",
            "17878. MGYP000297467889/10-613\n",
            "17879. MGYP000854639183/9-386\n",
            "17880. MGYP000854639183/513-596\n",
            "17881. MGYP000656007023/12-620\n",
            "17882. MGYP000880231073/5-626\n",
            "17883. MGYP000952117515/5-402\n",
            "17884. MGYP001048770724/2-397\n",
            "17885. MGYP001057869336/8-399\n",
            "17886. MGYP001034407674/12-407\n",
            "17887. MGYP001078814099/15-607\n",
            "17888. MGYP000100401418/32-623\n",
            "17889. MGYP000891206461/9-406\n",
            "17890. MGYP000595187574/9-643\n",
            "17891. MGYP000391952169/3-340\n",
            "17892. MGYP001044619415/11-609\n",
            "17893. MGYP000928984413/8-601\n",
            "17894. MGYP000003558505/14-612\n",
            "17895. MGYP000867961527/9-600\n",
            "17896. MGYP000872573780/10-425\n",
            "17897. MGYP000123492075/15-609\n",
            "17898. MGYP000706048513/13-434\n",
            "17899. MGYP001096026585/2-364\n",
            "17900. MGYP000990117529/15-622\n",
            "17901. MGYP000043234992/13-406\n",
            "17902. MGYP000917569049/6-449\n",
            "17903. MGYP000686099701/16-396\n",
            "17904. MGYP000165084637/23-440\n",
            "17905. MGYP000482005609/5-607\n",
            "17906. MGYP000505040198/13-418\n",
            "17907. MGYP000778992614/10-404\n",
            "17908. MGYP000635043589/14-422\n",
            "17909. MGYP000964637561/8-507\n",
            "17910. MGYP000821258031/2-397\n",
            "17911. MGYP000412659679/13-402\n",
            "17912. MGYP000863603501/10-604\n",
            "17913. MGYP000403499761/10-645\n",
            "17914. MGYP000522355803/7-394\n",
            "17915. MGYP000266846798/17-414\n",
            "17916. MGYP000101617946/12-609\n",
            "17917. MGYP001137883419/14-409\n",
            "17918. MGYP000713974958/7-394\n",
            "17919. MGYP000862972110/5-625\n",
            "17920. MGYP000833934251/13-396\n",
            "17921. MGYP000331808888/36-670\n",
            "17922. MGYP000807062459/13-412\n",
            "17923. MGYP000845960990/5-595\n",
            "17924. MGYP000570551917/11-385\n",
            "17925. MGYP000010449812/19-631\n",
            "17926. MGYP000904360839/28-637\n",
            "17927. MGYP000286321205/2-621\n",
            "17928. MGYP000468744175/3-578\n",
            "17929. MGYP000817108537/28-410\n",
            "17930. MGYP000624915523/8-587\n",
            "17931. MGYP000936424209/13-396\n",
            "17932. MGYP000404167819/9-445\n",
            "17933. MGYP000108088087/1-394\n",
            "17934. MGYP001109760867/40-674\n",
            "17935. MGYP000984648294/83-490\n",
            "17936. MGYP001093414505/61-657\n",
            "17937. MGYP000752246152/5-380\n",
            "17938. MGYP000735942994/6-592\n",
            "17939. MGYP000135172698/10-596\n",
            "17940. MGYP000415896859/7-412\n",
            "17941. MGYP000169001282/14-615\n",
            "17942. MGYP000434286766/15-428\n",
            "17943. MGYP000329899573/5-605\n",
            "17944. MGYP001048564387/30-641\n",
            "17945. MGYP001100561510/14-434\n",
            "17946. MGYP000805471138/4-395\n",
            "17947. MGYP000096867514/5-416\n",
            "17948. MGYP000237246387/3-379\n",
            "17949. MGYP001002783543/6-621\n",
            "17950. MGYP000702273879/5-342\n",
            "17951. MGYP001019799128/2-403\n",
            "17952. MGYP000146458560/5-599\n",
            "17953. MGYP000523984145/11-603\n",
            "17954. MGYP000685442331/11-612\n",
            "17955. MGYP001131163975/9-398\n",
            "17956. MGYP000900557200/17-513\n",
            "17957. MGYP001062674948/15-402\n",
            "17958. MGYP000463097100/2-397\n",
            "17959. MGYP000569680363/11-602\n",
            "17960. MGYP000491858591/22-634\n",
            "17961. MGYP000415713279/10-597\n",
            "17962. MGYP000515555370/15-618\n",
            "17963. MGYP000156384115/8-401\n",
            "17964. MGYP000689419636/8-601\n",
            "17965. MGYP000944173587/4-394\n",
            "17966. MGYP000377368292/11-507\n",
            "17967. MGYP000059913522/2-397\n",
            "17968. MGYP000258849422/5-394\n",
            "17969. MGYP000420344954/7-601\n",
            "17970. MGYP000298565379/13-606\n",
            "17971. MGYP000056222876/9-421\n",
            "17972. MGYP000323742869/19-405\n",
            "17973. MGYP000235440745/2-387\n",
            "17974. MGYP001034108398/57-655\n",
            "17975. MGYP001022293646/7-394\n",
            "17976. MGYP000188397917/10-604\n",
            "17977. MGYP000000604032/2-396\n",
            "17978. MGYP000339993471/7-394\n",
            "17979. MGYP000182875204/7-394\n",
            "17980. MGYP001050882628/42-636\n",
            "17981. MGYP001041543302/4-598\n",
            "17982. MGYP000942414392/51-440\n",
            "17983. MGYP000881378456/13-600\n",
            "17984. MGYP000766965307/18-600\n",
            "17985. MGYP000734161289/22-397\n",
            "17986. MGYP000566103833/17-602\n",
            "17987. MGYP000886996457/14-609\n",
            "17988. MGYP000049476004/4-611\n",
            "17989. MGYP000002012287/3-396\n",
            "17990. MGYP000986584853/6-620\n",
            "17991. MGYP000332903998/11-598\n",
            "17992. MGYP000928194130/19-441\n",
            "17993. MGYP000382266294/5-597\n",
            "17994. MGYP000530087424/2-593\n",
            "17995. MGYP001101436950/10-605\n",
            "17996. MGYP000110025385/38-633\n",
            "17997. MGYP000335690674/2-588\n",
            "17998. MGYP001030498840/7-405\n",
            "17999. MGYP000718087374/13-612\n",
            "18000. MGYP001105027361/8-612\n",
            "18001. MGYP000311760100/7-609\n",
            "18002. MGYP000623628646/20-618\n",
            "18003. MGYP000322970570/43-663\n",
            "18004. MGYP000070463783/8-355\n",
            "18005. MGYP000367268838/11-615\n",
            "18006. MGYP000343384254/20-433\n",
            "18007. MGYP000067820167/3-341\n",
            "18008. MGYP000448958967/3-396\n",
            "18009. MGYP001083415374/4-620\n",
            "18010. MGYP000996613181/23-418\n",
            "18011. MGYP001146602038/8-593\n",
            "18012. MGYP000509092128/7-404\n",
            "18013. MGYP000956738947/5-586\n",
            "18014. MGYP000359403030/11-615\n",
            "18015. MGYP000800946319/3-340\n",
            "18016. MGYP000983045859/8-399\n",
            "18017. MGYP000081866505/2-397\n",
            "18018. MGYP001047801080/9-613\n",
            "18019. MGYP000258570307/5-600\n",
            "18020. MGYP000861480217/7-393\n",
            "18021. MGYP000670722660/15-607\n",
            "18022. MGYP000756999727/6-340\n",
            "18023. MGYP000627320185/12-642\n",
            "18024. MGYP000778890104/15-607\n",
            "18025. MGYP000996410759/10-607\n",
            "18026. MGYP000183054793/29-632\n",
            "18027. MGYP000249021263/22-419\n",
            "18028. MGYP000227022276/8-597\n",
            "18029. MGYP000442205690/9-630\n",
            "18030. MGYP000198605026/4-378\n",
            "18031. MGYP000939247239/11-606\n",
            "18032. MGYP001021871282/7-438\n",
            "18033. MGYP001121175241/12-441\n",
            "18034. MGYP000891349572/6-420\n",
            "18035. MGYP000177491857/15-607\n",
            "18036. MGYP000321385194/14-606\n",
            "18037. MGYP000094456050/70-667\n",
            "18038. MGYP000053173803/10-597\n",
            "18039. MGYP000931669534/16-604\n",
            "18040. MGYP000001367605/7-596\n",
            "18041. MGYP000011825236/16-398\n",
            "18042. MGYP001058000876/6-592\n",
            "18043. MGYP000505509161/13-430\n",
            "18044. MGYP000988762071/7-616\n",
            "18045. MGYP000199151268/18-610\n",
            "18046. MGYP000860598940/6-383\n",
            "18047. MGYP000280084916/8-603\n",
            "18048. MGYP000849036308/23-609\n",
            "18049. MGYP000866742113/10-375\n",
            "18050. MGYP000558937676/27-631\n",
            "18051. MGYP000686591655/17-616\n",
            "18052. MGYP000035207556/12-612\n",
            "18053. MGYP001035110207/8-604\n",
            "18054. MGYP000716466272/8-396\n",
            "18055. MGYP000773175711/3-339\n",
            "18056. MGYP000070359546/12-600\n",
            "18057. MGYP000626215965/4-403\n",
            "18058. MGYP000179733163/7-599\n",
            "18059. MGYP000011170555/7-606\n",
            "18060. MGYP000852742881/12-611\n",
            "18061. MGYP000612104246/7-609\n",
            "18062. MGYP000480696874/2-385\n",
            "18063. MGYP000988654750/51-496\n",
            "18064. MGYP000489092275/23-429\n",
            "18065. MGYP000844872638/8-594\n",
            "18066. MGYP000958230473/4-608\n",
            "18067. MGYP000211147240/11-405\n",
            "18068. MGYP000400742140/6-606\n",
            "18069. MGYP000201898187/64-465\n",
            "18070. MGYP001055328132/20-420\n",
            "18071. MGYP000678847645/4-403\n",
            "18072. MGYP000852577956/20-605\n",
            "18073. MGYP000551133560/12-611\n",
            "18074. MGYP001141867419/6-384\n",
            "18075. MGYP000906813260/12-610\n",
            "18076. MGYP000063581629/198-540\n",
            "18077. MGYP000055594928/360-736\n",
            "18078. MGYP000661212730/12-605\n",
            "18079. MGYP000421393045/5-386\n",
            "18080. MGYP000487348482/20-430\n",
            "18081. MGYP000253118379/9-610\n",
            "18082. MGYP000762516653/15-607\n",
            "18083. MGYP000340054836/8-604\n",
            "18084. MGYP001124112398/50-641\n",
            "18085. MGYP001146146417/68-670\n",
            "18086. MGYP000021926694/60-488\n",
            "18087. MGYP000720521176/22-608\n",
            "18088. MGYP000325890868/3-606\n",
            "18089. MGYP000861569029/12-605\n",
            "18090. MGYP000906844717/6-418\n",
            "18091. MGYP000951482189/20-403\n",
            "18092. MGYP000613271582/7-610\n",
            "18093. MGYP001077667736/27-619\n",
            "18094. MGYP000164868322/6-368\n",
            "18095. MGYP000608382803/20-361\n",
            "18096. MGYP000861660925/14-626\n",
            "18097. MGYP000055851635/10-408\n",
            "18098. MGYP000979249761/10-613\n",
            "18099. MGYP000033051834/13-604\n",
            "18100. MGYP000879773046/5-389\n",
            "18101. MGYP001009863538/12-604\n",
            "18102. MGYP001057364167/4-394\n",
            "18103. MGYP000272993601/10-606\n",
            "18104. MGYP000299389227/10-607\n",
            "18105. MGYP001049993808/10-609\n",
            "18106. MGYP000039949432/11-614\n",
            "18107. MGYP000748525620/15-607\n",
            "18108. MGYP000994516693/6-527\n",
            "18109. MGYP000920425729/6-587\n",
            "18110. MGYP001092929216/7-581\n",
            "18111. MGYP000361472228/19-428\n",
            "18112. MGYP001102320973/15-610\n",
            "18113. MGYP000797132307/34-626\n",
            "18114. MGYP000237278704/45-446\n",
            "18115. MGYP000041744308/15-607\n",
            "18116. MGYP001149041377/10-400\n",
            "18117. MGYP000094732920/2-397\n",
            "18118. MGYP000556409231/14-607\n",
            "18119. MGYP000036742026/6-600\n",
            "18120. MGYP001132206183/9-605\n",
            "18121. MGYP000278266462/8-617\n",
            "18122. MGYP000459943415/28-631\n",
            "18123. MGYP000206646810/5-405\n",
            "18124. MGYP000708718194/5-398\n",
            "18125. MGYP000259193152/2-407\n",
            "18126. MGYP000365372150/25-401\n",
            "18127. MGYP001152852401/6-405\n",
            "18128. MGYP000255147008/3-632\n",
            "18129. MGYP000050501401/15-607\n",
            "18130. MGYP000935638843/12-613\n",
            "18131. MGYP000987244490/8-399\n",
            "18132. MGYP000431886359/7-596\n",
            "18133. MGYP000297354948/9-616\n",
            "18134. MGYP000050684218/5-606\n",
            "18135. MGYP000882249856/8-392\n",
            "18136. MGYP001140066612/7-594\n",
            "18137. MGYP000019201199/7-404\n",
            "18138. MGYP000918931795/6-426\n",
            "18139. MGYP000390616068/11-614\n",
            "18140. MGYP000885486399/15-403\n",
            "18141. MGYP000782876406/12-393\n",
            "18142. MGYP001078211732/9-588\n",
            "18143. MGYP000957075928/12-613\n",
            "18144. MGYP000429758685/5-589\n",
            "18145. MGYP000880847822/12-613\n",
            "18146. MGYP000042648422/6-595\n",
            "18147. MGYP000020166143/5-401\n",
            "18148. MGYP000148196255/36-629\n",
            "18149. MGYP000139252871/11-604\n",
            "18150. MGYP001062177046/7-609\n",
            "18151. MGYP000308430094/7-394\n",
            "18152. MGYP000948498123/3-392\n",
            "18153. MGYP000846750742/10-603\n",
            "18154. MGYP001094847123/6-599\n",
            "18155. MGYP000813060434/1-394\n",
            "18156. MGYP000141905997/8-586\n",
            "18157. MGYP000391447932/3-335\n",
            "18158. MGYP000009080497/32-622\n",
            "18159. MGYP000859613861/6-399\n",
            "18160. MGYP000333235587/2-595\n",
            "18161. MGYP000100003132/14-606\n",
            "18162. MGYP000451219967/2-397\n",
            "18163. MGYP000882385818/11-601\n",
            "18164. MGYP000025333307/3-345\n",
            "18165. MGYP000906372787/55-662\n",
            "18166. MGYP000044532996/15-615\n",
            "18167. MGYP000073840743/5-604\n",
            "18168. MGYP000974641854/4-395\n",
            "18169. MGYP000923234389/10-644\n",
            "18170. MGYP001049332195/15-603\n",
            "18171. MGYP000992267635/7-388\n",
            "18172. MGYP000619533036/12-414\n",
            "18173. MGYP000311273313/11-459\n",
            "18174. MGYP000009354750/15-607\n",
            "18175. MGYP000365250656/10-642\n",
            "18176. MGYP000626468941/11-421\n",
            "18177. MGYP000511783280/13-414\n",
            "18178. MGYP000847662974/12-643\n",
            "18179. MGYP000706513496/9-350\n",
            "18180. MGYP001063134022/15-603\n",
            "18181. MGYP000953159600/2-572\n",
            "18182. MGYP000871419261/10-609\n",
            "18183. MGYP000976926865/10-601\n",
            "18184. MGYP001121279665/35-630\n",
            "18185. MGYP000344282879/18-417\n",
            "18186. MGYP000371460759/8-609\n",
            "18187. MGYP001057989279/15-428\n",
            "18188. MGYP001004439082/9-408\n",
            "18189. MGYP000134881044/4-610\n",
            "18190. MGYP000229259437/103-501\n",
            "18191. MGYP000862963369/11-408\n",
            "18192. MGYP000848684560/7-411\n",
            "18193. MGYP000129334887/13-612\n",
            "18194. MGYP000536627590/31-621\n",
            "18195. MGYP000150813025/22-613\n",
            "18196. MGYP000862255353/4-411\n",
            "18197. MGYP000218483561/10-609\n",
            "18198. MGYP001035829408/8-605\n",
            "18199. MGYP000167721128/14-592\n",
            "18200. MGYP000281198528/14-407\n",
            "18201. MGYP000453633366/4-608\n",
            "18202. MGYP000001544760/8-406\n",
            "18203. MGYP000236619280/7-593\n",
            "18204. MGYP000689177540/9-605\n",
            "18205. MGYP000873564458/11-612\n",
            "18206. MGYP000228502674/15-396\n",
            "18207. MGYP000744506606/32-394\n",
            "18208. MGYP000958461158/11-440\n",
            "18209. MGYP000060551619/16-406\n",
            "18210. MGYP000161874053/10-412\n",
            "18211. MGYP001106683023/23-612\n",
            "18212. MGYP000297011454/10-401\n",
            "18213. MGYP000860248731/37-414\n",
            "18214. MGYP000617622226/14-613\n",
            "18215. MGYP000887899572/11-409\n",
            "18216. MGYP000402398153/9-583\n",
            "18217. MGYP000373054487/5-621\n",
            "18218. MGYP000735011028/5-602\n",
            "18219. MGYP000547629102/7-611\n",
            "18220. MGYP000123425692/23-439\n",
            "18221. MGYP000217417665/13-607\n",
            "18222. MGYP001028030047/17-431\n",
            "18223. MGYP000024946159/12-400\n",
            "18224. MGYP000706431717/7-592\n",
            "18225. MGYP001122057972/11-351\n",
            "18226. MGYP000611745781/16-409\n",
            "18227. MGYP000697609036/6-400\n",
            "18228. MGYP000871374570/6-607\n",
            "18229. MGYP000381748565/22-607\n",
            "18230. MGYP001150834056/262-854\n",
            "18231. MGYP000934696444/17-609\n",
            "18232. MGYP001134177062/13-408\n",
            "18233. MGYP001156062693/23-438\n",
            "18234. MGYP001121686467/36-434\n",
            "18235. MGYP000516177240/15-607\n",
            "18236. MGYP001025852540/13-605\n",
            "18237. MGYP000177741650/14-607\n",
            "18238. MGYP000242494230/14-437\n",
            "18239. MGYP001065890410/15-411\n",
            "18240. MGYP000109833077/37-430\n",
            "18241. MGYP000358474916/10-604\n",
            "18242. MGYP000527689887/12-400\n",
            "18243. MGYP000517351826/20-400\n",
            "18244. MGYP000091345511/10-610\n",
            "18245. MGYP000484971418/5-398\n",
            "18246. MGYP001126692152/6-563\n",
            "18247. MGYP000057539874/10-603\n",
            "18248. MGYP000415095512/4-607\n",
            "18249. MGYP000655169238/3-340\n",
            "18250. MGYP000872248625/6-599\n",
            "18251. MGYP000977354903/14-436\n",
            "18252. MGYP000242752818/11-417\n",
            "18253. MGYP000559497583/11-326\n",
            "18254. MGYP000517735772/6-411\n",
            "18255. MGYP000507488884/7-610\n",
            "18256. MGYP000134061170/18-611\n",
            "18257. MGYP000081285739/9-610\n",
            "18258. MGYP000821592911/12-610\n",
            "18259. MGYP000391263291/12-604\n",
            "18260. MGYP000570463091/11-607\n",
            "18261. MGYP001071293576/8-596\n",
            "18262. MGYP001140983892/13-423\n",
            "18263. MGYP000914532166/9-603\n",
            "18264. MGYP000846285015/11-423\n",
            "18265. MGYP000001522933/18-610\n",
            "18266. MGYP001135816226/7-603\n",
            "18267. MGYP000952595715/4-387\n",
            "18268. MGYP000459655313/10-607\n",
            "18269. MGYP001093606042/15-434\n",
            "18270. MGYP000350797097/5-403\n",
            "18271. MGYP001077030009/7-644\n",
            "18272. MGYP000185459945/7-392\n",
            "18273. MGYP000954917167/31-442\n",
            "18274. MGYP000666344486/10-614\n",
            "18275. MGYP000726520970/3-501\n",
            "18276. MGYP000152544888/22-607\n",
            "18277. MGYP000848222406/10-596\n",
            "18278. MGYP000281497435/15-399\n",
            "18279. MGYP000799644813/4-574\n",
            "18280. MGYP000073976115/97-695\n",
            "18281. MGYP000824696544/14-606\n",
            "18282. MGYP001065364424/82-500\n",
            "18283. MGYP000369923193/2-394\n",
            "18284. MGYP000808143756/15-607\n",
            "18285. MGYP000854703387/6-610\n",
            "18286. MGYP001068975739/74-467\n",
            "18287. MGYP000061873832/11-611\n",
            "18288. MGYP000570492542/22-413\n",
            "18289. MGYP000303108779/15-607\n",
            "18290. MGYP000994714539/13-607\n",
            "18291. MGYP000235152520/11-604\n",
            "18292. MGYP000623938919/2-397\n",
            "18293. MGYP000126049507/2-595\n",
            "18294. MGYP001005606956/4-419\n",
            "18295. MGYP000332900738/12-609\n",
            "18296. MGYP001002246715/63-658\n",
            "18297. MGYP000294139791/22-617\n",
            "18298. MGYP001043893182/16-625\n",
            "18299. MGYP000611006385/11-622\n",
            "18300. MGYP000147348617/7-607\n",
            "18301. MGYP000064518525/5-388\n",
            "18302. MGYP000427156057/16-617\n",
            "18303. MGYP000636744020/2-393\n",
            "18304. MGYP001098711573/11-597\n",
            "18305. MGYP000097018941/2-383\n",
            "18306. MGYP000887312920/12-605\n",
            "18307. MGYP000334183569/71-667\n",
            "18308. MGYP000163095681/1-391\n",
            "18309. MGYP001111508758/5-596\n",
            "18310. MGYP000867279829/52-439\n",
            "18311. MGYP000383019672/4-590\n",
            "18312. MGYP000110483572/3-581\n",
            "18313. MGYP000204029761/2-396\n",
            "18314. MGYP000341654034/12-612\n",
            "18315. MGYP000414974797/19-617\n",
            "18316. MGYP000068802281/8-362\n",
            "18317. MGYP000638252453/5-424\n",
            "18318. MGYP000312262045/7-389\n",
            "18319. MGYP000435661839/45-645\n",
            "18320. MGYP001113374770/69-459\n",
            "18321. MGYP000420754284/4-372\n",
            "18322. MGYP000703512252/7-397\n",
            "18323. MGYP000886296753/11-605\n",
            "18324. MGYP001049596807/20-419\n",
            "18325. MGYP000372928449/14-416\n",
            "18326. MGYP000660811102/4-609\n",
            "18327. MGYP000849070379/8-608\n",
            "18328. MGYP001100222610/5-398\n",
            "18329. MGYP000744077442/14-406\n",
            "18330. MGYP000362498350/9-415\n",
            "18331. MGYP001060049226/7-410\n",
            "18332. MGYP000324457466/5-586\n",
            "18333. MGYP000856283722/4-478\n",
            "18334. MGYP000007452897/8-601\n",
            "18335. MGYP001128127962/15-391\n",
            "18336. MGYP001041210181/12-427\n",
            "18337. MGYP000144592606/2-397\n",
            "18338. MGYP000608194581/4-395\n",
            "18339. MGYP001111160130/11-614\n",
            "18340. MGYP001029509546/13-601\n",
            "18341. MGYP001039659560/5-398\n",
            "18342. MGYP000164899037/16-612\n",
            "18343. MGYP000205050595/13-425\n",
            "18344. MGYP000163785157/4-592\n",
            "18345. MGYP000443445054/7-606\n",
            "18346. MGYP001120387996/9-614\n",
            "18347. MGYP000305006330/6-629\n",
            "18348. MGYP000444789967/11-613\n",
            "18349. MGYP001127462642/15-607\n",
            "18350. MGYP000575178088/72-472\n",
            "18351. MGYP001006297473/24-614\n",
            "18352. MGYP000486661216/7-405\n",
            "18353. MGYP000631805741/2-396\n",
            "18354. MGYP000581836123/12-617\n",
            "18355. MGYP000830385426/15-607\n",
            "18356. MGYP000894277994/10-404\n",
            "18357. MGYP000493955521/13-427\n",
            "18358. MGYP000519284085/12-605\n",
            "18359. MGYP001058137608/7-595\n",
            "18360. MGYP001118401178/6-592\n",
            "18361. MGYP000185298468/10-610\n",
            "18362. MGYP000438568110/13-613\n",
            "18363. MGYP000459944821/9-607\n",
            "18364. MGYP000581100455/104-504\n",
            "18365. MGYP001114825746/16-412\n",
            "18366. MGYP000057298628/16-613\n",
            "18367. MGYP000821572158/15-607\n",
            "18368. MGYP000395500936/8-416\n",
            "18369. MGYP000073404048/2-383\n",
            "18370. MGYP000076487952/16-601\n",
            "18371. MGYP001031820960/10-605\n",
            "18372. MGYP001081580541/11-607\n",
            "18373. MGYP000959340052/7-604\n",
            "18374. MGYP000597985387/15-607\n",
            "18375. MGYP000084735792/13-607\n",
            "18376. MGYP000069761164/8-609\n",
            "18377. MGYP000064457467/9-609\n",
            "18378. MGYP000878077758/20-387\n",
            "18379. MGYP000418396958/12-647\n",
            "18380. MGYP001133419382/13-620\n",
            "18381. MGYP001104963714/11-609\n",
            "18382. MGYP000951468659/8-395\n",
            "18383. MGYP000196258186/4-611\n",
            "18384. MGYP000129321338/28-426\n",
            "18385. MGYP000216550223/19-612\n",
            "18386. MGYP000021208438/9-603\n",
            "18387. MGYP000991052855/11-338\n",
            "18388. MGYP000635313899/7-584\n",
            "18389. MGYP001127615561/6-594\n",
            "18390. MGYP000968991559/4-611\n",
            "18391. MGYP000150276741/64-475\n",
            "18392. MGYP000716162462/13-619\n",
            "18393. MGYP000479135190/11-407\n",
            "18394. MGYP000409407704/12-615\n",
            "18395. MGYP000998317571/17-609\n",
            "18396. MGYP000023420442/18-411\n",
            "18397. MGYP000345217887/54-450\n",
            "18398. MGYP000155289187/7-604\n",
            "18399. MGYP000975286276/39-642\n",
            "18400. MGYP000138121801/14-619\n",
            "18401. MGYP001129303635/296-890\n",
            "18402. MGYP000706869298/17-605\n",
            "18403. MGYP000193026042/13-612\n",
            "18404. MGYP000982658971/2-395\n",
            "18405. MGYP000417314460/12-488\n",
            "18406. MGYP000589403621/16-614\n",
            "18407. MGYP001055980669/11-609\n",
            "18408. MGYP000473095214/5-337\n",
            "18409. MGYP001129831846/5-601\n",
            "18410. MGYP000371076849/20-616\n",
            "18411. MGYP000275703988/15-351\n",
            "18412. MGYP001101446299/10-604\n",
            "18413. MGYP000084681966/7-423\n",
            "18414. MGYP001154347184/5-596\n",
            "18415. MGYP001125258728/5-613\n",
            "18416. MGYP000039691649/16-402\n",
            "18417. MGYP000136317405/5-379\n",
            "18418. MGYP000086327926/13-395\n",
            "18419. MGYP000869692983/6-437\n",
            "18420. MGYP001090573024/24-607\n",
            "18421. MGYP000188656613/12-650\n",
            "18422. MGYP000563086771/5-349\n",
            "18423. MGYP000011125934/9-414\n",
            "18424. MGYP000025941790/24-413\n",
            "18425. MGYP000925765677/5-608\n",
            "18426. MGYP000681697093/2-397\n",
            "18427. MGYP000814154550/3-336\n",
            "18428. MGYP000057540307/9-614\n",
            "18429. MGYP000738504159/40-644\n",
            "18430. MGYP000826334890/3-340\n",
            "18431. MGYP000857303687/17-412\n",
            "18432. MGYP000962007466/2-372\n",
            "18433. MGYP000506920369/1-580\n",
            "18434. MGYP000326841810/17-389\n",
            "18435. MGYP000100242880/13-592\n",
            "18436. MGYP000592787044/3-341\n",
            "18437. MGYP000849556609/4-396\n",
            "18438. MGYP000222443448/1-552\n",
            "18439. MGYP000787359285/11-355\n",
            "18440. MGYP000465141221/5-589\n",
            "18441. MGYP000981579006/36-625\n",
            "18442. MGYP000143070002/4-597\n",
            "18443. MGYP001049303540/11-590\n",
            "18444. MGYP000885542137/9-619\n",
            "18445. MGYP000935893190/3-379\n",
            "18446. MGYP000058579464/9-485\n",
            "18447. MGYP000856552745/10-401\n",
            "18448. MGYP000898979245/9-587\n",
            "18449. MGYP000880885131/19-609\n",
            "18450. MGYP001053723382/10-605\n",
            "18451. MGYP000524400691/5-384\n",
            "18452. MGYP001070582876/11-373\n",
            "18453. MGYP001089639447/6-395\n",
            "18454. MGYP000564193363/15-607\n",
            "18455. MGYP000749714222/49-646\n",
            "18456. MGYP000672241248/18-617\n",
            "18457. MGYP001082661969/2-393\n",
            "18458. MGYP000633165687/13-604\n",
            "18459. MGYP000125393695/3-407\n",
            "18460. MGYP000890043772/21-414\n",
            "18461. MGYP000516839491/7-606\n",
            "18462. MGYP000381729132/4-592\n",
            "18463. MGYP001038032427/5-414\n",
            "18464. MGYP000363640425/8-350\n",
            "18465. MGYP000531445907/11-604\n",
            "18466. MGYP000381772315/68-505\n",
            "18467. MGYP000159509834/7-603\n",
            "18468. MGYP000030060412/8-603\n",
            "18469. MGYP000979378137/6-595\n",
            "18470. MGYP000066545284/12-606\n",
            "18471. MGYP000766201061/2-397\n",
            "18472. MGYP000244813585/15-606\n",
            "18473. MGYP000875227907/7-603\n",
            "18474. MGYP000718092774/18-652\n",
            "18475. MGYP000032348410/8-611\n",
            "18476. MGYP000473514297/42-632\n",
            "18477. MGYP000358942378/6-593\n",
            "18478. MGYP000512301695/9-611\n",
            "18479. MGYP000949099577/26-453\n",
            "18480. MGYP000986512546/9-370\n",
            "18481. MGYP000899666430/18-623\n",
            "18482. MGYP000040974141/9-610\n",
            "18483. MGYP000877678140/11-604\n",
            "18484. MGYP000309305601/6-607\n",
            "18485. MGYP000113412588/4-592\n",
            "18486. MGYP000993115579/7-612\n",
            "18487. MGYP001027586625/24-614\n",
            "18488. MGYP000044286750/15-352\n",
            "18489. MGYP000171207010/20-618\n",
            "18490. MGYP000537440180/30-621\n",
            "18491. MGYP000375851075/13-609\n",
            "18492. MGYP000184933129/4-389\n",
            "18493. MGYP000613303004/10-618\n",
            "18494. MGYP000095110920/7-600\n",
            "18495. MGYP001000383598/17-616\n",
            "18496. MGYP000066689983/2-592\n",
            "18497. MGYP000977632482/4-401\n",
            "18498. MGYP000316030006/2-398\n",
            "18499. MGYP001066410304/11-401\n",
            "18500. MGYP000601185705/6-414\n",
            "18501. MGYP000218567213/3-452\n",
            "18502. MGYP000902562029/37-638\n",
            "18503. MGYP000046818610/8-612\n",
            "18504. MGYP000176228357/4-394\n",
            "18505. MGYP000896816786/10-569\n",
            "18506. MGYP000924658222/5-615\n",
            "18507. MGYP001110687692/8-389\n",
            "18508. MGYP000885746744/15-601\n",
            "18509. MGYP001056302338/7-395\n",
            "18510. MGYP000881327726/9-611\n",
            "18511. MGYP000843890981/13-609\n",
            "18512. MGYP000550901527/130-496\n",
            "18513. MGYP000153619257/75-483\n",
            "18514. MGYP000172821386/13-621\n",
            "18515. MGYP000377058967/12-387\n",
            "18516. MGYP000381659255/18-611\n",
            "18517. MGYP000867467274/17-619\n",
            "18518. MGYP000851227495/11-430\n",
            "18519. MGYP000509352508/12-602\n",
            "18520. MGYP000976630994/8-405\n",
            "18521. MGYP000454599239/23-617\n",
            "18522. MGYP000377052381/10-402\n",
            "18523. MGYP000872177088/6-596\n",
            "18524. MGYP000565403136/6-588\n",
            "18525. MGYP000447902773/26-619\n",
            "18526. MGYP000323948214/9-491\n",
            "18527. MGYP000218046204/220-824\n",
            "18528. MGYP000513370958/7-604\n",
            "18529. MGYP000097276592/13-612\n",
            "18530. MGYP001119562792/8-424\n",
            "18531. MGYP000726810905/12-405\n",
            "18532. MGYP001101765829/1-578\n",
            "18533. MGYP000641280319/14-607\n",
            "18534. MGYP000920421384/5-608\n",
            "18535. MGYP001089368052/41-649\n",
            "18536. MGYP000170093942/6-606\n",
            "18537. MGYP000877780744/9-613\n",
            "18538. MGYP000029412064/6-606\n",
            "18539. MGYP000379544402/25-436\n",
            "18540. MGYP000901833511/6-628\n",
            "18541. MGYP000167385754/4-420\n",
            "18542. MGYP001047939561/14-604\n",
            "18543. MGYP001105676592/9-602\n",
            "18544. MGYP000990340192/5-407\n",
            "18545. MGYP000158443959/14-605\n",
            "18546. MGYP000975950226/11-612\n",
            "18547. MGYP000383895572/13-608\n",
            "18548. MGYP000979455341/36-439\n",
            "18549. MGYP000200200902/13-615\n",
            "18550. MGYP000642222068/11-355\n",
            "18551. MGYP000085429376/10-609\n",
            "18552. MGYP000368565561/12-635\n",
            "18553. MGYP000138998132/7-401\n",
            "18554. MGYP000674381743/8-603\n",
            "18555. MGYP000712993096/8-624\n",
            "18556. MGYP000951496142/8-592\n",
            "18557. MGYP000210160933/8-350\n",
            "18558. MGYP001058559485/13-405\n",
            "18559. MGYP000477317027/5-404\n",
            "18560. MGYP001026128069/9-606\n",
            "18561. MGYP000568360727/11-647\n",
            "18562. MGYP000109137684/2-397\n",
            "18563. MGYP000163017628/4-378\n",
            "18564. MGYP000376030874/7-605\n",
            "18565. MGYP000312225676/10-605\n",
            "18566. MGYP000632506579/18-399\n",
            "18567. MGYP000013189717/1-382\n",
            "18568. MGYP001029133624/34-636\n",
            "18569. MGYP000106225617/24-605\n",
            "18570. MGYP001022550431/8-406\n",
            "18571. MGYP000194143884/6-397\n",
            "18572. MGYP000058608710/4-598\n",
            "18573. MGYP000847933388/63-501\n",
            "18574. MGYP000022918330/3-607\n",
            "18575. MGYP000158275118/14-426\n",
            "18576. MGYP000944896772/8-609\n",
            "18577. MGYP000288628250/6-601\n",
            "18578. MGYP000171465750/15-407\n",
            "18579. MGYP000556049693/12-612\n",
            "18580. MGYP000028695256/8-604\n",
            "18581. MGYP000768123002/12-604\n",
            "18582. MGYP000059800467/12-430\n",
            "18583. MGYP000874023546/9-601\n",
            "18584. MGYP000918905853/6-596\n",
            "18585. MGYP000386207994/11-648\n",
            "18586. MGYP000859596823/6-609\n",
            "18587. MGYP000148129280/13-607\n",
            "18588. MGYP000197242669/12-612\n",
            "18589. MGYP001019640676/12-608\n",
            "18590. MGYP000849815950/8-607\n",
            "18591. MGYP000849546870/4-411\n",
            "18592. MGYP000236291483/7-406\n",
            "18593. MGYP001046628799/7-400\n",
            "18594. MGYP001102383173/9-611\n",
            "18595. MGYP000955538986/12-404\n",
            "18596. MGYP000988596579/10-345\n",
            "18597. MGYP000009609411/13-409\n",
            "18598. MGYP000204134848/11-606\n",
            "18599. MGYP001036939043/10-448\n",
            "18600. MGYP000647108840/12-617\n",
            "18601. MGYP000825457931/31-394\n",
            "18602. MGYP000296635220/12-608\n",
            "18603. MGYP000768202695/4-414\n",
            "18604. MGYP001044707276/69-501\n",
            "18605. MGYP000118722251/3-394\n",
            "18606. MGYP000302125481/10-600\n",
            "18607. MGYP000287792514/11-604\n",
            "18608. MGYP000021332580/7-613\n",
            "18609. MGYP000088385790/10-609\n",
            "18610. MGYP000680449364/12-418\n",
            "18611. MGYP001058588885/6-591\n",
            "18612. MGYP001118298753/11-609\n",
            "18613. MGYP000255540804/6-600\n",
            "18614. MGYP000085152418/13-614\n",
            "18615. MGYP000306600002/28-408\n",
            "18616. MGYP000775245676/3-335\n",
            "18617. MGYP001007988766/2-597\n",
            "18618. MGYP001091728117/29-427\n",
            "18619. MGYP000991411697/10-603\n",
            "18620. MGYP001045954581/6-610\n",
            "18621. MGYP000975805730/11-355\n",
            "18622. MGYP000859687435/7-611\n",
            "18623. MGYP000468537897/12-609\n",
            "18624. MGYP000874456478/9-396\n",
            "18625. MGYP000497390704/5-597\n",
            "18626. MGYP000802972685/12-617\n",
            "18627. MGYP001027760837/21-360\n",
            "18628. MGYP001059043168/9-406\n",
            "18629. MGYP000336242381/254-651\n",
            "18630. MGYP000103787326/8-349\n",
            "18631. MGYP000127714421/3-601\n",
            "18632. MGYP000488242567/12-392\n",
            "18633. MGYP000150141497/12-424\n",
            "18634. MGYP000686293680/7-418\n",
            "18635. MGYP000881975707/9-400\n",
            "18636. MGYP000874219232/6-411\n",
            "18637. MGYP000487863591/11-407\n",
            "18638. MGYP000715177410/8-399\n",
            "18639. MGYP000845654953/7-388\n",
            "18640. MGYP000866144394/8-406\n",
            "18641. MGYP001026581040/10-607\n",
            "18642. MGYP000817718471/3-339\n",
            "18643. MGYP000023846437/15-607\n",
            "18644. MGYP000149180438/15-608\n",
            "18645. MGYP000400557165/4-610\n",
            "18646. MGYP000797411625/4-343\n",
            "18647. MGYP000094248302/10-600\n",
            "18648. MGYP000049726013/9-388\n",
            "18649. MGYP000875544997/11-354\n",
            "18650. MGYP001086835892/2-415\n",
            "18651. MGYP001090448699/11-591\n",
            "18652. MGYP000742322545/9-603\n",
            "18653. MGYP001089758652/2-582\n",
            "18654. MGYP001047920766/2-394\n",
            "18655. MGYP000026454491/22-608\n",
            "18656. MGYP000076507818/12-609\n",
            "18657. MGYP000293296624/40-637\n",
            "18658. MGYP000684217603/7-389\n",
            "18659. MGYP000845580815/8-584\n",
            "18660. MGYP001038875249/5-597\n",
            "18661. MGYP000965626529/10-605\n",
            "18662. MGYP000454237727/4-591\n",
            "18663. MGYP000115236989/4-592\n",
            "18664. MGYP000446079750/14-604\n",
            "18665. MGYP000467551132/12-417\n",
            "18666. MGYP000957567928/30-616\n",
            "18667. MGYP000673834681/4-335\n",
            "18668. MGYP000755259169/2-397\n",
            "18669. MGYP001066858813/11-588\n",
            "18670. MGYP000925710046/5-408\n",
            "18671. MGYP000870450015/3-604\n",
            "18672. MGYP000993591222/6-587\n",
            "18673. MGYP001150122100/6-604\n",
            "18674. MGYP000114912230/54-641\n",
            "18675. MGYP000122255381/7-612\n",
            "18676. MGYP000998747494/11-429\n",
            "18677. MGYP000061469058/13-605\n",
            "18678. MGYP000041444049/52-448\n",
            "18679. MGYP000038590430/11-604\n",
            "18680. MGYP001087924121/188-810\n",
            "18681. MGYP001008435523/11-411\n",
            "18682. MGYP000331610646/11-407\n",
            "18683. MGYP000275645083/15-442\n",
            "18684. MGYP000498287602/8-604\n",
            "18685. MGYP001031492334/13-607\n",
            "18686. MGYP000908153359/6-609\n",
            "18687. MGYP001031451629/3-589\n",
            "18688. MGYP000713801294/12-603\n",
            "18689. MGYP000155994447/5-397\n",
            "18690. MGYP000414933135/15-396\n",
            "18691. MGYP000335003787/10-409\n",
            "18692. MGYP000685641993/4-411\n",
            "18693. MGYP001058629482/13-427\n",
            "18694. MGYP001113424312/6-616\n",
            "18695. MGYP000949875582/5-376\n",
            "18696. MGYP000636179461/9-414\n",
            "18697. MGYP000107743742/2-395\n",
            "18698. MGYP000220399746/11-412\n",
            "18699. MGYP000951237025/5-613\n",
            "18700. MGYP000674951445/8-397\n",
            "18701. MGYP001099738968/13-606\n",
            "18702. MGYP000906850439/4-612\n",
            "18703. MGYP000414774652/7-596\n",
            "18704. MGYP000015460325/22-421\n",
            "18705. MGYP001088385838/10-349\n",
            "18706. MGYP000492969863/4-392\n",
            "18707. MGYP000374655991/5-603\n",
            "18708. MGYP000772030790/17-406\n",
            "18709. MGYP000263206159/14-413\n",
            "18710. MGYP000182033425/5-602\n",
            "18711. MGYP000014952562/10-595\n",
            "18712. MGYP001044759646/5-597\n",
            "18713. MGYP000175263352/10-614\n",
            "18714. MGYP000004270598/2-396\n",
            "18715. MGYP000396409669/4-585\n",
            "18716. MGYP000186567896/12-604\n",
            "18717. MGYP000593985221/2-588\n",
            "18718. MGYP000081200376/12-610\n",
            "18719. MGYP000498952438/14-606\n",
            "18720. MGYP001080195939/7-405\n",
            "18721. MGYP000067354079/15-596\n",
            "18722. MGYP000878516337/6-597\n",
            "18723. MGYP000934164392/40-634\n",
            "18724. MGYP000589719138/19-609\n",
            "18725. MGYP000473160920/12-608\n",
            "18726. MGYP000905522630/8-399\n",
            "18727. MGYP000766016509/4-394\n",
            "18728. MGYP000827831690/16-382\n",
            "18729. MGYP000243180361/10-396\n",
            "18730. MGYP000207316222/4-369\n",
            "18731. MGYP000462469016/14-593\n",
            "18732. MGYP000012058233/8-398\n",
            "18733. MGYP000866199369/6-405\n",
            "18734. MGYP000034756972/4-394\n",
            "18735. MGYP001091085083/8-605\n",
            "18736. MGYP000660415060/8-399\n",
            "18737. MGYP001044452883/11-406\n",
            "18738. MGYP000657877701/15-403\n",
            "18739. MGYP000666999224/7-614\n",
            "18740. MGYP000359606766/55-650\n",
            "18741. MGYP000981479391/9-607\n",
            "18742. MGYP000689034368/24-612\n",
            "18743. MGYP001027093722/14-609\n",
            "18744. MGYP000300661201/18-417\n",
            "18745. MGYP000002300003/13-607\n",
            "18746. MGYP000072134286/12-608\n",
            "18747. MGYP000841333073/42-635\n",
            "18748. MGYP000046709864/7-614\n",
            "18749. MGYP000544686485/10-604\n",
            "18750. MGYP001051350911/93-693\n",
            "18751. MGYP000160752471/12-612\n",
            "18752. MGYP000547994724/9-611\n",
            "18753. MGYP001131894229/7-589\n",
            "18754. MGYP001040786415/8-600\n",
            "18755. MGYP001091626447/20-631\n",
            "18756. MGYP000305717498/26-404\n",
            "18757. MGYP000617033048/22-421\n",
            "18758. MGYP001002212459/15-403\n",
            "18759. MGYP000205714950/7-614\n",
            "18760. MGYP000290739665/7-599\n",
            "18761. MGYP000143022944/6-376\n",
            "18762. MGYP000427494884/16-392\n",
            "18763. MGYP000848436482/7-587\n",
            "18764. MGYP000989289124/6-594\n",
            "18765. MGYP000800206201/11-605\n",
            "18766. MGYP000793073719/16-609\n",
            "18767. MGYP000893872042/20-622\n",
            "18768. MGYP000849781261/5-390\n",
            "18769. MGYP000989620380/2-602\n",
            "18770. MGYP001083827114/13-640\n",
            "18771. MGYP000306591903/6-594\n",
            "18772. MGYP001040489561/13-421\n",
            "18773. MGYP001087587061/6-611\n",
            "18774. MGYP000482153389/6-615\n",
            "18775. MGYP000326689891/4-530\n",
            "18776. MGYP000187093094/12-599\n",
            "18777. MGYP000456906691/9-608\n",
            "18778. MGYP000898201276/13-612\n",
            "18779. MGYP000883583345/56-650\n",
            "18780. MGYP000335881235/2-605\n",
            "18781. MGYP000327887494/1-350\n",
            "18782. MGYP000852861762/10-431\n",
            "18783. MGYP000139429885/4-591\n",
            "18784. MGYP001050574698/13-603\n",
            "18785. MGYP000576681534/3-603\n",
            "18786. MGYP001062389117/9-604\n",
            "18787. MGYP000800851098/31-632\n",
            "18788. MGYP000646048004/10-591\n",
            "18789. MGYP001134180616/2-356\n",
            "18790. MGYP001047195812/12-612\n",
            "18791. MGYP000001358696/19-606\n",
            "18792. MGYP000909815749/9-620\n",
            "18793. MGYP000249093151/6-610\n",
            "18794. MGYP000518696223/25-400\n",
            "18795. MGYP000058367349/17-356\n",
            "18796. MGYP001052805084/15-400\n",
            "18797. MGYP000320581187/12-381\n",
            "18798. MGYP000976148774/52-490\n",
            "18799. MGYP000565333987/13-417\n",
            "18800. MGYP000312182613/8-604\n",
            "18801. MGYP000650792391/10-351\n",
            "18802. MGYP001045585585/9-612\n",
            "18803. MGYP000718683165/10-631\n",
            "18804. MGYP000113612286/8-400\n",
            "18805. MGYP000191257487/9-387\n",
            "18806. MGYP000956170670/5-609\n",
            "18807. MGYP000073202597/19-90\n",
            "18808. MGYP000073202597/118-585\n",
            "18809. MGYP000675751648/13-487\n",
            "18810. MGYP000720451527/21-613\n",
            "18811. MGYP000038244894/10-611\n",
            "18812. MGYP000581678003/13-604\n",
            "18813. MGYP000868812133/18-612\n",
            "18814. MGYP000480596997/3-345\n",
            "18815. MGYP000608853973/115-499\n",
            "18816. MGYP000721146480/9-394\n",
            "18817. MGYP000484264089/7-614\n",
            "18818. MGYP000407323804/11-604\n",
            "18819. MGYP000875925594/6-600\n",
            "18820. MGYP000462121734/14-608\n",
            "18821. MGYP000935014063/2-574\n",
            "18822. MGYP000628619842/16-608\n",
            "18823. MGYP000859909534/7-415\n",
            "18824. MGYP001029069207/12-615\n",
            "18825. MGYP000583941420/254-651\n",
            "18826. MGYP000137832327/3-587\n",
            "18827. MGYP000915913803/12-613\n",
            "18828. MGYP000243386429/6-602\n",
            "18829. MGYP000905329473/11-605\n",
            "18830. MGYP000987229801/12-614\n",
            "18831. MGYP001014338085/22-612\n",
            "18832. MGYP000948361947/9-417\n",
            "18833. MGYP001073185162/12-407\n",
            "18834. MGYP000221092252/4-591\n",
            "18835. MGYP000630887764/7-387\n",
            "18836. MGYP000760772184/26-432\n",
            "18837. MGYP001127987490/11-605\n",
            "18838. MGYP000984014242/11-380\n",
            "18839. MGYP000248486707/11-406\n",
            "18840. MGYP000326744220/7-402\n",
            "18841. MGYP000311098674/31-624\n",
            "18842. MGYP000105882795/7-615\n",
            "18843. MGYP001152549911/4-387\n",
            "18844. MGYP000931730855/5-594\n",
            "18845. MGYP001061680036/14-417\n",
            "18846. MGYP000157142617/12-406\n",
            "18847. MGYP001015168905/5-386\n",
            "18848. MGYP000580424370/12-432\n",
            "18849. MGYP000789587583/15-404\n",
            "18850. MGYP000035811171/6-412\n",
            "18851. MGYP000315166861/6-381\n",
            "18852. MGYP000956822139/6-609\n",
            "18853. MGYP000123786509/16-401\n",
            "18854. MGYP000124134349/9-604\n",
            "18855. MGYP000547636746/12-399\n",
            "18856. MGYP000910914905/8-602\n",
            "18857. MGYP000051800788/12-591\n",
            "18858. MGYP000443598591/12-592\n",
            "18859. MGYP000211387657/12-420\n",
            "18860. MGYP000711369535/14-443\n",
            "18861. MGYP000997385096/5-629\n",
            "18862. MGYP001135103752/6-604\n",
            "18863. MGYP000428945545/73-684\n",
            "18864. MGYP000572639786/15-403\n",
            "18865. MGYP000650725878/33-364\n",
            "18866. MGYP000689185562/10-631\n",
            "18867. MGYP000241325210/9-403\n",
            "18868. MGYP000432201103/10-607\n",
            "18869. MGYP000152086293/8-604\n",
            "18870. MGYP000090613844/11-384\n",
            "18871. MGYP000754836652/17-601\n",
            "18872. MGYP001035490872/12-604\n",
            "18873. MGYP000775040468/14-396\n",
            "18874. MGYP000240356655/13-619\n",
            "18875. MGYP000680901558/6-405\n",
            "18876. MGYP000515223827/6-602\n",
            "18877. MGYP000628016571/13-619\n",
            "18878. MGYP000099465550/16-602\n",
            "18879. MGYP000418641364/13-604\n",
            "18880. MGYP000138185084/8-603\n",
            "18881. MGYP001033254751/10-423\n",
            "18882. MGYP000491602190/13-433\n",
            "18883. MGYP000028243372/4-599\n",
            "18884. MGYP000644816308/10-614\n",
            "18885. MGYP001036498325/21-395\n",
            "18886. MGYP000338710302/103-705\n",
            "18887. MGYP001103098567/75-675\n",
            "18888. MGYP000595185618/227-633\n",
            "18889. MGYP000258655379/19-495\n",
            "18890. MGYP000048108463/14-407\n",
            "18891. MGYP000032472625/4-614\n",
            "18892. MGYP000040809701/10-608\n",
            "18893. MGYP000012377233/5-386\n",
            "18894. MGYP000427373074/23-411\n",
            "18895. MGYP000794196939/11-404\n",
            "18896. MGYP000849935459/6-606\n",
            "18897. MGYP000132103243/12-407\n",
            "18898. MGYP000050018900/23-607\n",
            "18899. MGYP000307680434/4-341\n",
            "18900. MGYP000753071799/11-406\n",
            "18901. MGYP000339638156/7-614\n",
            "18902. MGYP000937703429/11-604\n",
            "18903. MGYP000410034108/9-397\n",
            "18904. MGYP000526791425/4-594\n",
            "18905. MGYP000411176315/3-407\n",
            "18906. MGYP001010652915/39-432\n",
            "18907. MGYP000393922205/25-427\n",
            "18908. MGYP000232315318/14-418\n",
            "18909. MGYP000848597180/19-615\n",
            "18910. MGYP001095839226/43-638\n",
            "18911. MGYP001062762311/10-396\n",
            "18912. MGYP000845800613/20-618\n",
            "18913. MGYP000073209401/47-461\n",
            "18914. MGYP000583480839/10-384\n",
            "18915. MGYP000130822018/12-413\n",
            "18916. MGYP001143673601/8-598\n",
            "18917. MGYP000885975070/15-607\n",
            "18918. MGYP000793128706/17-352\n",
            "18919. MGYP000153058591/31-447\n",
            "18920. MGYP000434387443/11-604\n",
            "18921. MGYP000319249404/11-605\n",
            "18922. MGYP000240483537/9-583\n",
            "18923. MGYP000189434206/14-604\n",
            "18924. MGYP001153638085/8-427\n",
            "18925. MGYP000542449824/4-346\n",
            "18926. MGYP001020536275/6-419\n",
            "18927. MGYP000273592819/4-376\n",
            "18928. MGYP000672239068/14-411\n",
            "18929. MGYP001096767264/5-608\n",
            "18930. MGYP000189258355/21-446\n",
            "18931. MGYP000093522162/14-613\n",
            "18932. MGYP000244508989/13-414\n",
            "18933. MGYP000884344097/229-826\n",
            "18934. MGYP000147120617/9-602\n",
            "18935. MGYP000733967587/8-401\n",
            "18936. MGYP000995662163/11-409\n",
            "18937. MGYP000014921070/7-387\n",
            "18938. MGYP000968614861/6-586\n",
            "18939. MGYP000200838818/2-55\n",
            "18940. MGYP000200838818/56-681\n",
            "18941. MGYP000076656447/29-627\n",
            "18942. MGYP000924418143/33-427\n",
            "18943. MGYP000935665213/5-591\n",
            "18944. MGYP000002556486/23-422\n",
            "18945. MGYP000694007842/4-369\n",
            "18946. MGYP000870006850/8-610\n",
            "18947. MGYP000783061073/11-417\n",
            "18948. MGYP000317872825/13-606\n",
            "18949. MGYP000131035636/2-408\n",
            "18950. MGYP000845641440/7-600\n",
            "18951. MGYP000802964986/23-619\n",
            "18952. MGYP000194298271/12-644\n",
            "18953. MGYP000549832621/15-406\n",
            "18954. MGYP000050497639/15-402\n",
            "18955. MGYP001063602580/13-352\n",
            "18956. MGYP000268328626/6-614\n",
            "18957. MGYP000875943688/4-605\n",
            "18958. MGYP000926406019/3-609\n",
            "18959. MGYP001107489217/5-343\n",
            "18960. MGYP000229946900/13-395\n",
            "18961. MGYP000073329914/12-340\n",
            "18962. MGYP000409557134/10-357\n",
            "18963. MGYP001036982132/16-607\n",
            "18964. MGYP000782352768/10-399\n",
            "18965. MGYP001035623966/17-606\n",
            "18966. MGYP000832922124/13-418\n",
            "18967. MGYP000562437263/25-442\n",
            "18968. MGYP001139956023/14-406\n",
            "18969. MGYP000024276979/5-615\n",
            "18970. MGYP000175667888/8-606\n",
            "18971. MGYP000921139617/10-401\n",
            "18972. MGYP000631707121/14-406\n",
            "18973. MGYP000104989303/12-610\n",
            "18974. MGYP000575094546/14-414\n",
            "18975. MGYP001024754983/13-434\n",
            "18976. MGYP000073348342/9-600\n",
            "18977. MGYP000977667264/7-600\n",
            "18978. MGYP000817888980/62-656\n",
            "18979. MGYP001153803098/41-472\n",
            "18980. MGYP000183345036/6-382\n",
            "18981. MGYP000224034248/6-600\n",
            "18982. MGYP000788387712/5-596\n",
            "18983. MGYP000291125634/11-605\n",
            "18984. MGYP000577486316/10-423\n",
            "18985. MGYP000939936872/12-613\n",
            "18986. MGYP000503400532/7-592\n",
            "18987. MGYP000037495298/4-594\n",
            "18988. MGYP000459692936/9-335\n",
            "18989. MGYP000435415485/11-418\n",
            "18990. MGYP000301591007/8-350\n",
            "18991. MGYP000014771135/11-604\n",
            "18992. MGYP000722613232/14-607\n",
            "18993. MGYP001101808091/8-501\n",
            "18994. MGYP000928155167/9-397\n",
            "18995. MGYP000015044076/6-593\n",
            "18996. MGYP000253275956/6-609\n",
            "18997. MGYP000975983302/9-403\n",
            "18998. MGYP001110147059/8-411\n",
            "18999. MGYP000862135144/6-404\n",
            "19000. MGYP000820443501/10-596\n",
            "19001. MGYP000739546009/14-602\n",
            "19002. MGYP001079842560/28-422\n",
            "19003. MGYP000848452930/207-582\n",
            "19004. MGYP000389159701/3-587\n",
            "19005. MGYP000846012724/6-593\n",
            "19006. MGYP000956611523/11-603\n",
            "19007. MGYP000046380598/14-392\n",
            "19008. MGYP001064204080/6-595\n",
            "19009. MGYP001139093973/9-611\n",
            "19010. MGYP000594880302/14-625\n",
            "19011. MGYP000986091455/11-604\n",
            "19012. MGYP000981784867/10-611\n",
            "19013. MGYP000442374538/11-605\n",
            "19014. MGYP000479825160/6-603\n",
            "19015. MGYP000089783608/8-607\n",
            "19016. MGYP001055399357/12-611\n",
            "19017. MGYP000848049392/14-418\n",
            "19018. MGYP000930864798/12-418\n",
            "19019. MGYP000916330377/12-620\n",
            "19020. MGYP001112559069/12-606\n",
            "19021. MGYP001047678466/9-364\n",
            "19022. MGYP001028479218/60-400\n",
            "19023. MGYP001028479218/499-679\n",
            "19024. MGYP000271636531/6-588\n",
            "19025. MGYP000487937459/2-398\n",
            "19026. MGYP000400914528/9-348\n",
            "19027. MGYP000852225441/6-606\n",
            "19028. MGYP000683645781/1-563\n",
            "19029. MGYP000758846918/1-333\n",
            "19030. MGYP000865791749/40-417\n",
            "19031. MGYP000169457260/12-334\n",
            "19032. MGYP000504107938/1-563\n",
            "19033. MGYP000928645422/3-611\n",
            "19034. MGYP000030006104/12-605\n",
            "19035. MGYP000191198458/10-592\n",
            "19036. MGYP001039639895/15-407\n",
            "19037. MGYP000098359796/10-408\n",
            "19038. MGYP000206332851/7-613\n",
            "19039. MGYP000170952668/13-612\n",
            "19040. MGYP000476444049/276-885\n",
            "19041. MGYP000433730614/16-426\n",
            "19042. MGYP000704694393/1-312\n",
            "19043. MGYP000994477224/3-595\n",
            "19044. MGYP000365926534/4-394\n",
            "19045. MGYP001082985173/14-604\n",
            "19046. MGYP000091024410/12-615\n",
            "19047. MGYP001000430967/11-404\n",
            "19048. MGYP001058071899/15-611\n",
            "19049. MGYP000795388568/26-623\n",
            "19050. MGYP000539147881/11-596\n",
            "19051. MGYP000219138988/1-378\n",
            "19052. MGYP001066089352/5-605\n",
            "19053. MGYP001024827789/6-382\n",
            "19054. MGYP000622121920/17-616\n",
            "19055. MGYP000565690378/30-628\n",
            "19056. MGYP001088203141/7-407\n",
            "19057. MGYP000462120269/5-605\n",
            "19058. MGYP000528773246/57-405\n",
            "19059. MGYP000515553903/7-587\n",
            "19060. MGYP001024747144/7-614\n",
            "19061. MGYP000364915334/2-363\n",
            "19062. MGYP000911570793/7-584\n",
            "19063. MGYP000947962391/6-587\n",
            "19064. MGYP000719897530/36-446\n",
            "19065. MGYP000846684528/9-604\n",
            "19066. MGYP000046818337/11-613\n",
            "19067. MGYP000040658478/14-398\n",
            "19068. MGYP000743255566/19-619\n",
            "19069. MGYP000652520605/6-629\n",
            "19070. MGYP000279396861/21-616\n",
            "19071. MGYP000149161226/81-449\n",
            "19072. MGYP000435800573/11-604\n",
            "19073. MGYP000394544918/15-439\n",
            "19074. MGYP000684348582/12-603\n",
            "19075. MGYP001079756112/10-605\n",
            "19076. MGYP000048102870/14-406\n",
            "19077. MGYP000399814203/12-412\n",
            "19078. MGYP000914151851/11-406\n",
            "19079. MGYP001083218013/4-607\n",
            "19080. MGYP000969262111/10-613\n",
            "19081. MGYP001026663083/13-601\n",
            "19082. MGYP000403310618/2-344\n",
            "19083. MGYP000035333808/13-357\n",
            "19084. MGYP000865793128/6-611\n",
            "19085. MGYP000959416718/6-354\n",
            "19086. MGYP001134452883/12-617\n",
            "19087. MGYP000742090690/13-440\n",
            "19088. MGYP000667491699/1-362\n",
            "19089. MGYP000218591601/4-423\n",
            "19090. MGYP000741500008/11-593\n",
            "19091. MGYP000555724525/5-595\n",
            "19092. MGYP000200450466/14-398\n",
            "19093. MGYP000793716060/12-604\n",
            "19094. MGYP000108452513/17-616\n",
            "19095. MGYP000580841654/11-347\n",
            "19096. MGYP001066674962/8-613\n",
            "19097. MGYP000435825738/4-432\n",
            "19098. MGYP000993891582/44-439\n",
            "19099. MGYP000964906456/9-614\n",
            "19100. MGYP001005923348/25-388\n",
            "19101. MGYP001068405727/12-431\n",
            "19102. MGYP000836208613/8-397\n",
            "19103. MGYP000044648344/4-511\n",
            "19104. MGYP000554259349/15-353\n",
            "19105. MGYP000901211965/46-429\n",
            "19106. MGYP000388396451/7-591\n",
            "19107. MGYP000891919874/8-593\n",
            "19108. MGYP000831408255/8-414\n",
            "19109. MGYP000831135585/12-418\n",
            "19110. MGYP000021663602/5-586\n",
            "19111. MGYP000897820379/10-601\n",
            "19112. MGYP000797094100/16-610\n",
            "19113. MGYP000695856122/9-616\n",
            "19114. MGYP000117536181/10-424\n",
            "19115. MGYP000222773128/24-617\n",
            "19116. MGYP000716435846/7-599\n",
            "19117. MGYP000255948812/9-398\n",
            "19118. MGYP000224097426/14-404\n",
            "19119. MGYP000732061940/11-418\n",
            "19120. MGYP000583095346/6-595\n",
            "19121. MGYP000362621952/11-607\n",
            "19122. MGYP000623187524/9-610\n",
            "19123. MGYP000103391345/13-409\n",
            "19124. MGYP000357200658/9-613\n",
            "19125. MGYP000709930550/7-410\n",
            "19126. MGYP000357388565/9-595\n",
            "19127. MGYP000026480858/1-576\n",
            "19128. MGYP000797763969/3-406\n",
            "19129. MGYP001082317283/14-651\n",
            "19130. MGYP000382396306/4-421\n",
            "19131. MGYP000874453666/80-479\n",
            "19132. MGYP000770849236/3-329\n",
            "19133. MGYP000755676573/8-401\n",
            "19134. MGYP000189797141/2-340\n",
            "19135. MGYP000607131386/13-429\n",
            "19136. MGYP000545481948/9-675\n",
            "19137. MGYP000053158905/11-604\n",
            "19138. MGYP000350063846/12-614\n",
            "19139. MGYP001149352320/23-609\n",
            "19140. MGYP000385465736/11-406\n",
            "19141. MGYP000482716629/13-426\n",
            "19142. MGYP000846963285/5-606\n",
            "19143. MGYP000052647203/11-622\n",
            "19144. MGYP000361777741/14-418\n",
            "19145. MGYP000500927563/10-610\n",
            "19146. MGYP000158736438/13-607\n",
            "19147. MGYP000468872635/264-861\n",
            "19148. MGYP000474237201/12-406\n",
            "19149. MGYP001002421203/13-424\n",
            "19150. MGYP001030050033/6-589\n",
            "19151. MGYP000718144294/1-363\n",
            "19152. MGYP000580941567/15-606\n",
            "19153. MGYP001110190767/13-610\n",
            "19154. MGYP000415614872/94-693\n",
            "19155. MGYP000608451490/13-413\n",
            "19156. MGYP000307282525/2-397\n",
            "19157. MGYP000510047332/10-400\n",
            "19158. MGYP000798791590/3-330\n",
            "19159. MGYP001089645094/12-412\n",
            "19160. MGYP000802634437/6-599\n",
            "19161. MGYP000962521310/6-627\n",
            "19162. MGYP000013495975/17-616\n",
            "19163. MGYP000057262360/40-419\n",
            "19164. MGYP000159581347/11-604\n",
            "19165. MGYP000983815573/14-608\n",
            "19166. MGYP000845496180/25-138\n",
            "19167. MGYP000845496180/188-475\n",
            "19168. MGYP000275921013/13-609\n",
            "19169. MGYP000328228339/2-406\n",
            "19170. MGYP001052692895/3-331\n",
            "19171. MGYP000352809193/14-407\n",
            "19172. MGYP000059787321/34-443\n",
            "19173. MGYP000043469244/13-610\n",
            "19174. MGYP000029416477/14-608\n",
            "19175. MGYP000061090131/2-554\n",
            "19176. MGYP001031145844/10-401\n",
            "19177. MGYP000231758686/13-601\n",
            "19178. MGYP000188580641/12-351\n",
            "19179. MGYP000768294429/6-619\n",
            "19180. MGYP001037508448/7-606\n",
            "19181. MGYP001007718868/1-563\n",
            "19182. MGYP000720820449/13-406\n",
            "19183. MGYP000317856235/12-405\n",
            "19184. MGYP001117907806/11-610\n",
            "19185. MGYP001021122785/30-425\n",
            "19186. MGYP001052355900/198-819\n",
            "19187. MGYP000105849150/14-613\n",
            "19188. MGYP000710755708/3-618\n",
            "19189. MGYP000158463297/24-401\n",
            "19190. MGYP000724118634/15-597\n",
            "19191. MGYP000961379987/6-602\n",
            "19192. MGYP000807313664/13-405\n",
            "19193. MGYP000490870582/90-720\n",
            "19194. MGYP000710132077/12-407\n",
            "19195. MGYP000250054703/4-584\n",
            "19196. MGYP000704277946/26-408\n",
            "19197. MGYP001030695548/85-680\n",
            "19198. MGYP000043797558/2-604\n",
            "19199. MGYP000161245870/13-413\n",
            "19200. MGYP000563894277/11-609\n",
            "19201. MGYP000339444317/6-586\n",
            "19202. MGYP000987265108/28-626\n",
            "19203. MGYP000487102310/4-419\n",
            "19204. MGYP000649020597/11-602\n",
            "19205. MGYP000627579226/12-418\n",
            "19206. MGYP001154513291/12-360\n",
            "19207. MGYP000599237213/15-406\n",
            "19208. MGYP000691496088/17-440\n",
            "19209. MGYP000549156966/11-418\n",
            "19210. MGYP000244428744/17-614\n",
            "19211. MGYP000603592565/174-712\n",
            "19212. MGYP000864248667/15-406\n",
            "19213. MGYP000754606094/7-607\n",
            "19214. MGYP000343338071/4-585\n",
            "19215. MGYP000016109262/13-610\n",
            "19216. MGYP000866483214/11-613\n",
            "19217. MGYP000493625690/6-613\n",
            "19218. MGYP000816511715/11-404\n",
            "19219. MGYP001071773227/378-971\n",
            "19220. MGYP000986801732/5-390\n",
            "19221. MGYP001030290698/11-604\n",
            "19222. MGYP000480335001/4-430\n",
            "19223. MGYP000624353837/11-596\n",
            "19224. MGYP000103170387/10-593\n",
            "19225. MGYP000906025958/5-594\n",
            "19226. MGYP000242804788/3-609\n",
            "19227. MGYP000791720866/13-418\n",
            "19228. MGYP000613026618/15-403\n",
            "19229. MGYP000657517677/17-606\n",
            "19230. MGYP000214234901/7-617\n",
            "19231. MGYP000279671689/16-616\n",
            "19232. MGYP000059392996/5-600\n",
            "19233. MGYP000272815939/11-604\n",
            "19234. MGYP001017868140/24-629\n",
            "19235. MGYP000329621456/4-396\n",
            "19236. MGYP000089390524/11-604\n",
            "19237. MGYP001106542247/1-356\n",
            "19238. MGYP000838694865/12-417\n",
            "19239. MGYP000607099665/2-567\n",
            "19240. MGYP000962536765/9-607\n",
            "19241. MGYP000024658390/3-604\n",
            "19242. MGYP000473089368/6-393\n",
            "19243. MGYP000473089368/546-600\n",
            "19244. MGYP000958063870/14-459\n",
            "19245. MGYP000080023528/5-594\n",
            "19246. MGYP001049888457/9-602\n",
            "19247. MGYP000056544141/8-604\n",
            "19248. MGYP001097161083/6-589\n",
            "19249. MGYP001026327840/11-606\n",
            "19250. MGYP000014460443/11-612\n",
            "19251. MGYP000543531778/11-410\n",
            "19252. MGYP000515881849/34-628\n",
            "19253. MGYP000341710749/28-613\n",
            "19254. MGYP000571706700/18-609\n",
            "19255. MGYP000444609239/6-604\n",
            "19256. MGYP000215047197/3-340\n",
            "19257. MGYP000252756599/11-605\n",
            "19258. MGYP001116406680/8-588\n",
            "19259. MGYP000296665525/14-607\n",
            "19260. MGYP000464323876/13-599\n",
            "19261. MGYP000424562793/15-600\n",
            "19262. MGYP000933544767/8-422\n",
            "19263. MGYP000092206753/9-599\n",
            "19264. MGYP000070723235/7-421\n",
            "19265. MGYP000438099276/14-612\n",
            "19266. MGYP000574541371/68-477\n",
            "19267. MGYP000202236786/40-419\n",
            "19268. MGYP000168894866/6-618\n",
            "19269. MGYP001043095592/18-611\n",
            "19270. MGYP001062378256/12-605\n",
            "19271. MGYP000172094177/15-406\n",
            "19272. MGYP000895451009/19-614\n",
            "19273. MGYP000742342036/15-616\n",
            "19274. MGYP000822497044/11-403\n",
            "19275. MGYP000276408254/18-602\n",
            "19276. MGYP000711967738/12-418\n",
            "19277. MGYP000984576269/8-609\n",
            "19278. MGYP000739793083/7-605\n",
            "19279. MGYP000954937640/13-610\n",
            "19280. MGYP000581905697/13-404\n",
            "19281. MGYP001020788390/1-359\n",
            "19282. MGYP000294411411/9-396\n",
            "19283. MGYP000106097234/8-607\n",
            "19284. MGYP001084378320/10-607\n",
            "19285. MGYP000630719926/8-610\n",
            "19286. MGYP000902765494/4-602\n",
            "19287. MGYP000615109282/2-605\n",
            "19288. MGYP000666105585/8-614\n",
            "19289. MGYP000191160094/1-368\n",
            "19290. MGYP000450807278/5-350\n",
            "19291. MGYP000047053759/15-606\n",
            "19292. MGYP000069121075/15-611\n",
            "19293. MGYP001024334277/4-655\n",
            "19294. MGYP000785231565/13-406\n",
            "19295. MGYP000688766444/2-386\n",
            "19296. MGYP000964504363/3-408\n",
            "19297. MGYP000500009185/11-608\n",
            "19298. MGYP000976153372/11-600\n",
            "19299. MGYP000240380612/14-406\n",
            "19300. MGYP000481166387/2-395\n",
            "19301. MGYP000932459735/14-413\n",
            "19302. MGYP000094898085/8-604\n",
            "19303. MGYP000179427003/9-604\n",
            "19304. MGYP000231729376/35-420\n",
            "19305. MGYP000675197156/8-635\n",
            "19306. MGYP000245082318/13-427\n",
            "19307. MGYP000386216001/10-614\n",
            "19308. MGYP000202034483/10-614\n",
            "19309. MGYP000735620717/6-605\n",
            "19310. MGYP001044744862/11-604\n",
            "19311. MGYP000593550489/2-598\n",
            "19312. MGYP000129677021/15-396\n",
            "19313. MGYP000500936306/79-472\n",
            "19314. MGYP000306948650/10-405\n",
            "19315. MGYP000278807365/44-448\n",
            "19316. MGYP000997646404/22-408\n",
            "19317. MGYP000014262672/7-591\n",
            "19318. MGYP000956826115/8-400\n",
            "19319. MGYP000203576545/14-437\n",
            "19320. MGYP001123482961/8-593\n",
            "19321. MGYP000886904796/11-606\n",
            "19322. MGYP000432939125/217-811\n",
            "19323. MGYP000884409125/10-602\n",
            "19324. MGYP001038179051/3-408\n",
            "19325. MGYP000767937052/10-388\n",
            "19326. MGYP000875876666/29-636\n",
            "19327. MGYP000468378942/228-634\n",
            "19328. MGYP000099419988/10-608\n",
            "19329. MGYP000261759193/12-440\n",
            "19330. MGYP000791575814/12-400\n",
            "19331. MGYP000470198283/1-349\n",
            "19332. MGYP000821134073/8-350\n",
            "19333. MGYP000074668971/7-389\n",
            "19334. MGYP000034566752/5-596\n",
            "19335. MGYP000882747856/2-388\n",
            "19336. MGYP000715853396/11-404\n",
            "19337. MGYP000892494633/3-409\n",
            "19338. MGYP000320164834/7-333\n",
            "19339. MGYP001103794168/5-611\n",
            "19340. MGYP001077814516/12-390\n",
            "19341. MGYP000346374557/10-605\n",
            "19342. MGYP001120854299/7-398\n",
            "19343. MGYP000403519442/13-406\n",
            "19344. MGYP000904638835/31-405\n",
            "19345. MGYP000889266598/5-337\n",
            "19346. MGYP000635587046/9-400\n",
            "19347. MGYP000283727863/6-604\n",
            "19348. MGYP000415296952/1-567\n",
            "19349. MGYP000550897474/2-344\n",
            "19350. MGYP001121166824/6-400\n",
            "19351. MGYP000279668957/12-488\n",
            "19352. MGYP000036066138/13-435\n",
            "19353. MGYP000790331915/14-351\n",
            "19354. MGYP001124395367/11-600\n",
            "19355. MGYP000249866623/6-352\n",
            "19356. MGYP000298283108/6-372\n",
            "19357. MGYP000797070352/14-607\n",
            "19358. MGYP000032686695/109-504\n",
            "19359. MGYP000912412376/18-613\n",
            "19360. MGYP000464577947/8-397\n",
            "19361. MGYP000548892544/19-608\n",
            "19362. MGYP000919470835/8-612\n",
            "19363. MGYP001014617258/7-411\n",
            "19364. MGYP000070471432/13-402\n",
            "19365. MGYP000637438980/15-429\n",
            "19366. MGYP000059969858/14-605\n",
            "19367. MGYP000270135943/14-607\n",
            "19368. MGYP000094298554/15-442\n",
            "19369. MGYP000213456116/3-612\n",
            "19370. MGYP000800843082/13-434\n",
            "19371. MGYP000193171391/18-603\n",
            "19372. MGYP000528547398/19-410\n",
            "19373. MGYP000161276662/11-608\n",
            "19374. MGYP000898457375/7-599\n",
            "19375. MGYP000659188683/12-392\n",
            "19376. MGYP000742441178/4-610\n",
            "19377. MGYP001143773253/4-602\n",
            "19378. MGYP001038426180/11-608\n",
            "19379. MGYP000901778484/10-400\n",
            "19380. MGYP001037051774/14-335\n",
            "19381. MGYP000913440606/15-646\n",
            "19382. MGYP000761378167/11-604\n",
            "19383. MGYP000297300778/254-651\n",
            "19384. MGYP000595291418/4-374\n",
            "19385. MGYP000739123072/11-604\n",
            "19386. MGYP000832078332/4-326\n",
            "19387. MGYP000392782615/3-327\n",
            "19388. MGYP000801521448/8-401\n",
            "19389. MGYP000702799908/14-404\n",
            "19390. MGYP000108824537/13-612\n",
            "19391. MGYP000307295899/9-421\n",
            "19392. MGYP000046211887/34-452\n",
            "19393. MGYP000945213142/4-418\n",
            "19394. MGYP000173334046/14-406\n",
            "19395. MGYP000952131103/13-613\n",
            "19396. MGYP001045965320/7-320\n",
            "19397. MGYP001092850005/5-591\n",
            "19398. MGYP000174168411/6-605\n",
            "19399. MGYP000051873244/4-395\n",
            "19400. MGYP000597193101/6-397\n",
            "19401. MGYP000253685743/7-620\n",
            "19402. MGYP000441720591/10-413\n",
            "19403. MGYP000480342341/15-597\n",
            "19404. MGYP000441944574/10-428\n",
            "19405. MGYP000268946268/8-607\n",
            "19406. MGYP000334742642/7-586\n",
            "19407. MGYP000308824313/15-609\n",
            "19408. MGYP001144963945/32-439\n",
            "19409. MGYP000479863243/5-349\n",
            "19410. MGYP000284620050/6-614\n",
            "19411. MGYP000456993804/4-594\n",
            "19412. MGYP000149593250/8-596\n",
            "19413. MGYP000238254160/15-609\n",
            "19414. MGYP000459535123/12-434\n",
            "19415. MGYP001052154363/8-604\n",
            "19416. MGYP000719811364/8-582\n",
            "19417. MGYP001000970451/20-616\n",
            "19418. MGYP000764610700/10-421\n",
            "19419. MGYP000216844656/9-599\n",
            "19420. MGYP001040626743/1-567\n",
            "19421. MGYP000244108400/6-657\n",
            "19422. MGYP000978033704/5-394\n",
            "19423. MGYP000134363099/13-379\n",
            "19424. MGYP000023328200/77-679\n",
            "19425. MGYP000501522330/12-407\n",
            "19426. MGYP000914820282/13-432\n",
            "19427. MGYP000864155157/25-615\n",
            "19428. MGYP000352656585/4-379\n",
            "19429. MGYP000084485272/6-418\n",
            "19430. MGYP000403913954/8-389\n",
            "19431. MGYP001004755620/15-406\n",
            "19432. MGYP001072838695/10-409\n",
            "19433. MGYP001155488877/3-570\n",
            "19434. MGYP000797864783/2-376\n",
            "19435. MGYP001084225189/6-606\n",
            "19436. MGYP000354863896/20-616\n",
            "19437. MGYP000692904352/13-613\n",
            "19438. MGYP000168433201/5-597\n",
            "19439. MGYP000465555995/3-601\n",
            "19440. MGYP000025809459/2-391\n",
            "19441. MGYP000001550496/10-350\n",
            "19442. MGYP000267798458/5-607\n",
            "19443. MGYP000289168545/7-426\n",
            "19444. MGYP001118748979/6-611\n",
            "19445. MGYP000198308580/25-620\n",
            "19446. MGYP000034179488/15-605\n",
            "19447. MGYP000503852338/12-360\n",
            "19448. MGYP001109761039/29-623\n",
            "19449. MGYP000243896432/15-398\n",
            "\n",
            "Enter the number of the query.\n",
            "1\n",
            "Chosen query is query\n",
            "\n",
            "Bypassing Rate4Site. Initializing AI Protein Language Model (ProtBert)...\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "Loading weights:   0%|          | 0/487 [00:00<?, ?it/s]"
            ],
            "application/vnd.jupyter.widget-view+json": {
              "version_major": 2,
              "version_minor": 0,
              "model_id": "cf766ab06e0f43208f28fbf02f371aea"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "BertModel LOAD REPORT from: Rostlab/prot_bert\n",
            "Key                                        | Status     |  | \n",
            "-------------------------------------------+------------+--+-\n",
            "cls.predictions.transform.dense.weight     | UNEXPECTED |  | \n",
            "cls.seq_relationship.weight                | UNEXPECTED |  | \n",
            "cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | \n",
            "cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | \n",
            "cls.predictions.decoder.bias               | UNEXPECTED |  | \n",
            "cls.predictions.bias                       | UNEXPECTED |  | \n",
            "cls.predictions.decoder.weight             | UNEXPECTED |  | \n",
            "cls.predictions.transform.dense.bias       | UNEXPECTED |  | \n",
            "cls.seq_relationship.bias                  | UNEXPECTED |  | \n",
            "\n",
            "Notes:\n",
            "- UNEXPECTED\t:can be ignored when loading from different task/architecture; not ok if you expect identical arch.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Success! Populated gradesPE_Output with 1417 scores.\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "KeyError",
          "evalue": "'MSA_DENUM'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
            "\u001b[0;32m/tmp/ipython-input-565/249698956.py\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m   7726\u001b[0m         \u001b[0mLOG\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtraceback\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat_exc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   7727\u001b[0m         \u001b[0mLOG\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 7728\u001b[0;31m         \u001b[0;32mraise\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   7729\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   7730\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/tmp/ipython-input-565/249698956.py\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m   7702\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mvars\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'running_mode'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"_mode_pdb_no_msa\"\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mvars\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'running_mode'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"_mode_pdb_msa\"\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mvars\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'running_mode'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"_mode_pdb_msa_tree\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   7703\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 7704\u001b[0;31m         \u001b[0mconsurf_create_output\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   7705\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   7706\u001b[0m     \u001b[0;31m## mode : ConSeq - NO PDB\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/tmp/ipython-input-565/249698956.py\u001b[0m in \u001b[0;36mconsurf_create_output\u001b[0;34m()\u001b[0m\n\u001b[1;32m   6189\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6190\u001b[0m     \u001b[0matom_grades\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6191\u001b[0;31m     \u001b[0mcreate_gradesPE\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvars\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'gradesPE'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mr4s2pdb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvars\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'pdb_file_name'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mform\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'PDB_chain'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvars\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Used_PDB_Name'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvars\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'pdb_object'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0midentical_chains\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvars\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'cif_or_pdb'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0matom_grades\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6192\u001b[0m     \u001b[0mreplace_TmpFactor_Consurf_Scores\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0matom_grades\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mform\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'PDB_chain'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvars\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'pdb_file_name'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvars\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Used_PDB_Name'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# Create ATOMS section and replace the TempFactor Column with the ConSurf Grades (will create also isd file if relevant)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6193\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/tmp/ipython-input-565/249698956.py\u001b[0m in \u001b[0;36mcreate_gradesPE\u001b[0;34m(gradesPE, ref_match, pdb_file, chain, prefix, pdb_object, identical_chains, pdb_cif, atom_grades)\u001b[0m\n\u001b[1;32m   4806\u001b[0m             \u001b[0mPE\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"  %s\\t  %s\\t\"\u001b[0m \u001b[0;34m%\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0melem\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'B/E'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0melem\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'F/S'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4807\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4808\u001b[0;31m         \u001b[0mhomologs_in_pos\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0melem\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'MSA_NUM'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m\"/\"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0melem\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'MSA_DENUM'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;31m# number of homologs in the position\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4809\u001b[0m         \u001b[0mnum_spaces\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmsa_column_size\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhomologs_in_pos\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# number of spaces to add\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4810\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0mnum_spaces\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mKeyError\u001b[0m: 'MSA_DENUM'"
          ]
        }
      ],
      "source": [
        "#@title Run ConSurf\n",
        "\n",
        "\n",
        "import sys\n",
        "import traceback\n",
        "import socket\n",
        "import getpass\n",
        "import re\n",
        "import os\n",
        "import json\n",
        "import shutil\n",
        "import gzip\n",
        "import requests\n",
        "import subprocess\n",
        "import time\n",
        "import tempfile\n",
        "import urllib\n",
        "import tarfile\n",
        "import shutil\n",
        "\n",
        "\n",
        "os.environ[\"TRANSFORMERS_VERBOSITY\"] = \"error\"\n",
        "os.environ[\"HF_HUB_DISABLE_TELEMETRY\"] = \"1\"\n",
        "\n",
        "from zipfile import ZipFile\n",
        "from pathlib import Path\n",
        "from IPython.display import HTML, display, IFrame\n",
        "\n",
        "# Use HTML and CSS to style the input box\n",
        "display(HTML(\"\"\"\n",
        "<style>\n",
        "    .container { width: 100% !important; }\n",
        "    .CodeMirror, .output_subarea, .output_area { max-width: 100% !important; }\n",
        "    .input { width: 100% !important; }\n",
        "</style>\n",
        "\"\"\"))\n",
        "\n",
        "\n",
        "bayesInterval = 3\n",
        "ColorScale = {0 : 9, 1 : 8, 2 : 7, 3 : 6, 4 : 5, 5 : 4, 6 : 3, 7 : 2, 8 : 1}\n",
        "\n",
        "try:\n",
        "\n",
        "    # not first run\n",
        "    root_dir = vars['root_dir']\n",
        "\n",
        "except:\n",
        "\n",
        "    # first run\n",
        "    root_dir = os.getcwd()\n",
        "\n",
        "vars = {}\n",
        "form = {}\n",
        "\n",
        "job_name = 'test' #@param {type:\"string\"}\n",
        "\n",
        "vars['root_dir'] = root_dir\n",
        "os.chdir(vars['root_dir'])\n",
        "\n",
        "# check if directory already exists\n",
        "i = 1\n",
        "temp_job_name = job_name\n",
        "while os.path.isdir(temp_job_name):\n",
        "\n",
        "    temp_job_name = job_name + \"_\" + str(i)\n",
        "    i += 1\n",
        "\n",
        "vars['job_name'] = temp_job_name\n",
        "vars['working_dir'] = vars['root_dir'] + \"/\" + vars['job_name'] + \"/\"\n",
        "\n",
        "# create job directory\n",
        "os.mkdir(vars['job_name'])\n",
        "\n",
        "# directories for programs installed\n",
        "vars['prottest'] = vars['root_dir'] + \"/prottest-3.4.2/prottest-3.4.2.jar\"\n",
        "vars['cd_hit_dir'] = vars['root_dir'] + \"/cdhit/\"\n",
        "vars['rate4site_dir'] = vars['root_dir'] + \"/r4s_for_collab/\"\n",
        "vars['rate4site_slow_dir'] = vars['root_dir'] + \"/r4s_for_collab_slow/\"\n",
        "\n",
        "# weights\n",
        "if not os.path.isfile(\"WEIGHTS_READY\"):\n",
        "\n",
        "    WEIGHTS = open(\"WEIGHTS.BIN\", 'w')\n",
        "    WEIGHTS.write(\"\"\"0.079961\n",
        "-0.042471 0.284039 -0.115342 -0.059031 -0.057580 -0.221184 -0.622583 -0.632449 0.119240 -0.367390 -0.345141 -0.040769 -0.036835 -0.050943 -0.263045 -0.440234 0.092578 0.067303 -0.382988 -0.104377 0.076840 0.721150 -0.478015 0.034385 -0.033848 -0.226497 -0.232466 -0.785597 0.178759 -0.487048 -0.213410 -0.523454 -0.127242 -0.137284 0.171664 -0.079844 -0.104554 0.033901 -0.293857 -0.404099 -0.416435 0.157385 -0.414229 -0.296172 -0.471579 -0.306119 -0.431075 -0.313670 0.055607 0.035966 0.052927 -0.472520 -0.378305 -0.316330 0.063885 0.354014 -0.078238 0.464703 -0.377103 0.000405 -0.110874 -0.331173 -0.249899 -0.573167 -0.369664 -0.453246 -0.108283 -0.808977 0.148720 -0.042741 -0.036522 -0.367998 -0.844013 -0.248575 0.051849 0.234780 0.165519 0.469451 -0.324361 0.131725 0.080134 0.612314 0.227453 -0.125036 -0.005849 -0.156387 -0.428727 -0.896221 -0.095704 -0.251770 -0.248376 -0.462221 -0.299122 -0.144201 -0.555950 -0.263420 -0.268737 -0.528499 -0.255063 -0.139020 -0.611295 0.444935 -0.231921 -0.223407 -0.116494 -0.461209 -0.569035 -0.954768 0.150339 -0.036701 -0.129399 -0.293487 -0.253250 -0.325909 0.104611 0.242358 -0.045911 0.136231 -0.352202 -0.143680 -0.655280 -0.185911 -0.354350 -0.345799 -0.599220 -0.627056 -0.558502 1.093712 0.625240 0.160620 0.711126 0.442624 -1.440944 0.024459 -0.298804 1.393993 0.355034 0.151769 0.242346 0.240612 0.040324 0.103134 -0.648484 -0.387724 -0.268557 -0.333484 -0.346043 0.052173 0.104572 -1.094794 -0.322809 -0.557991 -0.511697 0.038476 0.007950 0.409273 0.243390 0.131082 0.008020 0.058760 0.003713 0.442034 -0.018652 -0.238415 -0.436684 -0.058566 -0.485227 -0.214224 -0.401662 0.004286 -0.148663 -0.161137 -0.139408 -0.223052 -0.306196 -0.093124 -0.418224 -0.341062 -0.532583 -0.232474 0.040335 -0.233600 -0.630569 -0.420837 -0.149225 -0.531155 -0.133665 -0.410447 -0.292123 0.145357 -0.476807 -0.399644 -0.283946 -0.224842 -0.174200 0.133730 0.139391 0.153037 -0.469379 -0.001143 -0.122711 -0.599835 -0.059480 -0.485474 -0.454370 -0.370191 -0.427881 -0.763049 -0.107038 0.125714 0.329847 -0.015233 -0.325263 0.003464 -0.073701 0.178642 -0.052665 0.244747 -0.477071 0.051825 0.461928 0.173730 0.134013 -0.363544 -0.036649 -0.180880 -0.213129 -0.774550 -0.142420 -0.313132 -0.439931 -0.169179 -0.018252 0.036593 -0.299105 0.000470 -0.200218 -0.123280 -0.425699 -0.251190 0.077114 -0.161261 -0.000334 -0.194496 -0.126588 -0.254180 -0.265544 -0.257622 0.072041 0.012121 -0.353757 -0.269024 -0.305191 -0.194474 -0.036296 -0.301452 -0.083759 -0.390128 -0.208091 -0.614116\n",
        "-0.037835\n",
        "-0.135652 -0.097861 -0.124032 -0.082832 -0.079981 -0.080759 -0.080005 -0.112538 -0.108645 -0.153970 -0.090287 -0.127751 -0.022577 -0.054848 -0.075064 -0.061371 -0.065418 -0.162476 -0.188244 -0.219030 -0.155061 -0.223714 -0.122331 -0.133745 -0.064488 -0.145680 -0.072198 -0.051936 -0.168912 -0.109149 -0.005762 -0.122343 0.009825 -0.030152 -0.041595 -0.035915 -0.116701 -0.141694 -0.121796 -0.110433 -0.099315 -0.301401 -0.101548 -0.148282 -0.130996 -0.176345 -0.193819 -0.148637 -0.113018 -0.033919 -0.106760 -0.110076 -0.002691 -0.183740 -0.105777 0.052121 0.015194 0.056617 -0.096041 -0.103306 -0.163592 -0.107840 -0.186439 -0.172847 -0.120203 -0.124392 -0.098864 -0.173437 -0.171424 -0.116866 -0.087735 -0.174578 -0.009553 -0.179103 -0.122766 0.140827 -0.090366 0.012798 -0.161375 -0.099884 -0.232814 -0.222946 -0.134413 -0.200724 -0.201677 -0.152266 -0.128065 -0.079498 -0.192186 -0.043459 -0.015164 -0.007513 -0.062827 -0.185702 -0.154283 -0.082762 -0.118400 -0.259073 0.041187 -0.003137 -0.178665 -0.287254 -0.193866 -0.255345 -0.185832 -0.194552 -0.085949 0.062150 -0.225319 0.075137 -0.002692 -0.054584 -0.052849 -0.220834 -0.100336 0.068750 -0.066442 -0.010368 -0.163883 -0.056490 -0.123533 -0.211744 -0.267799 -0.188097 -0.315561 -0.265491 -0.269421 0.140139 0.330403 0.097782 0.300753 0.379973 -0.374714 -0.150125 -0.128261 0.891383 0.040080 0.036730 0.082033 0.117802 -0.229061 -0.221868 -0.286277 -0.274466 -0.157755 -0.044367 -0.121318 0.132026 -0.180808 -0.205960 -0.037371 -0.210609 0.030780 -0.063130 -0.110591 -0.070211 0.036235 0.101944 0.018493 -0.016074 -0.146690 -0.240073 -0.220570 -0.141108 -0.141487 -0.243177 -0.190067 -0.044916 -0.178520 0.033824 -0.084478 -0.034626 -0.018853 -0.158008 -0.192055 0.039844 -0.106545 -0.020606 -0.007895 -0.255355 -0.207277 -0.350054 -0.202661 -0.213441 -0.223656 -0.177883 -0.191003 -0.089601 -0.213466 -0.018869 -0.042249 -0.096639 -0.022556 -0.214572 -0.145668 0.166606 -0.027856 0.005464 -0.074335 -0.091599 -0.138469 -0.187154 -0.212407 -0.275710 -0.092425 -0.171058 -0.103894 -0.093088 -0.200088 -0.083920 -0.121132 -0.114293 0.058626 -0.038506 -0.121911 0.116106 -0.020284 -0.024348 -0.136227 -0.008320 -0.058348 -0.102498 -0.092001 -0.116864 -0.054950 -0.081698 -0.053132 -0.136699 -0.070402 -0.129130 -0.070749 -0.144664 0.067491 -0.091968 -0.088276 -0.006360 -0.112374 -0.186199 -0.139677 -0.134099 -0.073819 -0.060381 -0.064273 -0.132142 -0.136076 -0.132970 -0.054528 -0.212359 -0.201456 -0.090024 -0.158139 -0.159991 -0.019921 -0.145357 -0.150928 -0.129886 -0.035867 -0.060223 -0.238624 -0.058521\n",
        "-0.078758\n",
        "-0.151740 -0.026767 -0.106500 -0.092961 -0.117010 -0.112364 -0.192851 -0.272185 -0.065880 -0.177882 -0.134294 -0.107798 -0.028889 -0.068127 -0.117018 -0.156986 -0.057003 -0.117327 -0.273595 -0.272451 -0.126404 -0.079874 -0.185071 -0.114151 -0.047380 -0.189966 -0.100589 -0.270994 -0.127206 -0.213702 -0.002939 -0.252945 0.014415 -0.038136 0.031142 -0.065138 -0.117974 -0.132185 -0.179534 -0.186496 -0.220966 -0.316856 -0.232701 -0.196354 -0.220504 -0.232680 -0.286300 -0.279093 -0.045867 0.022409 -0.056297 -0.185112 -0.082968 -0.233401 -0.074099 0.141940 0.029905 0.169479 -0.150252 -0.068862 -0.225708 -0.170601 -0.247159 -0.259554 -0.180183 -0.204830 -0.102752 -0.367822 -0.052640 -0.132881 -0.112569 -0.239671 -0.193336 -0.216353 -0.105227 0.182112 -0.041267 0.150941 -0.216789 -0.063747 -0.219734 -0.010456 -0.032662 -0.200514 -0.192833 -0.179559 -0.216095 -0.327812 -0.235238 -0.115399 -0.029876 -0.111954 -0.099778 -0.191547 -0.234776 -0.122411 -0.161412 -0.351707 -0.006454 -0.003536 -0.228784 -0.156429 -0.185776 -0.260639 -0.168103 -0.249610 -0.187173 -0.206471 -0.150917 0.032764 -0.067341 -0.102790 -0.062856 -0.257987 -0.088676 0.088700 -0.072901 0.012187 -0.224291 -0.133150 -0.274977 -0.237442 -0.321321 -0.222992 -0.419615 -0.387955 -0.338747 0.176174 0.563658 0.119931 0.396977 0.419047 -0.651059 -0.131797 -0.166447 1.066683 0.127136 0.110843 0.116229 0.164438 -0.243415 -0.111208 -0.390476 -0.321042 -0.219125 -0.085586 -0.186006 0.114595 -0.080241 -0.397100 -0.096225 -0.391717 -0.085629 -0.038488 -0.093712 -0.014351 0.112260 0.132180 0.005796 0.024345 -0.090793 -0.058719 -0.169727 -0.152258 -0.208756 -0.213912 -0.276227 -0.060369 -0.223690 -0.003176 -0.107536 -0.104605 -0.017045 -0.206944 -0.286270 -0.015647 -0.199786 -0.111666 -0.115525 -0.316241 -0.217539 -0.446536 -0.345833 -0.282246 -0.264920 -0.271352 -0.213021 -0.185402 -0.211945 0.019463 -0.088338 -0.161891 -0.072983 -0.227131 -0.175166 0.213077 0.035432 0.082961 -0.148847 -0.073226 -0.170132 -0.230370 -0.199540 -0.334150 -0.192919 -0.220681 -0.204650 -0.318956 -0.158520 -0.070792 -0.042598 -0.178701 0.026312 -0.033983 -0.119504 0.162432 -0.031498 0.019919 -0.228353 0.018548 0.023433 -0.067525 -0.076803 -0.172050 -0.083896 -0.098566 -0.083301 -0.284714 -0.099721 -0.186871 -0.138397 -0.163712 0.045302 -0.068164 -0.133595 0.011395 -0.125378 -0.198303 -0.214062 -0.205253 -0.065363 -0.073619 -0.079477 -0.167100 -0.169177 -0.173859 -0.077143 -0.268030 -0.169336 -0.060870 -0.222505 -0.168985 -0.096238 -0.180600 -0.157466 -0.160292 -0.017769 -0.116800 -0.260166 -0.170865\n",
        "-0.524720\n",
        "0.154513 0.203979 0.048294 0.090322 0.055480 0.078362 0.004012 -0.082407 0.090325 0.031930 -0.051573 0.137165 -0.021265 0.059140 -0.006711 -0.023714 0.100909 0.152038 0.022316 0.074661 0.124397 0.320426 0.032491 0.171462 0.113720 0.113248 0.068659 -0.198500 0.091383 -0.061092 0.021773 -0.082033 0.030394 0.013086 0.090308 0.069989 0.056663 0.073158 -0.027319 -0.093840 -0.077431 -0.005912 0.029071 0.074618 0.095422 0.104539 0.078844 -0.041097 0.130584 0.041358 0.193946 0.008914 -0.007856 0.134291 0.055873 0.031463 0.000300 -0.007945 0.058287 0.126511 0.035880 0.110994 0.150183 0.119831 0.053810 0.079044 0.115316 -0.174369 0.245411 0.051703 0.027911 0.052232 -0.087423 0.166628 0.102278 -0.132092 0.090646 0.000045 0.058830 0.121950 0.214049 0.370781 0.241658 0.247640 0.196595 0.131533 0.070703 -0.229280 0.144735 -0.102751 -0.014604 -0.041123 0.072117 0.152531 0.008596 -0.068417 0.056369 0.034000 -0.090817 -0.039916 0.030354 0.293707 0.187306 0.236166 0.080975 0.129781 -0.069191 -0.211208 0.275564 -0.034266 -0.003660 0.017415 0.050324 0.152961 0.124033 -0.167242 0.007611 0.018227 0.103696 -0.076929 -0.281846 -0.195073 0.075917 0.182325 -0.008900 0.102061 0.131555 -0.213223 -0.024959 -0.131552 -0.289637 -0.448343 -0.019283 0.229526 0.211105 -0.573386 0.142129 0.266419 0.108788 0.061394 0.125213 0.281386 0.138985 0.242696 0.115892 0.016850 0.106856 -0.105255 0.249325 -0.135292 -0.085143 -0.044688 -0.047135 0.132880 0.132071 0.028059 0.054803 -0.108401 -0.050485 -0.059982 0.160326 0.353662 0.329035 0.119034 0.111641 0.263509 0.133962 0.006345 0.123893 -0.138080 -0.018373 -0.055900 0.043093 0.126120 0.053354 -0.095226 -0.019028 -0.149699 -0.097878 0.036043 0.164260 0.118168 0.112959 0.134726 0.169293 0.096612 0.199498 -0.146965 0.218327 0.021544 -0.041629 0.040537 0.054411 0.192245 0.152940 -0.138435 0.077090 0.002609 -0.052208 -0.042528 0.129684 0.077185 0.186645 0.187370 0.038889 0.069162 0.061597 -0.197347 0.194808 0.041196 0.135066 0.028764 0.018190 0.100267 0.093109 -0.071298 -0.059168 -0.080016 -0.016150 -0.056207 0.124955 0.205059 0.164676 0.046918 0.070075 0.069983 0.097966 -0.200358 0.070685 -0.064280 -0.069341 0.017293 -0.039881 0.107536 0.071292 0.005646 0.025108 0.040541 0.042647 -0.043777 0.070598 0.044540 0.131586 0.078689 0.119421 0.067460 0.038218 -0.061153 0.186128 0.107766 0.024194 0.086200 -0.015730 0.051704 0.123282 0.057591 0.016769 -0.041757 0.156185 -0.064568\n",
        "-3.392679\n",
        "-0.045022 0.073616 -0.000408 0.160156 0.154186 0.251712 -0.033704 0.051172 0.428546 -0.027031 0.327703 0.021208 0.168122 0.315747 0.277537 0.200548 0.188944 0.206525 0.117451 -0.059984 0.010269 0.076054 -0.116692 0.055683 0.035898 -0.299356 0.019773 0.022074 0.151251 -0.128346 0.189414 0.179334 0.022655 0.420717 0.212447 -0.097653 0.027509 0.335759 0.155124 0.072035 0.129656 0.194478 0.083952 0.000398 0.097400 0.171972 -0.201141 -0.174791 0.170778 0.054058 0.211608 0.156615 0.046134 -0.399402 0.050863 0.202286 0.247494 0.491820 0.060935 0.276054 0.191318 0.260353 0.260102 -0.105599 0.190812 0.372459 0.169546 0.003022 -0.001262 -0.016418 0.311207 -0.183385 -0.150570 -0.070429 0.143529 0.720439 0.001143 0.513456 -0.137016 -0.023083 0.032894 0.193806 0.358182 -0.048925 -0.000153 0.156751 -0.013043 -0.082260 0.188459 0.415768 0.342516 0.154946 -0.068848 0.067780 -0.062038 0.250291 0.269551 0.155357 0.460478 0.548554 -0.044765 -0.001985 0.162424 -0.195813 -0.001534 0.126595 0.492953 -0.175303 0.008005 -0.043580 -0.012260 0.180904 -0.022612 0.128195 0.441880 0.617810 0.324884 0.616176 -0.232894 0.227519 1.505843 1.686601 1.598056 0.807867 1.273509 0.789232 0.309206 0.999822 2.621622 -0.367464 1.648606 1.451281 1.094744 -0.841376 -3.625247 -2.927335 -0.831452 -5.212661 -0.698117 -4.950808 0.041992 0.266384 0.049116 -0.118177 0.074138 0.288970 0.073334 -0.376111 0.019275 -0.747190 0.097097 -0.066781 0.178590 0.008632 0.435278 0.655038 0.419311 0.507827 0.217161 0.160956 0.462550 0.492895 0.473460 0.268630 0.258759 -0.016715 -0.217376 -0.238036 0.228603 -0.117788 0.100081 0.009037 0.361601 -0.429419 0.013617 0.574723 0.139799 0.557784 0.217710 -0.006795 0.500248 0.375827 0.344209 0.507256 0.275542 0.123528 0.290640 -0.282119 0.099417 -0.258653 -0.166433 -0.018108 0.248410 0.028849 -0.014118 0.820778 0.024199 0.359858 0.022365 0.035588 0.307911 0.280823 0.333445 0.036463 0.282302 0.114528 0.100341 0.043328 0.120032 -0.315219 -0.057145 -0.162257 0.410823 0.071707 -0.333178 0.150516 0.032312 0.305116 -0.204997 0.176935 0.351771 -0.063635 0.094086 -0.123456 0.088026 0.089180 -0.121424 -0.073208 0.128705 -0.018277 0.069866 -0.232850 0.223583 -0.072594 -0.028345 0.504329 0.003772 0.019249 -0.166544 0.014105 0.065732 0.069389 0.255167 0.082785 -0.132714 0.139074 0.188195 0.022171 0.127342 0.017475 0.309935 -0.111449 0.114716 0.092440 0.180858 -0.008111 0.451164 0.344322 -0.038933 -0.000417\n",
        "-0.041590\n",
        "-0.132178 -0.125809 -0.129481 -0.083871 -0.060361 -0.056246 -0.038208 -0.072517 -0.142981 -0.126048 -0.068757 -0.130837 -0.015395 -0.052602 -0.052098 -0.038161 -0.067340 -0.166669 -0.117403 -0.214311 -0.136092 -0.266236 -0.094254 -0.136102 -0.063053 -0.117778 -0.048600 0.028962 -0.214789 -0.074492 -0.005237 -0.048622 0.016390 -0.015761 -0.068684 -0.034536 -0.112518 -0.137535 -0.110367 -0.056196 -0.042167 -0.255057 -0.065883 -0.118502 -0.115106 -0.150318 -0.158210 -0.090196 -0.149499 -0.041800 -0.127829 -0.058796 0.023974 -0.148427 -0.118233 0.014338 0.010101 0.004113 -0.060837 -0.108300 -0.145160 -0.108116 -0.172288 -0.141194 -0.102221 -0.096769 -0.103515 -0.078920 -0.181267 -0.095226 -0.074094 -0.123623 0.051999 -0.163724 -0.133526 0.122587 -0.103489 -0.043332 -0.137697 -0.125252 -0.240598 -0.298775 -0.169125 -0.206577 -0.204941 -0.143264 -0.098425 0.020087 -0.163173 -0.020891 -0.005268 0.046792 -0.056783 -0.177118 -0.103821 -0.057102 -0.091597 -0.225117 0.068481 -0.006780 -0.138751 -0.330845 -0.199337 -0.242210 -0.177021 -0.158106 -0.035243 0.142816 -0.225250 0.075833 0.017431 -0.048766 -0.036734 -0.199296 -0.106929 0.040741 -0.058131 -0.022288 -0.155812 -0.031227 -0.030030 -0.211709 -0.248420 -0.174684 -0.245596 -0.217578 -0.236770 0.121377 0.206832 0.089238 0.261187 0.365860 -0.259989 -0.160865 -0.110200 0.787453 0.000391 -0.007308 0.059851 0.086278 -0.206551 -0.248948 -0.211574 -0.244159 -0.131569 -0.021120 -0.110736 0.136446 -0.210287 -0.106094 -0.024033 -0.140486 0.062489 -0.064616 -0.128236 -0.099440 0.015303 0.069268 0.027189 -0.045384 -0.174615 -0.322549 -0.252268 -0.129119 -0.127833 -0.235098 -0.152967 -0.047136 -0.158150 0.063726 -0.054265 0.015198 -0.009136 -0.140124 -0.157710 0.054738 -0.071698 0.022549 0.034080 -0.220900 -0.194509 -0.324938 -0.145107 -0.181066 -0.199487 -0.130370 -0.186619 -0.052283 -0.189933 -0.035983 -0.022248 -0.065733 -0.007002 -0.195613 -0.149219 0.138675 -0.048173 -0.020118 -0.044938 -0.084325 -0.112506 -0.171200 -0.196599 -0.250543 -0.058026 -0.142673 -0.070435 0.009976 -0.205047 -0.091593 -0.150539 -0.076336 0.059103 -0.046835 -0.120795 0.081649 -0.018661 -0.034827 -0.117086 -0.013402 -0.080044 -0.115887 -0.108078 -0.100791 -0.057655 -0.078371 -0.046533 -0.064284 -0.076863 -0.096889 -0.035407 -0.121175 0.063673 -0.101571 -0.052454 -0.023295 -0.112777 -0.163278 -0.093675 -0.088857 -0.081985 -0.065261 -0.066517 -0.121067 -0.103562 -0.103624 -0.052081 -0.152509 -0.211060 -0.095401 -0.140610 -0.123229 0.010304 -0.132269 -0.135124 -0.112554 -0.032559 -0.036863 -0.220164 -0.027542\n",
        "-0.271952\n",
        "-0.045395 -0.056557 -0.064905 -0.040018 -0.006837 -0.013241 -0.002977 -0.090237 -0.089319 -0.066436 -0.042569 -0.085973 -0.005157 -0.028067 -0.048009 -0.021940 -0.050250 -0.095640 -0.044304 -0.105895 -0.041250 -0.108397 -0.016584 -0.068507 -0.029634 -0.048570 -0.031929 0.018268 -0.087497 -0.023484 -0.047070 -0.010950 0.007556 -0.017551 -0.070870 -0.057478 -0.048746 -0.103563 -0.066915 -0.067710 -0.026611 -0.132975 -0.020912 -0.067232 -0.050926 -0.058340 -0.059234 -0.044551 -0.092157 -0.068601 -0.087231 -0.002799 0.012061 -0.051551 -0.060536 -0.032698 -0.002092 -0.029861 -0.018319 -0.062244 -0.070498 -0.115205 -0.099401 -0.060689 -0.049254 -0.042813 -0.073455 -0.008414 -0.112830 -0.041533 -0.041827 -0.050015 0.051119 -0.055389 -0.065314 0.017481 -0.052383 -0.074433 -0.046139 -0.079537 -0.120091 -0.190605 -0.137245 -0.090236 -0.083762 -0.063619 -0.050073 0.052787 -0.049213 0.015644 -0.052631 0.024157 -0.021599 -0.084328 -0.060639 -0.069323 -0.042671 -0.110230 0.014725 -0.044992 -0.106150 -0.181858 -0.108550 -0.122784 -0.075157 -0.059407 -0.020988 0.079720 -0.122836 0.049664 0.003600 -0.044370 -0.021544 -0.079783 -0.060791 -0.006372 -0.044595 -0.045142 -0.065355 -0.017842 -0.041831 -0.155656 -0.168019 -0.101678 -0.102315 -0.074573 -0.131361 0.022933 0.035834 0.043777 0.104031 0.168131 -0.092152 -0.065470 -0.024410 0.381073 -0.001802 -0.042548 0.017464 0.012254 -0.103151 -0.135851 -0.095654 -0.103571 -0.065123 -0.002239 -0.058437 0.034210 -0.110232 0.001864 -0.012001 -0.045981 0.032254 -0.020927 -0.085583 -0.072909 -0.029966 -0.009529 0.010964 -0.091342 -0.127509 -0.154757 -0.141770 -0.054333 -0.055695 -0.103479 -0.059540 -0.044984 -0.070868 0.055362 -0.033872 -0.006520 -0.013914 -0.048720 -0.080282 -0.010101 -0.035035 -0.012672 0.005685 -0.131039 -0.139054 -0.163211 -0.072477 -0.076171 -0.098375 -0.054348 -0.115966 -0.013842 -0.071572 -0.020503 -0.022137 -0.055190 -0.028280 -0.093389 -0.069000 0.040945 -0.040701 -0.046302 -0.011549 -0.051834 -0.092055 -0.173768 -0.130574 -0.106170 -0.020419 -0.049567 -0.052849 0.011113 -0.109798 -0.041580 -0.092304 -0.026687 0.001739 -0.020729 -0.043661 0.017232 0.005828 0.009025 -0.016535 -0.020974 -0.072331 -0.065299 -0.074565 -0.037504 -0.030196 -0.022728 -0.045894 -0.036832 -0.041925 -0.031327 -0.039065 -0.067165 0.034849 -0.052476 -0.047802 -0.044696 -0.066765 -0.072643 -0.038657 -0.039403 -0.057635 -0.022819 -0.038915 -0.046765 -0.039010 -0.033113 -0.027183 -0.082576 -0.095628 -0.068323 -0.079280 -0.095925 0.004589 -0.061938 -0.083236 -0.055389 -0.034972 -0.039288 -0.114143 -0.043573\n",
        "-0.182199\n",
        "0.333577 0.402240 0.057833 0.150660 0.067994 0.109537 -0.043435 -0.239277 0.232094 0.016439 -0.039315 0.207970 0.009558 0.084781 -0.023556 -0.029111 0.183485 0.319671 0.064607 0.161129 0.264343 0.735518 0.110168 0.278684 0.161790 0.137883 0.059641 -0.328873 0.367503 -0.086449 -0.046984 -0.139189 0.029268 -0.025356 0.131801 0.033863 0.147411 0.125243 0.004550 -0.235114 -0.132314 0.051398 -0.007926 0.075810 0.070909 0.147701 0.122150 0.012046 0.285163 0.015713 0.357777 0.038930 -0.072818 0.162763 0.147995 0.100531 0.057325 0.080997 0.079210 0.281238 0.098437 0.016878 0.152984 0.100850 0.041531 0.084748 0.127100 -0.193288 0.361176 0.108822 0.112392 0.073172 -0.181214 0.226990 0.284555 -0.096773 0.233745 0.090602 0.192730 0.233476 0.431908 0.729104 0.429869 0.399562 0.444897 0.193794 0.058353 -0.366542 0.364086 -0.126459 -0.076009 -0.163273 0.078494 0.222658 -0.029707 -0.134970 0.101285 0.106026 -0.213533 -0.056406 0.018651 0.622755 0.279328 0.340790 0.212750 0.140273 -0.155088 -0.404798 0.466519 -0.051646 -0.104733 -0.036386 0.054584 0.267557 0.321469 -0.082935 0.054593 0.039709 0.180017 -0.073799 -0.376128 -0.029779 0.164219 0.238586 0.112156 0.191306 0.179340 -0.227914 0.183345 -0.147179 -0.311659 -0.522449 -0.013863 0.323889 0.226280 -0.683743 0.210250 0.168602 0.018130 -0.038754 0.194055 0.470361 0.126607 0.379459 0.089224 0.030830 0.108482 -0.215503 0.433648 -0.166473 -0.065165 -0.038147 -0.157224 0.211466 0.299139 0.243043 0.103381 -0.082295 -0.046633 -0.061255 0.255138 0.860237 0.523026 0.214367 0.163951 0.405456 0.152334 0.174879 0.171473 -0.157955 -0.018311 -0.205462 0.041257 0.190565 0.067828 -0.194655 -0.000133 -0.252547 -0.211616 0.093334 0.211760 0.147077 0.068884 0.150826 0.222989 0.101185 0.252162 -0.155209 0.392221 0.118544 -0.049227 -0.039225 0.004824 0.314964 0.289233 -0.104307 0.184015 0.142527 -0.000708 0.040517 0.144526 -0.076148 0.178027 0.216719 0.025673 0.125074 -0.016286 -0.266400 0.307535 0.085370 0.338186 0.044601 -0.107020 0.144549 0.171706 -0.003326 0.030036 0.181491 0.090871 0.010198 0.214869 0.381655 0.163297 0.075862 0.153203 0.159458 0.067810 -0.268114 0.090157 -0.024632 -0.096695 0.016455 -0.091401 0.182737 0.012597 0.096837 0.052684 0.116400 0.082458 -0.003502 0.128264 0.216218 0.243622 0.131073 0.157021 0.104347 0.054909 0.027074 0.287800 0.085316 0.047755 0.047057 -0.081919 0.105615 0.173316 0.206435 0.063383 -0.045434 0.245405 -0.238126\n",
        "2.977369\n",
        "-0.163825 -0.148193 -0.539381 -0.053625 -0.014042 0.015580 0.040129 0.068626 0.104471 0.008693 0.014319 0.033902 0.256730 -0.058533 -0.127252 -0.063938 -0.003439 0.128616 0.053186 -0.006668 -0.150285 -0.028547 -0.300495 -0.203528 -0.170441 0.174345 -0.068260 -0.180999 0.186269 -0.091045 -0.238320 0.152270 0.089248 -0.026105 -0.231873 0.027399 0.388154 0.248230 0.127165 -0.086841 -0.626401 -0.181094 -0.447582 -0.457010 0.053507 0.244100 0.104963 -0.284865 -0.161025 -0.391923 0.086810 0.618345 0.125925 -0.360638 -0.114660 -0.102226 0.350878 -0.093910 0.197694 0.444243 -0.555014 -0.091094 -0.465243 -0.281863 -0.357215 -0.043547 -0.267070 -0.041294 -0.226283 -0.121937 0.144535 0.062620 0.054027 -0.176766 0.161581 0.297493 0.382143 -0.137483 0.252969 0.013659 -0.659448 -0.681797 -1.215761 -0.720721 -0.695361 -0.359942 -0.549934 -0.392816 0.119995 0.580115 0.033602 0.134216 -0.162930 -0.318197 0.094974 0.721420 0.646058 0.974587 0.316684 0.422577 -0.756116 -0.502022 -0.641515 -0.531720 -0.531165 -0.240401 -0.349240 -0.354057 0.247424 0.203356 -0.141061 -0.172185 -0.085899 -0.011301 0.130300 0.771358 0.565833 0.744581 0.153583 0.459617 -0.682627 -0.359130 -0.717771 -0.547726 -0.392511 -0.157887 0.321411 -1.404261 0.272510 -0.134869 -0.386654 -0.610925 -0.372502 -0.205246 1.897225 3.615124 1.110774 2.159541 0.224313 -0.388655 -0.113974 0.164249 -0.209217 -0.163456 -0.327062 -0.404606 -0.220949 -0.781521 0.040759 -0.335444 -0.059735 0.109019 -0.101773 -0.026008 0.233253 0.728030 0.116161 0.219268 0.355826 0.051124 -0.147746 0.101143 -0.212429 -0.295290 -0.128348 -0.466121 -0.374825 -0.650218 0.109741 -0.494414 -0.241089 -0.243229 -0.459396 -0.585304 0.106755 0.826389 0.559245 0.818260 0.041081 0.127557 0.468061 0.331585 0.376270 -0.100065 0.173053 0.126962 -0.103761 -0.836315 0.187833 -0.592226 -0.556322 -0.172010 -0.277741 -0.241643 -0.020999 0.305415 -0.197842 -0.305594 0.062626 -0.550501 0.379441 0.626009 0.327497 0.070365 0.379606 -0.183974 0.253020 -0.067718 0.175933 -0.707898 -0.573596 -0.234710 0.016759 -0.388430 -0.106722 -0.261563 -0.236886 -0.360963 0.076597 -0.519936 -0.150382 -0.240036 -0.380615 -0.270374 -0.140048 -0.096099 -0.374396 -0.330776 0.128389 0.167291 -0.072818 -0.287860 0.011989 -0.222849 -0.116449 0.615885 0.145824 0.333600 0.043666 0.012717 -0.486615 -0.210876 -0.086740 -0.130119 -0.358112 -0.059483 -0.073635 -0.065814 0.060688 -0.303054 0.165594 0.003233 -0.022032 -0.357019 -0.044611 0.365111 -0.125719 0.038804 -0.116673 -0.332629\n",
        "-0.094604\n",
        "-0.039338 0.296835 -0.109780 -0.053343 -0.079219 -0.197336 -0.545583 -0.718238 0.148056 -0.352939 -0.336613 -0.087467 -0.030405 -0.042069 -0.248009 -0.426245 0.051863 0.090482 -0.399861 -0.163716 0.004398 0.692705 -0.354803 0.045932 -0.069145 -0.220988 -0.197235 -0.726658 0.201940 -0.458321 -0.211128 -0.558260 -0.106761 -0.127248 0.157355 -0.142533 -0.090229 -0.020835 -0.253951 -0.499515 -0.450339 -0.042651 -0.385696 -0.284229 -0.508064 -0.297727 -0.407110 -0.370053 0.089062 0.032335 0.074856 -0.447646 -0.353307 -0.287445 0.079439 0.300494 -0.007943 0.438043 -0.313608 0.020245 -0.151971 -0.333631 -0.314983 -0.532355 -0.347810 -0.432814 -0.082438 -0.737444 0.164247 -0.083795 -0.049421 -0.410284 -0.752655 -0.207693 0.045166 0.190905 0.166991 0.418131 -0.275685 0.111588 -0.022886 0.641094 0.249008 -0.100379 -0.026861 -0.153531 -0.425699 -0.902560 -0.087920 -0.227595 -0.239064 -0.501777 -0.245282 -0.138420 -0.508208 -0.312420 -0.256226 -0.450817 -0.307563 -0.101317 -0.528765 0.423435 -0.112638 -0.202997 -0.057583 -0.414258 -0.510479 -0.977738 0.054880 -0.039762 -0.193027 -0.314843 -0.212870 -0.314162 0.113636 0.184343 -0.062394 0.082256 -0.311104 -0.229930 -0.661264 -0.150922 -0.317935 -0.281499 -0.559787 -0.580492 -0.517852 0.754533 0.734836 0.108148 0.630953 0.382296 -1.286025 0.008342 -0.267389 1.282460 0.345797 0.160099 0.182185 0.190976 -0.065601 0.117423 -0.604518 -0.337964 -0.280926 -0.307003 -0.337597 -0.071197 0.131602 -0.962672 -0.257850 -0.620464 -0.492981 0.059145 0.025409 0.354517 0.213128 0.175706 -0.026755 0.052323 0.026497 0.529440 0.002549 -0.186243 -0.385961 -0.054022 -0.442101 -0.028723 -0.454393 -0.045693 -0.196437 -0.298465 -0.114626 -0.209493 -0.377295 -0.189588 -0.369909 -0.338174 -0.517855 -0.311570 -0.066555 -0.233133 -0.648530 -0.385981 -0.213091 -0.507662 -0.138410 -0.423533 -0.238494 0.156420 -0.376262 -0.396881 -0.280833 -0.216257 -0.165199 0.119120 0.150084 0.216046 -0.424824 -0.035148 -0.135921 -0.606469 -0.104852 -0.452003 -0.381154 -0.342007 -0.485115 -0.764214 -0.079538 0.089738 0.275435 -0.094768 -0.302108 0.006434 -0.079492 0.172712 -0.033670 0.268834 -0.424918 0.086672 0.283847 0.132815 0.029089 -0.327448 -0.054655 -0.133993 -0.216832 -0.687770 -0.178209 -0.250897 -0.401552 -0.219359 -0.002567 0.030353 -0.290212 0.017752 -0.170495 -0.104610 -0.386807 -0.232563 0.036212 -0.050471 -0.053409 -0.164151 -0.169275 -0.232935 -0.259185 -0.329156 0.054224 -0.012304 -0.371517 -0.331643 -0.289799 -0.159334 -0.079343 -0.271762 -0.040660 -0.314192 -0.212018 -0.569574\n",
        "-0.392777\n",
        "0.203537 0.296640 0.065429 0.109822 0.067185 0.096914 -0.008482 -0.152957 0.140161 0.035327 -0.040236 0.183215 -0.012267 0.069784 -0.010210 -0.027530 0.128182 0.225686 0.043649 0.108298 0.185494 0.472725 0.061893 0.207828 0.128756 0.125855 0.068143 -0.245356 0.198891 -0.075380 0.013523 -0.101156 0.030953 0.006037 0.112339 0.062794 0.097405 0.093984 -0.013129 -0.142728 -0.097026 0.010063 0.019152 0.077966 0.093400 0.130726 0.098075 -0.042396 0.189559 0.042804 0.261221 0.018371 -0.019631 0.142844 0.085692 0.054201 0.023030 0.024954 0.080383 0.191031 0.064473 0.098640 0.149579 0.121460 0.051979 0.083557 0.119597 -0.195555 0.310305 0.087659 0.059454 0.069991 -0.130153 0.188555 0.172471 -0.118118 0.143640 0.029440 0.107274 0.165677 0.294418 0.524829 0.323565 0.312121 0.299076 0.153294 0.069785 -0.281318 0.232066 -0.115985 -0.035950 -0.089454 0.087497 0.183607 -0.005081 -0.081386 0.079269 0.061372 -0.121941 -0.041216 0.030541 0.419723 0.217438 0.284271 0.127450 0.138897 -0.095616 -0.282095 0.349096 -0.043567 -0.039506 0.012014 0.054252 0.209419 0.194034 -0.139690 0.029685 0.034301 0.141983 -0.070635 -0.312960 -0.136457 0.114074 0.207017 0.039599 0.142643 0.159224 -0.218020 0.058559 -0.135081 -0.310394 -0.495440 -0.006483 0.268749 0.220279 -0.635072 0.167186 0.228830 0.077136 0.024005 0.153210 0.357499 0.137760 0.305635 0.113599 0.016444 0.108366 -0.139975 0.335879 -0.148565 -0.076256 -0.036228 -0.085414 0.161532 0.188031 0.099451 0.079315 -0.105692 -0.042771 -0.050740 0.222864 0.540739 0.407701 0.158379 0.134144 0.324416 0.148800 0.067091 0.138713 -0.139825 -0.015136 -0.102607 0.037993 0.153182 0.058557 -0.131649 -0.016486 -0.182364 -0.135553 0.059650 0.186625 0.137912 0.101537 0.145659 0.182772 0.104982 0.220964 -0.158127 0.279868 0.059779 -0.042729 0.031355 0.042569 0.240120 0.203341 -0.124256 0.125841 0.060221 -0.032868 -0.010991 0.148098 0.042404 0.191024 0.210561 0.031927 0.097362 0.048109 -0.228724 0.241916 0.052664 0.209242 0.039070 -0.015631 0.126973 0.121133 -0.037404 -0.031303 0.004051 0.027034 -0.026682 0.165797 0.271204 0.181098 0.058164 0.109116 0.102160 0.096367 -0.232021 0.079374 -0.053306 -0.067261 0.030119 -0.066380 0.143831 0.052213 0.044223 0.043300 0.069748 0.054664 -0.033082 0.102198 0.112022 0.170247 0.095496 0.137250 0.082229 0.043658 -0.033703 0.241539 0.102650 0.042315 0.082426 -0.037443 0.078071 0.151551 0.115037 0.030598 -0.046591 0.194040 -0.118081\n",
        "-0.055820\n",
        "-0.121479 -0.133056 -0.115169 -0.083490 -0.057763 -0.056579 -0.029135 -0.102803 -0.142941 -0.111679 -0.048034 -0.134901 -0.009248 -0.059440 -0.065897 -0.035504 -0.079382 -0.156592 -0.118349 -0.177046 -0.129994 -0.277985 -0.083963 -0.136459 -0.073243 -0.119593 -0.043485 0.024679 -0.176581 -0.047406 -0.009394 -0.051934 0.007456 -0.019545 -0.062273 -0.033026 -0.097654 -0.150597 -0.101362 -0.067320 -0.054840 -0.248828 -0.063123 -0.125825 -0.102894 -0.147863 -0.153828 -0.096040 -0.131382 -0.044115 -0.119491 -0.059995 0.017253 -0.151601 -0.111128 0.013686 0.002520 0.008727 -0.055603 -0.103418 -0.144667 -0.128232 -0.164900 -0.146209 -0.099819 -0.092378 -0.107558 -0.057313 -0.188908 -0.095534 -0.066855 -0.115029 0.047725 -0.158264 -0.131815 0.117374 -0.087353 -0.044507 -0.128326 -0.115374 -0.225108 -0.290606 -0.198992 -0.197555 -0.197049 -0.129921 -0.107397 0.022246 -0.127625 -0.006417 -0.020481 0.038151 -0.043106 -0.172641 -0.112930 -0.072291 -0.089858 -0.216876 0.063296 -0.003437 -0.149940 -0.306912 -0.198909 -0.240335 -0.162010 -0.157916 -0.041190 0.152565 -0.211830 0.070833 0.020353 -0.030620 -0.047953 -0.192854 -0.108560 0.045081 -0.061613 -0.033767 -0.146886 -0.045043 -0.035122 -0.201860 -0.245936 -0.175989 -0.230188 -0.205728 -0.229009 0.099050 0.226996 0.090465 0.240876 0.352045 -0.237687 -0.159642 -0.108561 0.764744 -0.002452 -0.015068 0.056013 0.076670 -0.190420 -0.221864 -0.208151 -0.235902 -0.128749 -0.022118 -0.094027 0.109996 -0.180476 -0.097681 -0.008762 -0.129248 0.064832 -0.074999 -0.135937 -0.105771 0.003928 0.056951 0.009376 -0.058670 -0.177704 -0.303343 -0.256221 -0.120467 -0.118024 -0.228577 -0.146067 -0.057340 -0.124134 0.060037 -0.048466 0.011748 -0.022462 -0.130877 -0.150059 0.040134 -0.074255 0.009831 0.040874 -0.218470 -0.215642 -0.303035 -0.142924 -0.170101 -0.197026 -0.124626 -0.171419 -0.044217 -0.185641 -0.034296 -0.017797 -0.075835 -0.015893 -0.191592 -0.133849 0.132224 -0.047472 -0.018077 -0.034816 -0.087102 -0.143914 -0.194778 -0.202754 -0.247199 -0.052653 -0.123314 -0.071601 -0.000754 -0.183160 -0.089701 -0.139661 -0.078169 0.062691 -0.035014 -0.118674 0.084416 -0.020495 -0.033630 -0.078990 -0.008507 -0.094916 -0.107743 -0.103418 -0.095558 -0.056487 -0.067497 -0.047209 -0.042595 -0.068895 -0.096188 -0.062985 -0.132691 0.068711 -0.097243 -0.065401 -0.032095 -0.105592 -0.153127 -0.085402 -0.071613 -0.076555 -0.068316 -0.073121 -0.113681 -0.114025 -0.101642 -0.052198 -0.124898 -0.200499 -0.092317 -0.128149 -0.140340 0.003828 -0.126312 -0.140049 -0.119608 -0.043176 -0.036841 -0.199526 -0.004209\n",
        "-0.161078\n",
        "-0.172391 0.025239 -0.085512 -0.084904 -0.122630 -0.138075 -0.292537 -0.459631 0.024116 -0.212798 -0.222674 -0.103279 -0.027362 -0.054813 -0.110075 -0.200719 -0.020696 -0.041251 -0.308523 -0.242370 -0.094512 0.129123 -0.198249 -0.051768 -0.019994 -0.199654 -0.125130 -0.471003 -0.005567 -0.276041 -0.100492 -0.358800 -0.022249 -0.069341 0.055352 -0.095938 -0.110787 -0.126817 -0.196012 -0.316984 -0.297598 -0.301411 -0.285787 -0.199488 -0.296939 -0.276413 -0.291677 -0.375303 -0.008631 0.012930 -0.016319 -0.249079 -0.175029 -0.238590 -0.007563 0.183786 0.040158 0.248214 -0.169802 -0.014483 -0.274571 -0.271680 -0.275015 -0.334338 -0.222501 -0.275566 -0.106748 -0.428248 0.006957 -0.112649 -0.103941 -0.318722 -0.352560 -0.195494 -0.042672 0.205496 0.009432 0.261508 -0.195477 -0.020906 -0.137096 0.181666 0.084269 -0.154019 -0.105540 -0.171398 -0.291570 -0.539083 -0.185612 -0.166018 -0.099122 -0.235284 -0.160622 -0.169920 -0.300413 -0.226650 -0.191486 -0.361659 -0.089243 -0.057073 -0.276483 0.049469 -0.148107 -0.241054 -0.118095 -0.296521 -0.253199 -0.500722 -0.063161 -0.020703 -0.140950 -0.167000 -0.086346 -0.260960 -0.011717 0.099753 -0.071788 0.013291 -0.233057 -0.189618 -0.418945 -0.171773 -0.331955 -0.216415 -0.435723 -0.458900 -0.387893 0.206371 0.725318 0.107766 0.427584 0.385412 -0.821375 -0.091132 -0.184805 1.096339 0.196358 0.141818 0.112831 0.174096 -0.262809 -0.001910 -0.448283 -0.309763 -0.227441 -0.131561 -0.246966 0.046605 -0.017689 -0.518350 -0.149961 -0.514222 -0.209210 -0.027361 -0.029345 0.075802 0.158852 0.169768 -0.023767 0.014278 -0.041781 0.225495 -0.067278 -0.147090 -0.230889 -0.166383 -0.322870 -0.053692 -0.217476 -0.043260 -0.168465 -0.231783 -0.052413 -0.206351 -0.359356 -0.089411 -0.239557 -0.201603 -0.251934 -0.348514 -0.227247 -0.441685 -0.451012 -0.311771 -0.299639 -0.340731 -0.197510 -0.245530 -0.202427 0.058944 -0.171650 -0.227614 -0.121655 -0.209345 -0.139101 0.210194 0.075655 0.144711 -0.208947 -0.038906 -0.181463 -0.378369 -0.188499 -0.352619 -0.223904 -0.230327 -0.304854 -0.530162 -0.083519 -0.032368 0.045086 -0.210445 -0.063815 -0.029508 -0.111865 0.195861 -0.013328 0.123889 -0.263193 0.038298 0.067218 0.029397 -0.070134 -0.184194 -0.060511 -0.100832 -0.126483 -0.424602 -0.095108 -0.194774 -0.252977 -0.201346 0.027409 -0.032838 -0.172429 -0.021257 -0.136541 -0.186234 -0.268332 -0.199512 -0.088875 -0.003836 -0.071940 -0.145804 -0.186770 -0.188114 -0.109169 -0.273979 -0.140718 -0.060033 -0.275533 -0.243945 -0.156622 -0.164890 -0.157648 -0.180747 -0.032483 -0.147635 -0.252529 -0.284086\n",
        "-0.563571\n",
        "0.120322 0.154458 0.041098 0.073627 0.037233 0.076409 0.017270 -0.057808 0.073238 0.026444 -0.045548 0.100748 -0.022373 0.055925 -0.000089 -0.015441 0.072839 0.119565 0.023756 0.045131 0.083379 0.237394 0.033806 0.138231 0.102737 0.102833 0.064920 -0.162085 0.047248 -0.055964 0.027962 -0.067649 0.031920 0.019389 0.074054 0.065807 0.036460 0.046454 -0.028003 -0.077238 -0.068313 -0.020336 0.031672 0.073515 0.083638 0.088389 0.065469 -0.041788 0.096842 0.041880 0.159168 0.007666 0.006462 0.118759 0.035239 0.017411 -0.012948 -0.024826 0.049392 0.085123 0.023740 0.103529 0.128443 0.113459 0.052164 0.073035 0.100633 -0.152076 0.211742 0.039854 0.013456 0.049487 -0.070176 0.140208 0.065161 -0.141333 0.062139 -0.017834 0.037236 0.089601 0.167567 0.299501 0.190394 0.207793 0.151776 0.113968 0.065488 -0.202612 0.115226 -0.087068 -0.008540 -0.025654 0.066602 0.134856 0.015033 -0.063128 0.051069 0.023213 -0.072602 -0.040732 0.027346 0.224056 0.161321 0.211990 0.057096 0.115685 -0.055343 -0.171607 0.222655 -0.032219 0.004970 0.028330 0.044867 0.124676 0.078851 -0.174356 -0.011138 -0.000943 0.088707 -0.082074 -0.265033 -0.222822 0.046524 0.157241 -0.029840 0.078976 0.106708 -0.198657 -0.046291 -0.114275 -0.265609 -0.403534 -0.027848 0.198312 0.203056 -0.525759 0.129914 0.276387 0.124617 0.078650 0.104105 0.234809 0.138051 0.202483 0.118352 0.017503 0.097943 -0.077781 0.212668 -0.124081 -0.085158 -0.047266 -0.025869 0.109906 0.096959 -0.021738 0.041489 -0.098763 -0.055070 -0.068516 0.134757 0.252116 0.277225 0.103904 0.095756 0.226866 0.121194 -0.027503 0.126593 -0.134203 -0.010807 -0.036958 0.044039 0.111074 0.039732 -0.080888 -0.020815 -0.124666 -0.075622 0.015292 0.137525 0.095861 0.101984 0.115815 0.151238 0.080594 0.169109 -0.130559 0.174649 0.011270 -0.033394 0.048167 0.060185 0.161278 0.118346 -0.143063 0.057783 -0.017055 -0.057869 -0.045772 0.107045 0.087503 0.169693 0.169809 0.045307 0.062701 0.062737 -0.165204 0.166752 0.035014 0.102036 0.021505 0.033209 0.085936 0.077331 -0.080510 -0.067440 -0.108128 -0.028024 -0.062128 0.098899 0.161838 0.141706 0.036940 0.056320 0.052189 0.095025 -0.179686 0.064414 -0.067892 -0.067008 0.003530 -0.033432 0.077296 0.072190 -0.002364 0.027397 0.035145 0.035084 -0.050795 0.046409 0.010211 0.101925 0.078864 0.100554 0.059511 0.031974 -0.071298 0.157141 0.097137 0.017377 0.081699 -0.008919 0.047634 0.104277 0.027635 0.001932 -0.039838 0.128556 -0.039240\n",
        "-0.055524\n",
        "-0.121651 -0.101338 -0.121235 -0.084784 -0.067835 -0.064605 -0.037366 -0.123333 -0.143258 -0.119420 -0.053103 -0.127142 -0.018966 -0.056507 -0.050512 -0.037054 -0.069465 -0.167924 -0.118400 -0.195334 -0.131635 -0.272007 -0.104675 -0.138941 -0.071934 -0.126321 -0.051848 -0.004914 -0.191014 -0.065679 -0.003344 -0.049551 0.006970 -0.013819 -0.060370 -0.032778 -0.111196 -0.132110 -0.092390 -0.059479 -0.050894 -0.256448 -0.063320 -0.123688 -0.118327 -0.158227 -0.160815 -0.116985 -0.129793 -0.036013 -0.110564 -0.070994 0.013973 -0.150768 -0.112704 0.014372 0.002902 0.007730 -0.069568 -0.097175 -0.152530 -0.105592 -0.168924 -0.157034 -0.097092 -0.094271 -0.096795 -0.066025 -0.178140 -0.089373 -0.070440 -0.122396 0.028963 -0.149962 -0.129638 0.110476 -0.101048 -0.030802 -0.151373 -0.123410 -0.228702 -0.279224 -0.175215 -0.199674 -0.203795 -0.142241 -0.110343 0.025648 -0.163264 -0.019142 -0.010216 0.039434 -0.057636 -0.174898 -0.120690 -0.058461 -0.102664 -0.211775 0.067887 -0.006881 -0.141776 -0.319092 -0.205200 -0.239568 -0.168431 -0.169594 -0.033231 0.138793 -0.218081 0.074933 0.005742 -0.042981 -0.040221 -0.200093 -0.108539 0.043099 -0.062497 -0.024146 -0.150357 -0.044188 -0.021711 -0.208915 -0.257955 -0.179539 -0.240411 -0.210115 -0.245366 0.119732 0.233647 0.095043 0.261241 0.364013 -0.281277 -0.152620 -0.108577 0.787485 0.010426 -0.002798 0.062594 0.090302 -0.197515 -0.216695 -0.222525 -0.249882 -0.123530 -0.020037 -0.102914 0.121007 -0.189858 -0.106601 -0.022515 -0.151221 0.052090 -0.082054 -0.128048 -0.107045 0.016787 0.065745 0.007277 -0.045921 -0.166345 -0.301979 -0.251333 -0.135737 -0.122085 -0.236000 -0.158643 -0.054714 -0.141145 0.066686 -0.053127 0.004143 -0.016260 -0.143358 -0.157748 0.040642 -0.072479 0.010542 0.025920 -0.215266 -0.198187 -0.309701 -0.138896 -0.189426 -0.203630 -0.126080 -0.192996 -0.067037 -0.177639 -0.033326 -0.026226 -0.076226 -0.016337 -0.199065 -0.145941 0.134489 -0.043930 -0.017339 -0.041515 -0.084516 -0.123700 -0.191777 -0.191897 -0.257580 -0.058458 -0.129101 -0.081872 -0.019121 -0.200639 -0.095669 -0.137097 -0.080382 0.061683 -0.046181 -0.116035 0.074372 -0.028700 -0.025087 -0.104776 0.000581 -0.093153 -0.089673 -0.115359 -0.100979 -0.065522 -0.072421 -0.051436 -0.041600 -0.067092 -0.102080 -0.062083 -0.128460 0.060374 -0.094861 -0.067093 -0.019844 -0.105964 -0.171154 -0.111553 -0.088344 -0.077191 -0.062789 -0.067233 -0.121622 -0.118219 -0.107361 -0.056076 -0.156528 -0.208740 -0.082909 -0.115473 -0.138487 -0.002550 -0.135713 -0.140876 -0.112622 -0.031032 -0.039740 -0.222639 -0.021673\n",
        "-0.340880\n",
        "-0.010669 -0.008709 -0.032664 -0.018814 0.011350 0.008296 0.002915 -0.095287 -0.053276 -0.044164 -0.038849 -0.056359 -0.003695 -0.010153 -0.038092 -0.035090 -0.037073 -0.066615 -0.033802 -0.075677 -0.018758 -0.040032 0.003286 -0.029362 -0.009171 -0.017405 -0.014017 -0.012406 -0.052292 -0.015908 -0.046171 -0.019382 0.011681 -0.015143 -0.044084 -0.047031 -0.035571 -0.085328 -0.054307 -0.068596 -0.024859 -0.095742 -0.011174 -0.046443 -0.027005 -0.034031 -0.030518 -0.037408 -0.060673 -0.054100 -0.046651 -0.004083 0.000992 -0.018888 -0.040737 -0.031346 -0.008029 -0.026529 -0.009894 -0.042470 -0.036844 -0.090009 -0.050305 -0.028628 -0.026699 -0.022709 -0.045672 -0.020419 -0.060096 -0.022889 -0.029577 -0.038078 0.017980 -0.023075 -0.040808 -0.018548 -0.033046 -0.059204 -0.018927 -0.057977 -0.069024 -0.108314 -0.086055 -0.041126 -0.045787 -0.024992 -0.020627 0.013067 -0.017615 0.013456 -0.043768 0.020712 -0.014131 -0.040229 -0.046692 -0.067922 -0.018519 -0.079815 -0.009910 -0.057836 -0.086255 -0.100761 -0.063211 -0.060912 -0.042272 -0.033872 -0.024101 0.033391 -0.065751 0.034346 -0.004670 -0.039934 -0.006180 -0.044691 -0.040406 -0.032319 -0.045666 -0.043812 -0.037480 -0.015752 -0.080369 -0.178663 -0.142052 -0.058952 -0.082488 -0.049549 -0.088193 -0.003946 -0.016048 0.037912 0.058860 0.075167 -0.073451 -0.019209 0.018548 0.216351 0.017831 0.002890 0.035588 0.021159 -0.060321 -0.078895 -0.063021 -0.047834 -0.032695 -0.001212 -0.033330 0.023866 -0.048300 -0.002942 -0.018228 -0.041226 0.015202 -0.004359 -0.062784 -0.067143 -0.030500 -0.029204 -0.000549 -0.086299 -0.080078 -0.077281 -0.066569 -0.018943 -0.022992 -0.044136 -0.030067 -0.045453 -0.034200 0.032039 -0.030979 -0.009580 -0.015083 -0.019785 -0.055096 -0.022597 -0.029376 -0.047693 -0.014020 -0.100962 -0.092450 -0.107647 -0.040029 -0.038917 -0.041689 -0.030097 -0.068286 -0.023911 -0.032546 -0.009965 -0.017693 -0.038751 -0.021534 -0.043604 -0.035164 0.007090 -0.016169 -0.037198 -0.009088 -0.034636 -0.063924 -0.138150 -0.078362 -0.053076 -0.014407 -0.028168 -0.033091 -0.029238 -0.064291 -0.021520 -0.054481 -0.014342 -0.014393 -0.005188 -0.014237 -0.013621 -0.003359 0.011550 -0.008074 -0.025254 -0.047624 -0.025409 -0.025882 -0.028033 -0.011064 -0.004319 -0.028525 -0.051974 -0.019050 -0.024563 -0.044468 -0.055931 0.018788 -0.035872 -0.029706 -0.046159 -0.051239 -0.051328 -0.022407 -0.042540 -0.042409 -0.015712 -0.005086 -0.023235 -0.006871 -0.013907 -0.017394 -0.073153 -0.058088 -0.044144 -0.056008 -0.068301 -0.000228 -0.037227 -0.055310 -0.049773 -0.034661 -0.041731 -0.069481 -0.052783\n",
        "-0.055524\n",
        "-0.123188 -0.130947 -0.111958 -0.092500 -0.064006 -0.058321 -0.025669 -0.087031 -0.151760 -0.123802 -0.051260 -0.117508 -0.007170 -0.055864 -0.057568 -0.019901 -0.083637 -0.166667 -0.120907 -0.201315 -0.145045 -0.276746 -0.100759 -0.134633 -0.070106 -0.111931 -0.053883 0.022319 -0.210716 -0.061789 0.003330 -0.035003 0.022154 -0.012293 -0.055640 -0.035159 -0.109552 -0.143841 -0.093055 -0.049893 -0.031713 -0.241993 -0.061046 -0.121126 -0.109371 -0.146593 -0.162277 -0.098806 -0.152864 -0.046845 -0.127258 -0.062047 0.022331 -0.141547 -0.116173 0.007897 0.010447 0.006594 -0.051794 -0.100264 -0.132250 -0.102319 -0.159004 -0.138800 -0.099854 -0.089698 -0.110680 -0.064832 -0.173070 -0.095212 -0.061200 -0.106989 0.051641 -0.147268 -0.140468 0.112907 -0.097235 -0.045134 -0.135412 -0.132155 -0.241182 -0.305034 -0.194811 -0.198144 -0.205512 -0.135497 -0.104642 0.057546 -0.155201 -0.001192 -0.015004 0.039015 -0.050663 -0.174674 -0.112794 -0.054663 -0.090941 -0.215613 0.057098 -0.018033 -0.149497 -0.325976 -0.182822 -0.242581 -0.163308 -0.161563 -0.039461 0.161612 -0.215408 0.067102 0.023724 -0.033456 -0.044086 -0.190129 -0.108819 0.049387 -0.061385 -0.017780 -0.150605 -0.029813 -0.012017 -0.200655 -0.252967 -0.176066 -0.234370 -0.206350 -0.224511 0.113818 0.189642 0.091300 0.246149 0.363320 -0.248464 -0.159198 -0.108556 0.762741 -0.000994 -0.018037 0.055581 0.079560 -0.188300 -0.229577 -0.203075 -0.242469 -0.121178 -0.025496 -0.100213 0.126989 -0.209933 -0.089257 -0.003835 -0.131354 0.066194 -0.075523 -0.130078 -0.105788 0.000434 0.049314 0.011046 -0.059083 -0.176051 -0.339247 -0.247409 -0.123293 -0.121230 -0.233246 -0.149617 -0.058124 -0.145198 0.070631 -0.052749 0.024122 -0.014955 -0.128921 -0.154684 0.059350 -0.082790 0.017489 0.050102 -0.211435 -0.199979 -0.279635 -0.128665 -0.177572 -0.184523 -0.117800 -0.180919 -0.071180 -0.183225 -0.036130 -0.026228 -0.076075 -0.000932 -0.195691 -0.145029 0.133778 -0.057217 -0.034427 -0.035964 -0.080485 -0.126296 -0.173102 -0.193967 -0.245837 -0.058266 -0.135580 -0.068270 0.012207 -0.207528 -0.086007 -0.137100 -0.061534 0.058701 -0.039873 -0.118549 0.068637 -0.020781 -0.037527 -0.098883 -0.007877 -0.080812 -0.118788 -0.104510 -0.094973 -0.050723 -0.075839 -0.050990 -0.022879 -0.070608 -0.103792 -0.038852 -0.118345 0.072276 -0.100905 -0.060607 -0.034271 -0.111141 -0.161607 -0.096421 -0.093272 -0.071481 -0.065006 -0.063472 -0.117289 -0.105645 -0.107677 -0.052738 -0.144368 -0.198796 -0.089154 -0.131919 -0.131590 0.007530 -0.119763 -0.132352 -0.112089 -0.042553 -0.042068 -0.210420 -0.009216\n",
        "-0.486091\n",
        "0.064477 0.081504 0.026874 0.044634 0.037261 0.056560 0.015655 -0.034844 0.028678 0.007178 -0.043457 0.036630 -0.011737 0.036687 -0.011747 -0.011310 0.036552 0.047534 0.008210 0.001234 0.046964 0.102268 0.025401 0.077903 0.067433 0.066706 0.044558 -0.082250 -0.008746 -0.029429 0.009296 -0.032597 0.020852 0.014712 0.037189 0.021274 0.008153 0.005058 -0.031590 -0.047687 -0.026414 -0.036252 0.029871 0.043835 0.054545 0.040781 0.044192 -0.034354 0.040529 0.009457 0.067047 0.013326 0.007208 0.077851 0.010252 -0.013107 -0.024258 -0.030598 0.025934 0.018936 0.015891 0.053638 0.090974 0.078377 0.029254 0.044463 0.067352 -0.086567 0.110640 0.016358 -0.008995 0.011897 -0.029884 0.079599 0.011748 -0.116792 0.017392 -0.041552 0.022211 0.030854 0.076297 0.138927 0.096433 0.117447 0.069136 0.062304 0.044012 -0.105353 0.050478 -0.049578 -0.016938 0.015190 0.045340 0.072201 0.005880 -0.063869 0.029111 -0.015537 -0.035881 -0.048838 0.002865 0.091267 0.096272 0.122886 0.021172 0.079588 -0.025038 -0.075315 0.107730 0.001750 0.017535 0.014763 0.033558 0.060733 0.023844 -0.139122 -0.021014 -0.022912 0.043800 -0.069368 -0.199700 -0.222440 -0.027857 0.083226 -0.058485 0.035489 0.035407 -0.129288 -0.064453 -0.061740 -0.147288 -0.223603 -0.033822 0.120373 0.144032 -0.310629 0.078233 0.220361 0.117385 0.075370 0.056549 0.122092 0.085682 0.111980 0.077644 0.015114 0.054596 -0.018600 0.112969 -0.071757 -0.069830 -0.050383 -0.002183 0.063235 0.031887 -0.059578 0.002797 -0.082785 -0.037662 -0.087120 0.057080 0.106612 0.150975 0.057186 0.052134 0.131403 0.087739 -0.049076 0.078160 -0.083373 -0.009038 -0.012009 0.026432 0.072579 0.015223 -0.063027 -0.026930 -0.091203 -0.044165 -0.024702 0.057721 0.031649 0.066977 0.077464 0.092783 0.043605 0.089750 -0.082724 0.092782 -0.002469 -0.018213 0.038463 0.039650 0.083644 0.057403 -0.102030 0.016508 -0.033885 -0.049141 -0.037024 0.050683 0.027857 0.095785 0.100878 0.028308 0.032504 0.039158 -0.099253 0.089059 0.019542 0.029946 0.009274 0.027114 0.043419 0.039742 -0.071644 -0.044944 -0.082925 -0.015797 -0.067424 0.044159 0.084274 0.085030 0.017850 0.029234 0.025825 0.053502 -0.116168 0.042321 -0.057481 -0.052808 -0.016987 -0.004995 0.040141 0.047972 -0.027220 0.000804 0.005358 0.017293 -0.045098 0.014360 -0.008490 0.061663 0.046281 0.073928 0.035608 0.024990 -0.068028 0.086265 0.059148 -0.002630 0.042496 -0.001227 0.023778 0.052048 -0.013567 -0.013077 -0.037425 0.062453 -0.030843\n",
        "-0.022806\n",
        "-0.155360 -0.066650 -0.119927 -0.094214 -0.114445 -0.091846 -0.136155 -0.172902 -0.115423 -0.160518 -0.114837 -0.079660 -0.019202 -0.060908 -0.085990 -0.117745 -0.046584 -0.141824 -0.246103 -0.279503 -0.143058 -0.169634 -0.182320 -0.129732 -0.048653 -0.169455 -0.076733 -0.169466 -0.184966 -0.169890 0.002180 -0.172955 0.013730 -0.028397 0.004138 -0.042734 -0.116962 -0.134944 -0.167129 -0.125596 -0.168811 -0.282752 -0.171291 -0.168392 -0.172167 -0.202640 -0.248134 -0.217014 -0.086144 -0.001537 -0.083492 -0.157991 -0.029741 -0.220335 -0.097208 0.097354 0.015144 0.103367 -0.112372 -0.098831 -0.201880 -0.080104 -0.209565 -0.232628 -0.157016 -0.164620 -0.111527 -0.294555 -0.119790 -0.125459 -0.089252 -0.212894 -0.100429 -0.203062 -0.114994 0.192756 -0.069508 0.071803 -0.210157 -0.097825 -0.252878 -0.145768 -0.091934 -0.213814 -0.210380 -0.186967 -0.166408 -0.179349 -0.252828 -0.089035 -0.004230 -0.036768 -0.091384 -0.195063 -0.191896 -0.062071 -0.138017 -0.309404 0.031256 -0.004515 -0.216068 -0.256101 -0.202169 -0.267218 -0.190781 -0.231367 -0.115307 -0.060081 -0.188930 0.060493 -0.034551 -0.071494 -0.057059 -0.239721 -0.107063 0.087876 -0.059604 0.019725 -0.202176 -0.093566 -0.181354 -0.264459 -0.295692 -0.214066 -0.370961 -0.338641 -0.316944 0.234029 0.414957 0.120721 0.366519 0.432901 -0.568992 -0.149865 -0.162368 1.018977 0.081562 0.078209 0.118584 0.154448 -0.235470 -0.204636 -0.331409 -0.306707 -0.197640 -0.073851 -0.143376 0.185353 -0.169884 -0.321067 -0.066749 -0.286210 -0.022574 -0.054040 -0.122530 -0.061010 0.077675 0.097011 0.021096 0.015353 -0.124458 -0.230960 -0.206823 -0.169983 -0.187876 -0.244156 -0.240031 -0.063724 -0.267759 0.033642 -0.073610 -0.017757 -0.012209 -0.181368 -0.235108 0.063289 -0.153981 -0.050173 -0.037459 -0.267356 -0.171665 -0.417932 -0.273538 -0.254723 -0.247825 -0.224480 -0.193678 -0.145072 -0.256420 -0.011551 -0.077899 -0.105522 -0.025237 -0.228389 -0.181261 0.186442 -0.004426 0.027185 -0.116652 -0.084278 -0.121933 -0.164385 -0.178147 -0.323689 -0.137000 -0.203485 -0.119417 -0.211980 -0.209841 -0.076433 -0.108906 -0.116705 0.063359 -0.044561 -0.147860 0.125350 -0.038749 -0.037735 -0.218418 -0.003540 0.012043 -0.084678 -0.050313 -0.171633 -0.080256 -0.104211 -0.048266 -0.235514 -0.091574 -0.171429 -0.070577 -0.144370 0.069851 -0.090666 -0.099082 -0.005035 -0.135922 -0.234739 -0.190181 -0.174084 -0.045712 -0.082394 -0.066834 -0.160113 -0.146583 -0.165018 -0.069594 -0.271854 -0.228822 -0.068588 -0.170662 -0.124861 -0.048595 -0.175959 -0.146380 -0.140901 -0.029176 -0.105962 -0.275622 -0.098183\n",
        "-0.400644\n",
        "0.025863 0.037665 -0.000525 0.014878 0.021414 0.030382 0.009428 -0.061600 -0.020822 -0.012624 -0.047037 -0.018928 -0.000238 0.008169 -0.024889 -0.029390 -0.004440 -0.019274 -0.016174 -0.047775 0.011759 0.029397 0.023862 0.021848 0.033655 0.024533 0.016207 -0.043899 -0.034766 -0.022761 -0.025268 -0.021901 0.021216 -0.000587 -0.008505 -0.023175 -0.015196 -0.043818 -0.042678 -0.053814 -0.023551 -0.057142 0.012955 -0.000702 0.009950 0.010907 0.012455 -0.037645 -0.018585 -0.021604 -0.001600 0.007529 0.004206 0.035953 -0.015788 -0.024328 -0.014530 -0.024328 -0.000313 -0.024849 -0.010675 -0.025084 0.016239 0.018910 -0.000875 0.004668 0.005844 -0.039350 0.023010 -0.001101 -0.019054 -0.012512 -0.002170 0.026797 -0.021990 -0.066065 -0.011275 -0.054704 0.006118 -0.013365 -0.001913 0.011002 -0.005205 0.034078 0.014057 0.015080 0.014525 -0.029288 0.011781 -0.012597 -0.031152 0.020710 0.017857 0.013660 -0.025139 -0.066415 -0.000978 -0.050104 -0.022322 -0.059622 -0.043285 -0.012837 0.013283 0.027956 -0.011257 0.025595 -0.026572 -0.011584 0.013659 0.021266 0.003634 -0.015803 0.010474 0.004674 -0.015586 -0.083654 -0.032707 -0.034903 0.000139 -0.039439 -0.136470 -0.197467 -0.094446 -0.003130 -0.072347 -0.006393 -0.032628 -0.049245 -0.045201 -0.003011 -0.023094 -0.049691 -0.053097 0.046527 0.076725 -0.027360 0.041150 0.089747 0.073798 0.036583 -0.007736 0.013534 0.007252 0.030825 0.011398 0.012524 0.005745 0.012027 0.024996 -0.023544 -0.041737 -0.042142 0.004928 0.029623 -0.024135 -0.066536 -0.021604 -0.056537 -0.013476 -0.086867 -0.017341 0.011027 0.027187 0.018723 0.018834 0.041114 0.034509 -0.053383 0.016396 -0.014569 -0.030306 -0.013292 0.012070 0.028589 -0.025140 -0.050459 -0.032340 -0.071067 -0.031607 -0.066966 -0.019422 -0.045531 -0.001855 0.011790 0.020760 0.003760 0.002608 -0.036819 0.018927 0.000181 -0.014118 -0.001634 0.007490 0.013175 0.000599 -0.047058 -0.000328 -0.037104 -0.022015 -0.032448 -0.020274 -0.064078 -0.002927 0.019071 -0.000378 0.006713 -0.004564 -0.062295 0.007087 0.001102 -0.012645 -0.003796 -0.002150 0.021510 0.016576 -0.041921 -0.017699 -0.022329 -0.011881 -0.040682 -0.003955 0.029841 0.024216 -0.007419 -0.002425 0.008207 0.014743 -0.079612 0.010590 -0.035174 -0.051610 -0.032677 0.005272 0.000700 0.006142 -0.043504 -0.027036 -0.020666 -0.002434 -0.041886 -0.008996 -0.009557 0.022626 0.018796 0.035262 0.009018 0.003357 -0.071315 0.014485 0.007422 -0.028467 -0.017604 0.001882 -0.011179 -0.011990 -0.032207 -0.023198 -0.041724 -0.006724 -0.037791\n",
        "1.505785\n",
        "-2.427554 -0.948803 -1.203377 0.899563 4.836003 -0.879909 -0.390292 1.529532 -3.375759 -2.322754 1.113470 -0.850439 -1.474451 0.773085 -0.864312 -0.167740 -0.860402 0.443572 -1.095174 0.112439\n",
        "-1.503021\n",
        "2.424830 0.944486 1.203099 -0.894367 -4.835998 0.880314 0.384920 -1.533527 3.375663 2.326251 -1.114913 0.844994 1.476537 -0.774986 0.871945 0.166729 0.860142 -0.441613 1.096362 -0.111924\n",
        "\"\"\")\n",
        "\n",
        "    WEIGHTS.close()\n",
        "\n",
        "    os.system(\"touch WEIGHTS_READY\")\n",
        "\n",
        "# coloring scripts\n",
        "\n",
        "vars['pymol'] = \"pymol.py\"\n",
        "vars['pymol_CBS'] = \"pymol_CBS.py\"\n",
        "\n",
        "# write pymol script\n",
        "PYMOL = open(vars['working_dir'] + vars['pymol'], 'w')\n",
        "PYMOL.write(\"\"\"\n",
        "\n",
        "# Define a Python subroutine to colour atoms by B-factor, using predefined intervals\n",
        "\n",
        "\n",
        "def colour_consurf(selection=\"all\"):\n",
        "\n",
        "    # Colour other chains gray, while maintaining\n",
        "    # oxygen in red, nitrogen in blue and hydrogen in white\n",
        "    cmd.color(\"gray\", selection)\n",
        "    cmd.util.cnc()\n",
        "\n",
        "    # These are constants\n",
        "    minimum = 0.0\n",
        "    maximum = 9.0\n",
        "    n_colours = 9\n",
        "    # Colours are calculated by dividing the RGB colours by 255\n",
        "    # RGB = [[16,200,209],[140,255,255],[215,255,255],[234,255,255],[255,255,255],\n",
        "    #        [252,237,244],[250,201,222],[240,125,171],[160,37,96]]\n",
        "    colours = [\n",
        "                [0.039215686, 0.490196078, 0.509803922],\n",
        "                [0.294117647, 0.68627451, 0.745098039],\n",
        "                [0.647058824, 0.862745098, 0.901960784],\n",
        "                [0.843137255, 0.941176471, 0.941176471],\n",
        "                [1, 1, 1],\n",
        "                [0.980392157, 0.921568627, 0.960784314],\n",
        "                [0.980392157, 0.784313725, 0.862745098],\n",
        "                [0.941176471, 0.490196078, 0.666666667],\n",
        "                [0.62745098, 0.156862745, 0.37254902]]\n",
        "    bin_size = (maximum - minimum) / n_colours\n",
        "\n",
        "    # Loop through colour intervals\n",
        "    for i in range(n_colours):\n",
        "\n",
        "        lower = minimum + (i + 1) * bin_size\n",
        "        upper = lower + bin_size\n",
        "        colour = colours[i]\n",
        "\n",
        "        # Print out B-factor limits and the colour for this group\n",
        "        print(lower, \" - \", upper, \" = \", colour)\n",
        "\n",
        "        # Define a unique name for the atoms which fall into this group\n",
        "        group = selection + \"_group_\" + str(i + 1)\n",
        "\n",
        "        # Compose a selection command which will select all atoms which are\n",
        "        #\ta) in the original selection, AND\n",
        "        #\tb) have B factor in range lower <= b < upper\n",
        "        sel_string = selection + \" & ! b < \" + str(lower)\n",
        "\n",
        "        if(i < n_colours):\n",
        "            sel_string += \" & b < \" + str(upper)\n",
        "        else:\n",
        "            sel_string += \" & ! b > \" + str(upper)\n",
        "\n",
        "        # Select the atoms\n",
        "        cmd.select(group, sel_string)\n",
        "\n",
        "        # Create a new colour\n",
        "        colour_name = \"colour_\" + str(i + 1)\n",
        "        cmd.set_color(colour_name, colour)\n",
        "\n",
        "        # Colour them\n",
        "        cmd.color(colour_name, group)\n",
        "\n",
        "\n",
        "    # Create new colour for insufficient sequences\n",
        "    # RGB_colour = [255,255,150]\n",
        "    insuf_colour = [1, 1, 0.588235294]\n",
        "    cmd.set_color(\"insufficient_colour\", insuf_colour)\n",
        "\n",
        "    # Colour atoms with B-factor of 10 using the new colour\n",
        "    cmd.select(\"insufficient\", selection + \" & b = 10\")\n",
        "    cmd.color(\"insufficient_colour\", \"insufficient\")\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "# Make command available in PyMOL\n",
        "cmd.extend(\"colour_consurf\", colour_consurf)\n",
        "colour_consurf()\n",
        "\n",
        "# Hide external molecules (hetatm selection)\n",
        "cmd.select(\"het_atoms\", \"hetatm\")\n",
        "cmd.deselect()\n",
        "cmd.hide(selection=\"het_atoms\")\n",
        "cmd.show(selection=\"all\",representation=\"cartoon\")\n",
        "\n",
        "# Save PyMOL session to pse file\n",
        "cmd.save(\"consurf_pymol_session.pse\")\n",
        "\n",
        "\"\"\")\n",
        "\n",
        "PYMOL.close()\n",
        "\n",
        "# write pymol color blind script\n",
        "PYMOL_CBS = open(vars['working_dir'] + vars['pymol_CBS'], 'w')\n",
        "PYMOL_CBS.write(\"\"\"\n",
        "\n",
        "# Define a Python subroutine to colour atoms by B-factor, using predefined intervals\n",
        "\n",
        "\n",
        "def colour_consurf_CBS(selection=\"all\"):\n",
        "\n",
        "    # Colour other chains gray, while maintaining\n",
        "    # oxygen in red, nitrogen in blue and hydrogen in white\n",
        "    cmd.color(\"gray\", selection)\n",
        "    cmd.util.cnc()\n",
        "\n",
        "    # These are constants\n",
        "    minimum = 0.0\n",
        "    maximum = 9.0\n",
        "    n_colours = 9\n",
        "    # Colours are calculated by dividing the RGB colours by 255\n",
        "    # RGB = [[[27,120,55],[90,174,97],[166,219,160],[217,240,211],[255,255,255],\n",
        "    #        [231,212,232],[194,165,207],[153,112,171],[118,42,131]]\n",
        "    colours = [\n",
        "                [0.058823529, 0.352941176, 0.137254902],\n",
        "                [0.352941176, 0.68627451, 0.37254902],\n",
        "                [0.647058824, 0.862745098, 0.62745098],\n",
        "                [0.843137255, 0.941176471, 0.823529412],\n",
        "                [1, 1, 1],\n",
        "                [0.901960784, 0.823529412, 0.901960784],\n",
        "                [0.764705882, 0.647058824, 0.803921569],\n",
        "                [0.607843137, 0.431372549, 0.666666667],\n",
        "                [0.470588235, 0.156862745, 0.509803922]]\n",
        "    bin_size = (maximum - minimum) / n_colours\n",
        "\n",
        "    # Loop through colour intervals\n",
        "    for i in range(n_colours):\n",
        "\n",
        "        lower = minimum + (i + 1) * bin_size\n",
        "        upper = lower + bin_size\n",
        "        colour = colours[i]\n",
        "\n",
        "        # Print out B-factor limits and the colour for this group\n",
        "        print(lower, \" - \", upper, \" = \", colour)\n",
        "\n",
        "        # Define a unique name for the atoms which fall into this group\n",
        "        group = selection + \"_group_\" + str(i + 1)\n",
        "\n",
        "        # Compose a selection command which will select all atoms which are\n",
        "        #\ta) in the original selection, AND\n",
        "        #\tb) have B factor in range lower <= b < upper\n",
        "        sel_string = selection + \" & ! b < \" + str(lower)\n",
        "\n",
        "        if(i < n_colours):\n",
        "            sel_string += \" & b < \" + str(upper)\n",
        "        else:\n",
        "            sel_string += \" & ! b > \" + str(upper)\n",
        "\n",
        "        # Select the atoms\n",
        "        cmd.select(group, sel_string)\n",
        "\n",
        "        # Create a new colour\n",
        "        colour_name = \"colour_\" + str(i + 1)\n",
        "        cmd.set_color(colour_name, colour)\n",
        "\n",
        "        # Colour them\n",
        "        cmd.color(colour_name, group)\n",
        "\n",
        "\n",
        "    # Create new colour for insufficient sequences\n",
        "    # RGB_colour = [255,255,150]\n",
        "    insuf_colour = [1, 1, 0.588235294]\n",
        "    cmd.set_color(\"insufficient_colour\", insuf_colour)\n",
        "\n",
        "    # Colour atoms with B-factor of 10 using the new colour\n",
        "    cmd.select(\"insufficient\", selection + \" & b = 10\")\n",
        "    cmd.color(\"insufficient_colour\", \"insufficient\")\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "# Make command available in PyMOL\n",
        "cmd.extend(\"colour_consurf_CBS\", colour_consurf_CBS)\n",
        "colour_consurf_CBS()\n",
        "\n",
        "# Hide external molecules (hetatm selection)\n",
        "cmd.select(\"het_atoms\", \"hetatm\")\n",
        "cmd.deselect()\n",
        "cmd.hide(selection=\"het_atoms\")\n",
        "cmd.show(selection=\"all\",representation=\"cartoon\")\n",
        "\n",
        "# Save PyMOL session to pse file\n",
        "cmd.save(\"consurf_CBS_pymol_session.pse\")\n",
        "\n",
        "\"\"\")\n",
        "\n",
        "PYMOL_CBS.close()\n",
        "\n",
        "# get fonts\n",
        "if not os.path.isfile(\"FONTS_READY\"):\n",
        "\n",
        "    print(\"Getting fonts.\")\n",
        "    os.system(\"wget \\\"https://sourceforge.net/projects/dejavu/files/dejavu/2.37/dejavu-fonts-ttf-2.37.tar.bz2\\\" -O \\\"FONTS.tar.bz2\\\"\")\n",
        "    FONTS = tarfile.open(\"FONTS.tar.bz2\")\n",
        "    FONTS.extractall()\n",
        "    FONTS.close()\n",
        "    os.system(\"touch FONTS_READY\")\n",
        "\n",
        "# install py3dmol\n",
        "if not os.path.isfile(\"PY3DMOL_READY\"):\n",
        "\n",
        "    print(\"Installing py3dmol.\")\n",
        "    os.system(\"pip install py3Dmol\")\n",
        "    os.system(\"touch PY3DMOL_READY\")\n",
        "\n",
        "# install prank\n",
        "if not os.path.isfile(\"PRANK_READY\"):\n",
        "\n",
        "    print(\"Installing prank.\")\n",
        "    os.system(\"apt-get install prank\")\n",
        "    os.system(\"touch PRANK_READY\")\n",
        "\n",
        "# install clustalw\n",
        "if not os.path.isfile(\"CLUSTALW_READY\"):\n",
        "\n",
        "    print(\"Installing CLUSTALW.\")\n",
        "    os.system(\"apt-get -qq install -y clustalw\")\n",
        "    os.system(\"touch CLUSTALW_READY\")\n",
        "\n",
        "# install prottest\n",
        "if not os.path.isfile(\"PROTTEST_READY\"):\n",
        "\n",
        "    print(\"Installing prottest.\")\n",
        "    #os.system(\"git clone https://github.com/ddarriba/prottest3.git\")\n",
        "    response = urllib.request.urlopen(\"https://github.com/ddarriba/prottest3/releases/download/3.4.2-release/prottest-3.4.2-20160508.tar.gz\").read()\n",
        "\n",
        "    prottest_zip = \"prottest.tar.gz\"\n",
        "    PROTTEST_ZIP = open(prottest_zip, 'wb')\n",
        "    PROTTEST_ZIP.write(response)\n",
        "    PROTTEST_ZIP.close()\n",
        "\n",
        "    PROTTEST_EXTRACT = tarfile.open(prottest_zip)\n",
        "    PROTTEST_EXTRACT.extractall()\n",
        "    PROTTEST_EXTRACT.close()\n",
        "    os.system(\"touch PROTTEST_READY\")\n",
        "\n",
        "# install py3dmol\n",
        "if not os.path.isfile(\"PY3DMOL_READY\"):\n",
        "\n",
        "    print(\"Installing py3dmol.\")\n",
        "    os.system(\"pip install py3Dmol\")\n",
        "    os.system(\"touch PY3DMOL_READY\")\n",
        "\n",
        "# install biopython\n",
        "if not os.path.isfile(\"BIOPYTHON_READY\"):\n",
        "\n",
        "    print(\"Installing biopython.\")\n",
        "    os.system(\"pip install biopython\")\n",
        "    os.system(\"touch BIOPYTHON_READY\")\n",
        "\n",
        "# install fpdf\n",
        "if not os.path.isfile(\"FPDF_READY\"):\n",
        "\n",
        "    print(\"Installing fpdf.\")\n",
        "    os.system(\"pip install fpdf\")\n",
        "    os.system(\"touch FPDF_READY\")\n",
        "\n",
        "# install muscle\n",
        "if not os.path.isfile(\"MUSCLE_READY\"):\n",
        "\n",
        "    print(\"Installing muscle.\")\n",
        "    os.system(\"apt-get -qq install muscle\")\n",
        "    os.system(\"touch MUSCLE_READY\")\n",
        "\n",
        "# install mafft\n",
        "if not os.path.isfile(\"MAFFT_READY\"):\n",
        "\n",
        "    print(\"Installing mafft.\")\n",
        "    os.system(\"apt-get install mafft\")\n",
        "    os.system(\"touch MAFFT_READY\")\n",
        "\n",
        "# install rate4site\n",
        "if not os.path.isfile(\"RATE4SITE_READY\"):\n",
        "\n",
        "    print(\"Installing rate4site.\")\n",
        "\n",
        "    # create directory for rate4site\n",
        "    os.system(\"git clone https://github.com/barakav/r4s_for_collab.git\")\n",
        "\n",
        "    # create directory for rate4site slow\n",
        "    shutil.copytree(vars['rate4site_dir'], vars['rate4site_slow_dir'])\n",
        "\n",
        "    # make rate4site\n",
        "    try:\n",
        "\n",
        "        os.chdir(vars['rate4site_dir'])\n",
        "        os.system(\"make\")\n",
        "        os.chdir(vars['root_dir'])\n",
        "\n",
        "    except Exception as e:\n",
        "\n",
        "        print(e)\n",
        "        os.chdir(vars['root_dir'])\n",
        "        raise(\"Installing rate4site failed.\")\n",
        "\n",
        "    # change the make file and make rate4site slow\n",
        "    try:\n",
        "\n",
        "        os.chdir(vars['rate4site_slow_dir'])\n",
        "        os.remove(\"Makefile\") # delete regular file\n",
        "        os.rename(\"Makefile_slow\", \"Makefile\") # change to slow file\n",
        "        os.system(\"make\")\n",
        "        os.chdir(vars['root_dir'])\n",
        "\n",
        "    except Exception as e:\n",
        "\n",
        "        print(e)\n",
        "        os.chdir(vars['root_dir'])\n",
        "        raise(\"Installing rate4site failed.\")\n",
        "\n",
        "    os.system(\"touch RATE4SITE_READY\")\n",
        "\n",
        "\n",
        "# install cd-hit\n",
        "if not os.path.isfile(\"CDHIT_READY\"):\n",
        "\n",
        "    print(\"installing cd-hit.\")\n",
        "    os.system(\"git clone https://github.com/weizhongli/cdhit.git\")\n",
        "    \"\"\"\n",
        "    response = urllib.request.urlopen(\"https://github.com/weizhongli/cdhit/archive/refs/heads/master.zip\").read()\n",
        "\n",
        "    cd_hit_zip = \"cd-hit.zip\"\n",
        "    CD_HIT_ZIP = open(cd_hit_zip, 'wb')\n",
        "    CD_HIT_ZIP.write(response)\n",
        "    CD_HIT_ZIP.close()\n",
        "\n",
        "    CD_HIT_EXTRACT = ZipFile(cd_hit_zip, 'r')\n",
        "    CD_HIT_EXTRACT.extractall()\n",
        "    CD_HIT_EXTRACT.close()\n",
        "    \"\"\"\n",
        "    try:\n",
        "\n",
        "        os.chdir(\"cdhit\")\n",
        "        os.system(\"make\")\n",
        "        os.chdir(vars['root_dir'])\n",
        "\n",
        "    except Exception as e:\n",
        "\n",
        "        print(e)\n",
        "        os.chdir(vars['root_dir'])\n",
        "        raise(\"Installing cd-hit failed.\")\n",
        "\n",
        "    os.system(\"touch CDHIT_READY\")\n",
        "\n",
        "# install mmseqs2\n",
        "if not os.path.isfile(\"COLABFOLD_READY\"):\n",
        "\n",
        "    print(\"Installing colabfold.\")\n",
        "    os.system(\"pip install -q --no-warn-conflicts 'colabfold[alphafold-minus-jax] @ git+https://github.com/sokrypton/ColabFold'\")\n",
        "    os.system(\"pip install --upgrade dm-haiku\")\n",
        "    os.system(\"ln -s /usr/local/lib/python3.*/dist-packages/colabfold colabfold\")\n",
        "    os.system(\"ln -s /usr/local/lib/python3.*/dist-packages/alphafold alphafold\")\n",
        "    # patch for jax > 0.3.25\n",
        "    os.system(\"sed -i 's/weights = jax.nn.softmax(logits)/logits=jnp.clip(logits,-1e8,1e8);weights=jax.nn.softmax(logits)/g' alphafold/model/modules.py\")\n",
        "    os.system(\"pip install -q biopython==1.81\")\n",
        "    os.system(\"touch COLABFOLD_READY\")\n",
        "\n",
        "# create matrix for pairwise alignment\n",
        "if not os.path.isfile(\"MATRIX_READY\"):\n",
        "\n",
        "    MATRIX = open(\"matrix.txt\", 'w')\n",
        "    MATRIX.write(\"\"\"#  Matrix made by matblas from blosum62.iij\n",
        "#  * column uses minimum score\n",
        "#  BLOSUM Clustered Scoring Matrix in 1/2 Bit Units\n",
        "#  Blocks Database = /data/blocks_5.0/blocks.dat\n",
        "#  Cluster Percentage: >= 62\n",
        "#  Entropy =   0.6979, Expected =  -0.5209\n",
        "     A    R    N    D    C    Q    E    G    H    I    L    K    M    F    P    S    T    W    Y    V    B    Z    X    *\n",
        "A  4.0 -1.0 -2.0 -2.0  0.0 -1.0 -1.0  0.0 -2.0 -1.0 -1.0 -1.0 -1.0 -2.0 -1.0  1.0  0.0 -3.0 -2.0  0.0 -2.0 -1.0  0.0 -4.0\n",
        "R -1.0  5.0  0.0 -2.0 -3.0  1.0  0.0 -2.0  0.0 -3.0 -2.0  2.0 -1.0 -3.0 -2.0 -1.0 -1.0 -3.0 -2.0 -3.0 -1.0  0.0  0.0 -4.0\n",
        "N -2.0  0.0  6.0  1.0 -3.0  0.0  0.0  0.0  1.0 -3.0 -3.0  0.0 -2.0 -3.0 -2.0  1.0  0.0 -4.0 -2.0 -3.0  3.0  0.0  0.0 -4.0\n",
        "D -2.0 -2.0  1.0  6.0 -3.0  0.0  2.0 -1.0 -1.0 -3.0 -4.0 -1.0 -3.0 -3.0 -1.0  0.0 -1.0 -4.0 -3.0 -3.0  4.0  1.0  0.0 -4.0\n",
        "C  0.0 -3.0 -3.0 -3.0  9.0 -3.0 -4.0 -3.0 -3.0 -1.0 -1.0 -3.0 -1.0 -2.0 -3.0 -1.0 -1.0 -2.0 -2.0 -1.0 -3.0 -3.0  0.0 -4.0\n",
        "Q -1.0  1.0  0.0  0.0 -3.0  5.0  2.0 -2.0  0.0 -3.0 -2.0  1.0  0.0 -3.0 -1.0  0.0 -1.0 -2.0 -1.0 -2.0  0.0  3.0  0.0 -4.0\n",
        "E -1.0  0.0  0.0  2.0 -4.0  2.0  5.0 -2.0  0.0 -3.0 -3.0  1.0 -2.0 -3.0 -1.0  0.0 -1.0 -3.0 -2.0 -2.0  1.0  4.0  0.0 -4.0\n",
        "G  0.0 -2.0  0.0 -1.0 -3.0 -2.0 -2.0  6.0 -2.0 -4.0 -4.0 -2.0 -3.0 -3.0 -2.0  0.0 -2.0 -2.0 -3.0 -3.0 -1.0 -2.0  0.0 -4.0\n",
        "H -2.0  0.0  1.0 -1.0 -3.0  0.0  0.0 -2.0  8.0 -3.0 -3.0 -1.0 -2.0 -1.0 -2.0 -1.0 -2.0 -2.0  2.0 -3.0  0.0  0.0  0.0 -4.0\n",
        "I -1.0 -3.0 -3.0 -3.0 -1.0 -3.0 -3.0 -4.0 -3.0  4.0  2.0 -3.0  1.0  0.0 -3.0 -2.0 -1.0 -3.0 -1.0  3.0 -3.0 -3.0  0.0 -4.0\n",
        "L -1.0 -2.0 -3.0 -4.0 -1.0 -2.0 -3.0 -4.0 -3.0  2.0  4.0 -2.0  2.0  0.0 -3.0 -2.0 -1.0 -2.0 -1.0  1.0 -4.0 -3.0  0.0 -4.0\n",
        "K -1.0  2.0  0.0 -1.0 -3.0  1.0  1.0 -2.0 -1.0 -3.0 -2.0  5.0 -1.0 -3.0 -1.0  0.0 -1.0 -3.0 -2.0 -2.0  0.0  1.0  0.0 -4.0\n",
        "M -1.0 -1.0 -2.0 -3.0 -1.0  0.0 -2.0 -3.0 -2.0  1.0  2.0 -1.0  5.0  0.0 -2.0 -1.0 -1.0 -1.0 -1.0  1.0 -3.0 -1.0  0.0 -4.0\n",
        "F -2.0 -3.0 -3.0 -3.0 -2.0 -3.0 -3.0 -3.0 -1.0  0.0  0.0 -3.0  0.0  6.0 -4.0 -2.0 -2.0  1.0  3.0 -1.0 -3.0 -3.0  0.0 -4.0\n",
        "P -1.0 -2.0 -2.0 -1.0 -3.0 -1.0 -1.0 -2.0 -2.0 -3.0 -3.0 -1.0 -2.0 -4.0  7.0 -1.0 -1.0 -4.0 -3.0 -2.0 -2.0 -1.0  0.0 -4.0\n",
        "S  1.0 -1.0  1.0  0.0 -1.0  0.0  0.0  0.0 -1.0 -2.0 -2.0  0.0 -1.0 -2.0 -1.0  4.0  1.0 -3.0 -2.0 -2.0  0.0  0.0  0.0 -4.0\n",
        "T  0.0 -1.0  0.0 -1.0 -1.0 -1.0 -1.0 -2.0 -2.0 -1.0 -1.0 -1.0 -1.0 -2.0 -1.0  1.0  5.0 -2.0 -2.0  0.0 -1.0 -1.0  0.0 -4.0\n",
        "W -3.0 -3.0 -4.0 -4.0 -2.0 -2.0 -3.0 -2.0 -2.0 -3.0 -2.0 -3.0 -1.0  1.0 -4.0 -3.0 -2.0 11.0  2.0 -3.0 -4.0 -3.0  0.0 -4.0\n",
        "Y -2.0 -2.0 -2.0 -3.0 -2.0 -1.0 -2.0 -3.0  2.0 -1.0 -1.0 -2.0 -1.0  3.0 -3.0 -2.0 -2.0  2.0  7.0 -1.0 -3.0 -2.0  0.0 -4.0\n",
        "V  0.0 -3.0 -3.0 -3.0 -1.0 -2.0 -2.0 -3.0 -3.0  3.0  1.0 -2.0  1.0 -1.0 -2.0 -2.0  0.0 -3.0 -1.0  4.0 -3.0 -2.0  0.0 -4.0\n",
        "B -2.0 -1.0  3.0  4.0 -3.0  0.0  1.0 -1.0  0.0 -3.0 -4.0  0.0 -3.0 -3.0 -2.0  0.0 -1.0 -4.0 -3.0 -3.0  4.0  1.0  0.0 -4.0\n",
        "Z -1.0  0.0  0.0  1.0 -3.0  3.0  4.0 -2.0  0.0 -3.0 -3.0  1.0 -1.0 -3.0 -1.0  0.0 -1.0 -3.0 -2.0 -2.0  1.0  4.0  0.0 -4.0\n",
        "X  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  4.0 -4.0\n",
        "* -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0  1.0\n",
        "\"\"\")\n",
        "\n",
        "    MATRIX.close()\n",
        "\n",
        "    os.system(\"touch MATRIX_READY\")\n",
        "\n",
        "os.chdir(vars['job_name'])\n",
        "\n",
        "print(\"\\n\")\n",
        "\n",
        "from datetime import date\n",
        "from datetime import datetime\n",
        "from colabfold.batch import get_msa_and_templates\n",
        "from Bio import AlignIO\n",
        "from Bio import SeqIO\n",
        "from Bio import SearchIO\n",
        "from Bio import Phylo\n",
        "from Bio import Align\n",
        "from Bio.Align import substitution_matrices\n",
        "from google.colab import files, output\n",
        "from Bio.Blast import NCBIWWW\n",
        "import Bio\n",
        "import fpdf\n",
        "import math\n",
        "import py3Dmol\n",
        "import uuid\n",
        "\n",
        "def choose_chain(PDB_name):\n",
        "\n",
        "    # in the case of nmr we choose the first model\n",
        "    try:\n",
        "\n",
        "        PDB_FILE = open(PDB_name, 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        raise Exception(\"Can't read the PDB file.\")\n",
        "\n",
        "    # later we find the format (PDB/mmCIF) and we change the files name\n",
        "    temp_pdb = \"temp_file.txt\"\n",
        "    try:\n",
        "\n",
        "        PDB_TEMP = open(temp_pdb, 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        raise Exception(\"Error: choose_chain - Can't open file for writing.\")\n",
        "\n",
        "    line = PDB_FILE.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        if re.match(r'^MODEL\\s+2', line):\n",
        "\n",
        "            # nmr found\n",
        "            break\n",
        "\n",
        "        else:\n",
        "\n",
        "            PDB_TEMP.write(line)\n",
        "\n",
        "        line = PDB_FILE.readline()\n",
        "\n",
        "    PDB_TEMP.close()\n",
        "    PDB_FILE.close()\n",
        "\n",
        "    # we now find the chains\n",
        "    PDB_FILE = open(temp_pdb, 'r')\n",
        "\n",
        "\n",
        "\n",
        "    chains = []\n",
        "    last_chain_found = \"\"\n",
        "    chain_index = 0\n",
        "    nmr_struct = False\n",
        "\n",
        "    # for mmCIF format\n",
        "    found_auth_comp_id_column = False\n",
        "    found_auth_asym_id_column = False\n",
        "    found_label_comp_id_column = False\n",
        "    found_label_asym_id_column = False\n",
        "    found_column_numbers = False\n",
        "    auth_comp_id_column = 0\n",
        "    auth_asym_id_column = 0\n",
        "    label_comp_id_column = 0\n",
        "    label_asym_id_column = 0\n",
        "\n",
        "    line = PDB_FILE.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        if line[:4] == \"ATOM\":\n",
        "\n",
        "            if (found_auth_comp_id_column or found_label_comp_id_column) and (found_auth_asym_id_column or found_label_asym_id_column):\n",
        "\n",
        "                # mmCIF format\n",
        "                found_column_numbers = True\n",
        "                words = line.split()\n",
        "                num_columns = len(words)\n",
        "                if found_auth_comp_id_column:\n",
        "\n",
        "                    acid = words[num_columns - auth_comp_id_column]\n",
        "\n",
        "                else:\n",
        "\n",
        "                    acid = words[num_columns - label_comp_id_column]\n",
        "\n",
        "                if len(acid) == 3:\n",
        "\n",
        "                    # it's an amino acid\n",
        "                    if found_auth_asym_id_column:\n",
        "\n",
        "                        chain = words[num_columns - auth_asym_id_column]\n",
        "\n",
        "                    else:\n",
        "\n",
        "                        chain = words[num_columns - label_asym_id_column]\n",
        "\n",
        "                    if chain != last_chain_found:\n",
        "\n",
        "                        chains.append(chain)\n",
        "                        last_chain_found = chain\n",
        "\n",
        "            else:\n",
        "\n",
        "                # PDB format\n",
        "                acid = line[17:20]\n",
        "                if len(acid) == 3:\n",
        "\n",
        "                    # it's an amino acid\n",
        "                    chain = line[21:22]\n",
        "                    if chain == \" \":\n",
        "\n",
        "                        chains[0] = \"NONE\"\n",
        "\n",
        "                    elif chain != last_chain_found:\n",
        "\n",
        "                        chains.append(chain)\n",
        "                        last_chain_found = chain\n",
        "\n",
        "        # ATOM not found\n",
        "        # the format maybe mmCIF\n",
        "        # in this case we need to know what each column means\n",
        "        elif line.strip() == \"_atom_site.auth_comp_id\":\n",
        "\n",
        "            found_auth_comp_id_column = True\n",
        "\n",
        "        elif line.strip() == \"_atom_site.auth_asym_id\":\n",
        "\n",
        "            found_auth_asym_id_column = True\n",
        "\n",
        "        elif line.strip() == \"_atom_site.label_comp_id\":\n",
        "\n",
        "            found_label_comp_id_column = True\n",
        "\n",
        "        elif line.strip() == \"_atom_site.label_asym_id\":\n",
        "\n",
        "            found_label_asym_id_column = True\n",
        "\n",
        "        if found_auth_comp_id_column and not found_column_numbers:\n",
        "\n",
        "            auth_comp_id_column += 1\n",
        "\n",
        "        if found_auth_asym_id_column and not found_column_numbers:\n",
        "\n",
        "            auth_asym_id_column += 1\n",
        "\n",
        "        if found_label_comp_id_column and not found_column_numbers:\n",
        "\n",
        "            label_comp_id_column += 1\n",
        "\n",
        "        if found_label_asym_id_column and not found_column_numbers:\n",
        "\n",
        "            label_asym_id_column += 1\n",
        "\n",
        "        line = PDB_FILE.readline()\n",
        "\n",
        "    # we found the chains\n",
        "    # we change the file name according to the format\n",
        "    if found_column_numbers:\n",
        "\n",
        "        vars['cif_or_pdb'] = \"cif\"\n",
        "        vars['pdb_file_name'] = \"file.cif\"\n",
        "\n",
        "    else:\n",
        "\n",
        "        vars['cif_or_pdb'] = \"pdb\"\n",
        "        vars['pdb_file_name'] = \"pdb_file.ent\"\n",
        "\n",
        "    os.rename(temp_pdb, vars['pdb_file_name'])\n",
        "    return chains\n",
        "\n",
        "\n",
        "def check_msa_tree_match(ref_msa_seqs, ref_tree_nodes):\n",
        "\n",
        "    for node in ref_tree_nodes:\n",
        "\n",
        "        if not node in ref_msa_seqs:\n",
        "\n",
        "            raise Exception(\"The node %s is in the tree and not in the MSA.\" %node)\n",
        "\n",
        "    for seq_name in ref_msa_seqs: #check that all the msa nodes are in the tree\n",
        "\n",
        "        if not seq_name in ref_tree_nodes:\n",
        "\n",
        "            raise Exception(\"The sequence %s is in the msa and not in the tree\" %seq_name)\n",
        "\n",
        "    vars['unique_seqs'] = len(ref_msa_seqs)\n",
        "\n",
        "def check_validity_tree_file():\n",
        "\n",
        "\t  # checks validity of tree file and returns an array with the names of the nodes\n",
        "    try:\n",
        "\n",
        "        TREEFILE = open(vars['tree_file'], 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        raise Exception(\"Can't read the tree file.\")\n",
        "\n",
        "    tree = TREEFILE.read()\n",
        "    TREEFILE.close()\n",
        "    tree.replace(\"\\n\", \"\")\n",
        "    if tree[-1] != ';':\n",
        "\n",
        "        tree += ';'\n",
        "\n",
        "    leftBrackets = 0\n",
        "    rightBrackets = 0\n",
        "    noRegularFormatChar = \"\"\n",
        "    nodes = []\n",
        "    node_name = \"\"\n",
        "    in_node_name = False\n",
        "    in_node_score = False\n",
        "    for char in tree:\n",
        "\n",
        "        if char == ':':\n",
        "\n",
        "            if in_node_name:\n",
        "\n",
        "                nodes.append(node_name)\n",
        "\n",
        "            node_name = \"\"\n",
        "            in_node_name = False\n",
        "            in_node_score = True\n",
        "\n",
        "        elif char == '(':\n",
        "\n",
        "            leftBrackets += 1\n",
        "\n",
        "        elif char == ')':\n",
        "\n",
        "            rightBrackets += 1\n",
        "            in_node_score = False\n",
        "\n",
        "        elif char == ',':\n",
        "\n",
        "            in_node_score = False\n",
        "\n",
        "        elif char != ';':\n",
        "\n",
        "            if char in \"!@#$^&*~`{}'?<>\\\\\" and not char in noRegularFormatChar:\n",
        "\n",
        "                noRegularFormatChar += \" '\" + char + \"', \"\n",
        "\n",
        "            if not in_node_score:\n",
        "\n",
        "                node_name += char\n",
        "                in_node_name = True\n",
        "\n",
        "    if leftBrackets != rightBrackets:\n",
        "\n",
        "        raise Exception(\"The tree is missing parentheses.\")\n",
        "\n",
        "    if noRegularFormatChar != \"\":\n",
        "\n",
        "        raise Exception(\"The tree contains the following characters \" + noRegularFormatChar[:-2])\n",
        "\n",
        "    return nodes\n",
        "\n",
        "\n",
        "def check_msa_format():\n",
        "\n",
        "    MSA = open(vars['user_msa_file_name'], 'r')\n",
        "    line = MSA.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        line = line.strip()\n",
        "        if line == \"\":\n",
        "\n",
        "            line = MSA.readline()\n",
        "            continue\n",
        "\n",
        "        if line[:4] == \"MSF:\":\n",
        "\n",
        "            format = \"msf\"\n",
        "            break\n",
        "\n",
        "        elif line[0] == '>':\n",
        "\n",
        "            format = \"fasta\"\n",
        "            break\n",
        "\n",
        "        elif line[0] == '#':\n",
        "\n",
        "            format = \"nexus\"\n",
        "            break\n",
        "\n",
        "        elif line[0] == 'C':\n",
        "\n",
        "            format = \"clustal\"\n",
        "            break\n",
        "\n",
        "        elif line[0] == 'P':\n",
        "\n",
        "            format = \"gcg\"\n",
        "            break\n",
        "\n",
        "        else:\n",
        "\n",
        "            MSA.close()\n",
        "            raise Exception(\"Unknown format.\")\n",
        "\n",
        "        line = MSA.readline()\n",
        "\n",
        "    MSA.close()\n",
        "    return format\n",
        "\n",
        "def multiple_chains(chains):\n",
        "\n",
        "    print(\"Please choose a chain from this list:\")\n",
        "    i = 1\n",
        "    for chain in chains:\n",
        "\n",
        "        print(\"%d. %s\" %(i, chain))\n",
        "        i += 1\n",
        "\n",
        "    while True:\n",
        "\n",
        "        chain_index = input(\"Press the number of chain.\\n\")\n",
        "        if chain_index.isdigit():\n",
        "\n",
        "            chain_index = int(chain_index)\n",
        "            if chain_index > 0 and chain_index < i:\n",
        "\n",
        "                form['PDB_chain'] = chains[chain_index - 1]\n",
        "                print(\"You chose the chain %s.\\n\" %form['PDB_chain'])\n",
        "                break\n",
        "\n",
        "        print(\"Wrong input.\")\n",
        "\n",
        "def upload_PDB():\n",
        "\n",
        "    # we get the PDB file\n",
        "    while True:\n",
        "\n",
        "        PDB_uniprot = input(\"Do you have a PDB/uniprot ID? (Y/N):\\n\")\n",
        "        if PDB_uniprot.upper() == \"Y\":\n",
        "\n",
        "            ID = input(\"Please enter your ID:\\n\")\n",
        "            ID = ID.upper().strip()\n",
        "            form['pdb_ID'] = ID\n",
        "            if len(ID) == 4:\n",
        "\n",
        "                # PDB ID\n",
        "                pdb_url = \"https://files.rcsb.org/download/%s.pdb\" %ID\n",
        "                cif_url= \"https://files.rcsb.org/download/%s.cif\" %ID\n",
        "\n",
        "            else:\n",
        "\n",
        "                # uniprot ID\n",
        "                pdb_url = \"https://alphafold.ebi.ac.uk/files/AF-%s-F1-model_v6.pdb\" %ID\n",
        "                cif_url= \"https://alphafold.ebi.ac.uk/files/AF-%s-F1-model_v6.cif\" %ID\n",
        "\n",
        "            try:\n",
        "\n",
        "                response = urllib.request.urlopen(pdb_url).read()\n",
        "\n",
        "            except:\n",
        "\n",
        "                try:\n",
        "\n",
        "                    response = urllib.request.urlopen(cif_url).read()\n",
        "\n",
        "                except:\n",
        "\n",
        "                    raise Exception(\"Could not download model file.\")\n",
        "\n",
        "            PDB_name = \"temp_PDB.txt\"\n",
        "            PDB_FILE = open(PDB_name, 'wb')\n",
        "            PDB_FILE.write(response)\n",
        "            PDB_FILE.close()\n",
        "            vars['Used_PDB_Name'] = vars['job_name'] + \"_\" + ID\n",
        "            print()\n",
        "            break\n",
        "\n",
        "        elif PDB_uniprot.upper() == \"N\":\n",
        "\n",
        "            print(\"Please upload your model.\\n\")\n",
        "            PDB_file = files.upload()\n",
        "            PDB_name = (list(PDB_file))[0]\n",
        "            vars['Used_PDB_Name'] = PDB_name.replace(\" \", \"\")\n",
        "            match = re.search(r'(\\S+)\\.', vars['Used_PDB_Name'])\n",
        "            if match:\n",
        "\n",
        "                vars['Used_PDB_Name'] = vars['job_name'] + \"_\" + match.group(1)\n",
        "\n",
        "            break\n",
        "\n",
        "        else:\n",
        "\n",
        "            print(\"Wrong input.\")\n",
        "\n",
        "    # the user must choose a chain\n",
        "    chains = choose_chain(PDB_name)\n",
        "    if len(chains) == 1:\n",
        "\n",
        "        form['PDB_chain'] = chains[0]\n",
        "        print(\"The PDB has only one chain %s\\n\" %form['PDB_chain'])\n",
        "\n",
        "    else:\n",
        "\n",
        "        multiple_chains(chains)\n",
        "\n",
        "    os.rename(PDB_name, vars['pdb_file_name'])\n",
        "\n",
        "    if form['PDB_chain'] != \"NONE\":\n",
        "\n",
        "        vars['Used_PDB_Name'] += \"_\" + form['PDB_chain']\n",
        "\n",
        "def upload_sequence():\n",
        "\n",
        "    vars['query_string'] = \"Input_seq\"\n",
        "    vars['SEQRES_seq'] = \"\"\n",
        "    vars['ATOM_without_X_seq'] = \"\"\n",
        "    while True:\n",
        "\n",
        "        vars['protein_seq_string'] = input(\"Please enter your sequence.\\n\")\n",
        "        if '>' in vars['protein_seq_string']:\n",
        "\n",
        "            print(\"You should upload only one sequence, without the sequence name. If you want to upload a MSA, use the correct mode.\")\n",
        "\n",
        "        else:\n",
        "\n",
        "            break\n",
        "\n",
        "    # delete sequence name and white spaces\n",
        "    #vars['protein_seq_string'] = re.sub(r'>\\S*', \"\", vars['protein_seq_string'])\n",
        "    vars['protein_seq_string'] = re.sub(r'\\s', \"\", vars['protein_seq_string'])\n",
        "    print()\n",
        "    if re.match(r'^[actguACTGUNn]+$', vars['protein_seq_string']):\n",
        "\n",
        "        raise Exception(\"Your sequence is only composed of Nucleotides (i.e. :A,T,C,G).\")\n",
        "\n",
        "def upload_MSA():\n",
        "\n",
        "    print(\"Please upload your MSA.\")\n",
        "    MSA_file = files.upload()\n",
        "    MSA_name = (list(MSA_file))[0]\n",
        "\n",
        "    vars['user_msa_file_name'] = MSA_name\n",
        "    format = check_msa_format()\n",
        "    alignment = AlignIO.read(MSA_name, format)\n",
        "\n",
        "    MSA_FASTA = open(vars['msa_fasta'], 'w')\n",
        "    seq_names = []\n",
        "    seqs = []\n",
        "    num_of_seq = 0\n",
        "    for record in alignment:\n",
        "\n",
        "        num_of_seq += 1\n",
        "\n",
        "        seq_name = record.id\n",
        "        seq = record.seq\n",
        "\n",
        "        # we save the sequences and their name for two reasons\n",
        "        # to let the user choose the query\n",
        "        # to check if they much the tree, if there is one.\n",
        "        seq_names.append(seq_name)\n",
        "        seqs.append(seq)\n",
        "\n",
        "        # we write the msa in fasta format\n",
        "        MSA_FASTA.write(\">%s\\n%s\\n\" %(seq_name, seq))\n",
        "\n",
        "    MSA_FASTA.close()\n",
        "\n",
        "    if num_of_seq < 5:\n",
        "\n",
        "        raise Exception(\"There are %d sequences in the msa. There must be at least five.\" %num_of_seq)\n",
        "\n",
        "    vars['unique_seqs'] = num_of_seq\n",
        "    vars['final_number_of_homologoues'] = num_of_seq\n",
        "\n",
        "    print(\"\\nThe names of the sequences are:\\n\")\n",
        "    i = 1\n",
        "    for seq_name in seq_names:\n",
        "\n",
        "        print(\"%d. %s\" %(i, seq_name))\n",
        "        i += 1\n",
        "\n",
        "    while True:\n",
        "\n",
        "        query_number = input(\"\\nEnter the number of the query.\\n\")\n",
        "        if query_number.isdigit():\n",
        "\n",
        "            query_number = int(query_number)\n",
        "            if query_number > 0 and query_number < i:\n",
        "\n",
        "                vars['query_string'] = seq_names[query_number - 1]\n",
        "                vars['protein_seq_string'] = seqs[query_number - 1]\n",
        "                vars['protein_seq_string'] = vars['protein_seq_string'].replace(\"-\", \"\")\n",
        "                vars['protein_seq_string'] = vars['protein_seq_string'].upper()\n",
        "                break\n",
        "\n",
        "        print(\"Wrong input.\")\n",
        "\n",
        "    print(\"Chosen query is %s\\n\" %vars['query_string'])\n",
        "\n",
        "    try:\n",
        "\n",
        "        MSA_FASTA = open(vars['user_msa_file_name'], 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        raise Exception(\"Can't open file for writing\")\n",
        "\n",
        "    for i in range(len(seq_names)):\n",
        "\n",
        "        MSA_FASTA.write(\">%s\\n%s\\n\" %(seq_names[i], seqs[i]))\n",
        "\n",
        "    MSA_FASTA.close()\n",
        "\n",
        "    vars['msa_SEQNAME'] = vars['query_string']\n",
        "\n",
        "    # include Tree\n",
        "    if vars['running_mode'] == \"_mode_pdb_msa_tree\" or vars['running_mode'] == \"_mode_msa_tree\":\n",
        "\n",
        "        print(\"Please upload your tree.\")\n",
        "        Tree_file = files.upload()\n",
        "        Tree_name = (list(Tree_file))[0]\n",
        "        os.rename(Tree_name, vars['tree_file'])\n",
        "        nodes = check_validity_tree_file()\n",
        "        check_msa_tree_match(seq_names, nodes)\n",
        "        print()\n",
        "\n",
        "def create_MSA_parameters():\n",
        "\n",
        "    # the user may want to change the default parameters\n",
        "    form['E_VALUE'] = 0.0001\n",
        "    form['MAX_NUM_HOMOL'] = \"150\"\n",
        "    vars['hit_redundancy'] = 95\n",
        "    form['MSAprogram'] = \"MAFFT\"\n",
        "    form['MIN_IDENTITY'] = 35\n",
        "    form['best_uniform_sequences'] = \"sample\"\n",
        "    while True:\n",
        "\n",
        "        print(\"\\nMSA parameters:\")\n",
        "        print(\"1. Maximum number of homologs - %s\" %form['MAX_NUM_HOMOL'])\n",
        "        print(\"2. Maximum redundancy - %d\" %vars['hit_redundancy'])\n",
        "        print(\"3. MSA building program - %s\" %form['MSAprogram'])\n",
        "        print(\"4. Minimum identity - %d\" %form['MIN_IDENTITY'])\n",
        "        print(\"5. Homolog selection - %s\" %form['best_uniform_sequences'])\n",
        "        print(\"6. E-value cutoff - %.16g\" %form['E_VALUE'])\n",
        "        params = input(\"Would you like to change? (Y/N):\\n\")\n",
        "        if params.upper() == \"Y\":\n",
        "\n",
        "            number = input(\"Enter the number of the field you want to change.\\n\")\n",
        "            if number == \"1\":\n",
        "\n",
        "                MAX_NUM_HOMOL = input(\"What should the maximum number of homologs be?\\n\")\n",
        "                if MAX_NUM_HOMOL.isdigit() and int(MAX_NUM_HOMOL) > 0:\n",
        "\n",
        "                    form['MAX_NUM_HOMOL'] = MAX_NUM_HOMOL\n",
        "\n",
        "                else:\n",
        "\n",
        "                    print(\"Wrong input\")\n",
        "\n",
        "            elif number == \"2\":\n",
        "\n",
        "                MAX_REDUNDANCY = input(\"What should the maximum redundancy be?\\n\")\n",
        "                if MAX_REDUNDANCY.isdigit() and int(MAX_REDUNDANCY) > 0:\n",
        "\n",
        "                    if int(MAX_REDUNDANCY) <= 100:\n",
        "\n",
        "                        vars['hit_redundancy'] = int(MAX_REDUNDANCY)\n",
        "\n",
        "                    else:\n",
        "\n",
        "                        vars['hit_redundancy'] = 99.99999999999\n",
        "\n",
        "                else:\n",
        "\n",
        "                    print(\"Wrong input\")\n",
        "\n",
        "            elif number == \"3\":\n",
        "\n",
        "                print(\"The program to build the msa are:\\n1. mafft.\\n2. muscle.\\n3. CLUSTALW.\\n4. prank.\")\n",
        "                MSAprogram = input(\"Choose the number of the program.\\n\")\n",
        "                if MSAprogram == \"1\":\n",
        "\n",
        "                    form['MSAprogram'] = \"MAFFT\"\n",
        "\n",
        "                elif MSAprogram == \"2\":\n",
        "\n",
        "                    form['MSAprogram'] = \"MUSCLE\"\n",
        "\n",
        "                elif MSAprogram == \"3\":\n",
        "\n",
        "                    form['MSAprogram'] = \"CLUSTALW\"\n",
        "\n",
        "                elif MSAprogram == \"4\":\n",
        "\n",
        "                    form['MSAprogram'] = \"PRANK\"\n",
        "\n",
        "                else:\n",
        "\n",
        "                    print(\"Wrong input\")\n",
        "\n",
        "            elif number == \"4\":\n",
        "\n",
        "                MIN_IDENTITY = input(\"What should the minimum identity be?\\n\")\n",
        "                if MIN_IDENTITY.isdigit() and int(MIN_IDENTITY) > 0:\n",
        "\n",
        "                    form['MIN_IDENTITY'] = int(MIN_IDENTITY)\n",
        "\n",
        "                else:\n",
        "\n",
        "                    print(\"Wrong input.\")\n",
        "\n",
        "            elif number == \"5\":\n",
        "\n",
        "                print(\"We either sample from the list of homologs (sample) or take the homologs closest to the query (closest).\")\n",
        "                best_uniform = input(\"What the homolog selection method should be? (sample/closest):\\n\")\n",
        "                if best_uniform == \"closest\":\n",
        "\n",
        "                    form['best_uniform_sequences'] = \"closest\"\n",
        "\n",
        "                elif best_uniform == \"sample\":\n",
        "\n",
        "                    form['best_uniform_sequences'] = \"sample\"\n",
        "\n",
        "                else:\n",
        "\n",
        "                    print(\"Wrong input.\")\n",
        "\n",
        "            elif number == \"6\":\n",
        "\n",
        "                print(\"Sequences with higher e-value will be rejected.\")\n",
        "                form['E_VALUE'] = float(input(\"What should the e-value be?\\n\"))\n",
        "\n",
        "            else:\n",
        "\n",
        "                print(\"Wrong input.\")\n",
        "\n",
        "\n",
        "        elif params.upper() == \"N\":\n",
        "\n",
        "            print()\n",
        "            break\n",
        "\n",
        "        else:\n",
        "\n",
        "            print(\"Wrong input.\")\n",
        "\n",
        "\n",
        "mode = \"PDB_MSA\" #@param [\"PDB\", \"PDB_MSA\", \"PDB_MSA_Tree\", \"MSA\", \"MSA_Tree\", \"Sequence\"]\n",
        "\n",
        "if mode == \"PDB\":\n",
        "\n",
        "    vars['running_mode'] = \"_mode_pdb_no_msa\"\n",
        "\n",
        "elif mode == \"PDB_MSA\":\n",
        "\n",
        "    vars['running_mode'] = \"_mode_pdb_msa\"\n",
        "\n",
        "elif mode == \"PDB_MSA_Tree\":\n",
        "\n",
        "    vars['running_mode'] = \"_mode_pdb_msa_tree\"\n",
        "\n",
        "elif mode == \"MSA\":\n",
        "\n",
        "    vars['running_mode'] = \"_mode_msa\"\n",
        "\n",
        "elif mode == \"MSA_Tree\":\n",
        "\n",
        "    vars['running_mode'] = \"_mode_msa_tree\"\n",
        "\n",
        "elif mode == \"Sequence\":\n",
        "\n",
        "    vars['running_mode'] = \"_mode_no_pdb_no_msa\"\n",
        "\n",
        "vars['tree_file'] = vars['job_name'] + \"_Tree.txt\"\n",
        "vars['msa_fasta'] = vars['job_name'] + \"_msa_fasta.aln\" # msa copy in fasta format\n",
        "\"\"\"\n",
        "if vars['running_mode'] == \"_mode_pdb_no_msa\" or vars['running_mode'] == \"_mode_pdb_msa\" or vars['running_mode'] == \"_mode_pdb_msa_tree\":\n",
        "\n",
        "    upload_PDB()\n",
        "\n",
        "elif vars['running_mode'] == \"_mode_no_pdb_no_msa\":\n",
        "\n",
        "    upload_sequence()\n",
        "\n",
        "if vars['running_mode'] == \"_mode_pdb_msa\" or vars['running_mode'] == \"_mode_msa\" or vars['running_mode'] == \"_mode_pdb_msa_tree\" or vars['running_mode'] == \"_mode_msa_tree\":\n",
        "\n",
        "    upload_MSA()\n",
        "\n",
        "else:\n",
        "\n",
        "   create_MSA_parameters()\n",
        "\"\"\"\n",
        "# chose the best substitution model\n",
        "substitution_model = \"Dayhoff\" #@param [\"choose the best substitution model using porttest\",\"JTT\",\"LG\",\"mtREV\",\"cpREV\",\"WAG\",\"Dayhoff\"]\n",
        "if substitution_model == \"choose the best substitution model using porttest\":\n",
        "\n",
        "    form['SUB_MATRIX'] = \"BEST\"\n",
        "\n",
        "else:\n",
        "\n",
        "    form['SUB_MATRIX'] = substitution_model\n",
        "\n",
        "\"\"\"\n",
        "while True:\n",
        "\n",
        "    if form['SUB_MATRIX'] == \"BEST\":\n",
        "\n",
        "        print(\"The best substitution model is chosen using porttest.\")\n",
        "\n",
        "    else:\n",
        "\n",
        "        print(\"The substitution model is %s.\" %form['SUB_MATRIX'])\n",
        "\n",
        "    change_model = input(\"Would you like to change the substitution model? (Y/N)\\n\")\n",
        "    if change_model.upper() == \"Y\":\n",
        "\n",
        "        print(\"There are 6 models. The prottest program can choose the best model.\\n1. JTT\\n2. LG\\n3. mtREV (for mitochondrial proteins)\\n4. cpREV (for chloroplasts proteins)\\n5. WAG\\n6. Dayhoff\\n7. Use prottest to choose the best model.\")\n",
        "        model_num = input(\"Please select the number of the option you chose.\\n\")\n",
        "        if model_num == \"1\":\n",
        "\n",
        "            form['SUB_MATRIX'] = \"JTT\"\n",
        "\n",
        "        elif model_num == \"2\":\n",
        "\n",
        "            form['SUB_MATRIX'] = \"LG\"\n",
        "\n",
        "        elif model_num == \"3\":\n",
        "\n",
        "            form['SUB_MATRIX'] = \"mtREV\"\n",
        "\n",
        "        elif model_num == \"4\":\n",
        "\n",
        "            form['SUB_MATRIX'] = \"cpREV\"\n",
        "\n",
        "        elif model_num == \"5\":\n",
        "\n",
        "            form['SUB_MATRIX'] = \"WAG\"\n",
        "\n",
        "        elif model_num == \"6\":\n",
        "\n",
        "            form['SUB_MATRIX'] = \"Dayhoff\"\n",
        "\n",
        "        elif model_num == \"7\":\n",
        "\n",
        "            form['SUB_MATRIX'] = \"BEST\"\n",
        "\n",
        "        else:\n",
        "\n",
        "            print(\"Wrong input.\")\n",
        "\n",
        "    elif change_model.upper() == \"N\":\n",
        "\n",
        "        print()\n",
        "        break\n",
        "\n",
        "    else:\n",
        "\n",
        "        print(\"Wrong input.\")\n",
        "\"\"\"\n",
        "# choose rate4site algorithm, baysean or maximum likelihood\n",
        "rate4site_algorithm = \"maximum_likelihood\" #@param [\"baysean\",\"maximum_likelihood\"]\n",
        "if rate4site_algorithm == \"baysean\":\n",
        "\n",
        "    form['ALGORITHM'] = \"Bayes\"\n",
        "\n",
        "else:\n",
        "\n",
        "    form['ALGORITHM'] = \"Maximum\"\n",
        "\n",
        "\"\"\"\n",
        "while True:\n",
        "\n",
        "    if form['ALGORITHM'] == \"Bayes\":\n",
        "\n",
        "        print(\"The coloring algorithm is baysean.\")\n",
        "\n",
        "    else:\n",
        "\n",
        "        print(\"The coloring algorithm is maximum likelihood.\")\n",
        "\n",
        "    change_algorithm = input(\"Would you like to change the coloring algorithm? (Y/N):\\n\")\n",
        "    if change_algorithm.upper() == \"Y\":\n",
        "\n",
        "        coloring_algorithm = input(\"Should the algorithm be baysean (b) or maximum likelihood (m)? (b/m):\\n\")\n",
        "        if coloring_algorithm.upper() == \"B\":\n",
        "\n",
        "            form['ALGORITHM'] = \"Bayes\"\n",
        "\n",
        "        elif coloring_algorithm.upper() == \"M\":\n",
        "\n",
        "            form['ALGORITHM'] = \"Maximum\"\n",
        "\n",
        "        else:\n",
        "\n",
        "            print(\"Wrong input.\")\n",
        "\n",
        "    elif change_algorithm.upper() == \"N\":\n",
        "\n",
        "        break\n",
        "\n",
        "    else:\n",
        "\n",
        "        print(\"Wrong input.\")\n",
        "\"\"\"\n",
        "\n",
        "def print_instructions(pdb_with_grades, cif_pdb, pdb_with_grades_isd = False):\n",
        "\n",
        "    print(\"To create pymol with consurf:\")\n",
        "    if pdb_with_grades_isd:\n",
        "\n",
        "        create_download_link(pdb_with_grades_isd, \"Download the modified %s file showing insufficient data\" %cif_pdb)\n",
        "        print(\"or\")\n",
        "        create_download_link(pdb_with_grades, \"the %s file hiding insufficient data\" %cif_pdb)\n",
        "\n",
        "    else:\n",
        "\n",
        "        create_download_link(pdb_with_grades, \"Download the modified %s file\" %cif_pdb)\n",
        "\n",
        "    print(\"which contains ConSurf's color grades.\\nDownload the coloring script\")\n",
        "    create_download_link(vars['pymol'], \"regular version\")\n",
        "    print(\"or\")\n",
        "    create_download_link(vars['pymol_CBS'], \"color blind\")\n",
        "    print(\"1) Start the PyMOL program.\\n2) Drag the %s file to the pymol window.\\n3) Drag the pymol coloring acript to the window.\\n\" %cif_pdb)\n",
        "\n",
        "\n",
        "def print_legend(cbs):\n",
        "\n",
        "    text = \"\"\"<style>\n",
        ".scaleRowStrecher .scaleRowStrecherInner{display: flex;}\n",
        ".scaleRowStrecher .scaleRowStrecherInner .scaleColorRect{width: 50px;text-align: center;line-height: 32px;font-weight: 500;font-size: 20px;}\n",
        ".scaleRowStrecher .scaleRowStrecherInner .scaleColorRect.white{color: #ffffff;}\n",
        ".scaleRowStrecher .scaleRowStrecherInner .scaleColorRect span{width: 100%;}\n",
        ".scaleRowStrecher .label{display: flex;padding-right: 0;}\n",
        ".scaleRowStrecher .label div{width: 33.33%;text-align: center;font-weight: 500;font-size: 18px}\n",
        ".scaleRowStrecher .label .leftLabel{text-align: left;}\n",
        ".scaleRowStrecher .label .rightLabel{text-align: right;}\n",
        "#consrvScaleDiv a.download{margin-top: 20px; color: #400080;text-decoration: underline;transition: 0.5s all ease;-webkit-transition: 0.5s all ease;-o-transition: 0.5s all ease;-moz-transition: 0.5s all ease;}\n",
        ".scaleRowStrecher .insu_data{font-weight: 500;font-size: 18px;margin-top: 20px;}\n",
        ".scaleRowStrecher .insu_data span{display: inline-block;vertical-align: middle;height: 25px;width: 50px;background-color: #f8f499;margin-right: 10px;}\n",
        ".scaleRowStrecher {float: left; margin-top: 20px;margin-bottom: 20px;}\n",
        ".scaleRowStrecher .scaleRowStrecherInner{display: flex;}\n",
        ".scaleRowStrecher .scaleRowStrecherInner .scaleColorRect.white{color: #ffffff;}\n",
        ".scaleRowStrecher .scaleRowStrecherInner .scaleColorRect span{width: 100%;}\n",
        ".scaleRowStrecher .label{display: flex;}\n",
        ".scaleRowStrecher .label div{width: 33.33%;text-align: center;font-weight: 500;font-size: 18px}\n",
        ".scaleRowStrecher .label .leftLabel{text-align: left;}\n",
        ".scaleRowStrecher .label .rightLabel{text-align: right;}\n",
        "#consrvScaleDiv a.download{margin-top: 20px; color: #400080;text-decoration: underline;transition: 0.5s all ease;-webkit-transition: 0.5s all ease;-o-transition: 0.5s all ease;-moz-transition: 0.5s all ease;}\n",
        ".scaleRowStrecher .insu_data{font-weight: 500;font-size: 18px;margin-top: 20px;}\n",
        ".scaleRowStrecher .insu_data span{display: inline-block;vertical-align: middle;height: 25px;width: 50px;background-color: #f8f499;margin-right: 10px;}\n",
        ".result_c1{background-color: #0a7d82;color: #ffffff;}\n",
        ".result_c2{background-color: #44afbf;}\n",
        ".result_c3{background-color: #a5dce6;}\n",
        ".result_c4{background-color: #d7f0f0;}\n",
        ".result_c5{background-color: #ffffff;}\n",
        ".result_c6{background-color: #faebf5;}\n",
        ".result_c7{background-color: #fac8dc;}\n",
        ".result_c8{background-color: #f07daa;}\n",
        ".result_c9{background-color: #a0285f;color: #ffffff;}\n",
        ".result_c1_CBS{background-color: #0f5a23;color: #ffffff;}\n",
        ".result_c2_CBS{background-color: #5aaf5f;}\n",
        ".result_c3_CBS{background-color: #a5dca0;}\n",
        ".result_c4_CBS{background-color: #d7f0d2;}\n",
        ".result_c5_CBS{background-color: #ffffff;}\n",
        ".result_c6_CBS{background-color: #e6d2e6;}\n",
        ".result_c7_CBS{background-color: #c3a5cd;}\n",
        ".result_c8_CBS{background-color: #9b6eaa;}\n",
        ".result_c9_CBS{background-color: #782882;color: #ffffff;}\n",
        "    </style>\n",
        "    \"\"\"\n",
        "    if cbs:\n",
        "\n",
        "        text += \"\"\"\n",
        "    <div>\n",
        "        <div class=\"scaleRowStrecher\">\n",
        "            <div class=\"scaleRowStrecherInner\">\n",
        "            <div class=\"scaleColorRect result_c1_CBS\"><span>1</span></div>\n",
        "            <div class=\"scaleColorRect result_c2_CBS\"><span>2</span></div>\n",
        "            <div class=\"scaleColorRect result_c3_CBS\"><span>3</span></div>\n",
        "            <div class=\"scaleColorRect result_c4_CBS\"><span>4</span></div>\n",
        "            <div class=\"scaleColorRect result_c5_CBS\"><span>5</span></div>\n",
        "            <div class=\"scaleColorRect result_c6_CBS\"><span>6</span></div>\n",
        "            <div class=\"scaleColorRect result_c7_CBS\"><span>7</span></div>\n",
        "            <div class=\"scaleColorRect result_c8_CBS\"><span>8</span></div>\n",
        "            <div class=\"scaleColorRect result_c9_CBS\"><span>9</span></div>\n",
        "        </div>\n",
        "        <div class=\"label\">\n",
        "            <div class=\"leftLabel\">Variable</div>\n",
        "            <div class=\"centerLabel\">Average</div>\n",
        "            <div class=\"rightLabel\">Conserved</div>\n",
        "        </div>\n",
        "        <div class=\"insu_data\">\n",
        "            <span></span>Insufficient Data\n",
        "        </div>\n",
        "    </div>\n",
        "            \"\"\"\n",
        "\n",
        "    else:\n",
        "\n",
        "        text += \"\"\"\n",
        "    <div>\n",
        "        <div class=\"scaleRowStrecher\">\n",
        "            <div class=\"scaleRowStrecherInner\">\n",
        "            <div class=\"scaleColorRect result_c1\"><span>1</span></div>\n",
        "            <div class=\"scaleColorRect result_c2\"><span>2</span></div>\n",
        "            <div class=\"scaleColorRect result_c3\"><span>3</span></div>\n",
        "            <div class=\"scaleColorRect result_c4\"><span>4</span></div>\n",
        "            <div class=\"scaleColorRect result_c5\"><span>5</span></div>\n",
        "            <div class=\"scaleColorRect result_c6\"><span>6</span></div>\n",
        "            <div class=\"scaleColorRect result_c7\"><span>7</span></div>\n",
        "            <div class=\"scaleColorRect result_c8\"><span>8</span></div>\n",
        "            <div class=\"scaleColorRect result_c9\"><span>9</span></div>\n",
        "        </div>\n",
        "        <div class=\"label\">\n",
        "            <div class=\"leftLabel\">Variable</div>\n",
        "            <div class=\"centerLabel\">Average</div>\n",
        "            <div class=\"rightLabel\">Conserved</div>\n",
        "        </div>\n",
        "        <div class=\"insu_data\">\n",
        "            <span></span>Insufficient Data\n",
        "        </div>\n",
        "    </div>\n",
        "            \"\"\"\n",
        "\n",
        "    display(HTML(text))\n",
        "\n",
        "\n",
        "def print_msa_colors_FASTA_clustalwLike(grades, MSA, SeqName, cbs):\n",
        "\n",
        "    # Print the results: the colored sequence of the msa acording to the query sequence\n",
        "\n",
        "    Header = \"ConSurf Color-Coded MSA for Job:%s Date:%s\" %(vars['job_name'], vars['date'])\n",
        "\n",
        "    blockSize = 50\n",
        "\n",
        "    if form['DNA_AA'] == \"AA\":\n",
        "\n",
        "        unknownChar = 'X'\n",
        "\n",
        "    else:\n",
        "\n",
        "        unknownChar = 'N'\n",
        "\n",
        "    if cbs:\n",
        "\n",
        "        Out_file = vars['job_name'] + \"_colored_MSA_CBS.html\"\n",
        "\n",
        "    else:\n",
        "\n",
        "        Out_file = vars['job_name'] + \"_colored_MSA.html\"\n",
        "\n",
        "    try:\n",
        "\n",
        "        OUT = open(Out_file, 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error(\"sys_error\", \"could not open the file \" + Out_file + \" for writing.\")\n",
        "\n",
        "    OUT.write(\"<!DOCTYPE html>\\n\")\n",
        "    OUT.write(\"<html lang=\\\"en\\\">\\n\")\n",
        "    OUT.write(\"<head>\\n\")\n",
        "    OUT.write(\"<meta charset=\\\"utf-8\\\">\\n\")\n",
        "    OUT.write(\"<meta http-equiv=\\\"X-UA-Compatible\\\" content=\\\"IE=edge\\\">\\n\")\n",
        "    OUT.write(\"<meta name=\\\"viewport\\\" content=\\\"width=device-width, initial-scale=1\\\">\\n\")\n",
        "    #OUT.write(\"<link rel=\\\"stylesheet\\\" type=\\\"text/css\\\" href=\\\"%s\\\">\\n\" %CSS_File)\n",
        "    OUT.write(\"<style>\")\n",
        "    if cbs:\n",
        "\n",
        "        OUT.write(\"\"\"table {\n",
        "\t/*\ttable-layout: auto; */\n",
        "\t\ttable-layout: fixed;\n",
        "        margin-left: 0em;\n",
        "        margin-right: 0em;\n",
        "        padding:1em 1em 1em 1em;\n",
        "\t    margin:1em 1em 1em 1em;\n",
        "        border-collapse: collapse;\n",
        "      }\n",
        "td {\n",
        "        font-family: \"Courier New\", Courier, monospace;\n",
        "        font-size:1em;\n",
        "        font-weight: bold;\n",
        "        text-align: center;\n",
        "        overflow:hidden;\n",
        "     /*   white-space:nowrap; */\n",
        "\t\t white-space:pre;\n",
        "/*\t \tpadding:0.1em 0.1em 0.1em 0.1em; */\n",
        "/*\t \tmargin:0.5em 0.5em 0.5em 0.5em;*/\n",
        "\t\twidth: 1em;\n",
        "      }\n",
        "td.Seq_Name{\n",
        "        text-align: left;\n",
        "        width: 15em;\n",
        "        padding-right:1em;\n",
        "      }\n",
        "td.Score9{\n",
        "        color: #FFFFFF;\n",
        "        background: #782882;\n",
        "  \t}\n",
        "td.Score8{\n",
        "        background: #9B6EAA;\n",
        "        }\n",
        "td.Score7{\n",
        "        background: #C3A5CD;\n",
        "        }\n",
        "td.Score6{\n",
        "        background: #E6D2E6;\n",
        "        }\n",
        "td.Score5{\n",
        "        background: #FFFFFF;\n",
        "        }\n",
        "td.Score4{\n",
        "        background: #D7F0D2;\n",
        "\t}\n",
        "td.Score3{\n",
        "        background: #A5DCA0;\n",
        "\t}\n",
        "td.Score2{\n",
        "        background: #5AAF5F;\n",
        "        }\n",
        "td.Score1{\n",
        "\t\tcolor: #FFFFFF;\n",
        "        background: #0F5A23;\n",
        "        }\n",
        "td.ScoreNaN{\n",
        "\t    background: #808080;\n",
        "\t}\n",
        "\n",
        "/* ISD COLORES */\n",
        "td.Score_ISD{\n",
        "        background: #FFFF96;\n",
        "  \t}\n",
        "\n",
        "td.white{\n",
        "        background: #FFFFFF;\n",
        "        }\n",
        "/* GRAPH STYLING */\n",
        "\n",
        ".barGraph {\n",
        "/*\tbackground: url(images/horizontal_grid_line_50_pixel.png) bottom left;*/\n",
        "/*\tborder-bottom: 3px solid #333;*/\n",
        "/*\tfont: 9px Helvetica, Geneva, sans-serif; */\n",
        "\theight: 20em;\n",
        "\tmargin: 0em 0em;\n",
        "\tpadding: 0em;\n",
        "\tposition: relative;\n",
        "\t}\n",
        "\n",
        ".barGraph p {\n",
        "\tfont-size:1em;\n",
        "\t}\n",
        "\n",
        ".barGraph li {\n",
        "/*\tbackground: #666 url(images/bar_50_percent_highlight.png) repeat-y top right;*/\n",
        "/*\tborder: 0.2em solid #555;*/\n",
        "\tborder-bottom: none;\n",
        "\tbottom: 0em;\n",
        "\tcolor: #FFF;\n",
        "\tmargin: 0em;\n",
        "\tpadding: 0em 0em 0em 0em;\n",
        "\tposition: absolute;\n",
        "\tlist-style: none;\n",
        "\ttext-align: center;\n",
        "\twidth: 1em;\n",
        "\t}\n",
        "\n",
        ".barGraph li.p1{ background-color:#666666; }\n",
        ".barGraph li:hover {font-weight:bold;}\n",
        "\n",
        "                  \"\"\")\n",
        "\n",
        "    else:\n",
        "\n",
        "        OUT.write(\"\"\"table {\n",
        "\t/*\ttable-layout: auto; */\n",
        "\t\ttable-layout: fixed;\n",
        "        margin-left: 0em;\n",
        "        margin-right: 0em;\n",
        "        padding:1em 1em 1em 1em;\n",
        "\t    margin:1em 1em 1em 1em;\n",
        "        border-collapse: collapse;\n",
        "      }\n",
        "td {\n",
        "        font-family: \"Courier New\", Courier, monospace;\n",
        "        font-size:1em;\n",
        "        font-weight: bold;\n",
        "        text-align: center;\n",
        "        overflow:hidden;\n",
        "     /*   white-space:nowrap; */\n",
        "\t\t white-space:pre;\n",
        "/*\t \tpadding:0.1em 0.1em 0.1em 0.1em; */\n",
        "/*\t \tmargin:0.5em 0.5em 0.5em 0.5em;*/\n",
        "\t\twidth: 1em;\n",
        "      }\n",
        "td.Seq_Name{\n",
        "        text-align: left;\n",
        "        width: 15em;\n",
        "        padding-right:1em;\n",
        "      }\n",
        "td.Score9{\n",
        "        color: #FFFFFF;\n",
        "        background: #A0285F;\n",
        "  \t}\n",
        "td.Score8{\n",
        "        background: #F07DAA;\n",
        "        }\n",
        "td.Score7{\n",
        "        background: #FAC8DC;\n",
        "        }\n",
        "td.Score6{\n",
        "        background: #FAEBF5;\n",
        "        }\n",
        "td.Score5{\n",
        "        background: #FFFFFF;\n",
        "        }\n",
        "td.Score4{\n",
        "        background: #D7F0F0;\n",
        "\t}\n",
        "td.Score3{\n",
        "        background: #A5DCE6;\n",
        "\t}\n",
        "td.Score2{\n",
        "        background: #4BAFBE;\n",
        "        }\n",
        "td.Score1{\n",
        "\t\tcolor: #FFFFFF;\n",
        "        background: #0A7D82;\n",
        "        }\n",
        "td.ScoreNaN{\n",
        "\t    background: #808080;\n",
        "\t}\n",
        "\n",
        "/* ISD COLORES */\n",
        "td.Score_ISD{\n",
        "        background: #FFFF96;\n",
        "  \t}\n",
        "\n",
        "td.white{\n",
        "        background: #FFFFFF;\n",
        "        }\n",
        "/* GRAPH STYLING */\n",
        "\n",
        ".barGraph {\n",
        "/*\tbackground: url(images/horizontal_grid_line_50_pixel.png) bottom left;*/\n",
        "/*\tborder-bottom: 3px solid #333;*/\n",
        "/*\tfont: 9px Helvetica, Geneva, sans-serif; */\n",
        "\theight: 20em;\n",
        "\tmargin: 0em 0em;\n",
        "\tpadding: 0em;\n",
        "\tposition: relative;\n",
        "\t}\n",
        "\n",
        ".barGraph p {\n",
        "\tfont-size:1em;\n",
        "\t}\n",
        "\n",
        ".barGraph li {\n",
        "/*\tbackground: #666 url(images/bar_50_percent_highlight.png) repeat-y top right;*/\n",
        "/*\tborder: 0.2em solid #555;*/\n",
        "\tborder-bottom: none;\n",
        "\tbottom: 0em;\n",
        "\tcolor: #FFF;\n",
        "\tmargin: 0em;\n",
        "\tpadding: 0em 0em 0em 0em;\n",
        "\tposition: absolute;\n",
        "\tlist-style: none;\n",
        "\ttext-align: center;\n",
        "\twidth: 1em;\n",
        "\t}\n",
        "\n",
        ".barGraph li.p1{ background-color:#666666; }\n",
        ".barGraph li:hover {font-weight:bold;}\n",
        "\n",
        "                  \"\"\")\n",
        "\n",
        "    OUT.write(\"</style>\")\n",
        "    OUT.write(\"<title>%s</title>\\n\" %Header)\n",
        "    OUT.write(\"</head>\\n\")\n",
        "    OUT.write(\"<H1 align=center><u>%s</u></H1>\\n\\n\" %Header) # MSA color-coded by GAIN probability\n",
        "\n",
        "    [MSA_Hash, Seq_Names_In_Order] = ReadMSA(MSA)\n",
        "    NumOfBlocks = int(len(MSA_Hash[SeqName]) / blockSize) + 1\n",
        "\n",
        "    seq_pos = 0\n",
        "    ind = 0\n",
        "    for block in range(0, NumOfBlocks):\n",
        "\n",
        "        MSA_Pos = blockSize * block\n",
        "        OUT.write(\"<table>\\n\")\n",
        "        seqNum = 0\n",
        "        seq_pos += ind\n",
        "        for name in Seq_Names_In_Order:\n",
        "\n",
        "            seqNum += 1\n",
        "            OUT.write(\"<tr>\\n\")\n",
        "            if name == SeqName:\n",
        "\n",
        "                OUT.write(\"<td class=\\\"Seq_Name\\\"><u><b>%d %s</u></b></td>\\n\" %(seqNum, name))\n",
        "\n",
        "            else:\n",
        "\n",
        "                OUT.write(\"<td class=\\\"Seq_Name\\\">%d %s</td>\\n\" %(seqNum, name))\n",
        "\n",
        "            ind = 0\n",
        "            for (pos, char) in list(zip(MSA_Hash[SeqName], MSA_Hash[name]))[MSA_Pos : (block + 1) * blockSize]:\n",
        "\n",
        "                if pos != \"-\" and pos.upper() != unknownChar:\n",
        "\n",
        "                    ScoreClass = \"Score\" + str(grades[seq_pos + ind]['COLOR'])\n",
        "                    if grades[seq_pos + ind]['ISD'] == 1:\n",
        "\n",
        "                        ScoreClass = \"Score_ISD\"\n",
        "\n",
        "                    OUT.write(\"<td class=\\\"%s\\\">%s</td>\" %(ScoreClass, char))\n",
        "                    ind += 1\n",
        "\n",
        "                else:\n",
        "\n",
        "                    OUT.write(\"<td class=\\\"white\\\">%s</td>\" %char)\n",
        "\n",
        "            OUT.write(\"</tr>\\n\")\n",
        "\n",
        "        OUT.write(\"</table><br><br>\\n\")\n",
        "\n",
        "    # print the color scale\n",
        "    OUT.write(\"<table style = 'table-layout: auto;margin-left: 0em;margin-right: 0em;padding:1px 1px 1px 1px; margin:1px 1px 1px 1px; border-collapse: collapse;' border=0 cols=1 width=310>\\n<tr><td align=center>\\n<font face='Courier New' color='black' size=+1><center>\\n<tr>\")\n",
        "    for i in range(1,10):\n",
        "\n",
        "        OUT.write(\"<td class=\\\"%s\\\">%d</td>\\n\" %(\"Score\" + str(i), i))\n",
        "\n",
        "    OUT.write(\"</tr></font></center>\\n<center><table style = 'table-layout: auto;margin-left:0em;margin-right: 0em;padding:1px 1px 1px 1px; margin:1px 1px 1px 1px; border-collapse: collapse;' border=0 cols=3 width=310>\\n<tr>\\n<td align=left><td align=left><b>Variable</b></td><td></td><td align=center><b>Average</b></td><td></td>\\n<td align=right><b>Conserved</b></td>\\n</tr><tr></tr><tr></tr>\\n</table></center>\\n\")\n",
        "    OUT.write(\"<table><tr><b><td class=\\\"Score_ISD\\\">X</td><td class=\\\"white\\\"> - Insufficient data - the calculation for this site was performed on less than 10% of the sequences.</b><br></td></tr></table>\\n\")\n",
        "    OUT.write(\"</body>\\n</table>\\n\")\n",
        "    OUT.close()\n",
        "\n",
        "    create_download_link(Out_file, \"Download colored MSA\")\n",
        "\n",
        "    vars['zip_list'].append(Out_file)\n",
        "\n",
        "def ReadMSA(msa):\n",
        "\n",
        "    Seq_Names_In_Order = [] # array to hold sequences names in order\n",
        "    MSA_Hash = {} # hash to hold sequnces\n",
        "    Seq = \"\"\n",
        "    Seq_Name = \"\"\n",
        "\n",
        "    try:\n",
        "\n",
        "        MSA = open(msa, 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error(\"sys_error\", \"ReadMSA: Can't read the MSA: \" + msa)\n",
        "\n",
        "    line = MSA.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        line = line.rstrip()\n",
        "        match = re.match(r'^>(.*)', line)\n",
        "        if match:\n",
        "\n",
        "            if Seq != \"\":\n",
        "\n",
        "                MSA_Hash[Seq_Name] = Seq\n",
        "                Seq = \"\"\n",
        "                Seq_Name = \"\"\n",
        "\n",
        "            Seq_Name = match.group(1)\n",
        "            Seq_Names_In_Order.append(Seq_Name)\n",
        "\n",
        "        else:\n",
        "\n",
        "            Seq += line\n",
        "\n",
        "        line = MSA.readline()\n",
        "\n",
        "    MSA.close()\n",
        "    MSA_Hash[Seq_Name] = Seq # last sequence\n",
        "\n",
        "    return(MSA_Hash, Seq_Names_In_Order)\n",
        "\n",
        "\n",
        "def create_download_link(file, text):\n",
        "    callback_id = f\"download_{uuid.uuid4().hex}\"\n",
        "\n",
        "    def download_file():\n",
        "        files.download(vars['working_dir'] + file)\n",
        "\n",
        "    output.register_callback(callback_id, download_file)\n",
        "\n",
        "    display(HTML(f'''\n",
        "        <a href=\"#\" onclick=\"google.colab.kernel.invokeFunction('{callback_id}', [], {{}}); return false;\">\n",
        "           {text}\n",
        "        </a>\n",
        "    '''))\n",
        "\n",
        "def consurf_HTML_Output(cbs):\n",
        "\n",
        "    # Print the results: the colored sequence and the B/E information\n",
        "    if cbs:\n",
        "\n",
        "        consurf_html_colors = vars['color_array_CBS']\n",
        "    else:\n",
        "\n",
        "        consurf_html_colors = vars['color_array']\n",
        "\n",
        "    COLORS =\"<html>\\n<title>ConSurf Results</title>\\n\"\n",
        "    COLORS += \"<head>\\n<style>\\nb { float: left;}\\n</style>\\n</head>\\n\"\n",
        "    COLORS += \"<body bgcolor='white'>\\n\"\n",
        "    COLORS += \"\\n<table border=0 width=100%>\\n\"\n",
        "    COLORS += \"<tr><td>\\n\"\n",
        "\n",
        "    # print the colored sequence\n",
        "\n",
        "    count = 1\n",
        "    letter_str = \"\"\n",
        "\n",
        "    number_of_pos = len(vars['gradesPE_Output'])\n",
        "    for elem in vars['gradesPE_Output']:\n",
        "\n",
        "        # print the counter above the beginning of each 10 characters\n",
        "        if count % 50 == 1:\n",
        "\n",
        "            count_num = count\n",
        "            while count_num < count + 50:\n",
        "\n",
        "                if count_num <= number_of_pos:\n",
        "\n",
        "                    space_num = 11 - len(str(count_num))\n",
        "                    spaces = \"\"\n",
        "                    for i in range(0, space_num):\n",
        "\n",
        "                        spaces += \"&nbsp;\"\n",
        "\n",
        "                    COLORS += \"<font face='Courier New' color='black' size=+1>\" + str(count_num) + spaces + \"</font>\"\n",
        "\n",
        "                count_num += 10\n",
        "\n",
        "            COLORS += \"<br>\\n\"\n",
        "\n",
        "        # print the colored letters and 'e' for the exposed residues\n",
        "\n",
        "        # after 50 characters - print newline\n",
        "        if count % 50 == 0 or count == number_of_pos:\n",
        "\n",
        "            if elem['ISD'] == 1: # INSUFFICIENT DATA\n",
        "\n",
        "                letter_str += \"<b><font face='Courier New' color='black' size=+1><span style='background: %s;'>%s</span></font></b><br>\" %(consurf_html_colors['ISD'], elem['SEQ'])\n",
        "\n",
        "            elif elem['COLOR'] == 9 or elem['COLOR'] == 1: # MOST OR LEAST CONSERVED\n",
        "\n",
        "                letter_str += \"<b><font face='Courier New' color='white' size=+1><span style='background: %s;'>%s</span></font></b><br>\\n\" %(consurf_html_colors[elem['COLOR']], elem['SEQ'])\n",
        "\n",
        "            else:\n",
        "\n",
        "                letter_str += \"<b><font face='Courier New' color='black' size=+1><span style='background: %s;'>%s</span></font></b><br>\\n\" %(consurf_html_colors[elem['COLOR']], elem['SEQ'])\n",
        "\n",
        "            COLORS += letter_str\n",
        "            COLORS += \"</td></tr>\\n\"\n",
        "            COLORS += \"<tr><td>\\n\"\n",
        "\n",
        "            letter_str = \"\"\n",
        "\n",
        "        elif count % 10 == 1 and count % 50 != 1: # after 10 characters - print a space ('&nbsp;')\n",
        "\n",
        "            letter_str += \"<b><font face='Courier New' color='black' size=+1>&nbsp;</font></b>\"\n",
        "\n",
        "            if elem['ISD'] == 1:\n",
        "\n",
        "                letter_str += \"<b><font face='Courier New' color='black' size=+1><span style='background: %s;'>%s</span> </font></b>\\n\" %(consurf_html_colors['ISD'], elem['SEQ'])\n",
        "\n",
        "            elif elem['COLOR'] == 9 or elem['COLOR'] == 1: # MOST OR LEAST CONSERVED\n",
        "\n",
        "                letter_str += \"<b><font face='Courier New' color='white' size=+1><span style='background: %s;'>%s</span> </font></b>\\n\" %(consurf_html_colors[elem['COLOR']], elem['SEQ'])\n",
        "\n",
        "            else:\n",
        "\n",
        "                letter_str += \"<b><font face='Courier New' color='black' size=+1><span style='background: %s;'>%s</span> </font></b>\\n\" %(consurf_html_colors[elem['COLOR']], elem['SEQ'])\n",
        "\n",
        "        else:\n",
        "\n",
        "            if elem['ISD'] == 1:\n",
        "\n",
        "                letter_str += \"<b><font face='Courier New' color='black' size=+1><span style='background: %s;'>%s</span> </font></b>\\n\" %(consurf_html_colors['ISD'], elem['SEQ'])\n",
        "\n",
        "            elif elem['COLOR'] == 9 or elem['COLOR'] == 1: # MOST OR LEAST CONSERVED\n",
        "\n",
        "                letter_str += \"<b><font face='Courier New' color='white' size=+1><span style='background: %s;'>%s</span> </font></b>\\n\" %(consurf_html_colors[elem['COLOR']], elem['SEQ'])\n",
        "\n",
        "            else:\n",
        "\n",
        "                letter_str += \"<b><font face='Courier New' color='black' size=+1><span style='background: %s;'>%s</span> </font></b>\\n\" %(consurf_html_colors[elem['COLOR']], elem['SEQ'])\n",
        "\n",
        "        count += 1\n",
        "\n",
        "    COLORS += \"</td></tr>\\n</table><br>\\n\"\n",
        "    COLORS += \"</body>\\n</html>\\n\"\n",
        "    display(HTML(COLORS))\n",
        "\n",
        "def no_model_view(cbs = False):\n",
        "\n",
        "    if cbs:\n",
        "\n",
        "        print(\"------------------------------------------------------------------------------------------------------------------------\")\n",
        "        print(\"Color Blind View\")\n",
        "        # ConSurf color palette\n",
        "        pdf_file = vars['Colored_Seq_CBS_PDF']\n",
        "\n",
        "    else:\n",
        "\n",
        "        print(\"------------------------------------------------------------------------------------------------------------------------\")\n",
        "        print(\"Regular View\")\n",
        "        # ConSurf color palette\n",
        "        pdf_file = vars['Colored_Seq_PDF']\n",
        "\n",
        "\n",
        "    print_legend(cbs)\n",
        "    consurf_HTML_Output(cbs)\n",
        "    create_download_link(pdf_file, \"Download colored sequence pdf file\")\n",
        "    print_msa_colors_FASTA_clustalwLike(vars['gradesPE_Output'], vars['msa_fasta'], vars['msa_SEQNAME'], cbs)\n",
        "\n",
        "    if not cbs:\n",
        "\n",
        "        no_model_view(True)\n",
        "        print(\"------------------------------------------------------------------------------------------------------------------------\")\n",
        "\n",
        "def show_py3dmol(file, file_type, cbs = False):\n",
        "\n",
        "    # Read your ConSurf-colored PDB file\n",
        "    with open(file) as f:\n",
        "        pdb_data = f.read()\n",
        "\n",
        "    view = py3Dmol.view(width=800, height=600)\n",
        "    view.addModel(pdb_data, file_type)\n",
        "    view.setStyle({'cartoon': {}})\n",
        "\n",
        "    if cbs:\n",
        "\n",
        "        print(\"------------------------------------------------------------------------------------------------------------------------\")\n",
        "        print(\"Color Blind View\")\n",
        "        # ConSurf color palette\n",
        "        colors = vars['color_array_CBS']\n",
        "        pdf_file = vars['Colored_Seq_CBS_PDF']\n",
        "\n",
        "    else:\n",
        "\n",
        "        print(\"------------------------------------------------------------------------------------------------------------------------\")\n",
        "        print(\"Regular View\")\n",
        "        # ConSurf color palette\n",
        "        colors = vars['color_array']\n",
        "        pdf_file = vars['Colored_Seq_PDF']\n",
        "\n",
        "    # Apply each color to the appropriate B-factor range\n",
        "    for i in range(1, 10):\n",
        "\n",
        "        view.setStyle({'b': i}, {'cartoon': {'color': colors[i]}})\n",
        "\n",
        "    # Color residues with B = 10 (insufficient data) in yellow\n",
        "    view.setStyle({'b': 10}, {'cartoon': {'color': 'yellow'}})\n",
        "\n",
        "    view.zoomTo()\n",
        "    view.show()\n",
        "    print_legend(cbs)\n",
        "    #print(\"\\n\")\n",
        "    consurf_HTML_Output(cbs)\n",
        "    #print(\"\\n\\n\\n\\n\\n\\n\\n\\n\")\n",
        "    create_download_link(pdf_file, \"Download colored sequence pdf file\")\n",
        "    print_msa_colors_FASTA_clustalwLike(vars['gradesPE_Output'], vars['msa_fasta'], vars['msa_SEQNAME'], cbs)\n",
        "\n",
        "    if not cbs:\n",
        "\n",
        "        show_py3dmol(file, file_type, True)\n",
        "        print(\"------------------------------------------------------------------------------------------------------------------------\")\n",
        "\n",
        "\n",
        "def print_selected(arr_ref, print_for_pipe):\n",
        "\n",
        "    total_text = \"\"\n",
        "    string = \"\"\n",
        "    if print_for_pipe == \"yes\":\n",
        "\n",
        "        string = \"! select \"\n",
        "\n",
        "    else:\n",
        "\n",
        "        string = \"select \"\n",
        "\n",
        "    total_length = len(string)\n",
        "\n",
        "    if len(arr_ref) > 0:\n",
        "\n",
        "        for aa in arr_ref:\n",
        "\n",
        "            aa = aa.replace(\":\", \"\")\n",
        "            total_length += len(aa)\n",
        "            if total_length > 80:\n",
        "\n",
        "                if re.search(r', $', string):\n",
        "\n",
        "                    string = string[:-2]\n",
        "\n",
        "                total_text += string + \"\\n\"\n",
        "                if print_for_pipe == \"yes\":\n",
        "\n",
        "                    string = \"! select selected or %s, \" %aa\n",
        "\n",
        "                else:\n",
        "\n",
        "                    string = \"select selected or %s, \" %aa\n",
        "\n",
        "                total_length = len(string)\n",
        "\n",
        "            else:\n",
        "\n",
        "                string += aa + \", \"\n",
        "                total_length += 2\n",
        "\n",
        "    else:\n",
        "\n",
        "        total_text += string + \"none\"\n",
        "\n",
        "\n",
        "    if re.search(r', $', string):\n",
        "\n",
        "        string = string[:-2]\n",
        "        total_text += string\n",
        "\n",
        "    return total_text\n",
        "\n",
        "def print_rasmol(job_name, out_file, isd, ref_colors_array, chain, cbs):\n",
        "\n",
        "    # print out new format of rasmol\n",
        "\n",
        "\n",
        "    consurf_rasmol_colors = [\"\", \"[16,200,209]\", \"[140,255,255]\", \"[215,255,255]\", \"[234,255,255]\", \"[255,255,255]\", \"[252,237,244]\", \"[250,201,222]\", \"[240,125,171]\", \"[160,37,96]\", \"[255,255,150]\"]\n",
        "    consurf_rasmol_colors_CBS = [\"\", \"[27,120,55]\", \"[90,174,97]\", \"[166,219,160]\", \"[217,240,211]\", \"[247,247,247]\", \"[231,212,232]\", \"[194,165,207]\", \"[153,112,171]\", \"[118,42,131]\", \"[255,255,150]\"]\n",
        "\n",
        "    try:\n",
        "\n",
        "        OUT = open(out_file, 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"print_rasmol : Could not open the file \" + out_file + \" for writing.\")\n",
        "\n",
        "    OUT.write(\"ConSurfDB.tau.ac.il   %s   %s\\nAPD N.NN\\n\" %(vars['date'], job_name))\n",
        "    OUT.write(\"select all\\ncolor [200,200,200]\\n\\n\")\n",
        "\n",
        "    i = len(ref_colors_array) - 1\n",
        "    while i > 0:\n",
        "\n",
        "        if i == 10 and not isd:\n",
        "\n",
        "            i -= 1\n",
        "            continue\n",
        "\n",
        "        if len(ref_colors_array[i]) > 0:\n",
        "\n",
        "            OUT.write(print_selected(ref_colors_array[i], \"no\"))\n",
        "            OUT.write(\"\\nselect selected and :%s\\n\" %chain)\n",
        "            if cbs:\n",
        "\n",
        "                OUT.write(\"color %s\\nspacefill\\n\" %consurf_rasmol_colors_CBS[i])\n",
        "\n",
        "            else:\n",
        "\n",
        "                OUT.write(\"color %s\\nspacefill\\n\" %consurf_rasmol_colors[i])\n",
        "\n",
        "            OUT.write(\"define CON%d selected\\n\\n\" %i)\n",
        "\n",
        "        i -= 1\n",
        "\n",
        "    OUT.close()\n",
        "\n",
        "def create_rasmol(job_name, chain, ref_colors_array, ref_colors_array_isd):\n",
        "\n",
        "    RasMol_file = job_name + \"_jmol_consurf_colors.spt\"\n",
        "    RasMol_file_isd = job_name + \"_jmol_consurf_colors_isd.spt\"\n",
        "    RasMol_file_CBS = job_name + \"_jmol_consurf_colors_CBS.spt\"\n",
        "    RasMol_file_CBS_isd = job_name + \"_jmol_consurf_colors_CBS_isd.spt\"\n",
        "\n",
        "    if chain == \"NONE\":\n",
        "\n",
        "        chain = \" \"\n",
        "\n",
        "    print_rasmol(job_name, RasMol_file, False, ref_colors_array, chain, False)\n",
        "    print_rasmol(job_name, RasMol_file_CBS, False, ref_colors_array, chain, True)\n",
        "\n",
        "    if len(ref_colors_array_isd[10]) > 0: # there is isd\n",
        "\n",
        "        print_rasmol(job_name, RasMol_file_isd, True, ref_colors_array_isd, chain, False)\n",
        "        print_rasmol(job_name, RasMol_file_CBS_isd, True, ref_colors_array_isd, chain, True)\n",
        "\n",
        "def read_number(file):\n",
        "\n",
        "    # reads numbers from a numbers file without storing the whole file in a string\n",
        "    number = \"\"\n",
        "    char = file.read(1)\n",
        "    while char and char.isspace(): # first we skip white spaces\n",
        "\n",
        "        char = file.read(1)\n",
        "\n",
        "    while char and not char.isspace(): # we read the number\n",
        "\n",
        "        number += char\n",
        "        char = file.read(1)\n",
        "\n",
        "    if number != \"\":\n",
        "\n",
        "        number = float(number)\n",
        "\n",
        "    return number\n",
        "\n",
        "\n",
        "def print_pnet_file(num_pos):\n",
        "\n",
        "    # for each position in the msa we print percentage per positions for 6 positions before and after\n",
        "    window = 6\n",
        "    pad = []\n",
        "    pnet_file = open(\"p.net\", 'w')\n",
        "    acids = [\"V\", \"L\", \"I\", \"M\", \"F\", \"W\", \"Y\", \"G\", \"A\", \"P\", \"S\", \"T\", \"C\", \"H\", \"R\", \"K\", \"Q\", \"E\", \"N\", \"D\"]\n",
        "\n",
        "    for i in range(window):\n",
        "\n",
        "        pad.append({})\n",
        "        for acid in acids:\n",
        "\n",
        "            pad[-1][acid] = 0\n",
        "\n",
        "    padded_percentage_per_pos = pad + vars['percentage_per_pos'] + pad\n",
        "    for i in range(window, num_pos + window):\n",
        "\n",
        "        j = -window\n",
        "        while j < window + 1:\n",
        "\n",
        "            for acid  in acids:\n",
        "\n",
        "                if acid in padded_percentage_per_pos[i + j]:\n",
        "\n",
        "                    pnet_file.write(str(int(padded_percentage_per_pos[i + j][acid])) + \" \")\n",
        "\n",
        "                else:\n",
        "\n",
        "                    pnet_file.write(\"0 \")\n",
        "\n",
        "            j += 1\n",
        "\n",
        "        pnet_file.write(\"\\n\")\n",
        "\n",
        "    pnet_file.close()\n",
        "\n",
        "def reveal_buried_exposed(buried_exposed):\n",
        "\n",
        "    # this function reveals the buried and exposed atoms.\n",
        "    # We look at 6 positions before and 6 position after the current position, 260 numbers in total.\n",
        "    # We multiply each number by a unique weight and then we sum the numbers.\n",
        "    # The results to put in the formula 1 / (1 + e^(-x)).\n",
        "    # We do this twenty times with different weights. Now we have 20 numbers.\n",
        "    # We again multiply the numbers by different weights and sum them up and put them in the formula  1 / (1 + e^(-x)).\n",
        "    # We do this twice, again with different weights. Now we have two numbers.\n",
        "    # If number 1 > number2 we say the position is exposed, if not its buried.\n",
        "\n",
        "    num_pos = len(vars['percentage_per_pos'])\n",
        "    print_pnet_file(num_pos)\n",
        "    G_nl = 2\n",
        "    G_N = [260, 20, 2]\n",
        "\n",
        "    weights_file = open(\"/content/WEIGHTS.BIN\", 'r')\n",
        "    pnet_file = open(\"p.net\", 'r')\n",
        "\n",
        "\n",
        "    for i in range(num_pos):\n",
        "\n",
        "        G_o = [[], [],[]]\n",
        "        for j in range(G_N[0]):\n",
        "\n",
        "            G_o[0].append(read_number(pnet_file) / 100.0)\n",
        "\n",
        "        for s in range(1, G_nl + 1):\n",
        "\n",
        "            for k in range(G_N[s]):\n",
        "\n",
        "                weight = read_number(weights_file)\n",
        "                for j in range(G_N[s - 1]):\n",
        "\n",
        "                    weight += read_number(weights_file) * G_o[s - 1][j]\n",
        "\n",
        "                G_o[s].append(1 / (1 + math.exp(-weight)))\n",
        "\n",
        "\n",
        "        weights_file.seek(0) # each time we use the same weights, so we move the pointer to the start\n",
        "        if G_o[G_nl][0] < G_o[G_nl][1]:\n",
        "\n",
        "            buried_exposed.append(\"e\")\n",
        "\n",
        "        else:\n",
        "\n",
        "            buried_exposed.append(\"b\")\n",
        "\n",
        "    weights_file.close()\n",
        "    pnet_file.close()\n",
        "\n",
        "    return buried_exposed\n",
        "\n",
        "\n",
        "class pdbParser:\n",
        "\n",
        "    def __init__(self):\n",
        "\n",
        "        self.SEQRES = \"\"\n",
        "        self.ATOM = \"\"\n",
        "        self.ATOM_withoutX = {}\n",
        "        self.type = \"\"\n",
        "        self.MODIFIED_COUNT = 0\n",
        "        self.MODIFIED_LIST = \"\"\n",
        "        self.positions = {}\n",
        "        self.max_res_details = 0\n",
        "        self.num_known_atoms = 0\n",
        "        self.num_known_seqs = 0\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "    #def read(self, file, query_chain, DNA_AA, atom_position_filename):\n",
        "    def read(self, file, query_chain, DNA_AA):\n",
        "\n",
        "        #conversion_table = {\"ALA\" : \"A\", \"ARG\" : \"R\", \"ASN\" : \"N\", \"ASP\" : \"D\", \"CYS\" : \"C\", \"GLN\" : \"Q\", \"GLU\" : \"E\", \"GLY\" : \"G\", \"HIS\" : \"H\", \"ILE\" : \"I\", \"LEU\" : \"L\", \"LYS\" : \"K\", \"MET\" : \"M\", \"PHE\" : \"F\", \"PRO\" : \"P\", \"SER\" : \"S\", \"THR\" : \"T\", \"TRP\" : \"W\", \"TYR\" : \"Y\", \"VAL\" : \"V\", \"A\" : \"a\", \"T\" : \"t\", \"C\" : \"c\", \"G\" : \"g\", \"U\" : \"u\", \"I\" : \"i\", \"DA\" : \"a\", \"DT\" : \"t\", \"DC\" : \"c\", \"DG\" : \"g\", \"DU\" : \"u\", \"DI\" : \"i\", \"5CM\" : \"c\", \"5MU\" : \"t\", \"N\" : \"n\"}\n",
        "        #conversion_table = {\"ALA\" : \"A\", \"ARG\" : \"R\", \"ASN\" : \"N\", \"ASP\" : \"D\", \"CYS\" : \"C\", \"GLN\" : \"Q\", \"GLU\" : \"E\", \"GLY\" : \"G\", \"HIS\" : \"H\", \"ILE\" : \"I\", \"LEU\" : \"L\", \"LYS\" : \"K\", \"MET\" : \"M\", \"PHE\" : \"F\", \"PRO\" : \"P\", \"SER\" : \"S\", \"THR\" : \"T\", \"TRP\" : \"W\", \"TYR\" : \"Y\", \"VAL\" : \"V\", \"A\" : \"a\", \"T\" : \"t\", \"C\" : \"c\", \"G\" : \"g\", \"U\" : \"u\", \"I\" : \"i\", \"DA\" : \"a\", \"DT\" : \"t\", \"DC\" : \"c\", \"DG\" : \"g\", \"DU\" : \"u\", \"DI\" : \"i\", \"5CM\" : \"c\", \"N\" : \"n\"}\n",
        "        conversion_table = {\"ALA\" : \"A\", \"ARG\" : \"R\", \"ASN\" : \"N\", \"ASP\" : \"D\", \"CYS\" : \"C\", \"GLN\" : \"Q\", \"GLU\" : \"E\", \"GLY\" : \"G\", \"HIS\" : \"H\", \"ILE\" : \"I\", \"LEU\" : \"L\", \"LYS\" : \"K\", \"MET\" : \"M\", \"PHE\" : \"F\", \"PRO\" : \"P\", \"SER\" : \"S\", \"THR\" : \"T\", \"TRP\" : \"W\", \"TYR\" : \"Y\", \"VAL\" : \"V\", \"A\" : \"A\", \"T\" : \"T\", \"C\" : \"C\", \"G\" : \"G\", \"U\" : \"U\", \"I\" : \"I\", \"DA\" : \"A\", \"DT\" : \"T\", \"DC\" : \"C\", \"DG\" : \"G\", \"DU\" : \"U\", \"DI\" : \"I\", \"5CM\" : \"C\", \"N\" : \"N\"}\n",
        "        modified_residues = {\"MSE\" : \"MET\", \"MLY\" : \"LYS\", \"HYP\" : \"PRO\", \"CME\" : \"CYS\", \"CGU\" : \"GLU\", \"SEP\" : \"SER\", \"KCX\" : \"LYS\", \"MLE\" : \"LEU\", \"TPO\" : \"THR\", \"CSO\" : \"CYS\", \"PTR\" : \"TYR\", \"DLE\" : \"LEU\", \"LLP\" : \"LYS\", \"DVA\" : \"VAL\", \"TYS\" : \"TYR\", \"AIB\" : \"ALA\", \"OCS\" : \"CYS\", \"NLE\" : \"LEU\", \"MVA\" : \"VAL\", \"SEC\" : \"CYS\", \"PYL\" : \"LYS\"}\n",
        "        localMODRES = {}\n",
        "        FIRST = [] # first residue in chain\n",
        "        fas_pos = 0\n",
        "        chain = \"\" # current chain\n",
        "        ENDS = [] # end of chain reached and remaining HETATM should be skipped\n",
        "        last_residue_number = \"\"\n",
        "\n",
        "        if DNA_AA == \"Nuc\":\n",
        "\n",
        "            UnknownChar = \"N\"\n",
        "\n",
        "        else:\n",
        "\n",
        "            UnknownChar = \"X\"\n",
        "\n",
        "        # open file to read MODRES\n",
        "        try:\n",
        "\n",
        "            PDBFILE = open(file, 'r')\n",
        "\n",
        "        except:\n",
        "\n",
        "            return 0\n",
        "\n",
        "        try:\n",
        "\n",
        "            MODRES_FILE = open(file + \".MODRES\", 'w')\n",
        "\n",
        "        except:\n",
        "\n",
        "            return 0\n",
        "\n",
        "        # read the MODRES\n",
        "        line = PDBFILE.readline()\n",
        "        while line != \"\" and not re.match(r'^ATOM', line):\n",
        "\n",
        "            if re.match(r'^MODRES', line):\n",
        "\n",
        "                MODRES = line[12:15].strip() # strip spaces to support NUC\n",
        "                CHAIN = line[16:17]\n",
        "                if CHAIN == \" \":\n",
        "\n",
        "                    CHAIN = \"NONE\"\n",
        "\n",
        "                # we only look at the query chain\n",
        "                if CHAIN != query_chain:\n",
        "\n",
        "                    line = PDBFILE.readline()\n",
        "                    continue\n",
        "\n",
        "                RES = line[24:27].strip() # strip spaces to support NUC\n",
        "\n",
        "                if not MODRES in localMODRES:\n",
        "\n",
        "                    localMODRES[MODRES] = RES\n",
        "                    MODRES_FILE.write(MODRES + \"\\t\" + RES + \"\\n\")\n",
        "\n",
        "                elif localMODRES[MODRES] != RES:\n",
        "\n",
        "                    localMODRES[MODRES] = \"\" # two different values to the same residue\n",
        "\n",
        "            line = PDBFILE.readline()\n",
        "\n",
        "\n",
        "        MODRES_FILE.close()\n",
        "        PDBFILE.close()\n",
        "\n",
        "        # reopen file to read all the file\n",
        "        try:\n",
        "\n",
        "            PDBFILE = open(file, 'r')\n",
        "\n",
        "        except:\n",
        "\n",
        "            return 0\n",
        "\n",
        "        line = PDBFILE.readline()\n",
        "        while line != \"\":\n",
        "\n",
        "            line = line.strip()\n",
        "\n",
        "            if re.search(r'^SEQRES', line): # SEQRES record\n",
        "\n",
        "                chain_seqres = line[11:12] # get chain id\n",
        "                if chain_seqres == \" \":\n",
        "\n",
        "                    chain_seqres = \"NONE\"\n",
        "\n",
        "                # we skip the chain if it is not the query\n",
        "                if query_chain != chain_seqres:\n",
        "\n",
        "                    line = PDBFILE.readline()\n",
        "                    continue\n",
        "\n",
        "                # convert to one letter format\n",
        "                for acid in line[19:70].split():\n",
        "\n",
        "                    # regular conversion\n",
        "                    if acid in conversion_table:\n",
        "\n",
        "                        # add to chain\n",
        "                        self.SEQRES += conversion_table[acid]\n",
        "                        self.num_known_seqs += 1\n",
        "\n",
        "                    # modified residue\n",
        "                    else:\n",
        "\n",
        "                        # count this modified residue\n",
        "                        self.MODIFIED_COUNT += 1\n",
        "\n",
        "                        # check if residue is identified\n",
        "                        if acid in modified_residues and modified_residues[acid] in conversion_table:\n",
        "\n",
        "                            self.SEQRES += conversion_table[modified_residues[acid]]\n",
        "                            self.num_known_seqs += 1\n",
        "\n",
        "                            # add to modified residue list\n",
        "                            if not acid + \" > \" in self.MODIFIED_LIST:\n",
        "\n",
        "                                self.MODIFIED_LIST += acid + \" > \" + conversion_table[modified_residues[acid]] + \"\\n\"\n",
        "\n",
        "                        elif acid in localMODRES and localMODRES[acid] != \"\" and localMODRES[acid] in conversion_table:\n",
        "\n",
        "                            self.SEQRES += conversion_table[localMODRES[acid]]\n",
        "                            self.num_known_seqs += 1\n",
        "\n",
        "                            # add to modified residue list\n",
        "                            if not acid + \" > \" in self.MODIFIED_LIST:\n",
        "\n",
        "                                self.MODIFIED_LIST += acid + \" > \" + conversion_table[localMODRES[acid]] + \"\\n\"\n",
        "\n",
        "                        else:\n",
        "\n",
        "                            # set residue name to X or N\n",
        "                            self.SEQRES += UnknownChar\n",
        "\n",
        "                            # add message to front of modified residue list\n",
        "                            modified_changed_to_X_or_N_msg = \"Modified residue(s) in this chain were converted to the one letter representation '\" + UnknownChar + \"'\\n\"\n",
        "\n",
        "                            if not \"Modified residue\" in self.MODIFIED_LIST:\n",
        "\n",
        "                                self.MODIFIED_LIST = modified_changed_to_X_or_N_msg + self.MODIFIED_LIST\n",
        "\n",
        "            elif re.search(r'^ATOM', line):\n",
        "\n",
        "                # extract atom data\n",
        "                res = line[17:20].strip() # for DNA files there is only one or two letter code\n",
        "                chain = line[21:22]\n",
        "                pos = line[22:27].strip()\n",
        "\n",
        "                if chain == \" \":\n",
        "\n",
        "                    chain = \"NONE\"\n",
        "\n",
        "                # find modified residue\n",
        "                if res in modified_residues:\n",
        "\n",
        "                    mod_res = modified_residues[res]\n",
        "\n",
        "                elif res in localMODRES and localMODRES[res] != \"\" and localMODRES[res] in conversion_table:\n",
        "\n",
        "                    mod_res = localMODRES[res]\n",
        "\n",
        "                else:\n",
        "\n",
        "                    mod_res = res\n",
        "\n",
        "                # convert residue to one letter\n",
        "                if mod_res in conversion_table:\n",
        "\n",
        "                    oneLetter = conversion_table[mod_res]\n",
        "\n",
        "                else:\n",
        "\n",
        "                    oneLetter = UnknownChar\n",
        "\n",
        "                # check if we reached a new residue\n",
        "                if not chain in FIRST:\n",
        "\n",
        "                    FIRST.append(chain)\n",
        "                    last_pos = pos\n",
        "                    self.ATOM_withoutX[chain] = oneLetter\n",
        "\n",
        "                elif pos != last_pos:\n",
        "\n",
        "                    last_pos = pos\n",
        "                    self.ATOM_withoutX[chain] += oneLetter\n",
        "\n",
        "                else:\n",
        "\n",
        "                    line = PDBFILE.readline()\n",
        "                    continue\n",
        "\n",
        "                # if the chain is not the query we only extract the sequence\n",
        "                if query_chain != chain:\n",
        "\n",
        "                    line = PDBFILE.readline()\n",
        "                    continue\n",
        "\n",
        "                self.num_known_atoms += 1\n",
        "\n",
        "                # writing atom position file\n",
        "                fas_pos += 1\n",
        "                res_details = \"%s:%s:%s\" %(res, pos, chain)\n",
        "                self.positions[fas_pos] = res_details\n",
        "                if len(res_details) > self.max_res_details:\n",
        "\n",
        "                    self.max_res_details = len(res_details)\n",
        "\n",
        "                #CORR.write(\"%s\\t%d\\t%s\\n\" %(res, fas_pos, pos))\n",
        "\n",
        "                #residue_number = int(line[22:26].strip())\n",
        "\n",
        "                # check type\n",
        "                if self.type == \"\":\n",
        "\n",
        "                    if len(mod_res) < 3:\n",
        "\n",
        "                        self.type = \"Nuc\"\n",
        "\n",
        "                    else:\n",
        "\n",
        "                        self.type = \"AA\"\n",
        "                \"\"\"\n",
        "                if FIRST[chain]:\n",
        "\n",
        "                    FIRST[chain] = False\n",
        "\n",
        "                elif last_residue_number < residue_number:\n",
        "\n",
        "                    while residue_number != last_residue_number + 1: # For Disorder regions\n",
        "\n",
        "                        self.ATOM += UnknownChar\n",
        "                        last_residue_number += 1\n",
        "\n",
        "                self.ATOM += oneLetter\n",
        "                last_residue_number = residue_number\n",
        "                \"\"\"\n",
        "            elif re.search(r'^HETATM', line):\n",
        "\n",
        "                # extract hetatm data\n",
        "                res = line[17:20].strip() # for DNA files there is only one or two letter code\n",
        "                chain = line[21:22]\n",
        "                pos = line[22:27].strip()\n",
        "\n",
        "                if chain == \" \":\n",
        "\n",
        "                    chain = \"NONE\"\n",
        "\n",
        "                if chain in ENDS:\n",
        "\n",
        "                    line = PDBFILE.readline()\n",
        "                    continue\n",
        "\n",
        "                # find modified residue\n",
        "                if res in modified_residues:\n",
        "\n",
        "                    mod_res = modified_residues[res]\n",
        "\n",
        "                elif res in localMODRES and localMODRES[res] != \"\" and localMODRES[res] in conversion_table:\n",
        "\n",
        "                    mod_res = localMODRES[res]\n",
        "\n",
        "                else:\n",
        "\n",
        "                    mod_res = res\n",
        "\n",
        "                # convert residue to one letter\n",
        "                if mod_res in conversion_table:\n",
        "\n",
        "                    oneLetter = conversion_table[mod_res]\n",
        "\n",
        "                else:\n",
        "\n",
        "                    oneLetter = UnknownChar\n",
        "\n",
        "                # check if we reached a new residue\n",
        "                if not chain in FIRST:\n",
        "\n",
        "                    FIRST.append(chain)\n",
        "                    last_pos = pos\n",
        "                    self.ATOM_withoutX[chain] = oneLetter\n",
        "\n",
        "                elif pos != last_pos:\n",
        "\n",
        "                    last_pos = pos\n",
        "                    self.ATOM_withoutX[chain] += oneLetter\n",
        "\n",
        "                else:\n",
        "\n",
        "                    line = PDBFILE.readline()\n",
        "                    continue\n",
        "\n",
        "                # if the chain is not the query we only extract the sequence\n",
        "                if query_chain != chain:\n",
        "\n",
        "                    line = PDBFILE.readline()\n",
        "                    continue\n",
        "\n",
        "                self.num_known_atoms += 1\n",
        "\n",
        "                # writing atom position file\n",
        "                fas_pos += 1\n",
        "                res_details = \"%s:%s:%s\" %(res, pos, chain)\n",
        "                self.positions[fas_pos] = res_details\n",
        "                if len(res_details) > self.max_res_details:\n",
        "\n",
        "                    self.max_res_details = len(res_details)\n",
        "\n",
        "                \"\"\"\n",
        "                residue_number = int(line[22:26].strip())\n",
        "\n",
        "                if FIRST[chain]:\n",
        "\n",
        "                    last_residue_number = residue_number\n",
        "                    FIRST[chain] = False\n",
        "\n",
        "                elif last_residue_number < residue_number:\n",
        "\n",
        "                    while residue_number != last_residue_number + 1: # For Disorder regions\n",
        "\n",
        "                        self.ATOM += UnknownChar\n",
        "                        last_residue_number += 1\n",
        "\n",
        "                self.ATOM += oneLetter\n",
        "                last_residue_number = residue_number\n",
        "                \"\"\"\n",
        "            elif re.search(r'^TER', line):\n",
        "\n",
        "                if not chain in ENDS:\n",
        "\n",
        "                    ENDS.append(chain)\n",
        "\n",
        "            line = PDBFILE.readline()\n",
        "\n",
        "        PDBFILE.close()\n",
        "        #CORR.close()\n",
        "        return 1\n",
        "\n",
        "    def get_num_known_atoms(self):\n",
        "\n",
        "        return self.num_known_atoms\n",
        "\n",
        "    def get_num_known_seqs(self):\n",
        "\n",
        "        return self.num_known_seqs\n",
        "\n",
        "    def get_max_res_details(self):\n",
        "\n",
        "        return self.max_res_details\n",
        "\n",
        "    def get_positions(self):\n",
        "\n",
        "        return self.positions\n",
        "\n",
        "    def get_type(self):\n",
        "\n",
        "        return self.type\n",
        "\n",
        "    def get_SEQRES(self):\n",
        "\n",
        "        return self.SEQRES\n",
        "\n",
        "    def get_ATOM_withoutX(self):\n",
        "\n",
        "        return self.ATOM_withoutX\n",
        "\n",
        "    def get_MODIFIED_COUNT(self):\n",
        "\n",
        "\n",
        "        return self.MODIFIED_COUNT\n",
        "\n",
        "\n",
        "\n",
        "    def get_MODIFIED_LIST(self):\n",
        "\n",
        "        return self.MODIFIED_LIST\n",
        "\n",
        "\n",
        "\n",
        "class cifParser:\n",
        "\n",
        "    def __init__(self):\n",
        "\n",
        "        self.SEQRES = \"\"\n",
        "        #self.ATOM = \"\"\n",
        "        self.ATOM_withoutX = {}\n",
        "        self.type = \"\"\n",
        "        self.MODIFIED_COUNT = 0\n",
        "        self.MODIFIED_LIST = \"\"\n",
        "        self.positions = {}\n",
        "        self.max_res_details = 0\n",
        "        self.auth_seq_id_column = 0\n",
        "        self.auth_comp_id_column = 0\n",
        "        self.auth_asym_id_column = 0\n",
        "        self.B_iso_or_equiv = 0\n",
        "\n",
        "\n",
        "    #def read(self, file, query_chain, DNA_AA, atom_position_filename):\n",
        "    def read(self, file, query_chain, DNA_AA):\n",
        "\n",
        "        #conversion_table = {\"ALA\" : \"A\", \"ARG\" : \"R\", \"ASN\" : \"N\", \"ASP\" : \"D\", \"CYS\" : \"C\", \"GLN\" : \"Q\", \"GLU\" : \"E\", \"GLY\" : \"G\", \"HIS\" : \"H\", \"ILE\" : \"I\", \"LEU\" : \"L\", \"LYS\" : \"K\", \"MET\" : \"M\", \"PHE\" : \"F\", \"PRO\" : \"P\", \"SER\" : \"S\", \"THR\" : \"T\", \"TRP\" : \"W\", \"TYR\" : \"Y\", \"VAL\" : \"V\", \"A\" : \"a\", \"T\" : \"t\", \"C\" : \"c\", \"G\" : \"g\", \"U\" : \"u\", \"I\" : \"i\", \"DA\" : \"a\", \"DT\" : \"t\", \"DC\" : \"c\", \"DG\" : \"g\", \"DU\" : \"u\", \"DI\" : \"i\", \"5CM\" : \"c\", \"N\" : \"n\"}\n",
        "        conversion_table = {\"ALA\" : \"A\", \"ARG\" : \"R\", \"ASN\" : \"N\", \"ASP\" : \"D\", \"CYS\" : \"C\", \"GLN\" : \"Q\", \"GLU\" : \"E\", \"GLY\" : \"G\", \"HIS\" : \"H\", \"ILE\" : \"I\", \"LEU\" : \"L\", \"LYS\" : \"K\", \"MET\" : \"M\", \"PHE\" : \"F\", \"PRO\" : \"P\", \"SER\" : \"S\", \"THR\" : \"T\", \"TRP\" : \"W\", \"TYR\" : \"Y\", \"VAL\" : \"V\", \"A\" : \"A\", \"T\" : \"T\", \"C\" : \"C\", \"G\" : \"G\", \"U\" : \"U\", \"I\" : \"I\", \"DA\" : \"A\", \"DT\" : \"T\", \"DC\" : \"C\", \"DG\" : \"G\", \"DU\" : \"U\", \"DI\" : \"I\", \"5CM\" : \"C\", \"N\" : \"N\"}\n",
        "        #modified_residues = {\"MSE\" : \"MET\", \"MLY\" : \"LYS\", \"HYP\" : \"PRO\", \"CME\" : \"CYS\", \"CGU\" : \"GLU\", \"SEP\" : \"SER\", \"KCX\" : \"LYS\", \"MLE\" : \"LEU\", \"TPO\" : \"THR\", \"CSO\" : \"CYS\", \"PTR\" : \"TYR\", \"DLE\" : \"LEU\", \"LLP\" : \"LYS\", \"DVA\" : \"VAL\", \"TYS\" : \"TYR\", \"AIB\" : \"ALA\", \"OCS\" : \"CYS\", \"NLE\" : \"LEU\", \"MVA\" : \"VAL\", \"SEC\" : \"CYS\", \"PYL\" : \"LYS\"}\n",
        "\n",
        "        # find the portion of the file that contains the seqres\n",
        "        SEQRES_string = \"\"\n",
        "        SEQRES_string_found = False\n",
        "        in_fasta = False\n",
        "        fas_pos = 0\n",
        "        last_pos = 0\n",
        "        current_chain = \"\"\n",
        "        hetatm = \"\" # the part of the sequence that in the HETATM lines\n",
        "        hetatm_withoutX = \"\" # X is not added to fill the breaks in the sequence\n",
        "        hetatm_pos = {} # the positions of the residues in the HETATM\n",
        "        hetatm_max_res_details = 0 # maximum length of the details of the residues in the HETATM\n",
        "\n",
        "        if DNA_AA == \"Nuc\":\n",
        "\n",
        "            UnknownChar = \"N\"\n",
        "\n",
        "        else:\n",
        "\n",
        "            UnknownChar = \"X\"\n",
        "\n",
        "        try:\n",
        "\n",
        "            CIF = open(file, 'r')\n",
        "\n",
        "        except:\n",
        "\n",
        "            return 0\n",
        "\n",
        "        line = CIF.readline()\n",
        "        while line != \"\":\n",
        "\n",
        "            if re.match(r'^_atom_site.', line):\n",
        "\n",
        "                # we reached the atoms\n",
        "                break\n",
        "\n",
        "            if re.match(r'^_entity_poly.entity_id', line):\n",
        "\n",
        "                while line != \"\":\n",
        "\n",
        "                    if ';' in line:\n",
        "\n",
        "                        in_fasta = not in_fasta\n",
        "\n",
        "                    line = line.replace(\";\", \"\")\n",
        "\n",
        "                    if '#' in line:\n",
        "\n",
        "                        # end of _entity_poly.entity_id\n",
        "                        SEQRES_string_found = True\n",
        "                        break\n",
        "\n",
        "                    elif in_fasta:\n",
        "\n",
        "                        # delete white spaces in fasta\n",
        "                        SEQRES_string += line.strip()\n",
        "\n",
        "                    else:\n",
        "\n",
        "                        SEQRES_string += line\n",
        "\n",
        "                    line = CIF.readline()\n",
        "\n",
        "            if SEQRES_string_found:\n",
        "\n",
        "                break\n",
        "\n",
        "            line = CIF.readline()\n",
        "\n",
        "        if re.match(r'^_entity_poly.entity_id\\s+1', SEQRES_string):\n",
        "\n",
        "            # one seqres\n",
        "            match1 = re.search(r'_entity_poly.pdbx_seq_one_letter_code_can\\s+(\\S+)', SEQRES_string)\n",
        "            match2 = re.search(r'_entity_poly.pdbx_strand_id\\s+(\\S+)', SEQRES_string)\n",
        "            if match1 and match2:\n",
        "\n",
        "                seqres = match1.group(1)\n",
        "                for chain in (match2.group(1)).split(','):\n",
        "\n",
        "                    if chain == query_chain:\n",
        "\n",
        "                        self.SEQRES = seqres\n",
        "\n",
        "        else:\n",
        "\n",
        "            # more than one seqres\n",
        "            SEQRES_substrings = re.split(r'\\d+\\s+\\'?poly', SEQRES_string)\n",
        "\n",
        "            POLY = open(\"poly\", 'w')\n",
        "            for string in SEQRES_substrings:\n",
        "\n",
        "                POLY.write(string + \"\\n>\\n\")\n",
        "\n",
        "            POLY.close()\n",
        "\n",
        "            SEQRES_substrings = SEQRES_substrings[1:] # delete titles\n",
        "            for substring in SEQRES_substrings:\n",
        "\n",
        "                words = substring.split()\n",
        "                for chain in (words[5]).split(','):\n",
        "\n",
        "                    if chain == query_chain:\n",
        "\n",
        "                        self.SEQRES = words[4]\n",
        "\n",
        "        number_of_columns = 0\n",
        "        # we find which columns has what value\n",
        "        auth_seq_id_column = 0\n",
        "        auth_comp_id_column = 0\n",
        "        auth_asym_id_column = 0\n",
        "        label_seq_id_column = 0\n",
        "        label_comp_id_column = 0\n",
        "        label_asym_id_column = 0\n",
        "        B_iso_or_equiv = 0\n",
        "        found_auth_seq_id_column = False\n",
        "        found_auth_comp_id_column = False\n",
        "        found_auth_asym_id_column = False\n",
        "        found_label_seq_id_column = False\n",
        "        found_label_comp_id_column = False\n",
        "        found_label_asym_id_column = False\n",
        "        found_B_iso_or_equiv = False\n",
        "        while line != \"\":\n",
        "\n",
        "            line = line.strip()\n",
        "\n",
        "            if line == \"_atom_site.B_iso_or_equiv\":\n",
        "\n",
        "                found_B_iso_or_equiv = True\n",
        "\n",
        "            if line == \"_atom_site.auth_seq_id\":\n",
        "\n",
        "                found_auth_seq_id_column = True\n",
        "\n",
        "            if line == \"_atom_site.auth_comp_id\":\n",
        "\n",
        "                found_auth_comp_id_column = True\n",
        "\n",
        "            if line == \"_atom_site.auth_asym_id\":\n",
        "\n",
        "                found_auth_asym_id_column = True\n",
        "\n",
        "            if line == \"_atom_site.label_seq_id\":\n",
        "\n",
        "                found_label_seq_id_column = True\n",
        "\n",
        "            if line == \"_atom_site.label_comp_id\":\n",
        "\n",
        "                found_label_comp_id_column = True\n",
        "\n",
        "            if line == \"_atom_site.label_asym_id\":\n",
        "\n",
        "                found_label_asym_id_column = True\n",
        "\n",
        "            if not re.match(r'^_atom_site.', line) and found_B_iso_or_equiv and (found_auth_seq_id_column or found_label_seq_id_column) and (found_auth_comp_id_column or found_label_comp_id_column) and (found_auth_asym_id_column or found_label_asym_id_column):\n",
        "\n",
        "                # we identified the necessary columns\n",
        "                break\n",
        "\n",
        "            if found_B_iso_or_equiv:\n",
        "\n",
        "                B_iso_or_equiv -= 1\n",
        "\n",
        "            if found_auth_seq_id_column:\n",
        "\n",
        "                auth_seq_id_column -= 1\n",
        "\n",
        "            if found_auth_comp_id_column:\n",
        "\n",
        "                auth_comp_id_column -= 1\n",
        "\n",
        "            if found_auth_asym_id_column:\n",
        "\n",
        "                auth_asym_id_column -= 1\n",
        "\n",
        "            if found_label_seq_id_column:\n",
        "\n",
        "                label_seq_id_column -= 1\n",
        "\n",
        "            if found_label_comp_id_column:\n",
        "\n",
        "                label_comp_id_column -= 1\n",
        "\n",
        "            if found_label_asym_id_column:\n",
        "\n",
        "                label_asym_id_column -= 1\n",
        "\n",
        "            line = CIF.readline()\n",
        "\n",
        "        if not found_auth_seq_id_column:\n",
        "\n",
        "            auth_seq_id_column = label_seq_id_column\n",
        "\n",
        "        if not found_auth_comp_id_column:\n",
        "\n",
        "            auth_comp_id_column = label_comp_id_column\n",
        "\n",
        "        if not found_auth_asym_id_column:\n",
        "\n",
        "            auth_asym_id_column = label_asym_id_column\n",
        "\n",
        "        FIRST = []\n",
        "        while line.strip() != \"\":\n",
        "\n",
        "            words = line.split()\n",
        "            if words[0] == \"ATOM\" and words[1].isnumeric():\n",
        "\n",
        "                number_of_columns = len(words)\n",
        "\n",
        "                # extract atom data\n",
        "                pos = int(words[auth_seq_id_column])\n",
        "                res = words[auth_comp_id_column]\n",
        "                chain = words[auth_asym_id_column]\n",
        "\n",
        "                # if HETATM is not in the end of the chain we add it to the sequence\n",
        "                if chain == current_chain:\n",
        "\n",
        "                    self.ATOM_withoutX[chain] += hetatm_withoutX\n",
        "\n",
        "                else:\n",
        "\n",
        "                    current_chain = chain\n",
        "\n",
        "                hetatm_withoutX = \"\"\n",
        "\n",
        "                # convert residue to one letter\n",
        "                if res in conversion_table:\n",
        "\n",
        "                    oneLetter = conversion_table[res]\n",
        "\n",
        "                else:\n",
        "\n",
        "                    oneLetter = UnknownChar\n",
        "\n",
        "                # check if we reached a new residue\n",
        "                if not chain in FIRST:\n",
        "\n",
        "                    FIRST.append(chain)\n",
        "                    last_pos = pos\n",
        "                    self.ATOM_withoutX[chain] = oneLetter\n",
        "\n",
        "                elif pos != last_pos:\n",
        "\n",
        "                    self.ATOM_withoutX[chain] += oneLetter\n",
        "\n",
        "                else:\n",
        "\n",
        "                    line = CIF.readline()\n",
        "                    continue\n",
        "\n",
        "                # if the chain is not the query we only extract the sequence\n",
        "                if query_chain != chain:\n",
        "\n",
        "                    #hetatm = \"\"\n",
        "                    line = CIF.readline()\n",
        "                    continue\n",
        "                \"\"\"\n",
        "                else:\n",
        "\n",
        "                    self.ATOM += hetatm\n",
        "                    hetatm = \"\"\n",
        "                \"\"\"\n",
        "\n",
        "\n",
        "                # writing atom position file\n",
        "\n",
        "                self.positions.update(hetatm_pos)\n",
        "                if hetatm_max_res_details > self.max_res_details:\n",
        "\n",
        "                    self.max_res_details = hetatm_max_res_details\n",
        "\n",
        "                hetatm_pos = {}\n",
        "                hetatm_max_res_details = 0\n",
        "\n",
        "                #CORR.write(hetatm_pos)\n",
        "                #hetatm_pos = \"\"\n",
        "\n",
        "\n",
        "                fas_pos += 1\n",
        "\n",
        "                res_details = \"%s:%d:%s\" %(res, pos, chain)\n",
        "                self.positions[fas_pos] = res_details\n",
        "                if len(res_details) > self.max_res_details:\n",
        "\n",
        "                    self.max_res_details = len(res_details)\n",
        "\n",
        "                #CORR.write(\"%s\\t%d\\t%s\\n\" %(res, fas_pos, pos))\n",
        "\n",
        "                # check type\n",
        "                if self.type == \"\":\n",
        "\n",
        "                    if len(res) < 3:\n",
        "\n",
        "                        self.type = \"Nuc\"\n",
        "\n",
        "                    elif len(res) == 3:\n",
        "\n",
        "                        self.type = \"AA\"\n",
        "                \"\"\"\n",
        "                if FIRST[chain]:\n",
        "\n",
        "                    FIRST[chain] = False\n",
        "\n",
        "                elif last_pos < pos:\n",
        "\n",
        "                    while pos != last_pos + 1: # For Disorder regions\n",
        "\n",
        "                        self.ATOM += UnknownChar\n",
        "                        last_pos += 1\n",
        "\n",
        "                self.ATOM += oneLetter\n",
        "                \"\"\"\n",
        "                last_pos = pos\n",
        "\n",
        "            elif words[0] == \"HETATM\" and words[1].isnumeric():\n",
        "\n",
        "\n",
        "                # extract atom data\n",
        "                pos = int(words[auth_seq_id_column])\n",
        "                res = words[auth_comp_id_column]\n",
        "                chain = words[auth_asym_id_column]\n",
        "\n",
        "\n",
        "                # convert residue to one letter\n",
        "                if res in conversion_table:\n",
        "\n",
        "                    oneLetter = conversion_table[res]\n",
        "\n",
        "                else:\n",
        "\n",
        "                    oneLetter = UnknownChar\n",
        "\n",
        "                # check if we reached a new residue\n",
        "                if not chain in FIRST:\n",
        "\n",
        "                    FIRST.append(chain)\n",
        "                    last_pos = pos\n",
        "                    self.ATOM_withoutX[chain] = \"\"\n",
        "                    hetatm_withoutX = oneLetter\n",
        "\n",
        "                elif pos != last_pos:\n",
        "\n",
        "                    hetatm_withoutX += oneLetter\n",
        "\n",
        "                else:\n",
        "\n",
        "                    line = CIF.readline()\n",
        "                    continue\n",
        "\n",
        "                # if the chain is not the query we only extract the sequence\n",
        "                if query_chain != chain:\n",
        "\n",
        "                    line = CIF.readline()\n",
        "                    continue\n",
        "\n",
        "                # writing atom position file\n",
        "                fas_pos += 1\n",
        "\n",
        "                res_details = \"%s:%d:%s\" %(res, pos, chain)\n",
        "                hetatm_pos[fas_pos] = res_details\n",
        "                if len(res_details) > hetatm_max_res_details:\n",
        "\n",
        "                    hetatm_max_res_details = len(res_details)\n",
        "\n",
        "                #hetatm_pos += \"%s\\t%d\\t%s\\n\" %(res, fas_pos, pos)\n",
        "\n",
        "                # check type\n",
        "                if self.type == \"\":\n",
        "\n",
        "                    if len(res) < 3:\n",
        "\n",
        "                        self.type = \"Nuc\"\n",
        "\n",
        "                    elif len(res) == 3:\n",
        "\n",
        "                        self.type = \"AA\"\n",
        "                \"\"\"\n",
        "                if FIRST[chain]:\n",
        "\n",
        "                    FIRST[chain] = False\n",
        "\n",
        "                elif last_pos < pos:\n",
        "\n",
        "                    while pos != last_pos + 1: # For Disorder regions\n",
        "\n",
        "                        hetatm += UnknownChar\n",
        "                        last_pos += 1\n",
        "\n",
        "                hetatm += oneLetter\n",
        "                \"\"\"\n",
        "                last_pos = pos\n",
        "\n",
        "            line = CIF.readline()\n",
        "\n",
        "        CIF.close()\n",
        "        #CORR.close()\n",
        "\n",
        "        self.auth_seq_id_column = number_of_columns + auth_seq_id_column\n",
        "        self.auth_comp_id_column = number_of_columns + auth_comp_id_column\n",
        "        self.auth_asym_id_column = number_of_columns + auth_asym_id_column\n",
        "        self.B_iso_or_equiv = number_of_columns + B_iso_or_equiv\n",
        "        LOG.write(\"residue number column - %s\\nresidue name column - %s\\nchain id column - %s\\nb-factor column - %s\\n\" %(self.auth_seq_id_column + 1, self.auth_comp_id_column + 1, self.auth_asym_id_column + 1, self.B_iso_or_equiv + 1))\n",
        "\n",
        "        return 1\n",
        "\n",
        "\n",
        "    def get_max_res_details(self):\n",
        "\n",
        "        return self.max_res_details\n",
        "\n",
        "    def get_columns(self):\n",
        "\n",
        "        return self.auth_seq_id_column, self.auth_comp_id_column, self.auth_asym_id_column, self.B_iso_or_equiv\n",
        "\n",
        "    def get_type(self):\n",
        "\n",
        "        return self.type\n",
        "\n",
        "    def get_SEQRES(self):\n",
        "\n",
        "        return self.SEQRES\n",
        "\n",
        "    def get_ATOM_withoutX(self):\n",
        "\n",
        "        return self.ATOM_withoutX\n",
        "\n",
        "    def get_MODIFIED_COUNT(self):\n",
        "\n",
        "\n",
        "        return self.MODIFIED_COUNT\n",
        "\n",
        "\n",
        "\n",
        "    def get_MODIFIED_LIST(self):\n",
        "\n",
        "        return self.MODIFIED_LIST\n",
        "\n",
        "\n",
        "    def get_positions(self):\n",
        "\n",
        "        return self.positions\n",
        "\n",
        "\n",
        "\n",
        "bayesInterval = 3\n",
        "ColorScale = {0 : 9, 1 : 8, 2 : 7, 3 : 6, 4 : 5, 5 : 4, 6 : 3, 7 : 2, 8 : 1}\n",
        "\n",
        "\n",
        "\n",
        "def insert_spaces(original, spaced_lines, B_iso_or_equiv):\n",
        "\n",
        "    # insert extra spaces after the tmp column, to allow room for the rate4site scores.\n",
        "\n",
        "    try:\n",
        "\n",
        "        READ_PDB = open(original, 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        return(\"cp_rasmol_gradesPE_and_pipe.replace_tempFactor: Can't open '\" + original + \"' for reading.\")\n",
        "\n",
        "    lines = READ_PDB.readlines()\n",
        "    READ_PDB.close()\n",
        "\n",
        "\n",
        "    max_start = 0\n",
        "    max_end = 0\n",
        "    for line in lines:\n",
        "\n",
        "        line = line.strip()\n",
        "        if re.match(r'^ATOM', line):\n",
        "\n",
        "            match = re.match(r'^((\\S+\\s+){'+ str(B_iso_or_equiv) + r'}\\S+)\\s+(\\S.*)', line)\n",
        "            length_start = len(match.group(1))\n",
        "            length_end = len(match.group(3))\n",
        "            if length_start > max_start:\n",
        "\n",
        "                max_start = length_start\n",
        "\n",
        "            if length_end > max_end:\n",
        "\n",
        "                max_end = length_end\n",
        "\n",
        "    for line in lines:\n",
        "\n",
        "        if re.match(r'^ATOM', line):\n",
        "\n",
        "            line = line.strip()\n",
        "            match = re.match(r'^((\\S+\\s+){'+ str(B_iso_or_equiv) + r'}\\S+)\\s+(\\S.*)', line)\n",
        "            line_start = match.group(1)\n",
        "            line_end = match.group(3)\n",
        "            while len(line_start) < max_start:\n",
        "\n",
        "                line_start = line_start + \" \"\n",
        "\n",
        "            while len(line_end) < max_end:\n",
        "\n",
        "                line_end = \" \" + line_end\n",
        "\n",
        "            spaced_lines.append(line_start + \"     \" + line_end + \"\\n\")\n",
        "\n",
        "        else:\n",
        "\n",
        "            spaced_lines.append(line)\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "def check_if_rate4site_failed(r4s_log):\n",
        "\n",
        "    res_flag = vars['r4s_out']\n",
        "    if not os.path.exists(res_flag) or os.path.getsize(res_flag) == 0: # 1\n",
        "\n",
        "        LOG.write(\"check_if_rate4site_failed : the file \" + res_flag + \" either does not exist or is empty. \\n\")\n",
        "        return True\n",
        "\n",
        "    try:\n",
        "\n",
        "        R4S_RES = open(res_flag, 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        LOG.write(\"check_if_rate4site_failed : can not open file: \" + res_flag + \". aborting.\\n\")\n",
        "        return True\n",
        "\n",
        "    error = False\n",
        "    line = R4S_RES.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        if \"likelihood of pos was zero\" in line:\n",
        "\n",
        "            LOG.write(\"check_if_rate4site_failed : the line: \\\"likelihood of pos was zero\\\" was found in %s.\\n\" %r4s_log)\n",
        "            error = True\n",
        "            break\n",
        "\n",
        "        if re.match(r'rate of pos\\:\\s\\d\\s=', line):\n",
        "\n",
        "            # output found\n",
        "            break\n",
        "\n",
        "        if \"Bad format in tree file\" in line:\n",
        "\n",
        "            exit_on_error('user_error', \"check_if_rate4site_failed : There is an error in the tree file format. Please check that your tree is in the <a href = \\\"\" + GENERAL_CONSTANTS.CONSURF_TREE_FAQ + \"\\\">requested format</a> and reupload it to the server.<br>\")\n",
        "\n",
        "        line = R4S_RES.readline()\n",
        "\n",
        "    R4S_RES.close()\n",
        "    return error\n",
        "\n",
        "\n",
        "\n",
        "\"\"\"\n",
        "def check_validity_tree_file(nodes):\n",
        "\n",
        "\t# checks validity of tree file and returns an array with the names of the nodes\n",
        "    try:\n",
        "\n",
        "        TREEFILE = open(form['tree_name'], 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"check_validity_tree_file : can't open the file \" + form['tree_name'] + \" for reading.\")\n",
        "\n",
        "    tree = TREEFILE.read()\n",
        "    TREEFILE.close()\n",
        "    tree.replace(\"\\n\", \"\")\n",
        "    if tree[-1] != ';':\n",
        "\n",
        "\t    tree += ';'\n",
        "\n",
        "    try:\n",
        "\n",
        "        TREEFILE = open(vars['working_dir'] + vars['tree_file'], 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"check_validity_tree_file : can't open the file \" + vars['working_dir'] + vars['tree_file'] + \" for writing.\")\n",
        "\n",
        "    TREEFILE.write(tree)\n",
        "    TREEFILE.close()\n",
        "\n",
        "    leftBrackets = 0\n",
        "    rightBrackets = 0\n",
        "    noRegularFormatChar = \"\"\n",
        "    #nodes = []\n",
        "    node_name = \"\"\n",
        "    in_node_name = False\n",
        "    in_node_score = False\n",
        "    for char in tree:\n",
        "\n",
        "        if char == ':':\n",
        "\n",
        "            if in_node_name:\n",
        "\n",
        "                nodes.append(node_name)\n",
        "\n",
        "            node_name = \"\"\n",
        "            in_node_name = False\n",
        "            in_node_score = True\n",
        "\n",
        "        elif char == '(':\n",
        "\n",
        "            leftBrackets += 1\n",
        "\n",
        "        elif char == ')':\n",
        "\n",
        "            rightBrackets += 1\n",
        "            in_node_score = False\n",
        "\n",
        "        elif char == ',':\n",
        "\n",
        "            in_node_score = False\n",
        "\n",
        "        elif char != ';':\n",
        "\n",
        "            if char in \"!@#$^&*~`{}'?<>\" and not char in noRegularFormatChar:\n",
        "\n",
        "                noRegularFormatChar += \" '\" + char + \"', \"\n",
        "\n",
        "            if not in_node_score:\n",
        "\n",
        "                node_name += char\n",
        "                in_node_name = True\n",
        "\n",
        "    if leftBrackets != rightBrackets:\n",
        "\n",
        "        msg = \"The uploaded tree file, which appears to be in Newick format, is missing parentheses.\"\n",
        "        exit_on_error('user_error', msg)\n",
        "\n",
        "    if noRegularFormatChar != \"\":\n",
        "\n",
        "        msg = \"The uploaded tree file, which appears to be in Newick format, ontains the following non-standard characters: \" + noRegularFormatChar[:-2]\n",
        "        exit_on_error('user_error', msg)\n",
        "\n",
        "    LOG.write(\"check_validity_tree_file : tree is valid\\n\")\n",
        "\n",
        "    #return nodes\n",
        "\"\"\"\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "def get_query_seq_in_MSA():\n",
        "\n",
        "    # returns the query sequence with gaps as it's in the MSA\n",
        "    try:\n",
        "\n",
        "        FASTA = open(vars['msa_fasta'], 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"get_query_seq_in_MSA : can't open the file \" + vars['msa_fasta'] + \" for reading.\")\n",
        "\n",
        "    found = False\n",
        "    seq = \"\"\n",
        "    line = FASTA.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        first_word = line.split()[0]\n",
        "        if found:\n",
        "\n",
        "            if first_word[0] == '>':\n",
        "\n",
        "                break\n",
        "\n",
        "            else:\n",
        "\n",
        "                seq += first_word\n",
        "\n",
        "        elif first_word == \">\"  + vars['msa_SEQNAME']:\n",
        "\n",
        "                found = True\n",
        "\n",
        "        line = FASTA.readline()\n",
        "\n",
        "    FASTA.close()\n",
        "    return seq\n",
        "\n",
        "def get_positions_in_MSA():\n",
        "\n",
        "    # returns the positions in the MSA were the query is not a gap\n",
        "    seq = get_query_seq_in_MSA()\n",
        "    positions = get_seq_legal_positions(seq)\n",
        "    return positions\n",
        "\n",
        "def get_seq_legal_positions(seq):\n",
        "\n",
        "    # return a aaray with the positions of the legal chars\n",
        "    positions = []\n",
        "    if form['DNA_AA'] == \"AA\":\n",
        "\n",
        "        legal_chars = [\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\", \"X\"]\n",
        "\n",
        "    else:\n",
        "\n",
        "        legal_chars = [\"A\", \"C\", \"G\", \"T\", \"U\", \"N\"]\n",
        "\n",
        "    for i in range(len(seq)):\n",
        "\n",
        "        if seq[i].upper() in legal_chars:\n",
        "\n",
        "            positions.append(i)\n",
        "\n",
        "    return positions\n",
        "\n",
        "\n",
        "\n",
        "def write_MSA_percentage_file():\n",
        "\n",
        "    # writes a file with the precentage of each acid in each position\n",
        "\n",
        "    #index_of_pos = get_positions_in_MSA()\n",
        "    percentage_per_pos = [] # precentage of each acid in each position in the MSA\n",
        "    #unknown_per_pos = [] # precentage of unknown acid in each position in the MSA\n",
        "    number_of_positions = len(vars['protein_seq_string'])\n",
        "    #number_of_positions = len(index_of_pos)\n",
        "    query_with_gaps = get_query_seq_in_MSA() # query sequence with gaps\n",
        "\n",
        "    if form['DNA_AA'] == \"AA\":\n",
        "\n",
        "        acids = [\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\", \"X\"]\n",
        "        unknown = \"X\"\n",
        "\n",
        "        for i in range(number_of_positions):\n",
        "\n",
        "            percentage_per_pos.append({})\n",
        "            #unknown_per_pos.append(0)\n",
        "\n",
        "    else:\n",
        "\n",
        "        acids = [\"A\", \"C\", \"G\", \"T\", \"U\", \"N\"]\n",
        "        unknown = \"N\"\n",
        "\n",
        "        for i in range(number_of_positions):\n",
        "\n",
        "            percentage_per_pos.append({})\n",
        "            #unknown_per_pos.append(0)\n",
        "\n",
        "    try:\n",
        "\n",
        "        FASTA = open(vars['msa_fasta'], 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"write_MSA_percentage_file : can't open the file \" + vars['msa_fasta'] + \" for reading.\")\n",
        "\n",
        "    try:\n",
        "\n",
        "        PRECENTAGE_FILE = open(vars['Msa_percentageFILE'], 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"write_MSA_percentage_file : can't open the file \" + vars['Msa_percentageFILE'] + \" for writing.\")\n",
        "\n",
        "    # we find the precentage of each acid in each position\n",
        "    seq = \"\"\n",
        "    first = True\n",
        "    line = FASTA.readline()\n",
        "    while True:\n",
        "\n",
        "        if line == \"\" or line[0] == '>':\n",
        "\n",
        "            if first:\n",
        "\n",
        "                first = False\n",
        "\n",
        "            else:\n",
        "\n",
        "                #pos = 0\n",
        "                #for i in index_of_pos:\n",
        "                i = 0\n",
        "                for pos in range(number_of_positions):\n",
        "\n",
        "                    while i < len(query_with_gaps) and query_with_gaps[i] == '-':\n",
        "\n",
        "                        i += 1\n",
        "\n",
        "                    char = seq[i]\n",
        "                    if char in  acids:\n",
        "\n",
        "                        if char in percentage_per_pos[pos]:\n",
        "\n",
        "                            percentage_per_pos[pos][char] += 1\n",
        "\n",
        "                        else:\n",
        "\n",
        "                            percentage_per_pos[pos][char] = 1\n",
        "\n",
        "                    \"\"\"\n",
        "                    elif char != '-':\n",
        "\n",
        "                        unknown_per_pos[pos] += 1\n",
        "                    \"\"\"\n",
        "                    i += 1\n",
        "\n",
        "                seq = \"\"\n",
        "\n",
        "            # this is for the last sequence in the msa\n",
        "            if line == \"\":\n",
        "\n",
        "                break\n",
        "\n",
        "        else:\n",
        "\n",
        "            seq += line.strip().upper()\n",
        "\n",
        "        line = FASTA.readline()\n",
        "\n",
        "    FASTA.close()\n",
        "\n",
        "    # sort dictionaries\n",
        "    for pos in range(number_of_positions):\n",
        "\n",
        "        # we sort the amino acids but not the unknown character\n",
        "        unknown_percent = percentage_per_pos[pos].pop(unknown, None)\n",
        "        percentage_per_pos[pos] = dict(sorted(percentage_per_pos[pos].items(), key=lambda item: item[1], reverse=True))\n",
        "        if unknown_percent:\n",
        "\n",
        "            percentage_per_pos[pos][unknown] = unknown_percent\n",
        "\n",
        "    # calculate the percentage\n",
        "    for pos in range(number_of_positions):\n",
        "\n",
        "        sum = 0.0\n",
        "        for char in percentage_per_pos[pos]:\n",
        "\n",
        "            sum += percentage_per_pos[pos][char]\n",
        "\n",
        "        #sum += unknown_per_pos[pos]\n",
        "\n",
        "        for char in percentage_per_pos[pos]:\n",
        "\n",
        "            percentage_per_pos[pos][char] = 100 * (percentage_per_pos[pos][char] / sum)\n",
        "\n",
        "        #unknown_per_pos[pos] = 100 * (unknown_per_pos[pos] / sum)\n",
        "\n",
        "    # we write the file\n",
        "    PRECENTAGE_FILE.write(\"\\\"The table details the residue variety in % for each position in the query sequence.\\\"\\n\\\"Each column shows the % for that amino-acid, found in position ('pos') in the MSA.\\\"\\n\\\"In case there are residues which are not a standard amino-acid in the MSA, they are represented under column 'OTHER'\\\"\\n\\npos,A,C,D,E,F,G,H,I,K,L,M,N,P,Q,R,S,T,V,W,Y,OTHER,MAX ACID,ConSurf grade\\n\\n\")\n",
        "    pos = 0\n",
        "    for i in range(number_of_positions):\n",
        "\n",
        "        if vars['protein_seq_string'][i] == unknown:\n",
        "\n",
        "            continue\n",
        "\n",
        "\t\t# position\n",
        "        PRECENTAGE_FILE.write(\"%d\" %(pos + 1))\n",
        "\n",
        "\t\t# known acids\n",
        "        for char in acids:\n",
        "\n",
        "            PRECENTAGE_FILE.write(\",\")\n",
        "            if char in percentage_per_pos[pos]:\n",
        "\n",
        "                PRECENTAGE_FILE.write(\"%.3f\" %percentage_per_pos[pos][char])\n",
        "\n",
        "            else:\n",
        "\n",
        "                PRECENTAGE_FILE.write(\"0\")\n",
        "\n",
        "\t\t# unknown acids\n",
        "        #PRECENTAGE_FILE.write(\",%.3f\" %unknown_per_pos[pos])\n",
        "\n",
        "\t\t# max acid\n",
        "        keys = list(percentage_per_pos[pos].keys())\n",
        "        max_acid = \",%s %.3f\" %(keys[0], percentage_per_pos[pos][keys[0]])\n",
        "        PRECENTAGE_FILE.write(max_acid.rstrip(\"0\").rstrip(\".\"))\n",
        "\n",
        "        # ConSurf grade\n",
        "        PRECENTAGE_FILE.write(\",%s\\n\" %vars['gradesPE_Output'][pos]['COLOR'])\n",
        "\n",
        "        pos += 1\n",
        "\n",
        "    PRECENTAGE_FILE.close()\n",
        "\n",
        "    vars['percentage_per_pos'] = percentage_per_pos\n",
        "    #vars['unknown_per_pos'] = unknown_per_pos\n",
        "\n",
        "\n",
        "    if form['DNA_AA'] != \"AA\":\n",
        "\n",
        "        vars['B/E'] = False\n",
        "        return\n",
        "\n",
        "    vars['B/E'] = True\n",
        "    buried_exposed = []\n",
        "    reveal_buried_exposed(buried_exposed)\n",
        "\n",
        "    pos = 0\n",
        "    for element in vars['gradesPE_Output']:\n",
        "\n",
        "        while pos < number_of_positions and vars['protein_seq_string'][pos] == unknown:\n",
        "\n",
        "            pos += 1\n",
        "\n",
        "        element['B/E'] = buried_exposed[pos]\n",
        "        pos += 1\n",
        "        if element['B/E'] == \"e\":\n",
        "\n",
        "            if element['COLOR'] == 9 or element['COLOR'] == 8:\n",
        "\n",
        "                element['F/S'] = \"f\"\n",
        "\n",
        "            else:\n",
        "\n",
        "                element['F/S'] = \" \"\n",
        "\n",
        "        elif element['COLOR'] == 9:\n",
        "\n",
        "            element['F/S'] = \"s\"\n",
        "\n",
        "        else:\n",
        "\n",
        "            element['F/S'] = \" \"\n",
        "\n",
        "\n",
        "def database_address(name):\n",
        "\n",
        "    parts = name.split('|')\n",
        "    if parts[0] == \"ur\":\n",
        "\n",
        "        return(\"https://www.uniprot.org/uniref/UniRef90_\" + parts[1])\n",
        "\n",
        "    elif parts[0] == \"up\":\n",
        "\n",
        "        return(\"https://www.uniprot.org/uniprot/\" + parts[1])\n",
        "\n",
        "    else:\n",
        "\n",
        "        return(\"https://www.ncbi.nlm.nih.gov/nuccore/\" + parts[1])\n",
        "\n",
        "def get_details(first_line, second_line, third_line):\n",
        "\n",
        "    query_seq = (first_line.split())[2]\n",
        "    db_seq = (third_line.split())[2]\n",
        "\n",
        "    gaps = 0\n",
        "    positives = 0\n",
        "    identities = 0\n",
        "    length = len(query_seq)\n",
        "\n",
        "    for char in query_seq:\n",
        "\n",
        "        if char == \"-\":\n",
        "\n",
        "            gaps += 1\n",
        "\n",
        "    for char in db_seq:\n",
        "\n",
        "        if char == \"-\":\n",
        "\n",
        "            gaps += 1\n",
        "\n",
        "    for char in second_line:\n",
        "\n",
        "        if char == \"+\":\n",
        "\n",
        "            positives += 1\n",
        "\n",
        "        if char.isalpha():\n",
        "\n",
        "            identities += 1\n",
        "\n",
        "    details = \"\"\n",
        "\n",
        "    if gaps != 0:\n",
        "\n",
        "        details += \", Gaps = %d/%d (%d%%)\" %(gaps, length, int((float(gaps) / length) * 100))\n",
        "\n",
        "    else:\n",
        "\n",
        "        details += \", Gaps = 0\"\n",
        "\n",
        "    if positives != 0:\n",
        "\n",
        "        details += \", Positives = %d/%d (%d%%)\" %(positives, length, int((float(positives) / length) * 100))\n",
        "\n",
        "    else:\n",
        "\n",
        "        details += \", Positives = 0\"\n",
        "\n",
        "    if identities != 0:\n",
        "\n",
        "        details += \", Identities = %d/%d (%d%%)\" %(identities, length, int((float(identities) / length) * 100))\n",
        "\n",
        "    else:\n",
        "\n",
        "        details += \", Identities = 0\"\n",
        "\n",
        "    return details\n",
        "\n",
        "\n",
        "def replace_TmpFactor_Rate4Site_Scores_CIF(PdbFile, chain, HashRef, Out):\n",
        "\n",
        "    # replace the tempFactor column in the PDB file\n",
        "\n",
        "    [auth_seq_id_column, auth_comp_id_column, auth_asym_id_column, B_iso_or_equiv] = vars['pdb_object'].get_columns()\n",
        "\n",
        "    spaced_lines = []\n",
        "    insert_spaces(PdbFile, spaced_lines, B_iso_or_equiv)\n",
        "\n",
        "    # write the PDB file and replace the tempFactor column\n",
        "    # with the new one.\n",
        "    try:\n",
        "\n",
        "        WRITE_PDB = open(Out, 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"replace_TmpFactor_Rate4Site_Scores_CIF: Can't open '\" + Out + \"' for writing.\")\n",
        "\n",
        "    for line in spaced_lines:\n",
        "\n",
        "        if re.match(r'^ATOM', line):\n",
        "\n",
        "            words = line.split()\n",
        "            PDBchain = words[auth_asym_id_column]\n",
        "            ResNum = words[auth_seq_id_column]\n",
        "            if PDBchain == chain and ResNum in HashRef:\n",
        "\n",
        "                score = HashRef[ResNum]\n",
        "                match = re.match(r'^((\\S+\\s+){'+ str(B_iso_or_equiv) + r'})(\\S+\\s+)(.+)', line)\n",
        "                length_temp_fact = len(match.group(3))\n",
        "                while len(score) < length_temp_fact:\n",
        "\n",
        "                    score = score + \" \"\n",
        "\n",
        "                WRITE_PDB.write(match.group(1) + score + match.group(4) + \"\\n\")\n",
        "\n",
        "            else:\n",
        "\n",
        "                WRITE_PDB.write(line)\n",
        "\n",
        "        else:\n",
        "\n",
        "            WRITE_PDB.write(line)\n",
        "\n",
        "    WRITE_PDB.close()\n",
        "\n",
        "\n",
        "\n",
        "def replace_TmpFactor_Rate4Site_Scores_PDB(PdbFile, chain, HashRef, Out):\n",
        "\n",
        "    # replace the tempFactor column in the PDB file\n",
        "\n",
        "    # read the PDB file to an array\n",
        "    try:\n",
        "\n",
        "        READ_PDB = open(PdbFile, 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"replace_TmpFactor_Rate4Site_Scores_PDB : Can't open '\" + PdbFile + \"' for reading.\")\n",
        "\n",
        "    # write the PDB file and replace the tempFactor column\n",
        "    # with the new one.\n",
        "    try:\n",
        "\n",
        "        WRITE_PDB = open(Out, 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"replace_TmpFactor_Rate4Site_Scores_PDB : Can't open '\" + Out + \"' for writing.\")\n",
        "\n",
        "    line = READ_PDB.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        if re.match(r'^ATOM', line):\n",
        "\n",
        "            PDBchain = line[21:22]\n",
        "            ResNum = (line[22:27]).strip()\n",
        "\n",
        "            if PDBchain == chain and ResNum in HashRef:\n",
        "\n",
        "                while len(HashRef[ResNum]) < 6:\n",
        "\n",
        "                    HashRef[ResNum] = \" \" + HashRef[ResNum]\n",
        "\n",
        "                WRITE_PDB.write(line[:60] + HashRef[ResNum] + line[66:])\n",
        "\n",
        "            else:\n",
        "\n",
        "                WRITE_PDB.write(line[:60] + \"      \" + line[66:])\n",
        "\n",
        "        else:\n",
        "\n",
        "            WRITE_PDB.write(line)\n",
        "\n",
        "        line = READ_PDB.readline()\n",
        "\n",
        "    READ_PDB.close()\n",
        "    WRITE_PDB.close()\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "def read_Rate4Site_gradesPE(gradesPE_file, gradesPE_hash_ref):\n",
        "\n",
        "    # the routine matches each position in the gradesPE file its Rate4Site grade.\n",
        "\n",
        "    try:\n",
        "\n",
        "        GRADES = open(gradesPE_file, 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error(\"sys_error\", \"read_Rate4Site_gradesPE : Can't open '\" + gradesPE_file + \"' for reading.\")\n",
        "\n",
        "    line = GRADES.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        if re.match(r'^\\s*\\d+\\s+\\w', line):\n",
        "\n",
        "            grades = line.split()\n",
        "            #ResNum = re.sub(r'[a-z]', \"\", grades[2], flags=re.IGNORECASE)\n",
        "            if grades[2] != \"-\":\n",
        "\n",
        "                 grades[2] = (grades[2]).split(\":\")[1]\n",
        "\n",
        "            gradesPE_hash_ref[grades[2]] = grades[3]\n",
        "\n",
        "        line = GRADES.readline()\n",
        "\n",
        "    GRADES.close()\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "def read_ConSurf_gradesPE(gradesPE_file, gradesPE_hash_ref, gradesPE_ISD_hash_ref):\n",
        "\n",
        "    # the routine matches each position in the gradesPE file its ConSurf grade. In case there was a grade mark with *, we put it in a seperate hash with the grade 0.\n",
        "    # the routine returns \"yes\" if a * was found and \"no\" otherwise\n",
        "\n",
        "    insufficient = False\n",
        "\n",
        "    try:\n",
        "\n",
        "        GRADES = open(gradesPE_file, 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"read_ConSurf_gradesPE can't open '\" + gradesPE_file + \"' for reading.\")\n",
        "\n",
        "    line = GRADES.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        if re.match(r'^\\s*\\d+\\s+\\w', line):\n",
        "\n",
        "            grades = line.split()\n",
        "            if grades[2] != \"-\":\n",
        "\n",
        "                 grades[2] = ((grades[2]).split(':'))[1]\n",
        "\n",
        "            #grades[2] = re.sub(r'[a-z]', \"\", grades[2], flags=re.IGNORECASE)\n",
        "            if re.match(r'\\d\\*?', grades[4]):\n",
        "\n",
        "                # if it is insufficient color - we change its grade to 0, which will be read as light yellow\n",
        "                if re.match(r'(\\d)\\*', grades[4]):\n",
        "\n",
        "                    gradesPE_hash_ref[grades[2]] = grades[4]\n",
        "                    gradesPE_ISD_hash_ref[grades[2]] = \"10\"\n",
        "                    insufficient = True\n",
        "\n",
        "                else:\n",
        "\n",
        "                    gradesPE_hash_ref[grades[2]] = grades[4]\n",
        "                    gradesPE_ISD_hash_ref[grades[2]] = grades[4]\n",
        "\n",
        "        line = GRADES.readline()\n",
        "\n",
        "    GRADES.close()\n",
        "\n",
        "    return(insufficient)\n",
        "\n",
        "\n",
        "\n",
        "def replace_TmpFactor_Consurf_Scores_CIF(atom_grades, query_chain, pdb_file, prefix):\n",
        "\n",
        "    # Creates The ATOM section with ConSurf grades instead of the TempFactor column, creates PDB file with ConSurf grades\n",
        "\n",
        "    pdb_with_grades = prefix + \"_ATOMS_section_With_ConSurf.cif\"\n",
        "    pdb_with_grades_isd = prefix + \"_ATOMS_section_With_ConSurf_isd.cif\"\n",
        "    pdb_with_scores = prefix + \"_With_Conservation_Scores.cif\"\n",
        "\n",
        "    [auth_seq_id_column, auth_comp_id_column, auth_asym_id_column, B_iso_or_equiv] = vars['pdb_object'].get_columns()\n",
        "\n",
        "    try:\n",
        "\n",
        "        PDB = open(pdb_file, 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"could not open file '\" + pdb_file + \"' for reading.\\n\")\n",
        "\n",
        "    try:\n",
        "\n",
        "        GRADES = open(pdb_with_grades, 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"could not open the file '\" + pdb_with_grades + \"' for writing.\\n\")\n",
        "\n",
        "    try:\n",
        "\n",
        "        SCORES = open(pdb_with_scores, 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"could not open the file '\" + pdb_with_scores + \"' for writing.\\n\")\n",
        "\n",
        "    if vars['insufficient_data']:\n",
        "\n",
        "        try:\n",
        "\n",
        "            GRADES_ISD = open(pdb_with_grades_isd, 'w')\n",
        "\n",
        "        except:\n",
        "\n",
        "            exit_on_error('sys_error', \"could not open the file '\" + pdb_with_grades_isd + \"' for writing.\\n\")\n",
        "\n",
        "    line = PDB.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        if line[:4] == \"ATOM\" or line[:6] == \"HETATM\":\n",
        "\n",
        "            words = line.split()\n",
        "            chain = words[auth_asym_id_column]\n",
        "            residue_number = words[auth_seq_id_column]\n",
        "\n",
        "            grade = \"0\"\n",
        "            score = \"0\"\n",
        "            grade_isd = \"0\"\n",
        "\n",
        "            if residue_number in atom_grades and chain == query_chain:\n",
        "\n",
        "                # getting the grade\n",
        "\n",
        "                [grade, isd, score] = atom_grades[residue_number]\n",
        "\n",
        "                if vars['insufficient_data']:\n",
        "\n",
        "                    if isd == 1:\n",
        "\n",
        "                        grade_isd = \"10\"\n",
        "\n",
        "                    else:\n",
        "\n",
        "                        grade_isd = grade\n",
        "\n",
        "            match = re.match(r'^((\\S+\\s+){' + str(B_iso_or_equiv) + r'})(\\S+\\s+)(.+)', line)\n",
        "            length_temp_fact = len(match.group(3))\n",
        "            line_start = match.group(1)\n",
        "            line_end = match.group(4)\n",
        "            while len(grade) < length_temp_fact:\n",
        "\n",
        "                grade = grade + \" \"\n",
        "\n",
        "            while len(score) < length_temp_fact:\n",
        "\n",
        "                score = score + \" \"\n",
        "\n",
        "            while len(grade_isd) < length_temp_fact:\n",
        "\n",
        "                grade_isd = grade_isd + \" \"\n",
        "\n",
        "            GRADES.write(line_start + grade + line_end + \"\\n\")\n",
        "            SCORES.write(line_start + score + line_end + \"\\n\")\n",
        "            if vars['insufficient_data']:\n",
        "\n",
        "                GRADES_ISD.write(line_start + grade_isd + line_end + \"\\n\")\n",
        "\n",
        "\n",
        "        else:\n",
        "\n",
        "            GRADES.write(line)\n",
        "            SCORES.write(line)\n",
        "            if vars['insufficient_data']:\n",
        "\n",
        "                GRADES_ISD.write(line)\n",
        "\n",
        "        line = PDB.readline()\n",
        "\n",
        "    GRADES.close()\n",
        "    SCORES.close()\n",
        "    vars['zip_list'].append(pdb_with_grades)\n",
        "    vars['zip_list'].append(pdb_with_scores)\n",
        "    if vars['insufficient_data']:\n",
        "\n",
        "        show_py3dmol(pdb_with_grades_isd, \"cif\")\n",
        "        print_instructions(pdb_with_grades, \"CIF\", pdb_with_grades_isd)\n",
        "\n",
        "    else:\n",
        "\n",
        "        show_py3dmol(pdb_with_grades, \"cif\")\n",
        "        print_instructions(pdb_with_grades, \"CIF\")\n",
        "\n",
        "\n",
        "def replace_TmpFactor_Consurf_Scores_PDB(atom_grades, query_chain, pdb_file, prefix):\n",
        "\n",
        "    # Creates The ATOM section with ConSurf grades instead of the TempFactor column, creates PDB file with ConSurf grades\n",
        "\n",
        "\n",
        "    pdb_with_grades = prefix + \"_ATOMS_section_With_ConSurf.pdb\"\n",
        "    pdb_with_grades_isd = prefix + \"_ATOMS_section_With_ConSurf_isd.pdb\"\n",
        "    pdb_with_scores = prefix + \"_With_Conservation_Scores.pdb\"\n",
        "\n",
        "    try:\n",
        "\n",
        "        PDB = open(pdb_file, 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"replace_TmpFactor_Consurf_Scores_PDB : could not open file '\" + pdb_file + \"' for reading.\\n\")\n",
        "\n",
        "    try:\n",
        "\n",
        "        GRADES = open(pdb_with_grades, 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"replace_TmpFactor_Consurf_Scores_PDB : could not open the file '\" + pdb_with_grades + \"' for writing.\\n\")\n",
        "\n",
        "    try:\n",
        "\n",
        "        SCORES = open(pdb_with_scores, 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"replace_TmpFactor_Consurf_Scores_PDB : could not open the file '\" + pdb_with_scores + \"' for writing.\\n\")\n",
        "\n",
        "    if vars['insufficient_data']:\n",
        "\n",
        "        try:\n",
        "\n",
        "            GRADES_ISD = open(pdb_with_grades_isd, 'w')\n",
        "\n",
        "        except:\n",
        "\n",
        "            exit_on_error('sys_error', \"replace_TmpFactor_Consurf_Scores_PDB : could not open the file '\" + pdb_with_grades_isd + \"' for writing.\\n\")\n",
        "\n",
        "    line = PDB.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        if line.strip() == \"\":\n",
        "\n",
        "            # line is empty\n",
        "            line = PDB.readline()\n",
        "            continue\n",
        "\n",
        "        if line[:4] == \"ATOM\" or line[:6] == \"HETATM\":\n",
        "\n",
        "            chain = line[21:22]\n",
        "            if chain == \" \":\n",
        "\n",
        "                chain = \"NONE\"\n",
        "\n",
        "            residue = (line[22:27]).strip()\n",
        "            if residue in atom_grades and chain == query_chain:\n",
        "\n",
        "                [grade, isd, score] = atom_grades[residue]\n",
        "                while len(score) < 6:\n",
        "\n",
        "                    score = \" \" + score\n",
        "\n",
        "                # the TF is updated with the grades and scores\n",
        "                GRADES.write(line[:60] + \"     \" + grade + \"      \\n\")\n",
        "                SCORES.write(line[:60] + score + line[66:])\n",
        "\n",
        "                if vars['insufficient_data']:\n",
        "\n",
        "                    # the TF is updated with the number from gradesPE showing isd\n",
        "                    if isd == 1:\n",
        "\n",
        "                        GRADES_ISD.write(line[:60] + \"    10      \\n\")\n",
        "\n",
        "                    else:\n",
        "\n",
        "                        GRADES_ISD.write(line[:60] + \"    \" + grade + \"      \\n\")\n",
        "\n",
        "            else:\n",
        "\n",
        "                GRADES.write(line[:60] + \"            \\n\")\n",
        "                SCORES.write(line[:60] + \"            \\n\")\n",
        "                if vars['insufficient_data']:\n",
        "\n",
        "                    GRADES_ISD.write(line[:60] + \"            \\n\")\n",
        "\n",
        "        else:\n",
        "\n",
        "            GRADES.write(line)\n",
        "            SCORES.write(line)\n",
        "            if vars['insufficient_data']:\n",
        "\n",
        "                GRADES_ISD.write(line)\n",
        "\n",
        "        line = PDB.readline()\n",
        "\n",
        "    GRADES.close()\n",
        "    SCORES.close()\n",
        "    vars['zip_list'].append(pdb_with_grades)\n",
        "    vars['zip_list'].append(pdb_with_scores)\n",
        "    if vars['insufficient_data']:\n",
        "\n",
        "        show_py3dmol(pdb_with_grades_isd, \"pdb\")\n",
        "        print_instructions(pdb_with_grades, \"PDB\", pdb_with_grades_isd)\n",
        "\n",
        "    else:\n",
        "\n",
        "        show_py3dmol(pdb_with_grades, \"pdb\")\n",
        "        print_instructions(pdb_with_grades, \"PDB\")\n",
        "\n",
        "def design_string_with_spaces_for_pipe(part_input):\n",
        "\n",
        "    if part_input.strip() == \"\":\n",
        "\n",
        "        return \"\"\n",
        "\n",
        "    words = part_input.split()\n",
        "    newPart = \"! \\\"\" +words[0]\n",
        "    part = \"\"\n",
        "    for word in words[1:]:\n",
        "\n",
        "        # if adding another word to the string will yeild a too long string - we cut it.\n",
        "        if len(word) + 1 + len(newPart) > 76:\n",
        "\n",
        "            part += newPart + \" \\\" +\\n\"\n",
        "            newPart = \"! \\\"\" + word\n",
        "\n",
        "        else:\n",
        "\n",
        "            newPart += \" \" + word\n",
        "\n",
        "    part += newPart + \"\\\" ;\"\n",
        "\n",
        "    return part\n",
        "\n",
        "\n",
        "\n",
        "def add_pdb_data_to_pipe(pdb_file, pipe_file):\n",
        "\n",
        "    # create the file to be shown using FGiJ. read the pdb file and concat header pipe to it.\n",
        "\n",
        "    try:\n",
        "\n",
        "        PIPE = open(pipe_file, 'a')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"add_pdb_data_to_pipe: cannot open \" + pipe_file + \" for writing.\")\n",
        "\n",
        "    try:\n",
        "\n",
        "        PDB_FILE = open(pdb_file, 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"add_pdb_data_to_pipe: cannot open the \" + pdb_file + \" for reading.\")\n",
        "\n",
        "    line = PDB_FILE.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        if not re.match(r'^HEADER', line):\n",
        "\n",
        "            PIPE.write(line)\n",
        "\n",
        "        line = PDB_FILE.readline()\n",
        "\n",
        "    PIPE.close()\n",
        "    PDB_FILE.close()\n",
        "\n",
        "\n",
        "\n",
        "def print_selected(arr_ref, print_for_pipe):\n",
        "\n",
        "    total_text = \"\"\n",
        "    string = \"\"\n",
        "    if print_for_pipe == \"yes\":\n",
        "\n",
        "        string = \"! select \"\n",
        "\n",
        "    else:\n",
        "\n",
        "        string = \"select \"\n",
        "\n",
        "    total_length = len(string)\n",
        "\n",
        "    if len(arr_ref) > 0:\n",
        "\n",
        "        for aa in arr_ref:\n",
        "\n",
        "            aa = aa.replace(\":\", \"\")\n",
        "            total_length += len(aa)\n",
        "            if total_length > 80:\n",
        "\n",
        "                if re.search(r', $', string):\n",
        "\n",
        "                    string = string[:-2]\n",
        "\n",
        "                total_text += string + \"\\n\"\n",
        "                if print_for_pipe == \"yes\":\n",
        "\n",
        "                    string = \"! select selected or %s, \" %aa\n",
        "\n",
        "                else:\n",
        "\n",
        "                    string = \"select selected or %s, \" %aa\n",
        "\n",
        "                total_length = len(string)\n",
        "\n",
        "            else:\n",
        "\n",
        "                string += aa + \", \"\n",
        "                total_length += 2\n",
        "\n",
        "    else:\n",
        "\n",
        "        total_text += string + \"none\"\n",
        "\n",
        "\n",
        "    if re.search(r', $', string):\n",
        "\n",
        "        string = string[:-2]\n",
        "        total_text += string\n",
        "\n",
        "    return total_text\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "def create_consurf_pipe_new(results_dir, IN_pdb_id_capital, chain, ref_header_title, final_pipe_file, identical_chains, partOfPipe, current_dir, run_number, msa_filename, query_name_in_msa = \"\", tree_filename = \"\", submission_time = \"\", completion_time = \"\", run_date = \"\"):\n",
        "\n",
        "\n",
        "    # Create the pipe file for FGiJ\n",
        "\n",
        "    if chain == 'NONE':\n",
        "\n",
        "        chain = \"\"\n",
        "        identical_chains = \"\"\n",
        "\n",
        "    # read info from the pdb file\n",
        "    [header_line, title_line] = ref_header_title\n",
        "\n",
        "    if title_line == \"\":\n",
        "\n",
        "        title_line = \"! \\\"No title or compound description was found in the PDB file\\\";\"\n",
        "\n",
        "    else:\n",
        "\n",
        "        title_line = design_string_with_spaces_for_pipe(title_line)\n",
        "\n",
        "    # design the identical chains line\n",
        "    identical_chains_line = \"! consurf_identical_chains = \\\"%s\\\";\" %identical_chains\n",
        "\n",
        "    current_dir += \"_sourcedir\"\n",
        "    # in case there is a source dir - we determine the var query_name_in_msa\n",
        "    if os.path.exists(current_dir):\n",
        "\n",
        "        try:\n",
        "\n",
        "            SOURCEDIR = open(current_dir, 'r')\n",
        "\n",
        "        except:\n",
        "\n",
        "            exit_on_error('sys_error', \"create_consurf_pipe : cannot open \" + current_dir + \" for reading.\")\n",
        "\n",
        "        match = re.match(r'(\\d[\\d\\w]{3})\\/(\\w)', SOURCEDIR.readline())\n",
        "        if match:\n",
        "\n",
        "            query_name_in_msa = SOURCEDIR.group(1) + SOURCEDIR.group(2)\n",
        "            SOURCEDIR.close()\n",
        "\n",
        "    if query_name_in_msa == \"\":\n",
        "\n",
        "        query_name_in_msa = IN_pdb_id_capital + chain.upper()\n",
        "\n",
        "    # write to the pipe file\n",
        "    try:\n",
        "\n",
        "        PIPE_PART = open(partOfPipe, 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"create_consurf_pipe : cannot open \" + partOfPipe + \" for reading.\")\n",
        "\n",
        "    try:\n",
        "\n",
        "        PIPE = open(final_pipe_file, 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"create_consurf_pipe : cannot open \" + final_pipe_file + \" for writing.\")\n",
        "\n",
        "    if header_line != \"\":\n",
        "\n",
        "        PIPE.write(header_line + \"\\n\")\n",
        "\n",
        "    else:\n",
        "\n",
        "        PIPE.write(\"HEADER                                 [THIS LINE ADDED FOR JMOL COMPATIBILITY]\\n\")\n",
        "\n",
        "    PIPE.write(\"\"\"!! ====== IDENTIFICATION SECTION ======\n",
        "!js.init\n",
        "! consurf_server = \"consurf\";\n",
        "! consurf_version = \"3.0\";\n",
        "! consurf_run_number = \\\"%s\\\";\n",
        "! consurf_run_date = \\\"%s\\\";\n",
        "! consurf_run_submission_time = \\\"%s\\\";\n",
        "! consurf_run_completion_time = \\\"%s\\\";\n",
        "!\n",
        "! consurf_pdb_id = \\\"%s\\\";\n",
        "! consurf_chain = \\\"%s\\\";\n",
        "%s\n",
        "! consurf_msa_filename = \\\"%s\\\";\n",
        "! consurf_msa_query_seq_name = \\\"%s\\\";\n",
        "! consurf_tree_filename = \\\"%s\\\";\n",
        "!\n",
        "\"\"\" %(run_number, run_date, submission_time, completion_time, IN_pdb_id_capital, chain, identical_chains_line, msa_filename, query_name_in_msa, tree_filename))\n",
        "\n",
        "    titleFlag = 0\n",
        "    line = PIPE_PART.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        if re.match(r'^~~~+', line):\n",
        "\n",
        "            if titleFlag == 0:\n",
        "\n",
        "                PIPE.write(\"! pipe_title = \\\"<i>ConSurf View:</i> %s chain %s.\\\"\\n!! pipe_subtitle is from TITLE else COMPND\\n!!\\n! pipe_subtitle =\\n%s\\n\" %(IN_pdb_id_capital, chain, title_line))\n",
        "                titleFlag = 1\n",
        "\n",
        "            elif chain != \"\":\n",
        "\n",
        "                PIPE.write(\"! select selected and :%s\\n\" %chain)\n",
        "\n",
        "            else:\n",
        "\n",
        "                PIPE.write(\"! select selected and protein\\n\")\n",
        "\n",
        "        else:\n",
        "\n",
        "            PIPE.write(line)\n",
        "\n",
        "        line = PIPE_PART.readline()\n",
        "\n",
        "    PIPE_PART.close()\n",
        "    PIPE.close()\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "def freq_array(isd_residue_color, no_isd_residue_color):\n",
        "\n",
        "    # design the frequencies array\n",
        "\n",
        "    consurf_grade_freqs_isd = \"Array(\" + str(len(isd_residue_color[10]))\n",
        "    i = 1\n",
        "    while i < 10:\n",
        "\n",
        "        consurf_grade_freqs_isd += \",\" + str(len(isd_residue_color[i]))\n",
        "        i += 1\n",
        "\n",
        "    consurf_grade_freqs_isd += \")\"\n",
        "\n",
        "    consurf_grade_freqs = \"Array(0\"\n",
        "    i = 1\n",
        "    while i < 10:\n",
        "\n",
        "        consurf_grade_freqs += \",\" + str(len(no_isd_residue_color[i]))\n",
        "        i += 1\n",
        "\n",
        "    consurf_grade_freqs += \")\"\n",
        "\n",
        "    return(consurf_grade_freqs_isd, consurf_grade_freqs)\n",
        "\n",
        "\n",
        "def design_string_for_pipe(string_to_format):\n",
        "\n",
        "    # take a string aaaaaaa and returns it in this format: ! \"aaa\" +\\n! \"aa\";\\n\n",
        "\n",
        "    part = string_to_format\n",
        "    newPart = \"\"\n",
        "\n",
        "    while len(part) > 73:\n",
        "\n",
        "        newPart += \"! \\\"\" + part[:73] + \"\\\" +\\n\"\n",
        "        part = part[73:]\n",
        "\n",
        "    newPart += \"! \\\"\" + part + \"\\\" ;\"\n",
        "\n",
        "    return newPart\n",
        "\n",
        "\n",
        "def extract_data_from_pdb(input_pdb_file):\n",
        "\n",
        "    header = \"\"\n",
        "    title = \"\"\n",
        "    compnd = \"\"\n",
        "\n",
        "    try:\n",
        "\n",
        "        PDB = open(input_pdb_file, 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"extract_data_from_pdb : Could not open the file \" + input_pdb_file + \" for reading.\")\n",
        "\n",
        "    line = PDB.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        match1 = re.match(r'^HEADER', line)\n",
        "        if match1:\n",
        "\n",
        "            header = line.rstrip()\n",
        "\n",
        "        else:\n",
        "\n",
        "            match2 =re.match(r'^TITLE\\s+\\d*\\s(.*)', line)\n",
        "            if match2:\n",
        "\n",
        "                title += match2.group(1) + \" \"\n",
        "\n",
        "            else:\n",
        "\n",
        "                match3 = re.match(r'^COMPND\\s+\\d*\\s(.*)', line)\n",
        "                if match3:\n",
        "\n",
        "                    compnd += match3.group(1) + \" \"\n",
        "\n",
        "                elif re.match(r'^SOURCE', line) or re.match(r'^KEYWDS', line) or re.match(r'^AUTHOR', line) or re.match(r'^SEQRES', line) or re.match(r'^ATOM', line):\n",
        "\n",
        "                    break # no nead to go over all the pdb\n",
        "\n",
        "        line = PDB.readline()\n",
        "\n",
        "    PDB.close()\n",
        "    if title == \"\":\n",
        "\n",
        "        return header, compnd\n",
        "\n",
        "    else:\n",
        "\n",
        "        return header, title\n",
        "\n",
        "\n",
        "def create_part_of_pipe_new(pipe_file, unique_seqs, db, seq3d_grades_isd, seq3d_grades, length_of_seqres, length_of_atom, ref_isd_residue_color, ref_no_isd_residue_color, E_score, iterations, max_num_homol, MSAprogram, algorithm, matrix, Average_pairwise_distance, scale = \"legacy\"):\n",
        "\n",
        "    # creating part of the pipe file, which contains all the non-unique information.\n",
        "    # each chain will use this file to construct the final pdb_pipe file, to be viewed with FGiJ\n",
        "\n",
        "    if scale == \"legacy\":\n",
        "\n",
        "        scale_block = \"!color color_grade0 FFFF96 insufficient data yellow\\n!color color_grade1 10C8D1 turquoise variable\\n!color color_grade2 8CFFFF\\n!color color_grade3 D7FFFF\\n!color color_grade4 EAFFFF\\n!color color_grade5 FFFFFF\\n!color color_grade6 FCEDF4\\n!color color_grade7 FAC9DE\\n!color color_grade8 F07DAB\\n!color color_grade9 A02560 burgundy conserved\"\n",
        "\n",
        "    else:\n",
        "\n",
        "        scale_block = \"!color color_grade0 FFFF96 insufficient data yellow\\n!color color_grade1 1b7837 variable\\n!color color_grade2 5aae61\\n!color color_grade3 a6dba0\\n!color color_grade4 d9f0d3\\n!color color_grade5 f7f7f7\\n!color color_grade6 e7d4e8\\n!color color_grade7 c2a5cf\\n!color color_grade8 9970ab\\n!color color_grade9 762a83 conserved\"\n",
        "\n",
        "    # design the seq3d to be printed out to the pipe file\n",
        "    seq3d_grades_isd = design_string_for_pipe(seq3d_grades_isd)\n",
        "    seq3d_grades = design_string_for_pipe(seq3d_grades)\n",
        "\n",
        "    # creating the frequencies array which corresponds the number of residues in each grade\n",
        "    [consurf_grade_freqs_isd, consurf_grade_freqs] = freq_array(ref_isd_residue_color, ref_no_isd_residue_color)\n",
        "\n",
        "    # Taking Care of Strings\n",
        "    if max_num_homol == \"all\":\n",
        "\n",
        "        max_num_homol = \"\\\"all\\\"\"\n",
        "\n",
        "    # write to the pipe file\n",
        "    try:\n",
        "\n",
        "        PIPE = open(pipe_file, 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"create_part_of_pipe_new : cannot open the file \" + pipe_file + \" for writing.\", 'PANIC')\n",
        "\n",
        "    PIPE.write(\"\"\"! consurf_psi_blast_e_value = %s;\n",
        "! consurf_psi_blast_database = \"%s\";\n",
        "! consurf_psi_blast_iterations = %s;\n",
        "! consurf_max_seqs = %s;\n",
        "! consurf_apd = %.2f;\n",
        "! consurf_alignment = \"%s\";\n",
        "! consurf_method = \"%s\";\n",
        "! consurf_substitution_model =  \"%s\";\n",
        "!\n",
        "! consurf_seqres_length = %s;\n",
        "! consurf_atom_seq_length = %s;\n",
        "! consurf_unique_seqs = %s;\n",
        "! consurf_grade_freqs_isd = %s;\n",
        "! consurf_grade_freqs = %s;\n",
        "!\n",
        "! seq3d_grades_isd =\n",
        "%s\n",
        "!\n",
        "! seq3d_grades =\n",
        "%s\n",
        "!\n",
        "!\n",
        "!! ====== CONTROL PANEL OPTIONS SECTION ======\n",
        "!js.init\n",
        "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
        "! pipe_title_enlarged = false;\n",
        "! pipe_background_color = \"white\";\n",
        "!\n",
        "!! Specify the custom consurf control panel\n",
        "!!\n",
        "! pipe_cp1 = \"consurf/consurf.htm\";\n",
        "!\n",
        "!! If you want the frontispiece to be reset every time you enter this\n",
        "!! page, use false. If this is a one-page presentation (no contents)\n",
        "!! and you want to be able to return from QuickViews without resetting\n",
        "!! the view, use true.\n",
        "!!\n",
        "! frontispiece_conditional_on_return = true;\n",
        "!\n",
        "!! Open the command input slot/message box to 30%% of window height.\n",
        "!!\n",
        "! pipe_show_commands = true;\n",
        "! pipe_show_commands_pct = 30;\n",
        "!\n",
        "!! Don't show the PiPE presentation controls in the lower left frame.\n",
        "!!\n",
        "! pipe_hide_controls = true;\n",
        "!\n",
        "!! Hide development viewing mode links at the bottom of the control panel.\n",
        "!!\n",
        "! pipe_tech_info = false;\n",
        "!\n",
        "!! pipe_start_spinning = true; // default is PE's Preference setting.\n",
        "!! top.nonStopSpin = true; // default: spinning stops after 3 min.\n",
        "!!\n",
        "!! ====== COLORS SECTION ======\n",
        "!!\n",
        "!color color_carbon C8C8C8\n",
        "!color color_sulfur FFC832\n",
        "!\n",
        "!! Ten ConSurf color grades follow:\n",
        "!!\n",
        "%s\n",
        "!\n",
        "!\n",
        "!! ====== SCRIPTS SECTION ======\n",
        "!!----------------------------------------\n",
        "!!\n",
        "!spt #name=select_and_chain\n",
        "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
        "!\n",
        "!!----------------------------------------\n",
        "!!\n",
        "!spt #name=view01\n",
        "! @spt consurf_view_isd\n",
        "!\n",
        "!!----------------------------------------\n",
        "!!\n",
        "!spt #name=hide_all\n",
        "! restrict none\n",
        "! ssbonds off\n",
        "! hbonds off\n",
        "! dots off\n",
        "! list * delete\n",
        "!\n",
        "!!----------------------------------------\n",
        "!! common_spt uses CPK carbon gray (or phosphorus yellow) for backbones.\n",
        "!!\n",
        "!spt #name=common_spt\n",
        "! @spt hide_all\n",
        "! select all\n",
        "! color [xC8C8C8] # rasmol/chime carbon gray\n",
        "! select nucleic\n",
        "! color [xFFA500] # phosphorus orange\n",
        "! select hetero\n",
        "! color cpk\n",
        "! select not hetero\n",
        "! backbone 0.4\n",
        "! javascript top.water=0\n",
        "!\n",
        "! ssbonds 0.3\n",
        "! set ssbonds backbone\n",
        "! color ssbonds @color_sulfur\n",
        "!\n",
        "! select hetero and not water\n",
        "! spacefill 0.45\n",
        "! wireframe 0.15\n",
        "! dots 50\n",
        "!\n",
        "! select protein\n",
        "! center selected\n",
        "!\n",
        "!!----------------------------------------\n",
        "!!\n",
        "!spt #name=consurf_view_isd\n",
        "! @spt common_spt\n",
        "! @for $=0, 9\n",
        "! @spt select_isd_grade$\n",
        "! @spt select_and_chain\n",
        "! color @color_grade$\n",
        "! spacefill\n",
        "! @endfor\n",
        "! zoom 115\n",
        "!\n",
        "!!----------------------------------------\n",
        "\"\"\" %(E_score, db, iterations, max_num_homol, round(float(Average_pairwise_distance), 2), MSAprogram, algorithm, matrix, length_of_seqres, length_of_atom, unique_seqs, consurf_grade_freqs_isd, consurf_grade_freqs, seq3d_grades_isd, seq3d_grades, scale_block))\n",
        "\n",
        "    lineToPrint = \"\"\n",
        "    i = 9\n",
        "    while i > 0:\n",
        "\n",
        "        PIPE.write(\"!!\\n!spt #name=select_isd_grade%d\\n!\\n\" %i)\n",
        "        lineToPrint = print_selected(ref_isd_residue_color[i], \"yes\")\n",
        "        if re.search(r'select', lineToPrint):\n",
        "\n",
        "            PIPE.write(lineToPrint + \"\\n\")\n",
        "\n",
        "        PIPE.write(\"!\\n!\\n!!----------------------------------------\\n\")\n",
        "        i -= 1\n",
        "\n",
        "    PIPE.write(\"!!\\n!spt #name=select_isd_grade0\\n\")\n",
        "    lineToPrint = print_selected(ref_isd_residue_color[10], \"yes\")\n",
        "    if re.search(r'select', lineToPrint):\n",
        "\n",
        "        PIPE.write(lineToPrint + \"\\n\")\n",
        "\n",
        "    PIPE.write(\"!\\n!\\n!!----------------------------------------\\n\")\n",
        "\n",
        "    i = 9\n",
        "    while i > 0:\n",
        "\n",
        "        PIPE.write(\"!!\\n!spt #name=select_grade%d\\n!\\n\" %i)\n",
        "        lineToPrint = print_selected(ref_no_isd_residue_color[i], \"yes\")\n",
        "        if re.search(r'select', lineToPrint):\n",
        "\n",
        "            PIPE.write(lineToPrint + \"\\n\")\n",
        "\n",
        "        PIPE.write(\"!\\n!\\n!!----------------------------------------\\n\")\n",
        "        i -= 1\n",
        "\n",
        "    PIPE.write(\"!!\\n!spt #name=select_grade0\\n! select none\\n!!\\n\")\n",
        "    PIPE.write(\"!! ====== END OF CONSURF PiPE BLOCK ======\\n\")\n",
        "    PIPE.close()\n",
        "\n",
        "\n",
        "\n",
        "def create_gradesPE(gradesPE, ref_match = \"\", pdb_file = \"\", chain = \"\", prefix = \"\", pdb_object = \"\", identical_chains = \"\", pdb_cif = \"\", atom_grades = \"\"):\n",
        "\n",
        "\n",
        "    # printing the the ConSurf gradesPE file\n",
        "\n",
        "    if pdb_cif == \"pdb\": # this is for the pipe file\n",
        "\n",
        "        seq3d_grades_isd = \"\"\n",
        "        seq3d_grades = \"\"\n",
        "        # arrays showing how to color the residues. The subarrays holds residues of the same color\n",
        "        no_isd_residue_color = [[],[],[],[],[],[],[],[],[],[],[]] # no insufficient data\n",
        "        isd_residue_color = [[],[],[],[],[],[],[],[],[],[],[]] # insufficient data\n",
        "\n",
        "    try:\n",
        "\n",
        "        PE = open(gradesPE, 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"create_gradesPE : can't open '\" + gradesPE + \"' for writing.\")\n",
        "\n",
        "    if form['DNA_AA'] == \"AA\":\n",
        "\n",
        "        unknown_char = \"X\"\n",
        "        PE.write(\"\\t Amino Acid Conservation Scores\\n\")\n",
        "\n",
        "    else:\n",
        "\n",
        "        unknown_char = \"N\"\n",
        "        PE.write(\"\\t Nucleic Acid Conservation Scores\\n\")\n",
        "\n",
        "    PE.write(\"\\t=======================================\\n\\n\")\n",
        "    PE.write(\"The layers for assigning grades are as follows.\\n\")\n",
        "    for i in range(1, len(vars['layers_array'])):\n",
        "\n",
        "        if vars['layers_array'][i - 1] < 0:\n",
        "\n",
        "            left_end = \"%.3f\" %vars['layers_array'][i - 1]\n",
        "\n",
        "        else:\n",
        "\n",
        "            left_end = \" %.3f\" %vars['layers_array'][i - 1]\n",
        "\n",
        "\n",
        "        if vars['layers_array'][i] < 0:\n",
        "\n",
        "            right_end = \"%.3f\" %vars['layers_array'][i]\n",
        "\n",
        "        else:\n",
        "\n",
        "            right_end = \" %.3f\" %vars['layers_array'][i]\n",
        "\n",
        "        PE.write(\"from %s to %s the grade is %d\\n\" %(left_end, right_end, 10 - i))\n",
        "\n",
        "    PE.write(\"\\nIf the difference between the colors of the CONFIDENCE INTERVAL COLORS is more than 3 or the msa number (under the column titled MSA) is less than 6, there is insufficient data and an * appears in the COLOR column.\\n\")\n",
        "\n",
        "\n",
        "    PE.write(\"\\n- POS: The position of the acid in the sequence.\\n\")\n",
        "    PE.write(\"- SEQ: The acid one letter.\\n\")\n",
        "    PE.write(\"- ATOM: When there's a model, The ATOM derived sequence in three letter code, including the acid's positions as they appear in the PDB file and the chain identifier.\\n\")\n",
        "    PE.write(\"- SCORE: The normalized conservation scores.\\n\")\n",
        "    PE.write(\"- COLOR: The color scale representing the conservation scores (9 - conserved, 1 - variable).\\n\")\n",
        "    PE.write(\"- CONFIDENCE INTERVAL: When using the bayesian method for calculating rates, a confidence interval is assigned to each of the inferred evolutionary conservation scores, next to it are the colors of the lower and upper bounds of the confidence interval\\n\")\n",
        "    if vars['B/E']:\n",
        "\n",
        "        PE.write(\"- B/E: Burried (b) or Exposed (e) residue.\\n\")\n",
        "        PE.write(\"- F/S: functional (f) or structural (s) residue (f - highly conserved and exposed, s - highly conserved and burried).\\n\")\n",
        "\n",
        "    PE.write(\"- MSA DATA: The number of aligned sequences having an acid (non-gapped) from the overall number of sequences at each position.\\n\")\n",
        "    PE.write(\"- RESIDUE VARIETY: The residues variety at each position of the multiple sequence alignment.\\n\\n\")\n",
        "\n",
        "\n",
        "\n",
        "    if form['ALGORITHM'] == \"Bayes\":\n",
        "\n",
        "        CONFIDENCE_INTERVAL = \"CONFIDENCE INTERVAL\\t\"\n",
        "\n",
        "    else:\n",
        "\n",
        "        CONFIDENCE_INTERVAL = \"\"\n",
        "\n",
        "    # the size of the POS, ATOM and MSA DATA columns is variable\n",
        "\n",
        "    pos_number_size = len(str(len(vars['protein_seq_string']))) # size of the of the number of positions\n",
        "    pos_column_title_size = len(\"POS\") # size of the title of the column POS\n",
        "    pos_column_size = max(pos_column_title_size, pos_number_size) # size of the column POS\n",
        "    num_spaces = pos_column_size - pos_column_title_size # number of spaces to add\n",
        "    while num_spaces > 0:\n",
        "\n",
        "        PE.write(\" \")\n",
        "        num_spaces -= 1\n",
        "\n",
        "    PE.write(\"POS\\tSEQ\\t\")\n",
        "\n",
        "    if ref_match != \"\":\n",
        "\n",
        "        # consurf. in conseq there is no atom because there is no model\n",
        "        max_res_details = pdb_object.get_max_res_details()\n",
        "        atom_column_title_size = len(\"ATOM\") # size of the title of the column ATOM\n",
        "        atom_column_size = max(atom_column_title_size, max_res_details) # size of the column ATOM\n",
        "        num_spaces = atom_column_size - atom_column_title_size # number of spaces to add\n",
        "        while num_spaces > 0:\n",
        "\n",
        "            PE.write(\" \")\n",
        "            num_spaces -= 1\n",
        "\n",
        "        PE.write(\"ATOM\\t\")\n",
        "\n",
        "    if vars['B/E']:\n",
        "\n",
        "        PE.write(\" SCORE\\tCOLOR\\t%sB/E\\tF/S\\t\" %CONFIDENCE_INTERVAL)\n",
        "\n",
        "    else:\n",
        "\n",
        "        PE.write(\" SCORE\\tCOLOR\\t%s\" %CONFIDENCE_INTERVAL)\n",
        "\n",
        "    msa_size = len(str(vars['final_number_of_homologoues'])) # size of the number of homologs\n",
        "    msa_column_title_size = len(\"MSA DATA\") # size of the title of the column MSA DATA\n",
        "    msa_column_size = max(msa_column_title_size, 2 * msa_size + 1) # size of the column MSA DATA\n",
        "    num_spaces = msa_column_size - msa_column_title_size # number of spaces to add\n",
        "    while num_spaces > 0:\n",
        "\n",
        "        PE.write(\" \")\n",
        "        num_spaces -= 1\n",
        "\n",
        "    PE.write(\"MSA DATA\\tRESIDUE VARIETY\\n\\n\")\n",
        "\n",
        "    seq_index = 0 # the index of the position in the query sequence (the rate4site output doesn't contain unknown chars)\n",
        "    for elem in vars['gradesPE_Output']:\n",
        "\n",
        "        pos = elem['POS']\n",
        "        var = \"\"\n",
        "        num_spaces = pos_column_size - len(str(pos)) # number of spaces to add\n",
        "        while num_spaces > 0:\n",
        "\n",
        "            PE.write(\" \")\n",
        "            num_spaces -= 1\n",
        "\n",
        "        PE.write(\"%d\\t\" %pos)\n",
        "        PE.write(\"  %s\\t\" %elem['SEQ'])\n",
        "\n",
        "        if ref_match != \"\":\n",
        "\n",
        "            # consurf, in conseq there is no atom because there is no model\n",
        "            atom_3L = ref_match[pos]\n",
        "            num_spaces = atom_column_size - len(atom_3L) # number of spaces to add\n",
        "            while num_spaces > 0:\n",
        "\n",
        "                PE.write(\" \")\n",
        "                num_spaces -= 1\n",
        "\n",
        "            PE.write(\"%s\\t\" %atom_3L)\n",
        "\n",
        "            # save the grade of the residue inorder to write it on the pdb file\n",
        "            if atom_3L != '-':\n",
        "\n",
        "                residue_number = atom_3L.split(':')[1]\n",
        "                atom_grades[residue_number] = [str(elem['COLOR']), elem['ISD'], \"%6.3f\" %elem['GRADE']]\n",
        "\n",
        "        PE.write(\"%6.3f\\t\" %elem['GRADE'])\n",
        "        if elem['ISD'] == 1:\n",
        "\n",
        "            PE.write(\"   %d*\\t\" %elem['COLOR'])\n",
        "\n",
        "        else:\n",
        "\n",
        "            PE.write(\"   %d \\t\" %elem['COLOR'])\n",
        "\n",
        "        if form['ALGORITHM'] == \"Bayes\":\n",
        "\n",
        "            PE.write(\"%6.3f, \" %elem['INTERVAL_LOW'])\n",
        "            PE.write(\"%6.3f  \" %elem['INTERVAL_HIGH'])\n",
        "            PE.write(\"%d,\" %ColorScale[elem['INTERVAL_LOW_COLOR']])\n",
        "            PE.write(\"%d\\t\" %ColorScale[elem['INTERVAL_HIGH_COLOR']])\n",
        "\n",
        "        if vars['B/E']:\n",
        "\n",
        "            PE.write(\"  %s\\t  %s\\t\" %(elem['B/E'], elem['F/S']))\n",
        "\n",
        "        homologs_in_pos = str(elem['MSA_NUM']) + \"/\" + elem['MSA_DENUM'] # number of homologs in the position\n",
        "        num_spaces = msa_column_size - len(homologs_in_pos) # number of spaces to add\n",
        "        while num_spaces > 0:\n",
        "\n",
        "            PE.write(\" \")\n",
        "            num_spaces -= 1\n",
        "\n",
        "        PE.write(homologs_in_pos)\n",
        "\n",
        "\t\t# we write the acids percentage in the msa\n",
        "        while seq_index < len(vars['protein_seq_string']) and vars['protein_seq_string'][seq_index] == unknown_char:\n",
        "\n",
        "            seq_index += 1\n",
        "\n",
        "        for char in vars['percentage_per_pos'][seq_index]:\n",
        "\n",
        "            if vars['percentage_per_pos'][seq_index][char] == 100:\n",
        "\n",
        "                var += char\n",
        "\n",
        "            else:\n",
        "\n",
        "\t\t\t\t# if the precentage is less than one write <1%\n",
        "                if vars['percentage_per_pos'][seq_index][char] > 1:\n",
        "\n",
        "                    var += \"%s %2d%%, \" %(char, vars['percentage_per_pos'][seq_index][char])\n",
        "\n",
        "                else:\n",
        "\n",
        "                    var += \"%s <1%%, \" %char\n",
        "\n",
        "        \"\"\"\n",
        "        if vars['unknown_per_pos'][pos - 1] != 0:\n",
        "\n",
        "\t\t\t# if the precentage is less than one write <1%\n",
        "            if vars['unknown_per_pos'][pos - 1] > 1:\n",
        "\n",
        "                var += \"%s %2d%%\" %(unknown_char, vars['unknown_per_pos'][pos - 1])\n",
        "\n",
        "            else:\n",
        "\n",
        "                var += \"%s <1%%\" %unknown_char\n",
        "        \"\"\"\n",
        "        if len(vars['percentage_per_pos'][seq_index]) != 1:\n",
        "\n",
        "            var = var[:-2] # we delete the last comma\n",
        "\n",
        "        PE.write(\"\\t\" + var + \"\\n\")\n",
        "        seq_index += 1\n",
        "\n",
        "        # the amino-acid in that position, must be part of the residue variety in this column\n",
        "        if not re.search(elem['SEQ'], var, re.IGNORECASE):\n",
        "\n",
        "            PE.close()\n",
        "            exit_on_error('sys_error', \"create_gradesPE : in position %s, the amino-acid %s does not match the residue variety: %s.\" %(pos, elem['SEQ'], var))\n",
        "\n",
        "        if pdb_cif != \"pdb\": # the next part is for the pipe file\n",
        "\n",
        "            continue\n",
        "\n",
        "        # printing the residue to the rasmol script\n",
        "        # assigning grades to seq3d strings\n",
        "        if not '-' in atom_3L:\n",
        "\n",
        "            atom_3L = re.search(r'(.+):', atom_3L).group(1)\n",
        "            if form['DNA_AA'] == \"Nuc\":\n",
        "\n",
        "                atom_3L = \"D\" + atom_3L\n",
        "\n",
        "            color = elem['COLOR']\n",
        "            no_isd_residue_color[color].append(atom_3L)\n",
        "            if elem['ISD'] == 1:\n",
        "\n",
        "                isd_residue_color[10].append(atom_3L)\n",
        "                seq3d_grades_isd += \"0\"\n",
        "\n",
        "            else:\n",
        "\n",
        "                isd_residue_color[color].append(atom_3L)\n",
        "                seq3d_grades_isd += str(color)\n",
        "\n",
        "            seq3d_grades += str(color)\n",
        "\n",
        "        else:\n",
        "\n",
        "            seq3d_grades_isd += \".\"\n",
        "            seq3d_grades += \".\"\n",
        "\n",
        "    PE.write(\"\\n\\n*Below the confidence cut-off - The calculations for this site were performed on less than 6 non-gaped homologue sequences,\\n\")\n",
        "    PE.write(\"or the confidence interval for the estimated score is equal to- or larger than- 4 color grades.\\n\")\n",
        "    PE.close()\n",
        "\n",
        "    if pdb_cif != \"pdb\": # the next part is for the pipe file\n",
        "\n",
        "        return\n",
        "\n",
        "    if seq3d_grades_isd == \"\" or seq3d_grades == \"\":\n",
        "\n",
        "        exit_on_error('sys_error', \"create_gradesPE : there is no data in the returned values seq3d_grades_isd or seq3d_grades from the routine\")\n",
        "\n",
        "    \"\"\"\n",
        "    # This will create the pipe file for FGiJ\n",
        "    pipeFile = prefix + \"_consurf_firstglance.pdb\"\n",
        "    pipeFile_CBS = prefix + \"_consurf_firstglance_CBS.pdb\" # pipe for color blind friendly\n",
        "    create_pipe_file(pipeFile, pipeFile_CBS, seq3d_grades, seq3d_grades_isd, isd_residue_color, no_isd_residue_color, pdb_file, chain, (prefix).upper(), identical_chains, pdb_object)\n",
        "\n",
        "    # create RasMol files\n",
        "    create_rasmol(prefix, chain, no_isd_residue_color, isd_residue_color)\n",
        "    \"\"\"\n",
        "\n",
        "def match_pdb_to_seq(ref_fas2pdb, query_seq, pdbseq, pdb_object):\n",
        "\n",
        "    # matches the position in the seqres/msa sequence to the position in the pdb\n",
        "\n",
        "    UnKnownChar = \"\"\n",
        "    if form['DNA_AA'] == \"AA\":\n",
        "\n",
        "        UnKnownChar = \"X\"\n",
        "\n",
        "    else:\n",
        "\n",
        "        UnKnownChar = \"N\"\n",
        "\n",
        "    # creating the hash that matches the position in the ATOM fasta to its position\n",
        "    # in the pdb file and also the fasta ATOM position to the correct residue\n",
        "    match_ATOM = pdb_object.get_positions()\n",
        "    \"\"\"\n",
        "    # creating the hash that matches the position in the ATOM fasta to its position\n",
        "    # in the pdb file and also the fasta ATOM position to the correct residue\n",
        "    try:\n",
        "\n",
        "        MATCH = open(atom_position, 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"match_pdb_to_seq : Could not open the file \" + atom_position + \" for reading.\")\n",
        "\n",
        "    max_res_details = 0 # longest residue details string. Dictates the size of the ATOM column in the results summary page\n",
        "    match_ATOM = {}\n",
        "    line = MATCH.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        words = line.split()\n",
        "        if len(words) == 3:\n",
        "\n",
        "            res_details = words[0] + \":\" + words[2] + \":\" + chain\n",
        "            match_ATOM[int(words[1])] = res_details\n",
        "            if res_details > max_res_details:\n",
        "\n",
        "                max_res_details = res_details\n",
        "\n",
        "        line = MATCH.readline()\n",
        "\n",
        "    MATCH.close()\n",
        "\n",
        "    length_of_atoms = 0\n",
        "    for char in pdbseq:\n",
        "\n",
        "        if char != '-' and char != UnKnownChar:\n",
        "\n",
        "            length_of_atoms += 1\n",
        "\n",
        "    length_of_seqres = 0\n",
        "    for char in query_seq:\n",
        "\n",
        "        if char != '-' and char != UnKnownChar:\n",
        "\n",
        "            length_of_seqres += 1\n",
        "    \"\"\"\n",
        "\n",
        "    query_pos = 1\n",
        "    pdb_pos = 1\n",
        "    for pos in range(len(query_seq)):\n",
        "\n",
        "        if query_seq[pos] != '-' and query_seq[pos] != UnKnownChar:\n",
        "\n",
        "            if pdbseq[pos] == '-' or pdb_pos not in match_ATOM:\n",
        "\n",
        "                ref_fas2pdb[query_pos] = '-'\n",
        "\n",
        "            else:\n",
        "\n",
        "                ref_fas2pdb[query_pos] = match_ATOM[pdb_pos]\n",
        "                pdb_pos += 1\n",
        "\n",
        "            query_pos += 1\n",
        "\n",
        "        elif pdbseq[pos] != '-':\n",
        "\n",
        "            pdb_pos += 1\n",
        "\n",
        "    #return length_of_seqres, length_of_atoms, max_res_details\n",
        "\n",
        "\n",
        "\n",
        "def find_pdb_position(ref_fas2pdb, pdb_object):\n",
        "\n",
        "    # finds the position of the the sequence in the pdb\n",
        "\n",
        "    LOG.write(\"find_pdb_position(ref_fas2pdb, pdb_object)\\n\")\n",
        "\n",
        "    UnKnownChar = \"\"\n",
        "    if form['DNA_AA'] == \"AA\":\n",
        "\n",
        "        UnKnownChar = \"X\"\n",
        "\n",
        "    else:\n",
        "\n",
        "        UnKnownChar = \"N\"\n",
        "\n",
        "    match_ATOM = pdb_object.get_positions()\n",
        "    # rate4site deletes the unknown chars\n",
        "    rate4site_pos = 1\n",
        "    pdb_pos = 1\n",
        "    for char in vars['ATOM_without_X_seq']:\n",
        "\n",
        "        if char != UnKnownChar:\n",
        "\n",
        "            ref_fas2pdb[rate4site_pos] = match_ATOM[pdb_pos]\n",
        "            rate4site_pos += 1\n",
        "\n",
        "        pdb_pos += 1\n",
        "\n",
        "\n",
        "\n",
        "def assign_colors_according_to_r4s_layers():\n",
        "\n",
        "    LOG.write(\"assign_colors_according_to_r4s_layers(%s, %s)\\n\" %(vars['gradesPE_Output'], vars['r4s_out']))\n",
        "\n",
        "    vars['insufficient_data'] = False\n",
        "    \"\"\"\n",
        "    if form['DNA_AA'] == \"AA\":\n",
        "\n",
        "\t    #runs the PACC algorithm to calculate burried/exposed\n",
        "        ref_Solv_Acc_Pred = predict_solvent_accesibility()\n",
        "\t    # this array connects the position to its index in ref_Solv_Acc_Pred (positions don't include unknown characters, PACC does)\n",
        "        index_of_pos = get_seq_legal_positions(vars['protein_seq_string'])\n",
        "\n",
        "    else:\n",
        "\n",
        "        ref_Solv_Acc_Pred = \"\"\n",
        "\n",
        "    if ref_Solv_Acc_Pred != \"\":\n",
        "\n",
        "        vars['B/E'] = True\n",
        "\n",
        "    else:\n",
        "\n",
        "        vars['B/E'] = False\n",
        "\n",
        "    \"\"\"\n",
        "\t# we extract the data from the rate4site output\n",
        "    try:\n",
        "\n",
        "        RATE4SITE = open(vars['r4s_out'], 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"assign_colors_according_to_r4s_layers : can't open \" + vars['r4s_out'])\n",
        "\n",
        "    line = RATE4SITE.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        line.rstrip()\n",
        "\n",
        "        if form['ALGORITHM'] == \"Bayes\":\n",
        "\n",
        "            # baysean\n",
        "            match1 = re.match(r'^\\s*(\\d+)\\s+(\\w)\\s+(\\S+)\\s+\\[\\s*(\\S+),\\s*(\\S+)\\]\\s+\\S+\\s+(\\d+)\\/(\\d+)', line)\n",
        "            if match1:\n",
        "\n",
        "                vars['gradesPE_Output'].append({'POS' : int(match1.group(1)), 'SEQ' : match1.group(2), 'GRADE' : float(match1.group(3)), 'INTERVAL_LOW' : float(match1.group(4)), 'INTERVAL_HIGH' : float(match1.group(5)), 'MSA_NUM' : int(match1.group(6)), 'MSA_DENUM' : match1.group(7)})\n",
        "\n",
        "        else:\n",
        "\n",
        "            # Maximum likelihood\n",
        "            match2 = re.match(r'^\\s*(\\d+)\\s+(\\w)\\s+(\\S+)\\s+(\\d+)\\/(\\d+)', line)\n",
        "            if match2:\n",
        "\n",
        "                vars['gradesPE_Output'].append({'POS' : int(match2.group(1)), 'SEQ' : match2.group(2), 'GRADE' : float(match2.group(3)), 'INTERVAL_LOW' : float(match2.group(3)), 'INTERVAL_HIGH' : float(match2.group(3)), 'MSA_NUM' : int(match2.group(4)), 'MSA_DENUM' : match2.group(5)})\n",
        "\n",
        "        line = RATE4SITE.readline()\n",
        "\n",
        "    RATE4SITE.close()\n",
        "\n",
        "    # we find the maximum and the minimum scores\n",
        "    max_cons = vars['gradesPE_Output'][0]['GRADE']\n",
        "    min_cons = vars['gradesPE_Output'][0]['GRADE']\n",
        "    for element in vars['gradesPE_Output']:\n",
        "\n",
        "        if element['GRADE'] < max_cons:\n",
        "\n",
        "            max_cons = element['GRADE']\n",
        "\n",
        "        if element['GRADE'] > min_cons:\n",
        "\n",
        "            min_cons = element['GRADE']\n",
        "\n",
        "    # we divide the interval between min_cons to max_cons to nine intervals\n",
        "    # 4 intervals on the left side are of length 2 * min_cons / 9\n",
        "    # 4 intervals on the right side are of length 2 * max_cons / 9\n",
        "    NoLayers = 10\n",
        "    LeftLayers = 5\n",
        "    RightLayers = 5\n",
        "    ColorLayers = []\n",
        "    i = 0\n",
        "    while i < LeftLayers:\n",
        "\n",
        "        ColorLayers.append(max_cons * ((9 - 2 * i) / 9.0))\n",
        "        i += 1\n",
        "\n",
        "    i = 0\n",
        "    while i < RightLayers:\n",
        "\n",
        "        ColorLayers.append(min_cons * ((1 + 2 * i) / 9.0))\n",
        "        i += 1\n",
        "\n",
        "    # each position gets a grade according to the layer its score is in\n",
        "    for element in vars['gradesPE_Output']:\n",
        "\n",
        "        i = 0\n",
        "        while not 'INTERVAL_LOW_COLOR' in element or not 'INTERVAL_HIGH_COLOR' in element or not 'COLOR' in element:\n",
        "\n",
        "            if not 'INTERVAL_LOW_COLOR' in element:\n",
        "\n",
        "                if i == NoLayers - 1:\n",
        "\n",
        "                    element['INTERVAL_LOW_COLOR'] = 8\n",
        "\n",
        "                elif element['INTERVAL_LOW'] >= ColorLayers[i] and element['INTERVAL_LOW'] < ColorLayers[i + 1]:\n",
        "\n",
        "                    element['INTERVAL_LOW_COLOR'] = i\n",
        "\n",
        "                elif element['INTERVAL_LOW'] < ColorLayers[0]:\n",
        "\n",
        "                    element['INTERVAL_LOW_COLOR'] = 0\n",
        "\n",
        "            if not 'INTERVAL_HIGH_COLOR' in element:\n",
        "\n",
        "                if i == NoLayers - 1:\n",
        "\n",
        "                    element['INTERVAL_HIGH_COLOR'] = 8\n",
        "\n",
        "                elif element['INTERVAL_HIGH'] >= ColorLayers[i] and element['INTERVAL_HIGH'] < ColorLayers[i + 1]:\n",
        "\n",
        "                    element['INTERVAL_HIGH_COLOR'] = i\n",
        "\n",
        "                elif element['INTERVAL_HIGH'] < ColorLayers[0]:\n",
        "\n",
        "                    element['INTERVAL_HIGH_COLOR'] = 0\n",
        "\n",
        "            if not 'COLOR' in element:\n",
        "\n",
        "                if i == NoLayers - 1:\n",
        "\n",
        "                    element['COLOR'] = ColorScale[i - 1]\n",
        "\n",
        "                elif element['GRADE'] >= ColorLayers[i] and element['GRADE'] < ColorLayers[i + 1]:\n",
        "\n",
        "                    element['COLOR'] = ColorScale[i]\n",
        "\n",
        "            i += 1\n",
        "\n",
        "\t\t# there is insufficient data if there are more than 3 layers in the confidence interval or the number of homologs in the MSA where the position is not empty is less than 5\n",
        "        if element['INTERVAL_HIGH_COLOR'] - element['INTERVAL_LOW_COLOR'] > bayesInterval or element['MSA_NUM'] <= 5:\n",
        "\n",
        "            element['ISD'] = 1\n",
        "            vars['insufficient_data'] = True\n",
        "\n",
        "        else:\n",
        "\n",
        "            element['ISD'] = 0\n",
        "        \"\"\"\n",
        "        if vars['B/E']:\n",
        "\n",
        "            element['B/E'] = ref_Solv_Acc_Pred[index_of_pos[element['POS'] - 1]]\n",
        "            if element['B/E'] == \"e\":\n",
        "\n",
        "                if element['COLOR'] == 9 or element['COLOR'] == 8:\n",
        "\n",
        "                    element['F/S'] = \"f\"\n",
        "\n",
        "                else:\n",
        "\n",
        "                    element['F/S'] = \" \"\n",
        "\n",
        "            elif element['COLOR'] == 9:\n",
        "\n",
        "                element['F/S'] = \"s\"\n",
        "\n",
        "            else:\n",
        "\n",
        "                element['F/S'] = \" \"\n",
        "        \"\"\"\n",
        "    vars['layers_array'] = ColorLayers\n",
        "    LOG.write(\"assign_colors_according_to_r4s_layers : color layers are %s\\n\" %str(vars['layers_array']))\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "def extract_diversity_matrix_info(r4s_log_file):\n",
        "\n",
        "    # extracting diversity matrix info\n",
        "\n",
        "    matrix_disINFO = \"\\\"\\\"\"\n",
        "    matrix_lowINFO = \"\\\"\\\"\"\n",
        "    matrix_upINFO = \"\\\"\\\"\"\n",
        "\n",
        "    try:\n",
        "\n",
        "        RES_LOG = open(r4s_log_file, 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"extract_diversity_matrix_info: Can't open '\" + r4s_log_file + \"' for reading\\n\")\n",
        "\n",
        "    line = RES_LOG.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        line = line.rstrip()\n",
        "        match1 = re.match(r'\\#Average pairwise distance\\s*=\\s+(.+)', line)\n",
        "        if match1:\n",
        "\n",
        "            matrix_disINFO = match1.group(1)\n",
        "\n",
        "        else:\n",
        "\n",
        "            match2 = re.match(r'\\#lower bound\\s*=\\s+(.+)', line)\n",
        "            if match2:\n",
        "\n",
        "                matrix_lowINFO = match2.group(1)\n",
        "\n",
        "            else:\n",
        "\n",
        "                match3 = re.match(r'\\#upper bound\\s*=\\s+(.+)', line)\n",
        "                if match3:\n",
        "\n",
        "                    matrix_upINFO = match3.group(1)\n",
        "                    break\n",
        "\n",
        "        line = RES_LOG.readline()\n",
        "\n",
        "    RES_LOG.close()\n",
        "\n",
        "    vars['Average pairwise distance'] = matrix_disINFO\n",
        "\n",
        "\n",
        "def add_sequences_removed_by_cd_hit_to_rejected_report(cd_hit_clusters_file, rejected_fragments_file, num_rejected_homologs):\n",
        "\n",
        "    LOG.write(\"add_sequences_removed_by_cd_hit_to_rejected_report : running add_sequences_removed_by_cd_hit_to_rejected_report(%s, %s, %d)\\n\" %(cd_hit_clusters_file, rejected_fragments_file, num_rejected_homologs))\n",
        "\n",
        "    try:\n",
        "\n",
        "        REJECTED = open(rejected_fragments_file, 'a')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"extract_sequences_removed_by_cd_hit: Can't open '\" + rejected_fragments_file + \"' for writing.\")\n",
        "\n",
        "    try:\n",
        "\n",
        "        CDHIT = open(cd_hit_clusters_file, 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"extract_sequences_removed_by_cd_hit: Can't open '\" + cd_hit_clusters_file + \"' for reading.\\n\")\n",
        "\n",
        "    REJECTED.write(\"\\n\\t Sequences rejected in the clustering stage by CD-HIT\\n\\n\")\n",
        "\n",
        "    cluster_members = {}\n",
        "    cluster_head = \"\"\n",
        "\n",
        "    line = CDHIT.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        match = re.match(r'^>Cluster', line)\n",
        "        if match:\n",
        "\n",
        "            # New Cluster\n",
        "            for cluster_member in cluster_members.keys():\n",
        "\n",
        "                REJECTED.write(\"%d Fragment %s rejected: the sequence shares %s identity with %s (which was preserved)\\n\" %(num_rejected_homologs, cluster_member, cluster_members[cluster_member], cluster_head))\n",
        "                num_rejected_homologs += 1\n",
        "\n",
        "            cluster_members = {}\n",
        "            cluster_head = \"\"\n",
        "\n",
        "        else:\n",
        "\n",
        "            # Clusters Members\n",
        "            words = line.split()\n",
        "            if len(words) > 2:\n",
        "\n",
        "                x = words[2][1:-3] # delete the symbols > and ... from the beginning and and ending of the sequence name\n",
        "                if words[3] == \"*\":\n",
        "\n",
        "                    cluster_head  = x\n",
        "\n",
        "                elif len(words) > 3:\n",
        "\n",
        "                    cluster_members[x] = words[4]\n",
        "\n",
        "        line = CDHIT.readline()\n",
        "\n",
        "    vars['num_rejected_homologs'] = num_rejected_homologs\n",
        "\n",
        "\n",
        "\n",
        "def choose_homologoues_from_search_with_lower_identity_cutoff(searchType, query_seq_length, redundancyRate, frag_overlap, min_length_percent, min_id_percent, min_num_of_homologues, search_output, fasta_output, rejected_seqs, ref_search_hash, Nuc_or_AA):\n",
        "\n",
        "    # searchType: HMMER, BLAST or MMseqs2\n",
        "    # query_seq_length: Length of the query sequence\n",
        "    # redundancyRate: The allowed similarity between the query and the hit\n",
        "    # frag_overlap: The allowed overlap between the hits\n",
        "    # min_length_percent: The hit can't be smaller than this percent of the query\n",
        "    # min_id_percent: The minimum similarity between the query and the hit\n",
        "    # min_num_of_homologues: Minimum number of homologs\n",
        "    # search_output: Raw homolog search output\n",
        "    # fasta_output: Accepted hits\n",
        "    # rejected_seqs: Rejected hits\n",
        "    # ref_search_hash: Hash with the evalues of the accepted hits. This is later used when choosing the final hits after cid-hit\n",
        "    # Nuc_or_AA: Amino or nucleic acid\n",
        "\n",
        "    LOG.write(\"choose_homologoues_from_search_with_lower_identity_cutoff(%s, %d, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s);\\n\" %(searchType, query_seq_length, redundancyRate, frag_overlap, min_length_percent, min_id_percent, min_num_of_homologues, search_output, fasta_output, rejected_seqs, ref_search_hash, Nuc_or_AA))\n",
        "\n",
        "    # Defining the minimum length a homologue should have\n",
        "    # 60% the query's length\n",
        "    min_length = query_seq_length * min_length_percent\n",
        "\n",
        "    # Reading blast/hmmer output and collect the homologues\n",
        "    # Printing the selected homologues to a file and insert the e-value info to hash\n",
        "\n",
        "    OUT_REJECT = open(rejected_seqs, 'w')\n",
        "    OUT = open(fasta_output, 'w')\n",
        "    RAW_OUT = open(search_output, 'r')\n",
        "\n",
        "    num_homologoues = 0\n",
        "    num_rejected = 1\n",
        "    final_num_homologues = 0\n",
        "    OUT_REJECT.write(\"\\tSequences rejected for being too short, too similar, or not similar enough to the query sequence.\\n\\n\")\n",
        "    if searchType == \"MMseqs2\":\n",
        "\n",
        "        # we skip the first two lines which have the query\n",
        "        line = RAW_OUT.readline()\n",
        "        line = RAW_OUT.readline()\n",
        "        line = RAW_OUT.readline()\n",
        "        while line != \"\":\n",
        "\n",
        "            if line[0] == \">\":\n",
        "\n",
        "                # this line has the sequence details\n",
        "                words = line.split()\n",
        "\n",
        "                # new seq found\n",
        "\n",
        "                num_homologoues += 1\n",
        "                seq_name = words[0][1:]\n",
        "\n",
        "                seq_eval = float(words[3])\n",
        "                seq_beg = int(words[7])\n",
        "                seq_end = int(words[8])\n",
        "                seq_ident = float(words[2]) * 100\n",
        "\n",
        "            elif line[0] != \"\\x00\":\n",
        "\n",
        "                # this line has the sequence. We take it and save it with the details of the previous line\n",
        "                seq = line.strip() # the seq in fasta with gaps\n",
        "                seq = re.sub(r'-', \"\", seq) # delete gaps\n",
        "                seq_frag_name = \"%s|%d_%d|%s\" %(seq_name, seq_beg, seq_end, seq_eval)\n",
        "\n",
        "                # deciding if we take the fragment\n",
        "                ans = check_if_seq_valid_with_min_id(redundancyRate, min_length, min_id_percent, seq_ident, seq, seq_name, Nuc_or_AA)\n",
        "\n",
        "                if seq_eval > form['E_VALUE']:\n",
        "\n",
        "                    ans = \"The E-value %.16g is over the limit %.16g.\" %(seq_eval, form['E_VALUE'])\n",
        "\n",
        "                # after taking the info, check if the currecnt sequence is valid. If so - insert it to the hash\n",
        "                if ans == \"yes\":\n",
        "\n",
        "                    final_num_homologues += 1\n",
        "                    OUT.write(\">%s\\n%s\\n\" %(seq_frag_name, seq))\n",
        "                    ref_search_hash[seq_frag_name] = seq_eval\n",
        "\n",
        "                else:\n",
        "\n",
        "                    OUT_REJECT.write(\"%d Fragment %s rejected: %s\\n\" %(num_rejected, seq_frag_name, ans))\n",
        "                    num_rejected += 1\n",
        "\n",
        "            line = RAW_OUT.readline()\n",
        "\n",
        "    OUT_REJECT.close()\n",
        "    OUT.close()\n",
        "    # Checking that the number of homologues found is legal\n",
        "\n",
        "    if final_num_homologues < min_num_of_homologues:\n",
        "\n",
        "        message = \"Only %d unique sequences were chosen. The minimal number of sequences required for the calculation is %d. You can try to rerun with a multiple sequence alignment file of your own, increase the E value or decrease the minimal %%ID for homologs\" %(final_num_homologues, min_num_of_homologues)\n",
        "        exit_on_error(\"\", message)\n",
        "\n",
        "    vars['number_of_homologoues'] = num_homologoues\n",
        "    vars['number_of_homologoues_before_cd-hit'] = final_num_homologues\n",
        "    vars['num_rejected_homologs'] = num_rejected\n",
        "\n",
        "\n",
        "\n",
        "def change_name(s_name):\n",
        "\n",
        "    if s_name[:2] == \"sp\" or s_name[:2] == \"tr\":\n",
        "\n",
        "        words = s_name.split(\"|\")\n",
        "        return  \"up|\" + words[1]\n",
        "\n",
        "    elif s_name[:2] == \"gi\":\n",
        "\n",
        "        words = s_name.split(\"|\")\n",
        "        return  \"gi|\" + words[1]\n",
        "\n",
        "    elif s_name[:8] == \"UniRef90\":\n",
        "\n",
        "        words = s_name.split(\"_\")\n",
        "        return \"ur|\" + words[1]\n",
        "\n",
        "    elif \"|\" in s_name:\n",
        "\n",
        "        words = s_name.split(\"|\")\n",
        "        return \"up|\" + words[1]\n",
        "\n",
        "    else:\n",
        "\n",
        "        return \"gi|\" + s_name\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "def check_if_seq_valid_with_min_id(redundancyRate, min_length, min_id, ident_percent, aaSeq, seqName, Nuc_or_AA):\n",
        "\n",
        "    seq_length = len(aaSeq)\n",
        "\n",
        "    if ident_percent >= redundancyRate:\n",
        "\n",
        "        # the sequence identity is not too high\n",
        "        return \"identity percent %.2f is too big\" %ident_percent\n",
        "\n",
        "    if ident_percent < min_id:\n",
        "\n",
        "        # the sequence identity is higher than the minium idnentity percent that was defined for homologus\n",
        "        return \"identity percent %.2f is too low (below %d)\" %(ident_percent, min_id)\n",
        "\n",
        "    elif seq_length < min_length:\n",
        "\n",
        "        # the sequnece length is greater than the minimum sequence length\n",
        "        return \"the sequence length %d is too short. The minimum is %d\" %(seq_length, min_length)\n",
        "\n",
        "    return check_illegal_character(aaSeq, seqName, Nuc_or_AA)\n",
        "\n",
        "\n",
        "def check_illegal_character(aaSeq, seqName, Nuc_or_AA):\n",
        "\n",
        "    # the sequnece letters should be legal to rate4site\n",
        "\n",
        "    if Nuc_or_AA == \"AA\":\n",
        "\n",
        "        # AA seq\n",
        "        if not re.match(r'^[ACDEFGHIKLMNPQRSTVWYBZXacdefghiklmnpqrstvwybzx]+$', aaSeq):\n",
        "\n",
        "            return \"illegal character was found in sequence: \" + seqName\n",
        "\n",
        "    else:\n",
        "\n",
        "        # Nuc seq\n",
        "        if not re.match(r'^[ACGTUINacgtuin]+$', aaSeq):\n",
        "\n",
        "            return \"illegal character was found in sequence: \" + seqName\n",
        "\n",
        "    return \"yes\"\n",
        "\n",
        "def check_if_no_overlap(max_overlap, ref_seq_details, s_bgn, s_end):\n",
        "\n",
        "    ans = \"check_if_no_overlap : no ans was picked\"\n",
        "\n",
        "    i = 0\n",
        "    while i < len(ref_seq_details):\n",
        "\n",
        "        fragment_beg = ref_seq_details[i][0]\n",
        "        fragment_end = ref_seq_details[i][1]\n",
        "        fragment_length = int(fragment_end) - int(fragment_beg) + 1\n",
        "\n",
        "        if s_bgn <= fragment_beg and s_end >= fragment_end:\n",
        "\n",
        "            # fragment is inside subjct\n",
        "            return \"previous fragment found %s_%s is fully inside new fragment\" %(fragment_beg, fragment_end)\n",
        "\n",
        "        elif s_bgn >= fragment_beg and s_end <= fragment_end:\n",
        "\n",
        "            # subjct is inside fragment\n",
        "            return \"new fragment is fully inside previous fragment found \" + str(fragment_beg + fragment_end)\n",
        "\n",
        "        elif fragment_end < s_end and fragment_end > s_bgn:\n",
        "\n",
        "            # fragment begins before subjct\n",
        "            overlap_length = fragment_end - s_bgn + 1\n",
        "            if overlap_length > fragment_length * max_overlap:\n",
        "\n",
        "                return \"overlap length of fragment is %d which is greater than maximum overlap: %d\" %(overlap_length, fragment_length * max_overlap)\n",
        "\n",
        "            else:\n",
        "\n",
        "                # when the fragment might be a good match, we can only insert it if it did not match to all the fragments\n",
        "                if i == len(ref_seq_details) - 1:\n",
        "\n",
        "                    ans = \"yes\"\n",
        "\n",
        "        elif fragment_beg > s_bgn and  fragment_beg < s_end:\n",
        "\n",
        "            # fragment begins after subjct\n",
        "            overlap_length = s_end - fragment_beg + 1\n",
        "            if overlap_length > fragment_length * max_overlap:\n",
        "\n",
        "                return \"overlap length of fragment is %d which is greater than maximum overlap: %d\" %(overlap_length, fragment_length * max_overlap)\n",
        "\n",
        "            else:\n",
        "\n",
        "                # when the fragment might be a good match, we can only insert it if it did not match to all the fragments\n",
        "                if i == len(ref_seq_details) - 1:\n",
        "\n",
        "                    ans = \"yes\"\n",
        "\n",
        "        elif fragment_beg >= s_end or fragment_end <= s_bgn:\n",
        "\n",
        "            # no overlap\n",
        "            if i == len(ref_seq_details) - 1:\n",
        "\n",
        "                ans = \"yes\"\n",
        "\n",
        "        i += 1\n",
        "\n",
        "    return ans\n",
        "\n",
        "\n",
        "def count_letters(s):\n",
        "\n",
        "    l = 0\n",
        "    for c in s:\n",
        "\n",
        "        if c.isalpha():\n",
        "\n",
        "            l += 1\n",
        "\n",
        "    return l\n",
        "\n",
        "\n",
        "def write_alignment(first_seq, first_seq_name, second_seq, second_seq_name, middle_line, clustalw_aln):\n",
        "\n",
        "    # we want all the lines to begin at the same point\n",
        "    while len(first_seq_name) < 15:\n",
        "\n",
        "        first_seq_name += \" \"\n",
        "\n",
        "    while len(second_seq_name) < 15:\n",
        "\n",
        "        second_seq_name += \" \"\n",
        "\n",
        "    middle_line_beginning = \"\"\n",
        "    while len(middle_line_beginning) < 15:\n",
        "\n",
        "        middle_line_beginning += \" \"\n",
        "\n",
        "    try:\n",
        "\n",
        "        CLUSTALW_ALN = open(clustalw_aln, 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"write_alignment : could not open \" + clustalw_aln + \" for writing.\")\n",
        "\n",
        "    seq_too_long_for_page = True\n",
        "    while seq_too_long_for_page:\n",
        "\n",
        "        # we show only 60 chars of the seq in each line\n",
        "        if len(first_seq) <= 60:\n",
        "\n",
        "            seq_too_long_for_page = False\n",
        "\n",
        "        CLUSTALW_ALN.write(first_seq_name + first_seq[:60] + \"\\n\")\n",
        "        CLUSTALW_ALN.write(middle_line_beginning + middle_line[:60] + \"\\n\")\n",
        "        CLUSTALW_ALN.write(second_seq_name + second_seq[:60] + \"\\n\\n\")\n",
        "\n",
        "        first_seq = first_seq[60:]\n",
        "        middle_line = middle_line[60:]\n",
        "        second_seq = second_seq[60:]\n",
        "\n",
        "    CLUSTALW_ALN.close()\n",
        "\n",
        "def pairwise_alignment(first_seq, second_seq, clustalw_aln = \"\", seq_type = \"\"):\n",
        "\n",
        "\t# for new Bio\n",
        "    aligner = Align.PairwiseAligner()\n",
        "\n",
        "    # Pairwise Alignment Paramaters\n",
        "\n",
        "    aligner.mode = 'global' #Can be either global or local, if undetermined, biopython will choose optimal algorithem\n",
        "\n",
        "    if form['DNA_AA'] == \"AA\":\n",
        "\n",
        "        aligner.substitution_matrix = substitution_matrices.load(\"/content/matrix.txt\")\n",
        "\n",
        "    else:\n",
        "\n",
        "        aligner.substitution_matrix = substitution_matrices.load(\"/content/matrix-nuc.txt\")\n",
        "\n",
        "    #Default Gap extension and opening penalties for ClustalW are 0.2 and 10.0.\n",
        "    aligner.open_gap_score = -5.0 #-10.0\n",
        "    aligner.extend_gap_score = -0.20\n",
        "    #aligner.target_end_gap_score = 0.0\n",
        "    #aligner.query_end_gap_score = 0.0\n",
        "\n",
        "    alignments = aligner.align(first_seq, second_seq)\n",
        "    #[first_seq_with_gaps, middle_line, second_seq_with_gaps] = (str(alignments[0])).split() # old Bio\n",
        "    first_seq_with_gaps = \"\"\n",
        "    second_seq_with_gaps = \"\"\n",
        "    alignment_string = str(alignments[0])\n",
        "    lines = alignment_string.split('\\n')\n",
        "    for line in lines:\n",
        "\n",
        "        words = line.split()\n",
        "        if len(words) > 2:\n",
        "\n",
        "            if words[0] == \"target\":\n",
        "\n",
        "                first_seq_with_gaps += words[2]\n",
        "\n",
        "            elif words[0] == \"query\":\n",
        "\n",
        "                second_seq_with_gaps += words[2]\n",
        "\n",
        "    matches = 0\n",
        "    length_without_gaps = 0\n",
        "    for i in range(len(first_seq_with_gaps)):\n",
        "\n",
        "        if first_seq_with_gaps[i] != '-' and second_seq_with_gaps[i] != '-':\n",
        "\n",
        "            length_without_gaps += 1\n",
        "            if first_seq_with_gaps[i] == second_seq_with_gaps[i]:\n",
        "\n",
        "                matches += 1\n",
        "\n",
        "    identity = (matches * 100.0) / length_without_gaps\n",
        "\n",
        "    if clustalw_aln == \"\":\n",
        "\n",
        "        # we don't write the alignment\n",
        "        return identity\n",
        "\n",
        "    try:\n",
        "\n",
        "        CLUSTALW_ALN = open(clustalw_aln, 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"write_alignment : could not open \" + clustalw_aln + \" for writing.\")\n",
        "\n",
        "    if seq_type == \"SEQRES\":\n",
        "\n",
        "        CLUSTALW_ALN.write(\"target - Seqres sequence\\nguery - Atom sequence\\n\\n\" + alignment_string)\n",
        "\n",
        "    else:\n",
        "\n",
        "        CLUSTALW_ALN.write(\"target - MSA sequence\\nguery - Atom sequence\\n\\n\" + alignment_string)\n",
        "\n",
        "    CLUSTALW_ALN.close()\n",
        "    #write_alignment(first_seq_with_gaps, seq_type + \"_SEQ\", second_seq_with_gaps, \"ATOM_SEQ\", middle_line, clustalw_aln) # old Bio\n",
        "    return(first_seq_with_gaps, second_seq_with_gaps, identity)\n",
        "\n",
        "\n",
        "def pairwise_alignment_old(first_seq, second_seq, clustalw_aln = \"\", seq_type = \"\"):\n",
        "\n",
        "\t# for old Bio\n",
        "    aligner = Align.PairwiseAligner()\n",
        "\n",
        "    # Pairwise Alignment Paramaters\n",
        "\n",
        "    aligner.mode = 'global' #Can be either global or local, if undetermined, biopython will choose optimal algorithem\n",
        "\n",
        "    if form['DNA_AA'] == \"AA\":\n",
        "\n",
        "        aligner.substitution_matrix = substitution_matrices.load(\"/content/matrix.txt\")\n",
        "\n",
        "    else:\n",
        "\n",
        "        aligner.substitution_matrix = substitution_matrices.load(\"/content/matrix-nuc.txt\")\n",
        "\n",
        "    #Default Gap extension and opening penalties for ClustalW are 0.2 and 10.0.\n",
        "    aligner.open_gap_score = -5.0 #-10.0\n",
        "    aligner.extend_gap_score = -0.20\n",
        "    #aligner.target_end_gap_score = 0.0\n",
        "    #aligner.query_end_gap_score = 0.0\n",
        "\n",
        "    alignments = aligner.align(first_seq, second_seq)\n",
        "    #[first_seq_with_gaps, middle_line, second_seq_with_gaps] = (str(alignments[0])).split() # old Bio\n",
        "    alignment_string = str(alignments[0])\n",
        "    lines = alignment_string.split('\\n')\n",
        "    first_seq_with_gaps = lines[0]\n",
        "    middle_line = lines[1]\n",
        "    second_seq_with_gaps = lines[2]\n",
        "\n",
        "    matches = 0\n",
        "    length_without_gaps = 0\n",
        "    for i in range(len(first_seq_with_gaps)):\n",
        "\n",
        "        if first_seq_with_gaps[i] != '-' and second_seq_with_gaps[i] != '-':\n",
        "\n",
        "            length_without_gaps += 1\n",
        "            if first_seq_with_gaps[i] == second_seq_with_gaps[i]:\n",
        "\n",
        "                matches += 1\n",
        "\n",
        "    identity = (matches * 100.0) / length_without_gaps\n",
        "\n",
        "    if clustalw_aln == \"\":\n",
        "\n",
        "        # we don't write the alignment\n",
        "        return identity\n",
        "\n",
        "    write_alignment(first_seq_with_gaps, seq_type + \"_SEQ\", second_seq_with_gaps, \"ATOM_SEQ\", middle_line, clustalw_aln) # old Bio\n",
        "    return(first_seq_with_gaps, second_seq_with_gaps, identity)\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "class PDF(fpdf.FPDF):\n",
        "\n",
        "    def __init__(self):\n",
        "\n",
        "        super().__init__()\n",
        "        self.lasth = 0\n",
        "        self.cbs = False\n",
        "\n",
        "    def Cell(self, w, h = 0, txt = '', border = 0, ln = 0, align = '', fill = False):\n",
        "\n",
        "        self.cell(w, h, txt, border, ln, align, fill)\n",
        "        self.lasth = h\n",
        "\n",
        "    def Print_4_Lines_Element(self, Rows_Pos, Score, AA, Pos, Solv_Acc, font_size, Insufficient_Data, Funct_Struct = \"\"):\n",
        "\n",
        "\n",
        "        if Pos != \"\":\n",
        "\n",
        "            x = self.get_x()\n",
        "            if Pos < 9:\n",
        "\n",
        "                self.set_xy(self.get_x() + 0.5, self.get_y())\n",
        "\n",
        "            elif Pos < 99:\n",
        "\n",
        "                self.set_xy(self.get_x() + 1, self.get_y())\n",
        "\n",
        "            elif Pos < 9999:\n",
        "\n",
        "                self.set_xy(self.get_x() + 1.5, self.get_y())\n",
        "\n",
        "            elif Pos < 99999:\n",
        "\n",
        "                self.set_xy(self.get_x() + 2, self.get_y())\n",
        "\n",
        "            self.Print_BackgroundColor(str(Pos + 1), \"\", font_size, 5)\n",
        "            self.set_xy(x, self.get_y())\n",
        "\n",
        "        self.set_xy(self.get_x(), self.get_y() + self.lasth)\n",
        "\n",
        "        self.Print_BackgroundColor(AA, 'B', font_size, Score, 3, Insufficient_Data)\n",
        "        Col_Pos = self.get_x() # position on the line after printing\n",
        "        self.set_xy(self.get_x() - 2.5, self.get_y() + self.lasth + 0.2)\n",
        "\n",
        "        if Solv_Acc == \"e\":\n",
        "\n",
        "            self.Print_ForegroundColor(Solv_Acc, 'B', font_size, 1)\n",
        "\n",
        "        elif Solv_Acc == \"b\":\n",
        "\n",
        "            self.Print_ForegroundColor(Solv_Acc, 'B', font_size, 2)\n",
        "\n",
        "        self.set_xy(self.get_x() - 2, self.get_y() + self.lasth - 1.2)\n",
        "\n",
        "        if Funct_Struct == \"f\":\n",
        "\n",
        "            self.Print_ForegroundColor(Funct_Struct, 'B', font_size, 3)\n",
        "\n",
        "        elif Funct_Struct == \"s\":\n",
        "\n",
        "            self.Print_ForegroundColor(Funct_Struct, 'B', font_size, 4)\n",
        "\n",
        "        self.set_xy(Col_Pos, Rows_Pos)\n",
        "\n",
        "\n",
        "\n",
        "    def Print_BackgroundColor(self, txt, print_style, font_size, Color_Num, spacer = 2, isd = False):\n",
        "\n",
        "        cbs = self.cbs\n",
        "\n",
        "        if Color_Num == 0:\n",
        "\n",
        "            self.set_text_color(0, 0, 0)\n",
        "            self.set_fill_color(255, 255, 150)\n",
        "\n",
        "        elif Color_Num == 1:\n",
        "\n",
        "            if cbs:\n",
        "\n",
        "                self.set_fill_color(10, 125, 130)\n",
        "\n",
        "            else:\n",
        "\n",
        "                self.set_fill_color(15, 90, 35)\n",
        "\n",
        "            self.set_text_color(255, 255, 255)\n",
        "\n",
        "        elif Color_Num == 2:\n",
        "\n",
        "            if cbs:\n",
        "\n",
        "                self.set_fill_color(75, 175, 190)\n",
        "\n",
        "            else:\n",
        "\n",
        "                self.set_fill_color(90, 175, 95)\n",
        "\n",
        "        elif Color_Num == 3:\n",
        "\n",
        "            if cbs:\n",
        "\n",
        "                self.set_fill_color(165, 220, 230)\n",
        "\n",
        "            else:\n",
        "\n",
        "                self.set_fill_color(165, 220, 160)\n",
        "\n",
        "        elif Color_Num == 4:\n",
        "\n",
        "            if cbs:\n",
        "\n",
        "                self.set_fill_color(215, 240, 240)\n",
        "\n",
        "            else:\n",
        "\n",
        "                self.set_fill_color(215, 240, 210)\n",
        "\n",
        "        elif Color_Num == 5:\n",
        "\n",
        "            self.set_fill_color(255, 255, 255)\n",
        "\n",
        "        elif Color_Num == 6:\n",
        "\n",
        "            if cbs:\n",
        "\n",
        "                self.set_fill_color(250, 235, 245)\n",
        "\n",
        "            else:\n",
        "\n",
        "                self.set_fill_color(230, 210, 230)\n",
        "\n",
        "        elif Color_Num == 7:\n",
        "\n",
        "            if cbs:\n",
        "\n",
        "                self.set_fill_color(250, 200, 220)\n",
        "\n",
        "            else:\n",
        "\n",
        "                self.set_fill_color(195, 165, 205)\n",
        "\n",
        "        elif Color_Num == 8:\n",
        "\n",
        "            if cbs:\n",
        "\n",
        "                self.set_fill_color(240, 125, 170)\n",
        "\n",
        "            else:\n",
        "\n",
        "                self.set_fill_color(155, 110, 170)\n",
        "\n",
        "        elif Color_Num == 9:\n",
        "\n",
        "            if cbs:\n",
        "\n",
        "                self.set_fill_color(160, 40, 95)\n",
        "\n",
        "            else:\n",
        "\n",
        "                self.set_fill_color(120, 40, 130)\n",
        "\n",
        "            self.set_text_color(255, 255, 255)\n",
        "\n",
        "        if isd:\n",
        "\n",
        "            self.set_fill_color(255, 255, 150)\n",
        "\n",
        "        self.set_font(\"Courier\", print_style, font_size)\n",
        "        width = len(txt) - 1 + spacer\n",
        "        high = font_size / 2\n",
        "        self.Cell(width, high, txt, 0, 0, \"C\", True)\n",
        "        self.set_fill_color(255, 255, 255) # return to default background color (white)\n",
        "        self.set_text_color(0, 0, 0) # return to default text color (black)\n",
        "\n",
        "    def Print_NEW_Legend(self, IS_THERE_FUNCT_RES, IS_THERE_STRUCT_RES, IS_THERE_INSUFFICIENT_DATA, B_E_METHOD):\n",
        "\n",
        "        self.set_font(\"\", 'B', 12)\n",
        "        font_size = 12\n",
        "        self.ln()\n",
        "        self.Cell(40, 10, \"The conservation scale:\", 0, 1)\n",
        "        self.set_xy(18, self.get_y())\n",
        "        #self.Print_BackgroundColor('?', \"\", 12, 0, 4, 1)\n",
        "        self.Print_BackgroundColor('1', \"\", 12, 1, 4)\n",
        "        self.Print_BackgroundColor('2', \"\", 12, 2, 4)\n",
        "        self.Print_BackgroundColor('3', \"\", 12, 3, 4)\n",
        "        #Average_X = self.get_x()\n",
        "        self.Print_BackgroundColor('4', \"\", 12, 4, 4)\n",
        "        self.Print_BackgroundColor('5', \"\", 12, 5, 4)\n",
        "        self.Print_BackgroundColor('6', \"\", 12, 6, 4)\n",
        "        self.Print_BackgroundColor('7', \"\", 12, 7, 4)\n",
        "        #Conserved_X = self.get_x()\n",
        "        self.Print_BackgroundColor('8', '', 12, 8, 4)\n",
        "        self.Print_BackgroundColor('9', '', 12, 9, 4)\n",
        "        self.ln()\n",
        "        self.set_font(\"Times\", 'B', 9.5)\n",
        "        self.Cell(10, 6, \"Variable\", 0, 0, 'R', False)\n",
        "        self.set_xy(28, self.get_y())\n",
        "        self.Cell(15, 6, \"Average\", 0, 0, 'R', False)\n",
        "        self.set_xy(53, self.get_y())\n",
        "        self.Cell(15, 6, \"Conserved\", 0, 0, 'R', False)\n",
        "        self.ln()\n",
        "        self.ln()\n",
        "\n",
        "        if B_E_METHOD != \"no prediction\":\n",
        "\n",
        "\n",
        "            offset_1 = 0\n",
        "            offset_2 = 0\n",
        "            if B_E_METHOD == \"neural network algorithm\":\n",
        "\n",
        "                offset_1 = 63.5\n",
        "                offset_2 = 62\n",
        "\n",
        "            elif B_E_METHOD == \"NACSES algorithm\":\n",
        "\n",
        "                offset_1 = 57.5\n",
        "                offset_2 = 56\n",
        "\n",
        "\n",
        "            self.Print_ForegroundColor('e', 'B', font_size, 1)\n",
        "            self.set_xy(offset_1, self.get_y())\n",
        "            self.Print_ForegroundColor(\" - An exposed residue according to the %s.\" %B_E_METHOD, 'B', font_size, 9)\n",
        "            self.ln()\n",
        "            #self.set_y(self.get_y() + self.lasth + 5)\n",
        "            #self.set_x(self.left_margin)\n",
        "            self.Print_ForegroundColor('b', 'B', font_size, 2)\n",
        "            self.set_xy(offset_2, self.get_y())\n",
        "            self.Print_ForegroundColor(\" - A buried residue according to the %s.\" %B_E_METHOD, 'B', font_size, 9)\n",
        "            self.ln()\n",
        "\n",
        "        #self.set_xy(45, self.get_y() + self.lasth + 5)\n",
        "        #self.y += self.lasth + 5\n",
        "        #self.x = self.left_margin\n",
        "        if IS_THERE_FUNCT_RES:\n",
        "\n",
        "            self.Print_ForegroundColor('f', 'B', font_size, 3)\n",
        "            self.set_xy(64.5, self.get_y())\n",
        "            self.Print_ForegroundColor(\" - A predicted functional residue (highly conserved and exposed).\", 'B', font_size, 9)\n",
        "            self.ln()\n",
        "            #self.set_y(self.get_y() + self.lasth + 5)\n",
        "            #self.x = self.left_margin\n",
        "\n",
        "        if IS_THERE_STRUCT_RES:\n",
        "\n",
        "            self.Print_ForegroundColor('s', 'B', font_size, 4)\n",
        "            self.set_xy(64, self.get_y())\n",
        "            self.Print_ForegroundColor(\" - A predicted structural residue (highly conserved and buried).\", 'B', font_size, 9)\n",
        "            self.ln()\n",
        "            #self.set_y(self.get_y() + self.lasth + 5)\n",
        "            #self.x = self.left_margin\n",
        "\n",
        "        if IS_THERE_INSUFFICIENT_DATA:\n",
        "\n",
        "            self.Print_BackgroundColor('x', 'B', font_size, 0, 2, 1)\n",
        "            self.set_xy(58, self.get_y())\n",
        "            self.Print_ForegroundColor(\" - Insufficient data - the calculation for this site was\", 'B', font_size, 9)\n",
        "            self.ln()\n",
        "            #self.set_y(self.get_y() + self.lasth + 4)\n",
        "            #self.x = self.left_margin\n",
        "            self.set_xy(48, self.get_y())\n",
        "            self.Print_ForegroundColor(\"     performed on less than 10% of the sequences.\",'B', font_size, 9)\n",
        "            #self.set_y(self.get_y() + self.lasth + 5)\n",
        "            #self.x = self.left_margin\n",
        "\n",
        "\n",
        "    def Print_ForegroundColor(self, txt, print_style, font_size, Color, spacer = 2):\n",
        "\n",
        "        if Color == 1: # orange\n",
        "\n",
        "            self.set_text_color(255, 153, 0)\n",
        "\n",
        "        elif Color == 2: # green\n",
        "\n",
        "            self.set_text_color(0, 204, 0)\n",
        "\n",
        "        elif Color == 3: # red\n",
        "\n",
        "            self.set_text_color(255, 0, 0)\n",
        "\n",
        "        elif Color == 4: # blue\n",
        "\n",
        "            self.set_text_color(0, 0, 153)\n",
        "\n",
        "        else: # black\n",
        "\n",
        "            self.set_text_color(0, 0, 0)\n",
        "\n",
        "        self.set_font(\"Courier\", print_style, font_size)\n",
        "        width = len(txt) - 1 + spacer\n",
        "        high = font_size / 2\n",
        "        self.Cell(width, high, txt, 0, 0, 'C', True)\n",
        "        self.set_fill_color(255, 255, 255) # return to default background (white)\n",
        "        self.set_text_color(0, 0, 0) # return to default text color (black)\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "def create_pdf_regular_or_cbs(cbs, name):\n",
        "\n",
        "    #prediction_method = prediction_method.replace('-', ' ')\n",
        "    pdf = PDF()\n",
        "    pdf.add_page()\n",
        "    #pdf.set_font(\"Times\", \"B\", 20)\n",
        "    pdf.add_font('DejaVu', '', '/content/dejavu-fonts-ttf-2.37/ttf/DejaVuSans.ttf', uni=True)\n",
        "    pdf.set_font('DejaVu', '', 20)\n",
        "    pdf.Cell(0, 0, \"ConSurf Results. Date:\" + vars['date'], 0, 0, 'C')\n",
        "    pdf.set_font(\"Times\", \"B\", 20)\n",
        "    pdf.set_y(pdf.get_y() + 10)\n",
        "    pdf.cbs = cbs\n",
        "    Rows_Pos = 0\n",
        "\n",
        "    \"\"\"\n",
        "    ConSurf_Grades = []\n",
        "    Protein_Length = 0\n",
        "    IS_THERE_INSUFFICIENT_DATA = False\n",
        "    IS_THERE_FUNCT_RES = False\n",
        "    IS_THERE_STRUCT_RES = False\n",
        "    try:\n",
        "\n",
        "        GRADES = open(vars['gradesPE'], 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error(\"create_pdf: unable to open the file \" + vars['gradesPE'] + \" for reading.\")\n",
        "\n",
        "    line = GRADES.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        words = line.split()\n",
        "        if len(words) > 5 and words[0].isnumeric():\n",
        "\n",
        "            Protein_Length += 1\n",
        "            details = {}\n",
        "\n",
        "            details[\"AA\"] = words[1]\n",
        "\n",
        "            if '*' in words[color_column]:\n",
        "\n",
        "                details[\"COLOR\"] = 0\n",
        "                IS_THERE_INSUFFICIENT_DATA = True\n",
        "\n",
        "            else:\n",
        "\n",
        "                details[\"COLOR\"] = int(words[color_column])\n",
        "\n",
        "            if B_E_column is not None:\n",
        "\n",
        "                F_S_column = B_E_column + 1\n",
        "                if words[B_E_column] == 'b' or words[B_E_column] == 'e':\n",
        "\n",
        "                    details[\"B_E\"] = words[B_E_column]\n",
        "                    IS_THERE_STRUCT_RES = True\n",
        "                    if words[F_S_column] == 'f' or words[F_S_column] == 's':\n",
        "\n",
        "\n",
        "                        details[\"F_S\"] = words[F_S_column]\n",
        "                        IS_THERE_FUNCT_RES = True\n",
        "\n",
        "                    else:\n",
        "\n",
        "                       details[\"F_S\"] = \"\"\n",
        "\n",
        "                else:\n",
        "\n",
        "                    details[\"B_E\"] = \"\"\n",
        "                    details[\"F_S\"] = \"\"\n",
        "\n",
        "            else:\n",
        "\n",
        "                details[\"B_E\"] = \"\"\n",
        "                details[\"F_S\"] = \"\"\n",
        "\n",
        "            ConSurf_Grades.append(details)\n",
        "\n",
        "        line = GRADES.readline()\n",
        "\n",
        "    GRADES.close()\n",
        "    \"\"\"\n",
        "    maxPosPerPage = 600\n",
        "    for elem in vars['gradesPE_Output']:\n",
        "\n",
        "        Pos = elem['POS'] - 1\n",
        "        if vars['B/E']:\n",
        "\n",
        "            prediction_method = \"neural network algorithm\"\n",
        "            B_E = elem['B/E']\n",
        "            F_S = elem['F/S'].strip()\n",
        "\n",
        "        else:\n",
        "\n",
        "            prediction_method = \"no prediction\"\n",
        "            B_E = \"\"\n",
        "            F_S = \"\"\n",
        "\n",
        "        if Pos % maxPosPerPage == 0 and Pos != 0:\n",
        "\n",
        "            pdf.add_page()\n",
        "\n",
        "        if Pos % 50 == 0:\n",
        "\n",
        "            pdf.ln()\n",
        "            pdf.ln()\n",
        "            pdf.ln()\n",
        "            pdf.ln()\n",
        "\n",
        "            Rows_Pos = pdf.get_y()\n",
        "\n",
        "        elif Pos % 10 == 0:\n",
        "\n",
        "            pdf.Print_ForegroundColor(\"\", 'B', 10, 0, 4)\n",
        "\n",
        "        if Pos % 10 == 0:\n",
        "\n",
        "            pdf.Print_4_Lines_Element(Rows_Pos, elem['COLOR'], elem['SEQ'], Pos, B_E, 10, elem['ISD'], F_S)\n",
        "\n",
        "        else:\n",
        "\n",
        "            pdf.Print_4_Lines_Element(Rows_Pos, elem['COLOR'], elem['SEQ'], \"\", B_E, 10, elem['ISD'], F_S)\n",
        "\n",
        "    pdf.ln()\n",
        "    pdf.ln()\n",
        "    pdf.ln()\n",
        "    pdf.ln()\n",
        "    pdf.Print_NEW_Legend(vars['B/E'], vars['B/E'], vars['insufficient_data'],  prediction_method)\n",
        "\n",
        "    pdf.output(name)\n",
        "\n",
        "\n",
        "\n",
        "def conseq_create_output():\n",
        "\n",
        "    create_gradesPE(vars['gradesPE'])\n",
        "    create_pdf()\n",
        "    no_model_view()\n",
        "\n",
        "\n",
        "def consurf_create_output():\n",
        "\n",
        "\n",
        "    r4s2pdb = {} # key: poistion in SEQRES/MSA, value: residue name with position in atom (i.e: ALA:22:A)\n",
        "\n",
        "    if vars['running_mode'] == \"_mode_pdb_msa\" or vars['running_mode'] == \"_mode_pdb_msa_tree\" or vars['SEQRES_seq'] != \"\":\n",
        "\n",
        "        match_pdb_to_seq(r4s2pdb, vars['seqres_or_msa_seq_with_gaps'], vars['ATOM_seq_with_gaps'], vars['pdb_object'])\n",
        "\n",
        "    else: # no seqres and no msa\n",
        "\n",
        "        find_pdb_position(r4s2pdb, vars['pdb_object'])\n",
        "\n",
        "\n",
        "    identical_chains = find_identical_chains_in_PDB_file(vars['pdb_object'], form['PDB_chain'])\n",
        "\n",
        "    atom_grades = {}\n",
        "    create_gradesPE(vars['gradesPE'], r4s2pdb, vars['pdb_file_name'], form['PDB_chain'], vars['Used_PDB_Name'], vars['pdb_object'], identical_chains, vars['cif_or_pdb'], atom_grades)\n",
        "    replace_TmpFactor_Consurf_Scores(atom_grades, form['PDB_chain'], vars['pdb_file_name'], vars['Used_PDB_Name']) # Create ATOMS section and replace the TempFactor Column with the ConSurf Grades (will create also isd file if relevant)\n",
        "\n",
        "    create_pdf()\n",
        "\n",
        "def extract_data_from_MSA():\n",
        "\n",
        "    #vars['query_string'] = form['msa_SEQNAME']\n",
        "    #vars['protein_seq_string'] = vars['MSA_query_seq']\n",
        "\n",
        "    ## mode :  include msa and pdb\n",
        "\n",
        "    if vars['running_mode'] == \"_mode_pdb_msa\" or vars['running_mode'] == \"_mode_pdb_msa_tree\":\n",
        "\n",
        "        compare_atom_seqres_or_msa(\"MSA\")\n",
        "\n",
        "def no_MSA():\n",
        "\n",
        "    # if there is pdb : we compare the atom and seqres\n",
        "    if vars['running_mode'] ==  \"_mode_pdb_no_msa\" and ('SEQRES_seq' in vars and len(vars['SEQRES_seq']) > 0):\n",
        "\n",
        "        # align seqres and pdb sequences\n",
        "        compare_atom_seqres_or_msa(\"SEQRES\")\n",
        "\n",
        "    vars['max_homologues_to_display'] = 500\n",
        "\n",
        "    blast_hash = {}\n",
        "\n",
        "    try:\n",
        "\n",
        "        call_mmseqs2()\n",
        "\n",
        "    except:\n",
        "\n",
        "        os.chdir(vars['working_dir'])\n",
        "        exit_on_error(\"calling mmseqs2 failed.\")\n",
        "\n",
        "    # choosing homologs, create fasta file for all legal homologs\n",
        "    cd_hit_hash = {}\n",
        "    #vars['hit_redundancy'] = float(form['MAX_REDUNDANCY']) # Now taken as argument from user #OLD: #CONSURF_CONSTANTS.FRAGMENT_REDUNDANCY_RATE\n",
        "    vars['hit_overlap'] = 0.1\n",
        "    #vars['min_num_of_hits'] = GENERAL_CONSTANTS.MINIMUM_FRAGMENTS_FOR_MSA\n",
        "    vars['low_num_of_hits'] = 10\n",
        "    vars['HITS_fasta_file'] = \"query_homolougs.txt\"\n",
        "    vars['HITS_rejected_file'] = vars['job_name'] + \"_rejected_homolougs.txt\"\n",
        "\n",
        "    choose_homologoues_from_search_with_lower_identity_cutoff(form['Homolog_search_algorithm'], len(vars['protein_seq_string']), vars['hit_redundancy'], vars['hit_overlap'], vars['hit_min_length'], float(form['MIN_IDENTITY']), vars['min_num_of_hits'], vars['BLAST_out_file'], vars['HITS_fasta_file'], vars['HITS_rejected_file'], blast_hash, form['DNA_AA'])\n",
        "\n",
        "    vars['cd_hit_out_file'] = \"query_cdhit.out\"\n",
        "    vars['unique_seqs'] = cluster_homologoues(cd_hit_hash)\n",
        "    LOG.write(\"num_of_unique_seq: %d\\n\" %vars['unique_seqs'])\n",
        "    add_sequences_removed_by_cd_hit_to_rejected_report(vars['cd_hit_out_file'] + \".clstr\", vars['HITS_rejected_file'], vars['num_rejected_homologs'])\n",
        "    choose_final_homologoues(blast_hash, cd_hit_hash, float(form['MAX_NUM_HOMOL']) -1, form['best_uniform_sequences'], vars['FINAL_sequences'], vars['HITS_rejected_file'], vars['num_rejected_homologs'])\n",
        "    vars['zip_list'].append(vars['HITS_rejected_file'])\n",
        "\n",
        "    if form['DNA_AA'] == \"Nuc\":\n",
        "\n",
        "        # convert rna to dna\n",
        "\n",
        "        LOG.write(\"convert_rna_to_dna(%s, %s)\\n\" %(vars['FINAL_sequences'], vars['FINAL_sequences'] + \".dna\"))\n",
        "        ans = convert_rna_to_dna(vars['FINAL_sequences'], vars['FINAL_sequences'] + \".dna\")\n",
        "        if ans[0] == \"OK\":\n",
        "\n",
        "            vars['FINAL_sequences'] += \".dna\"\n",
        "            LOG.write(\"Seqs with u or U: \" + str(ans[1]))\n",
        "            for seq in ans[1]:\n",
        "\n",
        "                print(\"Warnning: The seqeunce '\" + seq + \"' contains a 'U' replaced by 'T'\")\n",
        "\n",
        "        else:\n",
        "\n",
        "            exit_on_error('sys_error', ans)\n",
        "\n",
        "    LOG.write(\"make_sequences_file_HTML(%s, %s)\\n\" %(vars['FINAL_sequences'], vars['FINAL_sequences_html']))\n",
        "    make_sequences_file_HTML(vars['FINAL_sequences'], vars['FINAL_sequences_html'])\n",
        "\n",
        "    # we save to copies of the msa, one in fasta format and another in clustal format.\n",
        "    #vars['msa_fasta'] = \"msa_fasta.aln\"\n",
        "    #vars['msa_clustal'] = \"msa_clustal.aln\"\n",
        "    create_MSA()\n",
        "    vars['msa_SEQNAME'] = vars['query_string']\n",
        "\n",
        "    print(\"%d homologues were collected.\" %vars['number_of_homologoues'])\n",
        "    create_download_link(vars['FINAL_sequences_html'], \"These %d are sequences used for creating the MSA.\" %vars['final_number_of_homologoues'])\n",
        "    create_download_link(vars['HITS_rejected_file'], \"Download the list of rejected homologues\")\n",
        "\n",
        "def call_mmseqs2():\n",
        "\n",
        "    os.chdir(vars['root_dir'])\n",
        "\n",
        "    msa_mode = \"mmseqs2_uniref\"\n",
        "    pair_mode = \"unpaired_paired\"\n",
        "    pairing_strategy = \"greedy\"\n",
        "    result_dir = vars['job_name']\n",
        "\n",
        "    csv_file = \"/content/%s/%s.csv\" %(vars['job_name'], vars['job_name'])\n",
        "\n",
        "    CSV = open(csv_file, 'w')\n",
        "\n",
        "    CSV.write(\"id,sequence\\n%s,%s\" %(vars['job_name'], vars['protein_seq_string']))\n",
        "    CSV.close()\n",
        "\n",
        "    result_dir = Path(result_dir)\n",
        "\n",
        "    get_msa_and_templates(vars['job_name'], vars['protein_seq_string'], None, result_dir, msa_mode, False, None, pair_mode, pairing_strategy, 'https://api.colabfold.com', 'colabfold/google-colab-main')\n",
        "\n",
        "    os.chdir(vars['working_dir'])\n",
        "\n",
        "\n",
        "def extract_data_from_model():\n",
        "\n",
        "\n",
        "\n",
        "    if vars['cif_or_pdb'] == \"pdb\":\n",
        "\n",
        "        vars['pdb_object'] = pdbParser()\n",
        "\n",
        "    else:\n",
        "\n",
        "        vars['pdb_object'] = cifParser()\n",
        "\n",
        "    vars['pdb_object'].read(vars['pdb_file_name'], form['PDB_chain'], form['DNA_AA'])\n",
        "\n",
        "\n",
        "    #[vars['SEQRES_seq'], vars['ATOM_seq'], vars['ATOM_without_X_seq']] = get_seqres_atom_seq(vars['pdb_object'], form['PDB_chain'], vars['pdb_file_name'])\n",
        "    vars['SEQRES_seq'] = vars['pdb_object'].get_SEQRES()\n",
        "    All_atoms = vars['pdb_object'].get_ATOM_withoutX()\n",
        "    if form['PDB_chain'] in All_atoms:\n",
        "\n",
        "        vars['ATOM_without_X_seq'] = All_atoms[form['PDB_chain']]\n",
        "\n",
        "    else:\n",
        "\n",
        "        exit_on_error('user_error', \"The chain is not in the PDB. Select the PDB chain using the flag --chain\")\n",
        "\n",
        "    analyse_seqres_atom()\n",
        "\n",
        "    try:\n",
        "\n",
        "        FAS = open(vars['protein_seq'], 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error',\"cannot open the file \" + vars['protein_seq'] + \" for writing!\")\n",
        "\n",
        "    # we write the sequence to a fasta file for the homologues search\n",
        "    # we save the name of the quey string for rate4site\n",
        "    if vars['SEQRES_seq'] == \"\":\n",
        "\n",
        "        vars['query_string'] = \"Input_seq_ATOM_\" + form['PDB_chain']\n",
        "        vars['protein_seq_string'] = vars['ATOM_without_X_seq']\n",
        "        FAS.write(\">\" + vars['query_string'] + \"\\n\" + vars['ATOM_without_X_seq'])\n",
        "\n",
        "    else:\n",
        "\n",
        "        vars['query_string'] = \"Input_seq_SEQRES_\" + form['PDB_chain']\n",
        "        vars['protein_seq_string'] = vars['SEQRES_seq']\n",
        "        FAS.write(\">\" + vars['query_string'] + \"\\n\" + vars['SEQRES_seq'])\n",
        "\n",
        "    FAS.close()\n",
        "\n",
        "\n",
        "def create_cd_hit_output(input_file, output_file, cutoff, ref_cd_hit_hash, type):\n",
        "\n",
        "\n",
        "    seq = \"\"\n",
        "    seq_name = \"\"\n",
        "    cmd = \"\"\n",
        "    n = 0\n",
        "\n",
        "    # running cd-hit\n",
        "\n",
        "    if type == \"AA\":\n",
        "\n",
        "        cmd += \"%scd-hit -i %s -o %s \" %(vars['cd_hit_dir'], input_file, output_file)\n",
        "        if cutoff > 0.7 and cutoff < 1:\n",
        "\n",
        "            n = 5\n",
        "\n",
        "        elif cutoff > 0.6 and cutoff <= 0.7:\n",
        "\n",
        "            n = 4\n",
        "\n",
        "        elif cutoff > 0.5 and cutoff <= 0.6:\n",
        "\n",
        "            n = 3\n",
        "\n",
        "        elif cutoff > 0.4 and cutoff <= 0.5:\n",
        "\n",
        "            n = 2\n",
        "\n",
        "    else:\n",
        "\n",
        "        # DNA\n",
        "        cmd += \"cd-hit-est -i %s -o %s \" %(input_file, output_file)\n",
        "        if cutoff > 0.9 and cutoff < 1:\n",
        "\n",
        "            n = 8\n",
        "\n",
        "        elif cutoff > 0.88 and cutoff <= 0.9:\n",
        "\n",
        "            n = 7\n",
        "\n",
        "        elif cutoff > 0.85 and cutoff <= 0.88:\n",
        "\n",
        "            n = 6\n",
        "\n",
        "        elif cutoff > 0.8 and cutoff <= 0.85:\n",
        "\n",
        "            n = 5\n",
        "\n",
        "        elif cutoff > 0.75 and cutoff <= 0.8:\n",
        "\n",
        "            n = 4\n",
        "\n",
        "    cmd += \"-c %f -n %d -d 0\" %(cutoff, n)\n",
        "\n",
        "    submit_job_to_Q(\"CD-HIT\", cmd)\n",
        "\n",
        "    if not os.path.exists(output_file) or os.path.getsize(output_file) == 0:\n",
        "\n",
        "        exit_on_error(\"sys_error\", \"create_cd_hit_output : \" + str(cmd) + \": CD-HIT produced no output!\\n\")\n",
        "\n",
        "    num_cd_hits = 0\n",
        "\n",
        "    try:\n",
        "\n",
        "        CDHIT_OUTPUT = open(output_file, 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"create_cd_hit_output : could not open the file \" + output_file + \" for writing.\")\n",
        "\n",
        "    # inserting chosen homologues to a hash\n",
        "    line = CDHIT_OUTPUT.readline()\n",
        "    seq_name = \"\"\n",
        "    seq = \"\"\n",
        "    while line != \"\":\n",
        "\n",
        "        line = line.strip()\n",
        "        if line[0] == \">\":\n",
        "\n",
        "            seq_name = line[1:]\n",
        "\n",
        "        else:\n",
        "\n",
        "            seq = line\n",
        "            if not seq_name in ref_cd_hit_hash:\n",
        "\n",
        "                num_cd_hits += 1\n",
        "                ref_cd_hit_hash[seq_name] =seq\n",
        "\n",
        "        line = CDHIT_OUTPUT.readline()\n",
        "\n",
        "    CDHIT_OUTPUT.close()\n",
        "    return num_cd_hits\n",
        "\n",
        "\n",
        "\n",
        "def make_sequences_file_HTML(plain_txt_sequences, HTML_sequences):\n",
        "\n",
        "    try:\n",
        "\n",
        "        HTML_SEQUENCES = open(HTML_sequences, 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"make_sequences_file_HTML : cannot open the file \" + HTML_sequences + \" for writing.\")\n",
        "\n",
        "    try:\n",
        "\n",
        "        TXT_SEQUENCES = open(plain_txt_sequences, 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"make_sequences_file_HTML : cannot open the file \" + plain_txt_sequences + \" for reading.\")\n",
        "\n",
        "    counter = 1\n",
        "    line = TXT_SEQUENCES.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        line = line.strip()\n",
        "        if line == \"\":\n",
        "\n",
        "            line = TXT_SEQUENCES.readline()\n",
        "            continue\n",
        "\n",
        "        if line[0] != \">\":\n",
        "\n",
        "            counter += 1\n",
        "            HTML_SEQUENCES.write(\"<FONT FACE=\\\"courier new\\\" SIZE=3>\" + line + \"</FONT><BR>\\n\")\n",
        "\n",
        "        else:\n",
        "\n",
        "            line = line[1:]\n",
        "            if line[:9] == \"Input_seq\":\n",
        "\n",
        "                HTML_SEQUENCES.write(\"<FONT FACE=\\\"courier new\\\" SIZE=3>>%d_%s</FONT><BR>\\n\" %(counter, line))\n",
        "\n",
        "            else:\n",
        "\n",
        "                name = line.split(\"|\")[0]\n",
        "                HTML_SEQUENCES.write(\"<FONT FACE=\\\"courier new\\\" SIZE=3><A HREF=\\\"https://www.uniprot.org/uniref/%s\\\">>%d_%s</A></FONT><BR>\\n\" %(name, counter, line))\n",
        "\n",
        "        line = TXT_SEQUENCES.readline()\n",
        "\n",
        "    HTML_SEQUENCES.close()\n",
        "    TXT_SEQUENCES.close()\n",
        "\n",
        "\n",
        "def convert_rna_to_dna(Seqs, Seqs_dna):\n",
        "\n",
        "    # replace the u with t and return the sequences names replaced\n",
        "\n",
        "    try:\n",
        "\n",
        "        OUT = open(Seqs_dna, 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        return(\"convert_rna_to_dna: Can't open file \" + Seqs_dna + \" for writing.\")\n",
        "\n",
        "    try:\n",
        "\n",
        "        SEQS = open(Seqs, 'r')\n",
        "\n",
        "    except:\n",
        "\n",
        "        return(\"convert_rna_to_dna: Can't open file \" + Seqs + \" for reading.\")\n",
        "\n",
        "    Seqs_Names = []\n",
        "    seq_name = \"\"\n",
        "\n",
        "    line = SEQS.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        line = line.rstrip()\n",
        "        match1 = re.match(r'^>(.*)', line)\n",
        "        if match1:\n",
        "\n",
        "            seq_name = match1.group(1)\n",
        "\n",
        "\n",
        "        elif 'u' in line or 'U' in line:\n",
        "\n",
        "            Seqs_Names.append(seq_name)\n",
        "            line = line.replace('u', 't')\n",
        "            line = line.replace('U', 'T')\n",
        "\n",
        "        OUT.write(line + \"\\n\")\n",
        "        line = SEQS.readline()\n",
        "\n",
        "    OUT.close()\n",
        "    SEQS.close()\n",
        "\n",
        "    return(\"OK\", Seqs_Names)\n",
        "\n",
        "\n",
        "def create_pdf():\n",
        "\n",
        "    create_pdf_regular_or_cbs(True, vars['Colored_Seq_PDF'])\n",
        "    create_pdf_regular_or_cbs(False, vars['Colored_Seq_CBS_PDF'])\n",
        "\n",
        "\n",
        "def create_pymol(input, prefix):\n",
        "\n",
        "    cmd = \"pymol -qc \" + input + \" -d \\\"run \" + vars['pymol_color_script_isd'] + \"\\\"\\n\"\n",
        "    cmd += \"pymol -qc \" + input + \" -d \\\"run \" + vars['pymol_color_script_CBS_isd'] + \"\\\"\\n\"\n",
        "\n",
        "    LOG.write(\"create_pymol : %s\\n\" %cmd)\n",
        "    submit_job_to_Q(\"PYMOL\", cmd)\n",
        "\n",
        "    pymol_session = \"consurf_pymol_session.pse\"\n",
        "    pymol_session_CBS = \"consurf_CBS_pymol_session.pse\"\n",
        "\n",
        "    if os.path.exists(pymol_session) and os.path.getsize(pymol_session) != 0:\n",
        "\n",
        "        os.chmod(pymol_session, 0o664)\n",
        "        os.rename(pymol_session, prefix + pymol_session)\n",
        "        vars['zip_list'].append(prefix + pymol_session)\n",
        "\n",
        "    if os.path.exists(pymol_session_CBS) and os.path.getsize(pymol_session_CBS) != 0:\n",
        "\n",
        "        os.chmod(pymol_session_CBS, 0o664)\n",
        "        os.rename(pymol_session_CBS, prefix + pymol_session_CBS)\n",
        "        vars['zip_list'].append(prefix + pymol_session_CBS)\n",
        "\n",
        "\n",
        "def create_chimera(input, prefix):\n",
        "\n",
        "    run_chimera(input, prefix + \"consurf_chimerax_session.cxs\", vars['chimera_color_script'])\n",
        "    run_chimera(input, prefix + \"consurf_CBS_chimerax_session.cxs\", vars['chimera_color_script_CBS'])\n",
        "\n",
        "def run_chimera(input, output, script):\n",
        "\n",
        "    cmd = \"chimerax --nogui --script '%s %s %s' --exit\\n\" %(script, input, output)\n",
        "    LOG.write(\"create_chimera : %s\\n\" %cmd)\n",
        "    submit_job_to_Q(\"CHIMERA\", cmd)\n",
        "\n",
        "    vars['zip_list'].append(output)\n",
        "\n",
        "\"\"\"\n",
        "def check_msa_tree_match(ref_msa_seqs):\n",
        "\n",
        "    ref_tree_nodes = []\n",
        "    check_validity_tree_file(ref_tree_nodes)\n",
        "    LOG.write(\"check_msa_tree_match : check if all the nodes in the tree are also in the MSA\\n\")\n",
        "\n",
        "    for node in ref_tree_nodes:\n",
        "\n",
        "        if not node in ref_msa_seqs:\n",
        "\n",
        "            exit_on_error('user_error', \"The uploaded tree file is inconsistant with the uploaded MSA file. The node '\" + node + \"' is found in the tree file, but there is no sequence in the MSA file with that exact name. Note that the search is case-sensitive!\")\n",
        "\n",
        "    LOG.write(\"check_msa_tree_match : check if all the sequences in the MSA are also in the tree\\n\")\n",
        "\n",
        "    for seq_name in ref_msa_seqs: #check that all the msa nodes are in the tree\n",
        "\n",
        "        if not seq_name in ref_tree_nodes:\n",
        "\n",
        "            exit_on_error('user_error', \"The uploaded MSA file is inconsistant with the uploaded tree file. The Sequence name '\" + seq_name + \"' is found in the MSA file, but there is no node with that exact name in the tree file. Note that the search is case-sensitive!\")\n",
        "\n",
        "    vars['unique_seqs'] = len(ref_msa_seqs)\n",
        "    LOG.write(\"There are \" + str(vars['unique_seqs']) + \" in the MSA.\\n\")\n",
        "\"\"\"\n",
        "\n",
        "def get_info_from_msa(seq_names):\n",
        "\n",
        "    # returns the name of the query sequence and fills the input array with the names of the sequences\n",
        "    query_seq = \"\"\n",
        "    msa_format = check_msa_format(vars['user_msa_file_name'])\n",
        "    try:\n",
        "\n",
        "        alignment = AlignIO.read(vars['user_msa_file_name'], msa_format)\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"get_info_from_msa : can't open the file \" + vars['user_msa_file_name'] + \" for reading.\")\n",
        "\n",
        "    try:\n",
        "\n",
        "        MSA = open(vars['working_dir'] + vars['msa_fasta'], 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"get_info_from_msa : can't open the file \" + vars['msa_fasta'] + \" for writing.\")\n",
        "\n",
        "    for record in alignment:\n",
        "\n",
        "        new_seq_name = str(record.id)\n",
        "        if new_seq_name in seq_names:\n",
        "\n",
        "            exit_on_error('user_error', \"The sequence %s appears more than once in the MSA.\" %new_seq_name)\n",
        "\n",
        "\n",
        "        seq_names.append(new_seq_name)\n",
        "\n",
        "        seq = str(record.seq)\n",
        "        if form['msa_SEQNAME'] == new_seq_name:\n",
        "\n",
        "            query_seq = seq\n",
        "\n",
        "\n",
        "        if form['DNA_AA'] == \"Nuc\" and ('u' in seq or 'U' in seq):\n",
        "\n",
        "            seq = seq.replace('u', 't')\n",
        "            seq = seq.replace('U', 'T')\n",
        "            print(\"Warnning: The seqeunce '\" + new_seq_name + \"' contains a 'U' replaced by 'T'\")\n",
        "\n",
        "        MSA.write(\">%s\\n%s\\n\" %(new_seq_name, seq))\n",
        "\n",
        "    MSA.close()\n",
        "\n",
        "\n",
        "    num_of_seq = len(seq_names)\n",
        "    vars['unique_seqs'] = num_of_seq\n",
        "    vars['final_number_of_homologoues'] = num_of_seq\n",
        "    LOG.write(\"MSA contains \" + str(num_of_seq) + \" sequences\\n\")\n",
        "    if num_of_seq < 5:\n",
        "\n",
        "        exit_on_error('user_error',\"The MSA file contains only \" + str(num_of_seq) + \" sequences. The minimal number of homologues required for the calculation is 5.\")\n",
        "\n",
        "    query_seq = query_seq.replace(\"-\", \"\")\n",
        "    query_seq = query_seq.upper()\n",
        "\n",
        "    if query_seq == \"\":\n",
        "\n",
        "        exit_on_error('user_error', \"The query sequence is not in the msa. Please choose the name of the query sequence by adding the flag --query\")\n",
        "\n",
        "    vars['msa_SEQNAME'] = form['msa_SEQNAME']\n",
        "    #vars['query_string'] = form['msa_SEQNAME']\n",
        "    vars['MSA_query_seq'] = query_seq\n",
        "    \"\"\"\n",
        "    # there is no input seq, use msa seq instead\n",
        "    vars['protein_seq_string'] = vars['MSA_query_seq']\n",
        "    try:\n",
        "\n",
        "        QUERY_FROM_MSA = open(vars['working_dir'] + vars['protein_seq'], 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"get_info_from_msa : Could not open %s for writing.\" %vars['protein_seq'])\n",
        "\n",
        "    QUERY_FROM_MSA.write(\">\" + form['msa_SEQNAME'] + \"\\n\")\n",
        "    QUERY_FROM_MSA.write(vars['MSA_query_seq'] + \"\\n\")\n",
        "    QUERY_FROM_MSA.close()\n",
        "    \"\"\"\n",
        "\n",
        "\n",
        "def create_pipe_file(pipeFile, pipeFile_CBS, seq3d_grades, seq3d_grades_isd, isd_residue_color_ArrRef, no_isd_residue_color_ArrRef, pdb_file_name, user_chain, IN_pdb_id_capital, identical_chains, pdb_object):\n",
        "\n",
        "    # CREATE PART of PIPE\n",
        "    partOfPipe = \"partOfPipe\"\n",
        "    partOfPipe_CBS = \"partOfPipe_CBS\"\n",
        "\n",
        "    length_of_seqres = pdb_object.get_num_known_seqs()\n",
        "    length_of_atom = pdb_object.get_num_known_atoms()\n",
        "\n",
        "    LOG.write(\"create_part_of_pipe_new(%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\\n\" %(partOfPipe, vars['unique_seqs'], \"user_DB\", \"seq3d_grades_isd\", \"seq3d_grades\", length_of_seqres, length_of_atom, \"isd_residue_color_ArrRef\", \"no_isd_residue_color_ArrRef\", form['E_VALUE'], form['ITERATIONS'], form['MAX_NUM_HOMOL'], form['MSAprogram'], form['ALGORITHM'], form['SUB_MATRIX'], \"legacy\"))\n",
        "    create_part_of_pipe_new(partOfPipe, vars['unique_seqs'], \"user_DB\", seq3d_grades_isd, seq3d_grades, length_of_seqres, length_of_atom, isd_residue_color_ArrRef, no_isd_residue_color_ArrRef, form['E_VALUE'], form['ITERATIONS'], form['MAX_NUM_HOMOL'], form['MSAprogram'], form['ALGORITHM'], form['SUB_MATRIX'], vars['Average pairwise distance'], \"legacy\")\n",
        "\n",
        "\n",
        "    # create the color blind friendly version\n",
        "    create_part_of_pipe_new(partOfPipe_CBS, vars['unique_seqs'], \"user_DB\", seq3d_grades_isd, seq3d_grades, length_of_seqres, length_of_atom, isd_residue_color_ArrRef, no_isd_residue_color_ArrRef, form['E_VALUE'], form['ITERATIONS'], form['MAX_NUM_HOMOL'], form['MSAprogram'], form['ALGORITHM'], form['SUB_MATRIX'], vars['Average pairwise distance'], \"cb\")\n",
        "\n",
        "\n",
        "    LOG.write(\"extract_data_from_pdb(%s)\\n\" %pdb_file_name)\n",
        "    header_pipe = extract_data_from_pdb(pdb_file_name)\n",
        "\n",
        "\n",
        "    # GET THE FILE NAMES\n",
        "    msa_filename = \"\"\n",
        "    msa_query_seq_name = \"\"\n",
        "    if vars['user_msa_file_name'] is not None:\n",
        "\n",
        "        msa_filename = vars['user_msa_file_name']\n",
        "        msa_query_seq_name = form['msa_SEQNAME']\n",
        "\n",
        "    tree_filename = \"\"\n",
        "    if form['tree_name'] is not None:\n",
        "\n",
        "        tree_filename = vars['tree_file']\n",
        "\n",
        "    # GET THE CURRENT TIME\n",
        "    completion_time = str(datetime.now().time())\n",
        "    run_date = str(datetime.now().date())\n",
        "\n",
        "    # USE THE CREATED PART of PIPE to CREATE ALL THE PIPE TILL THE PDB ATOMS (DELETE THE PART PIPE)\n",
        "    LOG.write(\"create_consurf_pipe_new(%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\\n\" %(vars['working_dir'], IN_pdb_id_capital, user_chain, \"header_pipe\", pipeFile, identical_chains, partOfPipe, vars['working_dir'], form['Run_Number'], msa_filename, msa_query_seq_name, tree_filename, vars['submission_time'], completion_time, run_date))\n",
        "    create_consurf_pipe_new(vars['working_dir'], IN_pdb_id_capital, user_chain, header_pipe, pipeFile, identical_chains, partOfPipe, vars['working_dir'], form['Run_Number'], msa_filename, msa_query_seq_name, tree_filename, vars['submission_time'], completion_time, run_date)\n",
        "\n",
        "    # USE THE CREATED PART of PIPE to CREATE ALL THE PIPE TILL THE PDB ATOMS (DELETE THE PART PIPE) - Color friendly version\n",
        "    LOG.write(\"create_consurf_pipe_new(%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\\n\" %(vars['working_dir'], IN_pdb_id_capital, user_chain, \"header_pipe\", pipeFile_CBS, identical_chains, partOfPipe_CBS, vars['working_dir'], form['Run_Number'], msa_filename, msa_query_seq_name, tree_filename, vars['submission_time'], completion_time, run_date))\n",
        "    create_consurf_pipe_new(vars['working_dir'], IN_pdb_id_capital, user_chain, header_pipe, pipeFile_CBS, identical_chains, partOfPipe_CBS, vars['working_dir'], form['Run_Number'], msa_filename, msa_query_seq_name, tree_filename, vars['submission_time'], completion_time, run_date)\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "    # Add the PDB data to the pipe\n",
        "    LOG.write(\"add_pdb_data_to_pipe(%s, %s)\\n\" %(pdb_file_name, pipeFile))\n",
        "    add_pdb_data_to_pipe(pdb_file_name, pipeFile)\n",
        "\n",
        "\n",
        "    # Add the PDB data to the pipe - color blind version\n",
        "    LOG.write(\"add_pdb_data_to_pipe(%s, %s)\\n\" %(pdb_file_name, pipeFile_CBS))\n",
        "    add_pdb_data_to_pipe(pdb_file_name, pipeFile_CBS)\n",
        "\n",
        "\n",
        "def compare_atom_to_query(Query_seq, ATOM_seq, pairwise_aln, PDB_Name):\n",
        "\n",
        "    # in case there are both seqres and atom fields, checks the similarity between the 2 sequences.\n",
        "\n",
        "    [first_seq, second_seq, score] = pairwise_alignment(Query_seq, ATOM_seq, pairwise_aln, \"QUERY\")\n",
        "    return(first_seq, second_seq)\n",
        "\n",
        "\n",
        "def upload_protein_sequence():\n",
        "\n",
        "\n",
        "    if os.path.exists(vars['uploaded_Seq']) and os.path.getsize(vars['uploaded_Seq']) != 0: # file fasta uploaded\n",
        "\n",
        "        try:\n",
        "\n",
        "            UPLOADED = open(vars['uploaded_Seq'], 'r')\n",
        "\n",
        "        except:\n",
        "\n",
        "            exit_on_error('sys_error', \"upload_protein_sequence : Cannot open the file \" + vars['protein_seq'] + \"for writing!\")\n",
        "\n",
        "        protein_seq_string = UPLOADED.read()\n",
        "        UPLOADED.close()\n",
        "\n",
        "        if protein_seq_string.count('>') > 1:\n",
        "\n",
        "            exit_on_error('user_error', \"The protein input <a href = \\\"%s\\\">sequence</a> contains more than one FASTA sequence. If you wish to upload MSA, please upload it as a file.\" %protein_seq_string)\n",
        "\n",
        "        # delete sequence name and white spaces\n",
        "        protein_seq_string = re.sub(r'>.*\\n', \"\", protein_seq_string)\n",
        "        protein_seq_string = re.sub(r'\\s', \"\", protein_seq_string)\n",
        "\n",
        "    else:\n",
        "\n",
        "        exit_on_error('sys_error', 'upload_protein_sequence : no user sequence.')\n",
        "\n",
        "    # we write the sequence to a file for the homologues search\n",
        "    try:\n",
        "\n",
        "        UPLOADED = open(vars['working_dir'] + vars['protein_seq'], 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"upload_protein_sequence : Cannot open the file \" + vars['protein_seq'] + \"for writing!\")\n",
        "\n",
        "    UPLOADED.write(\">Input_seq\\n\" + protein_seq_string)\n",
        "    UPLOADED.close()\n",
        "\n",
        "    amino_acids = [\"A\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"K\", \"L\", \"M\", \"N\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"V\", \"W\", \"Y\", \"X\"]\n",
        "    nucleic_acids = [\"A\", \"C\", \"G\", \"T\", \"U\", \"N\"]\n",
        "    if form['DNA_AA'] == \"AA\":\n",
        "\n",
        "        amino = False\n",
        "        for char in protein_seq_string:\n",
        "\n",
        "            if not char.upper() in amino_acids:\n",
        "\n",
        "                exit_on_error('user_error', \"The input sequence contains the illegal character %s.\" %char)\n",
        "\n",
        "            elif not char.upper() in nucleic_acids:\n",
        "\n",
        "                amino = True\n",
        "\n",
        "        if not amino:\n",
        "\n",
        "            exit_on_error('user_error',\"It seems that the protein input is only composed of Nucleotides (i.e. :A,T,C,G). Please note that you chose to run the server based on amino acids sequnce and not DNA / RNA sequence.<br />You may translate your sequence to amino acids and resubmit your query, or alternatively choose to analyze nucleotides.<br />\")\n",
        "\n",
        "    else:\n",
        "\n",
        "\n",
        "        for char in protein_seq_string:\n",
        "\n",
        "            if not char.upper() in amino_acids:\n",
        "\n",
        "                exit_on_error('user_error', \"The input sequence contains the illegal character %s.\" %char)\n",
        "\n",
        "            elif not char.upper() in nucleic_acids:\n",
        "\n",
        "                exit_on_error('user_error',\"It seems that the input sequence contains the Amino Acid %s. Please note that you chose to run the server based on nucleotides sequnce and not protein sequence.<br />You may resubmit your query and choose to analyze Amino Acids.<br />\" %char)\n",
        "\n",
        "    vars['protein_seq_string'] = protein_seq_string\n",
        "    vars['query_string'] = \"Input_seq\" # name of the sequence is saved for rate4site\n",
        "\n",
        "\n",
        "def array_to_string(array):\n",
        "\n",
        "    string = \"\"\n",
        "    for word in array:\n",
        "\n",
        "        string += \" \" +word\n",
        "\n",
        "    return string\n",
        "\n",
        "\n",
        "def zip_all_outputs():\n",
        "\n",
        "    zipObj = ZipFile(vars['All_Outputs_Zip'], 'w')\n",
        "\n",
        "    for file in vars['zip_list']:\n",
        "\n",
        "        if os.path.exists(file):\n",
        "\n",
        "            zipObj.write(file)\n",
        "\n",
        "    zipObj.close()\n",
        "    create_download_link(vars['All_Outputs_Zip'], \"Download zip with all the files.\")\n",
        "\n",
        "\n",
        "\n",
        "def find_identical_chains_in_PDB_file(pdb_Object, query_chain):\n",
        "\n",
        "    # Looking in the PDB for chains identical to the original chain\n",
        "    ATOMS = pdb_Object.get_ATOM_withoutX()\n",
        "\n",
        "    # string with identical chains\n",
        "    identicalChains = query_chain\n",
        "\n",
        "    # looking for chains identical to the original chain\n",
        "    for chain in ATOMS:\n",
        "\n",
        "        if query_chain != chain:\n",
        "\n",
        "            other_seq = ATOMS[chain]\n",
        "            query_seq = ATOMS[query_chain]\n",
        "            chain_length = len(other_seq)\n",
        "            OrgChain_length = len(query_seq)\n",
        "\n",
        "            # if length not similar, skip\n",
        "            if min(OrgChain_length, chain_length)/max(OrgChain_length, chain_length) <= 0.9:\n",
        "\n",
        "                continue\n",
        "\n",
        "            # compare the two chains\n",
        "            try:\n",
        "\n",
        "                if pairwise_alignment(other_seq, query_seq) > 0.95:\n",
        "\n",
        "                    identicalChains += \" \" + chain\n",
        "\n",
        "            except:\n",
        "\n",
        "                LOG.write(\"find_identical_chains_in_PDB_file: Error comparing the chains %s and %s\\n\" %(query_chain, chain))\n",
        "\n",
        "    return identicalChains\n",
        "\n",
        "\n",
        "\n",
        "def replace_TmpFactor_Rate4Site_Scores(chain, pdb_file, gradesPE, pdb_file_with_score_at_TempFactor):\n",
        "\n",
        "    # This will create a PDB file that contains the Rate4Site scores instead of the TempFactor Column\n",
        "\n",
        "    Rate4Site_Grades = {}\n",
        "    LOG.write(\"read_Rate4Site_gradesPE(%s, %s)\\n\" %(gradesPE, str(Rate4Site_Grades)))\n",
        "    read_Rate4Site_gradesPE(gradesPE, Rate4Site_Grades)\n",
        "\n",
        "    if vars['cif_or_pdb'] == \"pdb\":\n",
        "\n",
        "        LOG.write(\"replace_TmpFactor_Rate4Site_Scores_PDB(%s, %s, %s, %s)\\n\" %(pdb_file, chain, \"Rate4Site_Grades\", pdb_file_with_score_at_TempFactor))\n",
        "        replace_TmpFactor_Rate4Site_Scores_PDB(pdb_file, chain, Rate4Site_Grades, pdb_file_with_score_at_TempFactor)\n",
        "\n",
        "    else:\n",
        "\n",
        "        LOG.write(\"replace_TmpFactor_Rate4Site_Scores_CIF(%s, %s, %s, %s)\\n\" %(pdb_file, chain, \"Rate4Site_Grades\", pdb_file_with_score_at_TempFactor))\n",
        "        replace_TmpFactor_Rate4Site_Scores_CIF(pdb_file, chain, Rate4Site_Grades, pdb_file_with_score_at_TempFactor)\n",
        "\n",
        "\n",
        "\n",
        "def replace_TmpFactor_Consurf_Scores(atom_grades, chain, pdb_file, prefix):\n",
        "\n",
        "    # This Will create a File containing the ATOMS records with the ConSurf grades instead of the TempFactor column\n",
        "\n",
        "    if vars['cif_or_pdb'] == \"pdb\":\n",
        "\n",
        "        LOG.write(\"replace_TmpFactor_Consurf_Scores_PDB(atom_grades, %s, %s, %s);\\n\" %(chain, pdb_file, prefix))\n",
        "        replace_TmpFactor_Consurf_Scores_PDB(atom_grades, chain, pdb_file, prefix)\n",
        "\n",
        "    else:\n",
        "        LOG.write(\"replace_TmpFactor_Consurf_Scores_CIF(atom_grades, %s, %s, %s);\\n\" %(chain, pdb_file, prefix))\n",
        "        replace_TmpFactor_Consurf_Scores_CIF(atom_grades, chain, pdb_file, prefix)\n",
        "\n",
        "\n",
        "def install_rate4site(rate4site_dir, rate4site_slow_dir):\n",
        "\n",
        "    LOG.write(\"Installing rate4site.\\n\")\n",
        "\n",
        "    # create directory for rate4site\n",
        "    submit_job_to_Q(\"download_rate4site\", \"git clone https://github.com/barakav/r4s_for_collab.git\")\n",
        "\n",
        "    # create directory for rate4site slow\n",
        "    shutil.copytree(rate4site_dir, rate4site_slow_dir)\n",
        "\n",
        "    # make rate4site\n",
        "    submit_job_to_Q(\"install_rate4site\", \"cd %s\\nmake\\nchmod 755 rate4site\" %rate4site_dir)\n",
        "\n",
        "    # change the make file\n",
        "    os.remove(rate4site_slow_dir + \"Makefile\")\n",
        "    os.rename(rate4site_slow_dir + \"Makefile_slow\", rate4site_slow_dir + \"Makefile\")\n",
        "\n",
        "    # make rate4site\n",
        "    submit_job_to_Q(\"install_rate4site\", \"cd %s\\nmake\\nchmod 755 rate4site\" %rate4site_slow_dir)\n",
        "\n",
        "\n",
        "import os\n",
        "import torch\n",
        "import numpy as np\n",
        "from Bio import SeqIO\n",
        "from transformers import BertModel, BertTokenizer\n",
        "\n",
        "\n",
        "def run_rate4site():\n",
        "    print(\"Bypassing Rate4Site. Initializing AI Protein Language Model (ProtBert)...\")\n",
        "\n",
        "    vars['r4s_log'] = \"r4s.log\"\n",
        "    vars['r4s_out'] = \"r4s.res\"\n",
        "    vars['gradesPE_Output'] = [] # Initialize early\n",
        "\n",
        "    try:\n",
        "        # 1. Extract Sequence\n",
        "        records = list(SeqIO.parse(vars['msa_fasta'], \"fasta\"))\n",
        "        query_seq_raw = str(records[0].seq).replace(\"-\", \"\")\n",
        "        query_seq_spaced = \" \".join(list(query_seq_raw))\n",
        "\n",
        "        # 2. Model Inference\n",
        "        tokenizer = BertTokenizer.from_pretrained(\"Rostlab/prot_bert\", do_lower_case=False)\n",
        "        model = BertModel.from_pretrained(\"Rostlab/prot_bert\", output_attentions=True)\n",
        "\n",
        "        inputs = tokenizer(query_seq_spaced, return_tensors=\"pt\")\n",
        "        with torch.no_grad():\n",
        "            outputs = model(**inputs)\n",
        "\n",
        "        # 3. Handle Attention & Slicing\n",
        "        # ProtBert adds [CLS] at index 0 and [SEP] at the end. We must remove them.\n",
        "        attentions = outputs.attentions\n",
        "        last_layer_attn = attentions[-1].squeeze(0).mean(dim=0)\n",
        "        importance_scores = last_layer_attn.sum(dim=0).numpy()\n",
        "\n",
        "        # SLICE: Remove the first and last tokens to match original sequence length\n",
        "        importance_scores = importance_scores[1:-1]\n",
        "\n",
        "        # 4. Normalize (1-9)\n",
        "        s_min, s_max = importance_scores.min(), importance_scores.max()\n",
        "        consurf_style_scores = 1 + 8 * (importance_scores - s_min) / (s_max - s_min)\n",
        "\n",
        "       # 5. Populate gradesPE_Output with ALL legacy naming conventions\n",
        "        for i, score in enumerate(consurf_style_scores):\n",
        "            grade = int(np.clip(round(score), 1, 9))\n",
        "            residue_letter = query_seq_raw[i]\n",
        "\n",
        "            entry = {\n",
        "                'POS': i + 1,                  # Fixes the previous KeyError: 'POS'\n",
        "                'SEQ': residue_letter,         # ADD THIS: Fixes KeyError: 'SEQ'\n",
        "                'RES': residue_letter,         # Keep for compatibility\n",
        "                'AA': residue_letter,          # Added just in case\n",
        "                'SEQ_POS': i + 1,\n",
        "                'GRADE': grade,\n",
        "                'SCORE': round(float(score), 3),\n",
        "                'COLOR': grade,\n",
        "                'INTERVAL_LOW': round(float(score), 3),\n",
        "                'INTERVAL_HIGH': round(float(score), 3),\n",
        "                'INTERVAL_LOW_COLOR': grade,\n",
        "                'INTERVAL_HIGH_COLOR': grade,\n",
        "                'MSA_NUM': 100,\n",
        "                'ISD': 0,\n",
        "                'MSA_DATA': 'AI_PREDICTED'\n",
        "            }\n",
        "            vars['gradesPE_Output'].append(entry)\n",
        "\n",
        "        # 6. File Mocking\n",
        "        with open(vars['r4s_out'], \"w\") as f:\n",
        "            f.write(\"# POS  RES  SCORE\\n\")\n",
        "            for item in vars['gradesPE_Output']:\n",
        "                f.write(f\"{item['SEQ_POS']}  {item['RES']}  {item['SCORE']}\\n\")\n",
        "\n",
        "        with open(vars['r4s_log'], \"w\") as f:\n",
        "            f.write(\"AI-Inference successful.\")\n",
        "\n",
        "        if not os.path.exists(vars['tree_file']):\n",
        "            with open(vars['tree_file'], \"w\") as f:\n",
        "                f.write(\"(Query_AI:0.00001);\")\n",
        "\n",
        "        print(f\"Success! Populated gradesPE_Output with {len(vars['gradesPE_Output'])} scores.\")\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"AI Optimization failed: {e}\")\n",
        "        # If this fails, the downstream code will still see an empty list and error out,\n",
        "        # which is better than proceeding with \"garbage\" data.\n",
        "\n",
        "\n",
        "\n",
        "def run_rate4site_oldkevin():\n",
        "\n",
        "    rate4site_dir = \"/content/r4s_for_collab/\"\n",
        "    rate4site_slow_dir = \"/content/r4s_for_collab_slow/\"\n",
        "    vars['r4s_log'] = \"r4s.log\" # log file\n",
        "    vars['r4s_out'] = \"r4s.res\" # output file\n",
        "\n",
        "    #install_rate4site(rate4site_dir, rate4site_slow_dir)\n",
        "\n",
        "    MatrixHash = {'JTT' : '-Mj', 'MTREV' : '-Mr', 'CPREV' : '-Mc', 'WAG' : '-Mw', 'DAYHOFF' : '-Md', 'T92' : '-Mt', 'HKY' : '-Mh', 'GTR' : '-Mg', 'JC_NUC' : '-Mn', 'JC_AA' : '-Ma', 'LG' : '-Ml'}\n",
        "\n",
        "    params = \"rate4site -a '%s' -s %s -zn %s -bn -o %s -v 9\" %(vars['query_string'], vars['msa_fasta'], MatrixHash[(form['SUB_MATRIX']).upper()], vars['r4s_out'])\n",
        "    if vars['running_mode'] == \"_mode_pdb_msa_tree\" or vars['running_mode'] == \"_mode_msa_tree\":\n",
        "\n",
        "        params += \" -t %s\" %vars['tree_file']\n",
        "\n",
        "    if form['ALGORITHM'] == \"Bayes\":\n",
        "\n",
        "        params +=  \" -ib\"\n",
        "\n",
        "    else:\n",
        "\n",
        "        params +=  \" -im\"\n",
        "\n",
        "    r4s_comm = rate4site_dir + params + \" -l \" + vars['r4s_log']\n",
        "\n",
        "    LOG.write(\"run_rate4site : running command: %s\\n\" %r4s_comm)\n",
        "    print(\"The conservation scores are being calculated. Please wait.\")\n",
        "    submit_job_to_Q(\"rate4site\", r4s_comm)\n",
        "\n",
        "    # if the run failed - we rerun using the slow verion\n",
        "    if check_if_rate4site_failed(vars['r4s_log']):\n",
        "\n",
        "        LOG.write(\"run_rate4site : The run of rate4site failed. Sending warning message to output.\\nThe same run will be done using the SLOW version of rate4site.\\n\")\n",
        "        #print(\"Warning: The given MSA is very large, therefore it will take longer for ConSurf calculation to finish. The results will be sent to the e-mail address provided.<br>The calculation continues nevertheless.\")\n",
        "        vars['r4s_log'] = \"r4s_slow.log\"\n",
        "        r4s_comm = rate4site_slow_dir + params + \" -l \" + vars['r4s_log']\n",
        "        LOG.write(\"run_rate4site : running command: %s\\n\" %str(r4s_comm))\n",
        "        submit_job_to_Q(\"rate4siteSlow\", r4s_comm)\n",
        "\n",
        "        if check_if_rate4site_failed(vars['r4s_log']):\n",
        "\n",
        "            exit_on_error('sys_error', \"Both rate4site and rate4site slow failed.\")\n",
        "\n",
        "    extract_diversity_matrix_info(vars['r4s_log'])\n",
        "\n",
        "    if not os.path.exists(vars['tree_file']):\n",
        "\n",
        "        os.rename(\"TheTree.txt\", vars['tree_file'])\n",
        "\n",
        "def run_rate4site_old():\n",
        "\n",
        "    rate4s = vars['script_dir'] + \"/rate4site_bioseq/rate4site\"\n",
        "    rate4s_ML = vars['script_dir'] + \"/rate4site_bioseq/rate4site.24Mar2010\"\n",
        "    rate4s_slow = vars['script_dir'] + \"/rate4site_bioseq/rate4site.doubleRep\"\n",
        "\n",
        "    vars['r4s_log'] = \"r4s.log\" # log file\n",
        "    vars['r4s_out'] = \"r4s.res\" # output file\n",
        "    MatrixHash = {'JTT' : '-Mj', 'MTREV' : '-Mr', 'CPREV' : '-Mc', 'WAG' : '-Mw', 'DAYHOFF' : '-Md', 'T92' : '-Mt', 'HKY' : '-Mh', 'GTR' : '-Mg', 'JC_NUC' : '-Mn', 'JC_AA' : '-Ma', 'LG' : '-Ml'}\n",
        "\n",
        "    params = \" -a '%s' -s %s -zn %s -bn -o %s\" %(vars['query_string'], vars['msa_fasta'], MatrixHash[(form['SUB_MATRIX']).upper()], vars['r4s_out'])\n",
        "    if vars['running_mode'] == \"_mode_pdb_msa_tree\" or vars['running_mode'] == \"_mode_msa_tree\":\n",
        "\n",
        "        params += \" -t %s\" %vars['tree_file']\n",
        "\n",
        "    if form['ALGORITHM'] == \"Bayes\":\n",
        "\n",
        "        params += \" -ib -n 32 -v 9\"\n",
        "        r4s_comm = rate4s + params\n",
        "\n",
        "    else:\n",
        "\n",
        "        params += \" -im -v 9\"\n",
        "        r4s_comm = rate4s_ML + params\n",
        "\n",
        "    r4s_comm += \" -l \" + vars['r4s_log']\n",
        "    LOG.write(\"run_rate4site : running command: %s\\n\" %r4s_comm)\n",
        "    submit_job_to_Q(\"rate4site\", r4s_comm)\n",
        "\n",
        "    # if the run failed - we rerun using the slow verion\n",
        "    if check_if_rate4site_failed(vars['r4s_log']):\n",
        "\n",
        "        vars['r4s_log'] = \"r4s_slow.log\"\n",
        "        LOG.write(\"run_rate4site : The run of rate4site failed. Sending warning message to output.\\nThe same run will be done using the SLOW version of rate4site.\\n\")\n",
        "        print(\"Warning: The given MSA is very large, therefore it will take longer for ConSurf calculation to finish. The results will be sent to the e-mail address provided.<br>The calculation continues nevertheless.\")\n",
        "        r4s_comm = rate4s_slow + params + \" -l \" + vars['r4s_log']\n",
        "        LOG.write(\"run_rate4site : running command: %s\\n\" %r4s_comm)\n",
        "        submit_job_to_Q(\"rate4siteSlow\", r4s_comm)\n",
        "\n",
        "        if check_if_rate4site_failed(vars['r4s_log']):\n",
        "\n",
        "            exit_on_error('sys_error', \"Both rate4site and rate4site slow failed.\")\n",
        "\n",
        "    extract_diversity_matrix_info(vars['r4s_log'])\n",
        "\n",
        "def find_best_substitution_model():\n",
        "\n",
        "    try:\n",
        "\n",
        "        # convert fasta to phylip\n",
        "        msa_phy_filepath = \"input_msa.phy\"\n",
        "        #convert_msa_format(vars['msa_fasta'], \"fasta\", msa_phy_filepath, \"phylip-relaxed\")\n",
        "        AlignIO.convert(vars['msa_fasta'], \"fasta\", msa_phy_filepath, \"phylip-relaxed\")\n",
        "        os.chmod(vars['msa_fasta'], 0o644)\n",
        "        os.chmod(msa_phy_filepath, 0o644)\n",
        "\n",
        "        if form['DNA_AA'] == \"Nuc\":\n",
        "\n",
        "            run_jmt(msa_phy_filepath)\n",
        "\n",
        "        else:\n",
        "\n",
        "            run_prottest(msa_phy_filepath)\n",
        "\n",
        "    except:\n",
        "\n",
        "        vars['best_fit'] = \"model_search_failed\"\n",
        "        form['SUB_MATRIX'] = \"JTT\"\n",
        "        print(\"The evolutionary model search has failed. The JTT model is chosen by default.\")\n",
        "\n",
        "\n",
        "def run_prottest(msa_file_path):\n",
        "\n",
        "    output_file_path = vars['job_name'] + \"_model_selection.txt\"\n",
        "    cmd = \"java -jar %s -log disabled -i %s -AICC -o %s -S 1 -JTT -LG -MtREV -Dayhoff -WAG -CpREV -threads 1\" %(vars['prottest'], msa_file_path, output_file_path)\n",
        "    submit_job_to_Q(\"protest\", cmd)\n",
        "    LOG.write(\"run_protest: %s\\n\" %cmd)\n",
        "\n",
        "    f = open(output_file_path, 'r')\n",
        "\n",
        "    match = re.search(r\"(?<=Best model according to AICc: ).*\", f.read())\n",
        "    f.close()\n",
        "    if match:\n",
        "\n",
        "        vars['best_fit'] = \"model_found\"\n",
        "        model = match.group()\n",
        "        model = model.strip('()')\n",
        "        print(\"The best evolutionary model was selected to be: \" + model)\n",
        "        create_download_link(output_file_path, \"See details\")\n",
        "        vars['zip_list'].append(output_file_path)\n",
        "\n",
        "    else:\n",
        "\n",
        "        vars['best_fit'] = \"model_search_failed\"\n",
        "        model = \"JTT\"\n",
        "        print(\"The evolutionary model search has failed. The JTT model is chosen by default.\")\n",
        "\n",
        "    form['SUB_MATRIX'] = model\n",
        "\n",
        "def run_jmt(msa_file_path):\n",
        "\n",
        "    JMT_JAR_FILE = GENERAL_CONSTANTS.JMODELTEST2\n",
        "    output_file_path = \"model_selection.txt\"\n",
        "    cmd = \"java -Djava.awt.headless=true -jar %s -d %s -t BIONJ -AICc -f -o %s\" %(JMT_JAR_FILE, msa_file_path, output_file_path)\n",
        "    submit_job_to_Q(\"jmt\", cmd)\n",
        "    LOG.write(\"run_jmt: %s\\n\" %cmd)\n",
        "\n",
        "    f = open(output_file_path, 'r')\n",
        "\n",
        "    start_reading = False\n",
        "    JMT_VALID_MODELS = [\"JC\",\"HKY\",\"GTR\"]\n",
        "\n",
        "    # extract best model from table\n",
        "    line = f.readline()\n",
        "    while line != \"\":\n",
        "\n",
        "        if start_reading:\n",
        "\n",
        "            line = line.strip()\n",
        "            split_row = line.split()\n",
        "            model = split_row[0]\n",
        "            if model in JMT_VALID_MODELS:\n",
        "\n",
        "                f.close()\n",
        "                #model = model.strip('()')\n",
        "                if model == \"JC\":\n",
        "\n",
        "                    form['SUB_MATRIX'] = \"JC_Nuc\"\n",
        "\n",
        "                else:\n",
        "\n",
        "                    form['SUB_MATRIX'] = model\n",
        "\n",
        "                vars['best_fit'] = \"model_found\"\n",
        "                print(\"The best evolutionary model was selected to be: \" + model)\n",
        "                return\n",
        "\n",
        "        elif re.search(r'Model             -lnL    K     AICc       delta       weight   cumWeight', line, re.M):\n",
        "\n",
        "            start_reading = True\n",
        "\n",
        "        line = f.readline()\n",
        "\n",
        "    vars['best_fit'] = \"model_search_failed\"\n",
        "    form['SUB_MATRIX'] = \"JC_Nuc\"\n",
        "    print(\"The evolutionary model search has failed. The JC model is chosen by default\")\n",
        "    f.close()\n",
        "\n",
        "\n",
        "def convert_msa_format(infile, infileformat, outfile, outfileformat):\n",
        "\n",
        "    try:\n",
        "\n",
        "        AlignIO.convert(infile, infileformat, outfile, outfileformat)\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"convert_msa_format : exception\")\n",
        "\n",
        "def create_MSA():\n",
        "\n",
        "    if form['MSAprogram'] == \"CLUSTALW\":\n",
        "\n",
        "        cmd = \"clustalw -infile=%s -outfile=%s\" %(vars['FINAL_sequences'], vars['msa_clustal'])\n",
        "        LOG.write(\"create_MSA : run %s\\n\" %cmd)\n",
        "        submit_job_to_Q(\"clustalw\", cmd)\n",
        "        convert_msa_format(vars['msa_clustal'], \"clustal\", vars['msa_fasta'], \"fasta\")\n",
        "\n",
        "    elif form['MSAprogram'] == \"MAFFT\":\n",
        "\n",
        "        cmd = \"mafft --localpair --maxiterate 1000 --quiet %s > %s\" %(vars['FINAL_sequences'], vars['msa_fasta'])\n",
        "        LOG.write(\"create_MSA : run %s\\n\" %cmd)\n",
        "        submit_job_to_Q(\"MAFFT\", cmd)\n",
        "        #convert_msa_format(vars['msa_fasta'], \"fasta\", vars['msa_clustal'], \"clustal\")\n",
        "\n",
        "    elif form['MSAprogram'] == \"PRANK\":\n",
        "\n",
        "        cmd = \"prank -d=%s -o=%s -F\" %(vars['FINAL_sequences'], vars['msa_fasta'])\n",
        "        print(\"Warning: PRANK is accurate but slow MSA program, please be patient.\")\n",
        "        LOG.write(\"create_MSA : run %s\\n\" %cmd)\n",
        "        submit_job_to_Q(\"PRANK\", cmd)\n",
        "\n",
        "        if os.path.exists(vars['msa_fasta'] + \".2.fas\"):\n",
        "\n",
        "            vars['msa_fasta'] += \".2.fas\"\n",
        "\n",
        "        elif os.path.exists(vars['msa_fasta'] + \".1.fas\"):\n",
        "\n",
        "            vars['msa_fasta'] += \".1.fas\"\n",
        "\n",
        "        elif os.path.exists(vars['msa_fasta'] + \".best.fas\"):\n",
        "\n",
        "            vars['msa_fasta'] +=  \".best.fas\"\n",
        "\n",
        "        #convert_msa_format(vars['msa_fasta'], \"fasta\", vars['msa_clustal'], \"clustal\")\n",
        "\n",
        "    elif form['MSAprogram'] == \"MUSCLE\":\n",
        "\n",
        "        #cmd = \"muscle -align %s -output %s\" %(vars['FINAL_sequences'], vars['msa_fasta'])\n",
        "        cmd = \"muscle -in %s -out %s -quiet\" %(vars['FINAL_sequences'], vars['msa_fasta'])\n",
        "        LOG.write(\"create_MSA : run %s\\n\" %cmd)\n",
        "        submit_job_to_Q(\"MUSCLE\", cmd)\n",
        "        #convert_msa_format(vars['msa_clustal'], \"clustal\", vars['msa_fasta'], \"fasta\")\n",
        "\n",
        "    else:\n",
        "\n",
        "        exit_on_error('user_error', \"Choose one of the programs for creating the msa: clustalw, mafft, prank or muscle.\")\n",
        "\n",
        "    if not os.path.exists(vars['msa_fasta']) or os.path.getsize(vars['msa_fasta']) == 0:\n",
        "\n",
        "        exit_on_error('user_error', \"The %s program failed to create the MSA. Choose a different program to create the MSA.\" %form['MSAprogram'])\n",
        "\n",
        "def choose_final_homologoues(ref_to_seqs_hash, ref_to_cd_hash, max_num_homologs, witch_unifrom, output_file, rejected_file, num_rejected_homologs):\n",
        "\n",
        "    LOG.write(\"sort_sequences_from_eval(%s ,%s , %f, %s, %s, %s, %d)\\n\" %(\"ref_to_seqs_hash\", \"ref_to_cd_hash\", max_num_homologs, witch_unifrom, output_file, rejected_file, num_rejected_homologs))\n",
        "\n",
        "\n",
        "    query_name = \"\"\n",
        "    query_AAseq = \"\"\n",
        "    counter = 1\n",
        "\n",
        "\n",
        "    try:\n",
        "\n",
        "        FINAL = open(vars['FINAL_sequences'], 'w')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error',\"choose_final_homologoues : cannot open the file %s for writing\" %vars['FINAL_sequences'])\n",
        "\n",
        "    # we write the query sequence to the file of the final homologs\n",
        "    FINAL.write(\">%s\\n%s\\n\" %(vars['query_string'], vars['protein_seq_string']))\n",
        "\n",
        "    final_file_size = os.path.getsize(vars['FINAL_sequences']) # take the size of the file before we add more sequences to it\n",
        "\n",
        "    try:\n",
        "\n",
        "        REJECTED = open(rejected_file, 'a')\n",
        "\n",
        "    except:\n",
        "\n",
        "        exit_on_error('sys_error', \"Can't open '\" + rejected_file + \"' for writing.\")\n",
        "\n",
        "    # write query details\n",
        "    if query_AAseq != \"\":\n",
        "\n",
        "        FINAL.write(\">%s\\n%s\\n\" %(query_name, query_AAseq))\n",
        "\n",
        "    size_cd_hit_hash = len(ref_to_cd_hash)\n",
        "    uniform = 1\n",
        "    jump = 1\n",
        "\n",
        "    if witch_unifrom == \"sample\":\n",
        "\n",
        "        uniform = int(size_cd_hit_hash / max_num_homologs)\n",
        "        if uniform == 0:\n",
        "\n",
        "            uniform = 1\n",
        "\n",
        "    final_number_of_homologoues = 1\n",
        "    REJECTED.write(\"\\n\\tSequences rejected because of the requirement to select only %d representative homologs\\n\\n\" %(max_num_homologs + 1))\n",
        "    # write homologs\n",
        "    for s_name in sorted(ref_to_seqs_hash.keys(), key = ref_to_seqs_hash.get):\n",
        "\n",
        "        # write next homolog\n",
        "        if s_name in ref_to_cd_hash: # and 'SEQ' in ref_to_cd_hash[s_name]:\n",
        "\n",
        "            if counter != jump or counter > max_num_homologs * uniform:\n",
        "\n",
        "\n",
        "\n",
        "                counter += 1\n",
        "                REJECTED.write(\"%d %s\\n\" %(num_rejected_homologs, s_name))\n",
        "                num_rejected_homologs += 1\n",
        "                continue\n",
        "\n",
        "            final_number_of_homologoues += 1\n",
        "            FINAL.write(\">%s\\n%s\\n\" %(s_name, ref_to_cd_hash[s_name]))\n",
        "            counter += 1\n",
        "            jump += uniform\n",
        "\n",
        "    FINAL.close()\n",
        "    REJECTED.close()\n",
        "\n",
        "    vars['final_number_of_homologoues'] = final_number_of_homologoues\n",
        "    # check that more sequences were added to the file\n",
        "    if not final_file_size < os.path.getsize(vars['FINAL_sequences']):\n",
        "\n",
        "        exit_on_error('sys_error', \"choose_final_homologoues : the file \" + vars['FINAL_sequences'] + \" doesn't contain sequences\")\n",
        "\n",
        "def cluster_homologoues(ref_cd_hit_hash):\n",
        "\n",
        "    msg = \"\"\n",
        "    LOG.write(\"cluster_homologoues : create_cd_hit_output(%s, %s, %f, %s, %s);\\n\" %(vars['HITS_fasta_file'], vars['cd_hit_out_file'], vars['hit_redundancy']/100, ref_cd_hit_hash, form['DNA_AA']))\n",
        "    total_num_of_hits = create_cd_hit_output(vars['HITS_fasta_file'], vars['cd_hit_out_file'], vars['hit_redundancy']/100, ref_cd_hit_hash, form['DNA_AA'])\n",
        "\n",
        "    if form['MAX_NUM_HOMOL'].upper() == 'ALL':\n",
        "\n",
        "        form['MAX_NUM_HOMOL'] = total_num_of_hits\n",
        "\n",
        "    if total_num_of_hits < vars['min_num_of_hits']: # less seqs than the minimum: exit\n",
        "\n",
        "        if total_num_of_hits <= 1:\n",
        "\n",
        "            msg = \"There is only 1 \"\n",
        "\n",
        "        else:\n",
        "\n",
        "            msg = \"There are only %d \" %total_num_of_hits\n",
        "\n",
        "        msg += \"unique hits. The minimal number of sequences required for the calculation is %d. You may try to: \" %vars['min_num_of_hits']\n",
        "        msg += \"Re-run the server with a multiple sequence alignment file of your own. Increase the Evalue. Decrease the Minimal %ID For Homologs\"\n",
        "\n",
        "        if int(form['ITERATIONS']) < 5:\n",
        "\n",
        "            msg += \" Increase the number of \" + form['Homolog_search_algorithm'] + \" iterations.\"\n",
        "\n",
        "        msg += \"\\n\"\n",
        "        exit_on_error('user_error',msg)\n",
        "\n",
        "    elif total_num_of_hits + 1 < vars['low_num_of_hits']: # less seqs than 10 : output a warning.\n",
        "\n",
        "        msg = \"Warning: There are \"\n",
        "\n",
        "        if total_num_of_hits + 1 < vars['number_of_homologoues_before_cd-hit']: # because we will add the query sequence itself to all the unique sequences.\n",
        "\n",
        "            msg += \"%d hits, only %d of them are\" %(vars['number_of_homologoues_before_cd-hit'], total_num_of_hits+1)\n",
        "\n",
        "        else:\n",
        "\n",
        "            msg += str(total_num_of_hits + 1)\n",
        "\n",
        "        msg += \" unique sequences. The calculation is performed on the %d unique sequences, but it is recommended to run the program with a multiple sequence alignment file containing at least %s sequences.\" %(total_num_of_hits + 1, vars['low_num_of_hits'])\n",
        "\n",
        "    else:\n",
        "\n",
        "        msg = \"There are %d %s hits. %d of them are unique, including the query. The calculation is performed on \" %(vars['number_of_homologoues_before_cd-hit'], form['Homolog_search_algorithm'], total_num_of_hits + 1)\n",
        "\n",
        "        if total_num_of_hits <= int(form['MAX_NUM_HOMOL']):\n",
        "\n",
        "            msg += \"%d unique sequences.\" %(total_num_of_hits + 1)\n",
        "\n",
        "        elif form['best_uniform_sequences'] == \"best\":\n",
        "\n",
        "            msg += \"the %s <a href=\\\"<?=$orig_path?>/%s\\\" style=\\\"color: #400080; text-decoration:underline;\\\">sequences</a> closest to the query (with the lowest E-value).\" %(form['MAX_NUM_HOMOL'], vars['FINAL_sequences_html'])\n",
        "\n",
        "        else:\n",
        "\n",
        "            msg += \"a sample of %s sequences that represent the list of homologues to the query.\" %form['MAX_NUM_HOMOL']\n",
        "            #msg += \"a sample of %s <a href=\\\"<?=$orig_path?>/%s\\\" style=\\\"color: #400080; text-decoration:underline;\\\">sequences</a> that represent the list of homologues to the query.\" %(form['MAX_NUM_HOMOL'], vars['FINAL_sequences_html'])\n",
        "\n",
        "    print(msg)\n",
        "\n",
        "    #if os.path.exists(vars['HITS_rejected_file']) and os.path.getsize(vars['HITS_rejected_file']) != 0:\n",
        "\n",
        "        #print_message_to_output(\"Here is the <a href=\\\"<?=$orig_path?>/\" + vars['HITS_rejected_file'] + \"\\\" TARGET=Rejected_Seqs style=\\\"color: #400080; text-decoration:underline;\\\">list of sequences</a> that produced significant alignments, but were not chosen as hits.\")\n",
        "        #print_message_to_output(\"<a href=\\\"<?=$orig_path?>/\" + vars['HITS_rejected_file'] + \"\\\" TARGET=Rejected_Seqs style=\\\"color: #400080; text-decoration:underline;\\\">Click here</a> if you wish to view the list of sequences which produced significant alignments, but were not chosen as hits.\")\n",
        "\n",
        "    return(total_num_of_hits + 1)\n",
        "\n",
        "\n",
        "\n",
        "def submit_job_to_Q(job_name_prefix, cmd):\n",
        "\n",
        "    os.system(\"cd %s\\n%s\" %(vars['working_dir'], cmd))\n",
        "\n",
        "\n",
        "def compare_atom_seqres_or_msa(what_to_compare):\n",
        "\n",
        "    # in case there is a msa and pdb, we check the similarity between the atom and the msa sequences\n",
        "    # in case there is no msa and there are both atom and seqres sequences, we check the similarity between them\n",
        "\n",
        "    pairwise_aln = \"PDB_\" + what_to_compare + \".aln\"\n",
        "    atom_length = len(vars['ATOM_without_X_seq'])\n",
        "    alignment_score = 0\n",
        "    other_query_length = len(vars['protein_seq_string'])\n",
        "    query_line = {}\n",
        "    atom_line = \"sequence extracted from the ATOM field of the PDB file\"\n",
        "    query_line['SEQRES'] = \"sequence extracted from the SEQRES field of the PDB file\"\n",
        "    query_line['MSA'] = \"sequence extracted from the MSA file\"\n",
        "\n",
        "    # compare the length of sequences. output a message accordingly\n",
        "    if other_query_length != 0 and other_query_length < atom_length:\n",
        "\n",
        "        print(\"The %s is shorter than the %s. The %s sequence has %d residues and the ATOM sequence has %d residues. The calculation continues nevertheless.\" %(query_line[what_to_compare],atom_line ,what_to_compare, other_query_length, atom_length))\n",
        "\n",
        "    if atom_length < other_query_length:\n",
        "\n",
        "        if atom_length < other_query_length * 0.2:\n",
        "\n",
        "            print(\"Warning: The %s is significantly shorter than the %s. The %s sequence has %d residues and the ATOM sequence has only %d residues. The calculation continues nevertheless.\" %(atom_line, query_line[what_to_compare], what_to_compare, other_query_length, atom_length))\n",
        "\n",
        "        else:\n",
        "\n",
        "            print(\"The %s is shorter than the %s. The %s sequence has %d residues and the ATOM sequence has %d residues. The calculation continues nevertheless.\" %(atom_line, query_line[what_to_compare], what_to_compare, other_query_length, atom_length))\n",
        "\n",
        "    # match the sequences\n",
        "    LOG.write(\"compare_atom_seqres_or_msa : Align ATOM and \" + what_to_compare + \" sequences\\n\")\n",
        "    [vars['seqres_or_msa_seq_with_gaps'], vars['ATOM_seq_with_gaps'], alignment_score] = pairwise_alignment(vars['protein_seq_string'], vars['ATOM_without_X_seq'], pairwise_aln, what_to_compare)\n",
        "\n",
        "    if alignment_score < 100:\n",
        "\n",
        "        if alignment_score < 30:\n",
        "\n",
        "            exit_on_error('user_error',\"The Score of the alignment between the %s and the %s is ONLY %d%% identity.<br>See <a href=\\\"<?=$orig_path?>/%s\\\" style=\\\"color: #400080; text-decoration:underline;\\\" TARGET=PairWise_Align>pairwise alignment</a>.\" %(query_line[what_to_compare], atom_line, alignment_score, pairwise_aln))\n",
        "\n",
        "        else:\n",
        "\n",
        "            print(\"The Score of the alignment between the %s and the %s is %d%% identity. The calculation continues nevertheless.\" %(query_line[what_to_compare], atom_line, alignment_score))\n",
        "\n",
        "\n",
        "def exit_on_error(which_error, error_msg):\n",
        "\n",
        "    complete_msg = \"\\n\\nEXIT on error:\\n\\n\" + error_msg + \"\\n\"\n",
        "    LOG.write(complete_msg)\n",
        "    print(complete_msg)\n",
        "    raise Exception(\"The error is not an exception\")\n",
        "\n",
        "\n",
        "\n",
        "def analyse_seqres_atom():\n",
        "\n",
        "    # there is no ATOM field in the PDB\n",
        "\n",
        "    if vars['ATOM_without_X_seq'] == \"\":\n",
        "\n",
        "        exit_on_error('user_error', \"There is no ATOM derived information in the PDB file.<br>Please refer to the OVERVIEW for detailed information about the PDB format.\")\n",
        "\n",
        "    # there is no SEQRES field in the PDB\n",
        "\n",
        "    if vars['SEQRES_seq'] == \"\":\n",
        "\n",
        "        msg = \"Warning: There is no SEQRES derived information in the PDB file. The calculation will be based on the ATOM derived sequence. \"\n",
        "\n",
        "        if vars['running_mode'] == \"_mode_pdb_no_msa\":\n",
        "\n",
        "            msg += \"If this sequence is incomplete, we recommend to re-run the server using an external multiple sequence alignment file, which is based on the complete protein sequence.\"\n",
        "\n",
        "        LOG.write(\"analyse_seqres_atom : There is no SEQRES derived information in the PDB file.\\n\")\n",
        "        print(msg)\n",
        "\n",
        "    if form['DNA_AA'] == \"AA\":\n",
        "\n",
        "        # check if seqres contains nucleic acid\n",
        "        if vars['pdb_object'].get_type() == \"Nuc\":\n",
        "\n",
        "            exit_on_error('user_error', \"The selected chain: \" + form['PDB_chain'] + \" contains nucleic acid, and you have selected amino acid\")\n",
        "\n",
        "    else:\n",
        "\n",
        "        # check if seqres contains amino acid\n",
        "        #type_SEQRES = vars['pdb_object'].get_type_SEQRES()\n",
        "        #if form['PDB_chain'] in type_SEQRES and type_SEQRES[form['PDB_chain']] == \"AA\":\n",
        "        if vars['pdb_object'].get_type() == \"AA\":\n",
        "\n",
        "            exit_on_error('user_error', \"The selected chain: \" + form['PDB_chain'] + \" contains amino acid, and you have selected nucleic acid\")\n",
        "\n",
        "    # if modified residues exists, print them to the screen\n",
        "\n",
        "    MAXIMUM_MODIFIED_PERCENT = 0.15\n",
        "    MODIFIED_COUNT = vars['pdb_object'].get_MODIFIED_COUNT()\n",
        "    if MODIFIED_COUNT > 0:\n",
        "\n",
        "        if form['DNA_AA'] == \"AA\":\n",
        "\n",
        "            if len(vars['SEQRES_seq']) > 0 and MODIFIED_COUNT / len(vars['SEQRES_seq']) > MAXIMUM_MODIFIED_PERCENT:\n",
        "\n",
        "                LOG.write(\"MODIFIED_COUNT %d\\nSEQRES_seq %s\\n\" %(MODIFIED_COUNT, vars['SEQRES_seq']))\n",
        "                exit_on_error('user_error', \"Too many modified residues were found in SEQRES field; %0.3f%% of the residues are modified, the maximum is %0.3f%%.\" %(MODIFIED_COUNT / len(vars['SEQRES_seq']) ,MAXIMUM_MODIFIED_PERCENT))\n",
        "\n",
        "            LOG.write(\"analyse_seqres_atom : modified residues found\\n\")\n",
        "            print(\"Please note: Before the analysis took place, modified residues read from SEQRES field were converted back to the original residues:\\n\" + vars['pdb_object'].get_MODIFIED_LIST() + \".\")\n",
        "\n",
        "        else:\n",
        "\n",
        "            LOG.write(\"analyse_seqres_atom : modified residues found\\n\")\n",
        "            print(\"Please note: Before the analysis took place, modified nucleotides read from SEQRES field were converted back to the original nucleotides:\\n\" + vars['pdb_object'].get_MODIFIED_LIST() + \".\")\n",
        "\n",
        "\n",
        "def get_seqres_atom_seq(PDB_Obj, query_chain, pdb_file_name, model = False):\n",
        "\n",
        "    # extract the sequences from the pdb\n",
        "\n",
        "    seqres = PDB_Obj.get_SEQRES() # seqres sequence\n",
        "    atom = PDB_Obj.get_ATOM() # atom sequence with gaps filled with X for amino acids and N for nucleic acids. This is only used to find the residues' place in the pdb\n",
        "    atom_without_X = PDB_Obj.get_ATOM_withoutX()[query_chain] # atom sequence with gaps not filled\n",
        "\n",
        "    if seqres == \"\" and atom == \"\":\n",
        "\n",
        "        if model:\n",
        "\n",
        "            return('user_error', \"The protein sequence for chain '%s' was not found in SEQRES nor ATOM fields in the <a href=\\\"%s\\\">PDB file</a>.\" %(query_chain, pdb_file_name), \"1\")\n",
        "\n",
        "        else:\n",
        "\n",
        "            exit_on_error('user_error', \"The protein sequence for chain '%s' was not found in SEQRES nor ATOM fields in the PDB file.\" %query_chain)\n",
        "\n",
        "    return [seqres, atom, atom_without_X]\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "vars['run_log'] = \"log.txt\"\n",
        "form['Homolog_search_algorithm'] = \"MMseqs2\"\n",
        "form['DNA_AA'] = \"AA\"\n",
        "\n",
        "\n",
        "LOG = open(vars['run_log'], 'w')\n",
        "\n",
        "vars['BLAST_out_file'] = \"%s/%s_all/uniref.a3m\" %(vars['working_dir'], vars['job_name'])\n",
        "\n",
        "\n",
        "\n",
        "vars['Msa_percentageFILE'] = vars['job_name'] + \"_msa_aa_variety_percentage.csv\"\n",
        "\n",
        "\n",
        "vars['All_Outputs_Zip'] = \"Consurf_Outputs_\" + vars['job_name'] + \".zip\"\n",
        "\n",
        "\n",
        "vars['hit_min_length'] = 0.60 # minimum length of homologs\n",
        "vars['min_num_of_hits'] = 5 # minimum number of homologs\n",
        "vars['FINAL_sequences'] = \"query_final_homolougs.fasta\" # finial homologs for creating the MSA\n",
        "vars['FINAL_sequences_html'] = vars['job_name'] + \"_final_homolougs.html\" # html files showing the finial homologs to the user\n",
        "vars['r4s_log'] = \"r4s.log\" # rate4site log\n",
        "vars['r4s_out'] = \"r4s.res\" # rate4site output\n",
        "vars['r4s_slow_log'] = \"r4s_slow.log\" # rate4site slow log\n",
        "vars['gradesPE'] = vars['job_name'] + \"_consurf_grades.txt\" # file with consurf output\n",
        "vars['zip_list'] = []\n",
        "vars['date'] = date.today().strftime(\"%d/%m/%Y\")\n",
        "vars['color_array'] = {1 : \"#0A7D82\", 2 : \"#4BAFBE\", 3 : \"#A5DCE6\", 4 : \"#D7F0F0\", 5 : \"#FFFFFF\", 6 : \"#FAEBF5\", 7 : \"#FAC8DC\", 8 : \"#F07DAA\", 9 : \"#A0285F\", 'ISD' : \"#FFFF96\"}\n",
        "vars['color_array_CBS'] = {1 : \"#0F5A23\", 2 : \"#5AAF5F\", 3 : \"#A5DCA0\", 4 : \"#D7F0D2\", 5 : \"#FFFFFF\", 6 : \"#E6D2E6\", 7 : \"#C3A5CD\", 8 : \"#9B6EAA\", 9 : \"#782882\", 'ISD' : \"#FFFF96\"}\n",
        "\n",
        "vars['Colored_Seq_PDF'] = vars['job_name'] + \"_consurf_colored_seq.pdf\"\n",
        "vars['Colored_Seq_CBS_PDF'] = vars['job_name'] + \"_consurf_colored_seq_CBS.pdf\"\n",
        "\n",
        "vars['msa_clustal'] = \"msa_clustal.aln\" # if the file is not in clustal format, we create a clustal copy of it\n",
        "\n",
        "vars['protein_seq'] = \"protein_seq.fas\" # a fasta file with the protein sequence from PDB or from protein seq input\n",
        "\n",
        "vars['gradesPE_Output'] = [] # an array to hold all the information that should be printed to gradesPE\n",
        "# in each array's cell there is a hash for each line from r4s.res.\n",
        "# POS: position of that aa in the sequence ; SEQ : aa in one letter ;\n",
        "# GRADE : the given grade from r4s output ; COLOR : grade according to consurf's scale\n",
        "\n",
        "\n",
        "\n",
        "try:\n",
        "\n",
        "    vars['zip_list'].append(vars['tree_file'])\n",
        "    vars['zip_list'].append(vars['gradesPE'])\n",
        "    vars['zip_list'].append(vars['Msa_percentageFILE'])\n",
        "    vars['zip_list'].append(vars['Colored_Seq_PDF'])\n",
        "    vars['zip_list'].append(vars['Colored_Seq_CBS_PDF'])\n",
        "    vars['zip_list'].append(vars['msa_fasta'])\n",
        "    vars['zip_list'].append(vars['pymol'])\n",
        "    vars['zip_list'].append(vars['pymol_CBS'])\n",
        "    vars['zip_list'].append(vars['FINAL_sequences_html'])\n",
        "\n",
        "    ## mode : include pdb\n",
        "\n",
        "    # create a pdbParser, to get various info from the pdb file\n",
        "    if vars['running_mode'] == \"_mode_pdb_no_msa\" or vars['running_mode'] == \"_mode_pdb_msa\" or vars['running_mode'] == \"_mode_pdb_msa_tree\":\n",
        "\n",
        "        upload_PDB()\n",
        "        extract_data_from_model()\n",
        "\n",
        "\n",
        "    ## mode : only protein sequence\n",
        "\n",
        "    # if there is only protein sequence: we upload it.\n",
        "    elif vars['running_mode'] == \"_mode_no_pdb_no_msa\":\n",
        "\n",
        "        upload_sequence()\n",
        "        #upload_protein_sequence()\n",
        "\n",
        "    ## mode : no msa - with PDB or without PDB\n",
        "\n",
        "    if vars['running_mode'] == \"_mode_pdb_no_msa\" or vars['running_mode'] == \"_mode_no_pdb_no_msa\":\n",
        "\n",
        "        create_MSA_parameters()\n",
        "        no_MSA()\n",
        "\n",
        "    ## mode : include msa\n",
        "\n",
        "    elif vars['running_mode'] == \"_mode_pdb_msa\" or vars['running_mode'] == \"_mode_msa\" or vars['running_mode'] == \"_mode_pdb_msa_tree\" or vars['running_mode'] == \"_mode_msa_tree\":\n",
        "\n",
        "        upload_MSA()\n",
        "        extract_data_from_MSA()\n",
        "\n",
        "    if form['SUB_MATRIX'] == \"BEST\":\n",
        "\n",
        "        #vars['best_fit'] = True\n",
        "        find_best_substitution_model()\n",
        "\n",
        "    else:\n",
        "\n",
        "        vars['best_fit'] = \"model_chosen\"\n",
        "\n",
        "    run_rate4site()\n",
        "    assign_colors_according_to_r4s_layers()\n",
        "    write_MSA_percentage_file()\n",
        "\n",
        "    ## mode : include pdb\n",
        "\n",
        "    if vars['running_mode'] == \"_mode_pdb_no_msa\" or vars['running_mode'] == \"_mode_pdb_msa\" or vars['running_mode'] == \"_mode_pdb_msa_tree\":\n",
        "\n",
        "        consurf_create_output()\n",
        "\n",
        "    ## mode : ConSeq - NO PDB\n",
        "\n",
        "    if vars['running_mode'] == \"_mode_msa\" or vars['running_mode'] == \"_mode_no_pdb_no_msa\" or vars['running_mode'] == \"_mode_msa_tree\":\n",
        "\n",
        "        conseq_create_output()\n",
        "\n",
        "    zip_all_outputs()\n",
        "    create_download_link(vars['msa_fasta'], \"Download Multiple Sequence Alignment in FASTA format\")\n",
        "    create_download_link(vars['gradesPE'], \"Download Per-Residue Details\")\n",
        "    create_download_link(vars['Msa_percentageFILE'], \"Download csv file showing residue variety per position in the MSA \")\n",
        "    create_download_link(vars['tree_file'], \"Download Phylogenetic Tree\")\n",
        "\n",
        "    ## Arrange The HTML Output File\n",
        "    print(\"The calculation is done.\")\n",
        "    LOG.close()\n",
        "\n",
        "except Exception as e:\n",
        "\n",
        "    if str(e) != \"The error is not an exception\":\n",
        "\n",
        "        LOG.write(traceback.format_exc())\n",
        "        LOG.close()\n",
        "        raise(e)\n",
        "\n",
        "    else:\n",
        "\n",
        "        LOG.close()\n",
        "\n",
        "\n",
        "\n"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    },
    "widgets": {
      "application/vnd.jupyter.widget-state+json": {
        "cf766ab06e0f43208f28fbf02f371aea": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_b3aa93f6e4a94a369ec1dfa8f3dbb688",
              "IPY_MODEL_0d1cdfdf33ae4d2292a6b191554e6e94",
              "IPY_MODEL_4d1c4ce75b924083bb9ee7ebec04b6d1"
            ],
            "layout": "IPY_MODEL_a6e7cb42440b4dcb8b464ba99c556cdc"
          }
        },
        "b3aa93f6e4a94a369ec1dfa8f3dbb688": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_799435b1fda343db866c5fdde712931b",
            "placeholder": "",
            "style": "IPY_MODEL_87508c5ca57e4b36a40b54f57f43a18c",
            "value": "Loadingweights:100%"
          }
        },
        "0d1cdfdf33ae4d2292a6b191554e6e94": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FloatProgressModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "success",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_8bd9d7a10e69492193c32945b7d1c972",
            "max": 487,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_0f8180bdfdfd4e9fb96b36b025f4ca5c",
            "value": 487
          }
        },
        "4d1c4ce75b924083bb9ee7ebec04b6d1": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "HTMLModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_16af9184a54848c685bfb3d380d74fd5",
            "placeholder": "",
            "style": "IPY_MODEL_7f5a34d7ce184d5cb5e3ce619924c4a5",
            "value": "487/487[00:00&lt;00:00,721.32it/s,Materializingparam=pooler.dense.weight]"
          }
        },
        "a6e7cb42440b4dcb8b464ba99c556cdc": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "799435b1fda343db866c5fdde712931b": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "87508c5ca57e4b36a40b54f57f43a18c": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "8bd9d7a10e69492193c32945b7d1c972": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "0f8180bdfdfd4e9fb96b36b025f4ca5c": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ProgressStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        },
        "16af9184a54848c685bfb3d380d74fd5": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "7f5a34d7ce184d5cb5e3ce619924c4a5": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        }
      }
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}